diff --git a/src/Framework/MockObject/Matcher.php b/src/Framework/MockObject/Matcher.php
index 55763024d..9cb6b87cd 100644
--- a/src/Framework/MockObject/Matcher.php
+++ b/src/Framework/MockObject/Matcher.php
@@ -153,6 +153,7 @@ public function matches(Invocation $invocation): bool
             throw new MethodNameNotConfiguredException;
         }
 
+dump($this->invocationRule, $invocation);
         if (!$this->invocationRule->matches($invocation)) {
             return false;
         }
