diff --git a/src/Framework/MockObject/Rule/Parameters.php b/src/Framework/MockObject/Rule/Parameters.php
index 3761973c9..1f926ed15 100644
--- a/src/Framework/MockObject/Rule/Parameters.php
+++ b/src/Framework/MockObject/Rule/Parameters.php
@@ -102,7 +102,7 @@ private function doVerify(): bool
         if ($this->invocation === null) {
             throw new ExpectationFailedException('Mocked method does not exist.');
         }
-
+dump([__METHOD__, $this->invocation->parameters(), count($this->parameters)]);
         if (count($this->invocation->parameters()) < count($this->parameters)) {
             $message = 'Parameter count for invocation %s is too low.';


