diff --git a/src/Framework/MockObject/InvocationHandler.php b/src/Framework/MockObject/InvocationHandler.php
index 2c4c5a864..139f4343c 100644
--- a/src/Framework/MockObject/InvocationHandler.php
+++ b/src/Framework/MockObject/InvocationHandler.php
@@ -103,6 +103,7 @@ public function invoke(Invocation $invocation): mixed
         $returnValue    = null;
 
         foreach ($this->matchers as $match) {
+dump([__METHOD__, $match]);
             try {
                 if ($match->matches($invocation)) {
                     $value = $match->invoked($invocation);
@@ -143,6 +144,7 @@ public function invoke(Invocation $invocation): mixed
     public function matches(Invocation $invocation): bool
     {
         foreach ($this->matchers as $matcher) {
+dump([__METHOD__, $matcher]);
             if (!$matcher->matches($invocation)) {
                 return false;
             }
