diff --git a/src/Framework/Exception/Exception.php b/src/Framework/Exception/Exception.php
index 0b21e6de3..4722bbe61 100644
--- a/src/Framework/Exception/Exception.php
+++ b/src/Framework/Exception/Exception.php
@@ -44,6 +44,11 @@ class Exception extends RuntimeException implements \PHPUnit\Exception
      */
     protected $serializableTrace;
 
+    /**
+     * @var string
+     */
+    protected $previousTrace;
+
     public function __construct($message = '', $code = 0, Throwable $previous = null)
     {
         parent::__construct($message, $code, $previous);
@@ -53,6 +58,10 @@ public function __construct($message = '', $code = 0, Throwable $previous = null
         foreach (array_keys($this->serializableTrace) as $key) {
             unset($this->serializableTrace[$key]['args']);
         }
+
+        if ($previous) {
+            $this->previousTrace = Filter::getFilteredStacktrace($previous);
+        }
     }
 
     public function __toString(): string
@@ -63,6 +72,10 @@ public function __toString(): string
             $string .= "\n" . $trace;
         }
 
+        if ($this->previousTrace) {
+            $string .= "\nCaused by\n" . $this->previousTrace;
+        }
+
         return $string;
     }
 
