diff --git a/src/Framework/MockObject/Rule/Parameters.php b/src/Framework/MockObject/Rule/Parameters.php
index cb71b271c..603fa3642 100644
--- a/src/Framework/MockObject/Rule/Parameters.php
+++ b/src/Framework/MockObject/Rule/Parameters.php
@@ -132,6 +132,7 @@ private function doVerify(): bool
         }

         foreach ($this->parameters as $i => $parameter) {
+dump([__METHOD__, $i, $parameter, $this->invocation->getParameters()[$i]]);
             $parameter->evaluate(
                 $this->invocation->getParameters()[$i],
                 sprintf(

