diff --git a/src/Framework/MockObject/Rule/Parameters.php b/src/Framework/MockObject/Rule/Parameters.php
index cb71b271c..bd919766a 100644
--- a/src/Framework/MockObject/Rule/Parameters.php
+++ b/src/Framework/MockObject/Rule/Parameters.php
@@ -113,7 +113,7 @@ private function doVerify(): bool
         if ($this->invocation === null) {
             throw new ExpectationFailedException('Mocked method does not exist.');
         }
-
+dump([__METHOD__, $this->invocation->getParameters(), $this->parameters]);
         if (count($this->invocation->getParameters()) < count($this->parameters)) {
             $message = 'Parameter count for invocation %s is too low.';


