diff --git a/src/Framework/Exception/AssertionFailedError.php b/src/Framework/Exception/AssertionFailedError.php
index 0ba25286f..aaef0f12b 100644
--- a/src/Framework/Exception/AssertionFailedError.php
+++ b/src/Framework/Exception/AssertionFailedError.php
@@ -9,11 +9,58 @@
  */
 namespace PHPUnit\Framework;
 
+use function get_class;
+use PHPUnit\Util\Filter;
+use Throwable;
+
 /**
  * @internal This class is not covered by the backward compatibility promise for PHPUnit
  */
 class AssertionFailedError extends Exception implements SelfDescribing
 {
+    /**
+     * @var string|null
+     */
+    protected $previousThrowableClass;
+
+    /**
+     * @var string|null
+     */
+    protected $previousThrowableMessage;
+
+    /**
+     * @var int|null
+     */
+    protected $previousThrowableCode;
+
+    /**
+     * @var string|null
+     */
+    protected $previousThrowableTrace;
+
+    public function __construct($message = '', $code = 0, Throwable $previous = null)
+    {
+        parent::__construct($message, $code, $previous);
+
+        if ($previous) {
+            $this->previousThrowableClass = get_class($previous);
+            $this->previousThrowableMessage = $previous->getMessage();
+            $this->previousThrowableCode = $previous->getCode();
+            $this->previousThrowableTrace = Filter::getFilteredStacktrace($previous);
+        }
+    }
+
+    public function __toString(): string
+    {
+        $string = parent::__toString();
+
+        if ($this->previousThrowableClass) {
+            $string .= "\nCaused by " . $this->previousThrowableClass . ": " . $this->previousThrowableMessage . "\n" . $this->previousThrowableTrace;
+        }
+
+        return $string;
+    }
+
     /**
      * Wrapper for getMessage() which is declared as final.
      */
