diff --git a/src/Symfony/Component/DependencyInjection/ContainerBuilder.php b/src/Symfony/Component/DependencyInjection/ContainerBuilder.php
index 10e17d4a40..30009cf6a8 100644
--- a/src/Symfony/Component/DependencyInjection/ContainerBuilder.php
+++ b/src/Symfony/Component/DependencyInjection/ContainerBuilder.php
@@ -785,7 +785,10 @@ class ContainerBuilder extends Container implements TaggedContainerInterface
      */
     public function compile(bool $resolveEnvPlaceholders = false): void
     {
+global $xxx;
+if ($xxx) dump('compile.1');
         $compiler = $this->getCompiler();
+if ($xxx) dump('compile.2', get_class($compiler));
 
         if ($this->trackResources) {
             foreach ($compiler->getPassConfig()->getPasses() as $pass) {
@@ -793,12 +796,21 @@ class ContainerBuilder extends Container implements TaggedContainerInterface
             }
         }
         $bag = $this->getParameterBag();
+if ($xxx) dump('compile.3');
 
         if ($resolveEnvPlaceholders && $bag instanceof EnvPlaceholderParameterBag) {
             $compiler->addPass(new ResolveEnvPlaceholdersPass(), PassConfig::TYPE_AFTER_REMOVING, -1000);
         }
 
-        $compiler->compile($this);
+if ($xxx) dump('compile.4');
+        try {
+            $compiler->compile($this);
+        }
+        catch (\Throwable $t) {
+if ($xxx) dump('compile.5.t', $t);
+            throw $t;
+        }
+if ($xxx) dump('compile.5');
 
         foreach ($this->definitions as $id => $definition) {
             if ($this->trackResources && $definition->isLazy()) {
@@ -816,7 +828,9 @@ class ContainerBuilder extends Container implements TaggedContainerInterface
             $this->envPlaceholders = $bag->getEnvPlaceholders();
         }
 
+if ($xxx) dump('compile.6');
         parent::compile();
+if ($xxx) dump('compile.7');
 
         foreach ($this->definitions + $this->aliasDefinitions as $id => $definition) {
             if ('.' === ($id[0] ?? '-')) {
