diff --git a/src/Symfony/Component/ErrorHandler/DebugClassLoader.php b/src/Symfony/Component/ErrorHandler/DebugClassLoader.php
index 4bf8467e2..ebfb5a874 100644
--- a/src/Symfony/Component/ErrorHandler/DebugClassLoader.php
+++ b/src/Symfony/Component/ErrorHandler/DebugClassLoader.php
@@ -332,7 +332,10 @@ class DebugClassLoader
             }
 
             $deprecations = $this->checkAnnotations($refl, $name);
-
+if (count($deprecations)) {
+    //dump(\Drupal\Core\Utility\Error::formatBacktrace(debug_backtrace()));
+    dump($deprecations[0]);
+}
             foreach ($deprecations as $message) {
                 @trigger_error($message, \E_USER_DEPRECATED);
             }
