diff --git a/src/Symfony/Component/ErrorHandler/DebugClassLoader.php b/src/Symfony/Component/ErrorHandler/DebugClassLoader.php
index 4bf8467e2..d75c25644 100644
--- a/src/Symfony/Component/ErrorHandler/DebugClassLoader.php
+++ b/src/Symfony/Component/ErrorHandler/DebugClassLoader.php
@@ -332,7 +332,12 @@ class DebugClassLoader
             }
 
             $deprecations = $this->checkAnnotations($refl, $name);
-
+if (count($deprecations)) {
+    $handler = set_error_handler('var_dump');
+    restore_error_handler();
+    //dump(\Drupal\Core\Utility\Error::formatBacktrace(debug_backtrace()));
+    dump([get_class($handler), $deprecations[0]]);
+}
             foreach ($deprecations as $message) {
                 @trigger_error($message, \E_USER_DEPRECATED);
             }
