diff --git a/src/Symfony/Component/ErrorHandler/DebugClassLoader.php b/src/Symfony/Component/ErrorHandler/DebugClassLoader.php
index 4bf8467e2..47f06527d 100644
--- a/src/Symfony/Component/ErrorHandler/DebugClassLoader.php
+++ b/src/Symfony/Component/ErrorHandler/DebugClassLoader.php
@@ -114,6 +114,8 @@ class DebugClassLoader
     private array $loaded = [];
     private array $patchTypes = [];
 
+    private static bool $holdDeprecationErrors = FALSE;
+    private static array $heldDeprecations = [];
     private static int $caseCheck;
     private static array $checkedClasses = [];
     private static array $final = [];
@@ -177,7 +179,7 @@ class DebugClassLoader
     /**
      * Wraps all autoloaders.
      */
-    public static function enable(): void
+    public static function enable(bool $holdDeprecationErrors = FALSE): void
     {
         // Ensures we don't hit https://bugs.php.net/42098
         class_exists(\Symfony\Component\ErrorHandler\ErrorHandler::class);
@@ -198,6 +200,9 @@ class DebugClassLoader
 
             spl_autoload_register($function);
         }
+
+        self::$heldDeprecations = [];
+        self::$holdDeprecationErrors = $holdDeprecationErrors;
     }
 
     /**
@@ -334,7 +339,11 @@ class DebugClassLoader
             $deprecations = $this->checkAnnotations($refl, $name);
 
             foreach ($deprecations as $message) {
-                @trigger_error($message, \E_USER_DEPRECATED);
+                if (self::$holdDeprecationErrors) {
+                    self::$heldDeprecations[] = $message;
+                } else {
+                    @trigger_error($message, \E_USER_DEPRECATED);
+                }
             }
         }
 
