diff --git a/src/Symfony/Component/DependencyInjection/ContainerBuilder.php b/src/Symfony/Component/DependencyInjection/ContainerBuilder.php
index 10e17d4a40..9c0c7300b5 100644
--- a/src/Symfony/Component/DependencyInjection/ContainerBuilder.php
+++ b/src/Symfony/Component/DependencyInjection/ContainerBuilder.php
@@ -785,6 +785,8 @@ class ContainerBuilder extends Container implements TaggedContainerInterface
      */
     public function compile(bool $resolveEnvPlaceholders = false): void
     {
+global $xxx;
+if ($xxx) dump('compile.1');
         $compiler = $this->getCompiler();
 
         if ($this->trackResources) {
@@ -816,7 +818,9 @@ class ContainerBuilder extends Container implements TaggedContainerInterface
             $this->envPlaceholders = $bag->getEnvPlaceholders();
         }
 
+if ($xxx) dump('compile.2');
         parent::compile();
+if ($xxx) dump('compile.3');
 
         foreach ($this->definitions + $this->aliasDefinitions as $id => $definition) {
             if ('.' === ($id[0] ?? '-')) {
