diff --git a/core/modules/views/src/Plugin/views/argument/ArgumentPluginBase.php b/core/modules/views/src/Plugin/views/argument/ArgumentPluginBase.php
index 9e39de0410..9a45bf0c3e 100644
--- a/core/modules/views/src/Plugin/views/argument/ArgumentPluginBase.php
+++ b/core/modules/views/src/Plugin/views/argument/ArgumentPluginBase.php
@@ -952,6 +952,7 @@ public function summaryBasics($count_field = TRUE) {
     $distinct = ($this->view->display_handler->getOption('distinct') && empty($this->query->no_distinct));
 
     $count_alias = $this->query->addField($this->view->storage->get('base_table'), $this->view->storage->get('base_field'), 'num_records', ['count' => TRUE, 'distinct' => $distinct]);
+dump([__METHOD__, $this->name_alias]);
     $this->query->addGroupBy($this->name_alias);
 
     if ($count_field) {
diff --git a/core/modules/views/src/Plugin/views/query/Sql.php b/core/modules/views/src/Plugin/views/query/Sql.php
index 035df8b593..a5e243d3a7 100644
--- a/core/modules/views/src/Plugin/views/query/Sql.php
+++ b/core/modules/views/src/Plugin/views/query/Sql.php
@@ -1348,6 +1348,7 @@ public function query($get_count = FALSE) {
       $this->hasAggregate = $this->view->display_handler->getOption('group_by');
     }
     $groupby = [];
+dump([__METHOD__, $this->hasAggregate, $this->groupby, $non_aggregates]);
     if ($this->hasAggregate && (!empty($this->groupby) || !empty($non_aggregates))) {
       $groupby = array_unique(array_merge($this->groupby, $non_aggregates));
     }
@@ -1382,6 +1383,7 @@ public function query($get_count = FALSE) {
         if ($field == $this->view->storage->get('base_field')) {
           $field = $this->view->storage->get('base_table') . '.' . $field;
         }
+dump([__METHOD__, $groupby, $field]);
         $query->groupBy($field);
       }
       if (!empty($this->having) && $condition = $this->buildCondition('having')) {
diff --git a/core/modules/views/tests/src/Kernel/QueryGroupByTest.php b/core/modules/views/tests/src/Kernel/QueryGroupByTest.php
index 171688321e..c2654a852e 100644
--- a/core/modules/views/tests/src/Kernel/QueryGroupByTest.php
+++ b/core/modules/views/tests/src/Kernel/QueryGroupByTest.php
@@ -230,7 +230,8 @@ public function testGroupByBaseField() {
     $view->displayHandlers->get('default')->options['fields']['name']['group_type'] = 'min';
     unset($view->displayHandlers->get('default')->options['fields']['id']['group_type']);
     $this->executeView($view);
-    $this->assertStringContainsString('GROUP BY entity_test.id', (string) $view->build_info['query'], 'GROUP BY field includes the base table name when grouping on the base field.');
+dump([__METHOD__, $view->build_info]);
+    $this->assertMatchesRegularExpression('/GROUP BY .entity_test.\..id./', (string) $view->build_info['query'], 'GROUP BY field includes the base table name when grouping on the base field.');
   }
 
   /**
