diff --git a/core/modules/node/tests/src/Functional/NodeTitleTest.php b/core/modules/node/tests/src/Functional/NodeTitleTest.php
index fe3213902d..af6836988a 100644
--- a/core/modules/node/tests/src/Functional/NodeTitleTest.php
+++ b/core/modules/node/tests/src/Functional/NodeTitleTest.php
@@ -90,7 +90,7 @@ public function testNodeTitle() {
     $node = $this->drupalCreateNode($settings);
     // Test that 0 appears as <title>.
     $this->drupalGet('node/' . $node->id());
-    $this->assertSession()->titleEquals('0 | Drupal');
+    $this->assertSession()->titleEquals('0 | Drupalx');
     // Test that 0 appears in the template <h1>.
     $this->assertSession()->elementTextEquals('xpath', '//h1', '0');
 
diff --git a/core/modules/system/tests/src/Functional/Bootstrap/DrupalMessengerServiceTest.php b/core/modules/system/tests/src/Functional/Bootstrap/DrupalMessengerServiceTest.php
index 0671fdf0c4..72e7ac36a8 100644
--- a/core/modules/system/tests/src/Functional/Bootstrap/DrupalMessengerServiceTest.php
+++ b/core/modules/system/tests/src/Functional/Bootstrap/DrupalMessengerServiceTest.php
@@ -2,9 +2,9 @@
 
 namespace Drupal\Tests\system\Functional\Bootstrap;
 
+use Behat\Mink\Exception\ExpectationException;
 use Drupal\Core\Url;
 use Drupal\Tests\BrowserTestBase;
-use PHPUnit\Framework\AssertionFailedError;
 
 /**
  * Tests the Messenger service.
@@ -112,7 +112,7 @@ public function testStatusMessageAssertions(): void {
     try {
       $this->assertSession()->statusMessageContains('This message is not real');
     }
-    catch (AssertionFailedError $e) {
+    catch (ExpectationException $e) {
       $expected_failure_occurred = TRUE;
     }
     $this->assertTrue($expected_failure_occurred, 'WebAssert::statusMessageContains() did not fail when it should have failed.');
@@ -121,7 +121,7 @@ public function testStatusMessageAssertions(): void {
     try {
       $this->assertSession()->statusMessageNotContains('markup');
     }
-    catch (AssertionFailedError $e) {
+    catch (ExpectationException $e) {
       $expected_failure_occurred = TRUE;
     }
     $this->assertTrue($expected_failure_occurred, 'WebAssert::statusMessageNotContains() did not fail when it should have failed.');
@@ -130,7 +130,7 @@ public function testStatusMessageAssertions(): void {
     try {
       $this->assertSession()->statusMessageExists('error');
     }
-    catch (AssertionFailedError $e) {
+    catch (ExpectationException $e) {
       $expected_failure_occurred = TRUE;
     }
     $this->assertTrue($expected_failure_occurred, 'WebAssert::statusMessageExists() did not fail when it should have failed.');
@@ -139,7 +139,7 @@ public function testStatusMessageAssertions(): void {
     try {
       $this->assertSession()->statusMessageNotExists();
     }
-    catch (AssertionFailedError $e) {
+    catch (ExpectationException $e) {
       $expected_failure_occurred = TRUE;
     }
     $this->assertTrue($expected_failure_occurred, 'WebAssert::statusMessageNotExists() did not fail when it should have failed.');
diff --git a/core/tests/Drupal/FunctionalTests/WebAssertTest.php b/core/tests/Drupal/FunctionalTests/WebAssertTest.php
index 6b13dc3184..2e1b837f74 100644
--- a/core/tests/Drupal/FunctionalTests/WebAssertTest.php
+++ b/core/tests/Drupal/FunctionalTests/WebAssertTest.php
@@ -5,7 +5,6 @@
 use Behat\Mink\Exception\ExpectationException;
 use Drupal\Core\Url;
 use Drupal\Tests\BrowserTestBase;
-use Behat\Mink\Exception\ResponseTextException;
 use PHPUnit\Framework\AssertionFailedError;
 
 /**
@@ -81,7 +80,7 @@ public function testPageTextContainsOnce() {
     $this->drupalGet('test-page');
     $this->assertSession()->pageTextContainsOnce('Test page text.');
 
-    $this->expectException(ResponseTextException::class);
+    $this->expectException(AssertionFailedError::class);
     $this->expectExceptionMessage("Failed asserting that the page matches the pattern '/does\\-not\\-exist/ui' 1 time(s), 0 found.");
     $this->assertSession()->pageTextContainsOnce('does-not-exist');
   }
diff --git a/core/tests/Drupal/Tests/BrowserTestBase.php b/core/tests/Drupal/Tests/BrowserTestBase.php
index cd75578c4e..705f2b1134 100644
--- a/core/tests/Drupal/Tests/BrowserTestBase.php
+++ b/core/tests/Drupal/Tests/BrowserTestBase.php
@@ -4,6 +4,7 @@
 
 use Behat\Mink\Driver\BrowserKitDriver;
 use Behat\Mink\Element\Element;
+use Behat\Mink\Exception\Exception as MinkException;
 use Behat\Mink\Mink;
 use Behat\Mink\Selector\SelectorsHandler;
 use Behat\Mink\Session;
@@ -16,6 +17,7 @@
 use Drupal\Tests\block\Traits\BlockCreationTrait;
 use Drupal\Tests\node\Traits\ContentTypeCreationTrait;
 use Drupal\Tests\node\Traits\NodeCreationTrait;
+use Drupal\Tests\Traits\MinkExceptionTrait;
 use Drupal\Tests\Traits\PhpUnitWarnings;
 use Drupal\Tests\user\Traits\UserCreationTrait;
 use Drupal\TestTools\Comparator\MarkupInterfaceComparator;
@@ -68,6 +70,7 @@ abstract class BrowserTestBase extends TestCase {
     createUser as drupalCreateUser;
   }
   use XdebugRequestTrait;
+  use MinkExceptionTrait;
   use PhpUnitWarnings;
   use PhpUnitCompatibilityTrait;
   use ExpectDeprecationTrait;
@@ -450,6 +453,16 @@ protected function tearDown(): void {
     }
   }
 
+  /**
+   * {@inheritdoc}
+   */
+  protected function onNotSuccessfulTest(\Throwable $t): void {
+    if ($t instanceof MinkException) {
+      $this->convertMinkExceptionToPhpUnitFailure($t);
+    }
+    parent::onNotSuccessfulTest($t);
+  }
+
   /**
    * Returns Mink session.
    *
diff --git a/core/tests/Drupal/Tests/Traits/MinkExceptionTrait.php b/core/tests/Drupal/Tests/Traits/MinkExceptionTrait.php
new file mode 100644
index 0000000000..f2551a9f56
--- /dev/null
+++ b/core/tests/Drupal/Tests/Traits/MinkExceptionTrait.php
@@ -0,0 +1,28 @@
+<?php
+
+namespace Drupal\Tests\Traits;
+
+use Behat\Mink\Exception\Exception as MinkException;
+use PHPUnit\Framework\AssertionFailedError;
+use PHPUnit\Util\Filter;
+
+/**
+ * Post-processes Drupal tests' failures.
+ *
+ * @internal
+ */
+trait MinkExceptionTrait {
+
+  /**
+   * Re-throws Mink exceptions as PHPUnit assertion failures.
+   *
+   * @param \Behat\Mink\Exception\Exception $minkException
+   *   The Mink exception that caused the assertion failure.
+   */
+  protected function convertMinkExceptionToPhpUnitFailure(MinkException $minkException): void {
+    $message = $minkException->getMessage() . PHP_EOL . 'Caused by ' . get_class($minkException) . PHP_EOL;
+    $message .= Filter::getFilteredStacktrace($minkException) . PHP_EOL;
+    throw new AssertionFailedError($message, $minkException->getCode());
+  }
+
+}
diff --git a/core/tests/Drupal/Tests/WebAssert.php b/core/tests/Drupal/Tests/WebAssert.php
index 404796e973..9c20fcbda7 100644
--- a/core/tests/Drupal/Tests/WebAssert.php
+++ b/core/tests/Drupal/Tests/WebAssert.php
@@ -3,7 +3,6 @@
 namespace Drupal\Tests;
 
 use Behat\Mink\Exception\ExpectationException;
-use Behat\Mink\Exception\ResponseTextException;
 use Behat\Mink\WebAssert as MinkWebAssert;
 use Behat\Mink\Element\ElementInterface;
 use Behat\Mink\Element\TraversableElement;
@@ -12,7 +11,6 @@
 use Drupal\Component\Utility\Html;
 use Drupal\Core\Url;
 use PHPUnit\Framework\Assert;
-use PHPUnit\Framework\AssertionFailedError;
 use PHPUnit\Framework\Constraint\ArrayHasKey;
 use PHPUnit\Framework\Constraint\IsIdentical;
 use PHPUnit\Framework\Constraint\IsEqual;
@@ -694,12 +692,7 @@ public function hiddenFieldValueNotEquals($field, $value, TraversableElement $co
   public function pageTextContainsOnce($text) {
     assert(func_num_args() === 1);
     $regex = '/' . preg_quote($text, '/') . '/ui';
-    try {
-      $this->pageTextMatchesCount(1, $regex);
-    }
-    catch (AssertionFailedError $e) {
-      throw new ResponseTextException($e->getMessage(), $this->session->getDriver());
-    }
+    $this->pageTextMatchesCount(1, $regex);
   }
 
   /**
@@ -1028,13 +1021,7 @@ public function checkboxNotChecked($field, TraversableElement $container = NULL)
    *   The optional message type: status, error, or warning.
    */
   public function statusMessageExists(string $type = NULL): void {
-    $selector = $this->buildStatusMessageSelector(NULL, $type);
-    try {
-      $this->elementExists('xpath', $selector);
-    }
-    catch (ExpectationException $e) {
-      Assert::fail($e->getMessage());
-    }
+    $this->elementExists('xpath', $this->buildStatusMessageSelector(NULL, $type));
   }
 
   /**
@@ -1044,13 +1031,7 @@ public function statusMessageExists(string $type = NULL): void {
    *   The optional message type: status, error, or warning.
    */
   public function statusMessageNotExists(string $type = NULL): void {
-    $selector = $this->buildStatusMessageSelector(NULL, $type);
-    try {
-      $this->elementNotExists('xpath', $selector);
-    }
-    catch (ExpectationException $e) {
-      Assert::fail($e->getMessage());
-    }
+    $this->elementNotExists('xpath', $this->buildStatusMessageSelector(NULL, $type));
   }
 
   /**
@@ -1062,13 +1043,7 @@ public function statusMessageNotExists(string $type = NULL): void {
    *   The optional message type: status, error, or warning.
    */
   public function statusMessageContains(string $message, string $type = NULL): void {
-    $selector = $this->buildStatusMessageSelector($message, $type);
-    try {
-      $this->elementExists('xpath', $selector);
-    }
-    catch (ExpectationException $e) {
-      Assert::fail($e->getMessage());
-    }
+    $this->elementExists('xpath', $this->buildStatusMessageSelector($message, $type));
   }
 
   /**
@@ -1080,13 +1055,7 @@ public function statusMessageContains(string $message, string $type = NULL): voi
    *   The optional message type: status, error, or warning.
    */
   public function statusMessageNotContains(string $message, string $type = NULL): void {
-    $selector = $this->buildStatusMessageSelector($message, $type);
-    try {
-      $this->elementNotExists('xpath', $selector);
-    }
-    catch (ExpectationException $e) {
-      Assert::fail($e->getMessage());
-    }
+    $this->elementNotExists('xpath', $this->buildStatusMessageSelector($message, $type));
   }
 
   /**
