diff --git a/core/lib/Drupal/Core/Cache/DatabaseBackend.php b/core/lib/Drupal/Core/Cache/DatabaseBackend.php
index d5c9150dd38..42864c8b4c4 100644
--- a/core/lib/Drupal/Core/Cache/DatabaseBackend.php
+++ b/core/lib/Drupal/Core/Cache/DatabaseBackend.php
@@ -296,7 +296,7 @@ protected function doSetMultiple(array $items) {
       // merges.
       $query = $this->connection
         ->upsert($this->bin)
-        ->key('cid')
+        ->keyColumns(['cid'])
         ->fields(['cid', 'expire', 'created', 'tags', 'checksum', 'data', 'serialized']);
       foreach ($values as $fields) {
         // Only pass the values since the order of $fields matches the order of
diff --git a/core/lib/Drupal/Core/Config/DatabaseStorage.php b/core/lib/Drupal/Core/Config/DatabaseStorage.php
index 8b1c0a6e762..560e282a11a 100644
--- a/core/lib/Drupal/Core/Config/DatabaseStorage.php
+++ b/core/lib/Drupal/Core/Config/DatabaseStorage.php
@@ -164,9 +164,13 @@ public function write($name, array $data) {
    *   TRUE when the write was successful, FALSE otherwise.
    */
   protected function doWrite($name, $data) {
-    return (bool) $this->connection->merge($this->table, $this->options)
-      ->keys(['collection', 'name'], [$this->collection, $name])
-      ->fields(['data' => $data])
+    return (bool) $this->connection->upsert($this->table, $this->options)
+      ->keyColumns(['collection', 'name'])
+      ->fields([
+        'collection' => $this->collection,
+        'name' => $name,
+        'data' => $data,
+      ])
       ->execute();
   }
 
diff --git a/core/lib/Drupal/Core/Database/Query/Upsert.php b/core/lib/Drupal/Core/Database/Query/Upsert.php
index 2b845fcccc4..b8a19a103dc 100644
--- a/core/lib/Drupal/Core/Database/Query/Upsert.php
+++ b/core/lib/Drupal/Core/Database/Query/Upsert.php
@@ -5,12 +5,12 @@
 use Drupal\Core\Database\Connection;
 
 /**
- * General class for an abstracted "Upsert" (UPDATE or INSERT) query operation.
+ * General class for an abstracted "UPSERT" (UPDATE or INSERT) query operation.
  *
- * This class can only be used with a table with a single unique index.
- * Often, this will be the primary key. On such a table this class works like
- * Insert except the rows will be set to the desired values even if the key
- * existed before.
+ * This query can only be used on a table where the field(s) passed to the
+ * ::keys() method match a unique or primary key constraint of that table.
+ * On such a table, this query works like Insert except the rows will be set to
+ * the desired values even if the key(s) existed before.
  */
 abstract class Upsert extends Query implements \Countable {
 
@@ -20,9 +20,21 @@ abstract class Upsert extends Query implements \Countable {
    * The unique or primary key of the table.
    *
    * @var string
+   *
+   * @deprecated in drupal:11.4.0 and is removed from drupal:12.0.0. Use
+   *   $keyColumns instead.
+   *
+   * @see https://www.drupal.org/node/3565717
    */
   protected $key;
 
+  /**
+   * The unique or primary key of the table.
+   *
+   * @var list<string>
+   */
+  protected array $keyColumns = [];
+
   /**
    * Constructs an Upsert object.
    *
@@ -45,10 +57,48 @@ public function __construct(Connection $connection, $table, array $options = [])
    *   The name of the field to set.
    *
    * @return $this
+   *
+   * @deprecated in drupal:11.4.0 and is removed from drupal:12.0.0. Use
+   *  keyColumns() instead.
+   *
+   * @see https://www.drupal.org/node/3565717
    */
   public function key($field) {
+    @trigger_error(__METHOD__ . '() is deprecated in drupal:11.4.0 and is removed from drupal:12.0.0. Use ::keyColumns() instead. See https://www.drupal.org/node/3565717', E_USER_DEPRECATED);
     $this->key = $field;
+    $this->keyColumns = [$field];
+    $this->fields([$field]);
+
+    return $this;
+  }
 
+  /**
+   * Sets the unique/primary key columns to be used as condition for this query.
+   *
+   * @param list<string> $columns
+   *   The name of the fields to set.
+   *
+   * @return $this
+   */
+  public function keyColumns(array $columns): static {
+    $this->keyColumns = $columns;
+    $this->fields(array_values($columns));
+
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function fields(array $fields, array $values = []): static {
+    if (empty($values) && !is_numeric(key($fields))) {
+      $values = array_values($fields);
+      $fields = array_keys($fields);
+    }
+    $this->insertFields = array_unique(array_merge($this->insertFields, $fields));
+    if (!empty($values)) {
+      $this->insertValues[] = $values;
+    }
     return $this;
   }
 
@@ -64,8 +114,8 @@ public function key($field) {
    */
   protected function preExecute() {
     // Confirm that the user set the unique/primary key of the table.
-    if (!$this->key) {
-      throw new NoUniqueFieldException('There is no unique field specified.');
+    if (!$this->keyColumns) {
+      throw new NoUniqueFieldException('No columns provided for key of the UPSERT operation.');
     }
 
     // Confirm that the user did not try to specify an identical
diff --git a/core/lib/Drupal/Core/KeyValueStore/DatabaseStorage.php b/core/lib/Drupal/Core/KeyValueStore/DatabaseStorage.php
index ea5247bc750..e7ba05d6be5 100644
--- a/core/lib/Drupal/Core/KeyValueStore/DatabaseStorage.php
+++ b/core/lib/Drupal/Core/KeyValueStore/DatabaseStorage.php
@@ -132,12 +132,13 @@ public function getAll() {
    *   The data to store.
    */
   protected function doSet($key, $value) {
-    $this->connection->merge($this->table)
-      ->keys([
-        'name' => $key,
+    $this->connection->upsert($this->table)
+      ->keyColumns(['collection', 'name'])
+      ->fields([
         'collection' => $this->collection,
+        'name' => $key,
+        'value' => $this->serializer->encode($value),
       ])
-      ->fields(['value' => $this->serializer->encode($value)])
       ->execute();
   }
 
diff --git a/core/lib/Drupal/Core/KeyValueStore/DatabaseStorageExpirable.php b/core/lib/Drupal/Core/KeyValueStore/DatabaseStorageExpirable.php
index 8813d9e2abe..bc5e3abf62f 100644
--- a/core/lib/Drupal/Core/KeyValueStore/DatabaseStorageExpirable.php
+++ b/core/lib/Drupal/Core/KeyValueStore/DatabaseStorageExpirable.php
@@ -111,12 +111,11 @@ public function getAll() {
    *   The time to live for items, in seconds.
    */
   protected function doSetWithExpire($key, $value, $expire) {
-    $this->connection->merge($this->table)
-      ->keys([
-        'name' => $key,
-        'collection' => $this->collection,
-      ])
+    $this->connection->upsert($this->table)
+      ->keyColumns(['collection', 'name'])
       ->fields([
+        'collection' => $this->collection,
+        'name' => $key,
         'value' => $this->serializer->encode($value),
         'expire' => $this->time->getRequestTime() + $expire,
       ])
diff --git a/core/modules/mysql/src/Driver/Database/mysql/Upsert.php b/core/modules/mysql/src/Driver/Database/mysql/Upsert.php
index 0e5f7d3b50e..9e485e52f8c 100644
--- a/core/modules/mysql/src/Driver/Database/mysql/Upsert.php
+++ b/core/modules/mysql/src/Driver/Database/mysql/Upsert.php
@@ -18,6 +18,7 @@ public function __toString() {
 
     // Default fields are always placed first for consistency.
     $insert_fields = array_merge($this->defaultFields, $this->insertFields);
+    $insert_fields = array_combine($insert_fields, $insert_fields);
     $insert_fields = array_map(function ($field) {
       return $this->connection->escapeField($field);
     }, $insert_fields);
@@ -27,8 +28,10 @@ public function __toString() {
     $values = $this->getInsertPlaceholderFragment($this->insertValues, $this->defaultFields);
     $query .= implode(', ', $values);
 
-    // Updating the unique / primary key is not necessary.
-    unset($insert_fields[$this->key]);
+    // Updating the unique / primary key fields is not necessary.
+    foreach ($this->keyColumns as $key) {
+      unset($insert_fields[$key]);
+    }
 
     $update = [];
     foreach ($insert_fields as $field) {
diff --git a/core/modules/pgsql/src/Driver/Database/pgsql/Upsert.php b/core/modules/pgsql/src/Driver/Database/pgsql/Upsert.php
index 35823a270b1..3e2eeeba743 100644
--- a/core/modules/pgsql/src/Driver/Database/pgsql/Upsert.php
+++ b/core/modules/pgsql/src/Driver/Database/pgsql/Upsert.php
@@ -27,6 +27,7 @@ public function execute() {
     $max_placeholder = 0;
     $blobs = [];
     $blob_count = 0;
+dump($this->table, $this->insertValues, $this->insertFields);
     foreach ($this->insertValues as $insert_values) {
       foreach ($this->insertFields as $idx => $field) {
         if (isset($table_information->blob_fields[$field]) && $insert_values[$idx] !== NULL) {
@@ -97,8 +98,13 @@ public function __toString() {
     // Create a sanitized comment string to prepend to the query.
     $comments = $this->connection->makeComment($this->comments);
 
+    $keys = array_map(function ($key) {
+      return $this->connection->escapeField($key);
+    }, $this->keyColumns);
+
     // Default fields are always placed first for consistency.
     $insert_fields = array_merge($this->defaultFields, $this->insertFields);
+    $insert_fields = array_combine($insert_fields, $insert_fields);
     $insert_fields = array_map(function ($field) {
       return $this->connection->escapeField($field);
     }, $insert_fields);
@@ -108,8 +114,10 @@ public function __toString() {
     $values = $this->getInsertPlaceholderFragment($this->insertValues, $this->defaultFields);
     $query .= implode(', ', $values);
 
-    // Updating the unique / primary key is not necessary.
-    unset($insert_fields[$this->key]);
+    // Updating the unique / primary key fields is not necessary.
+    foreach ($this->keyColumns as $key) {
+      unset($insert_fields[$key]);
+    }
 
     $update = [];
     foreach ($insert_fields as $field) {
@@ -118,7 +126,7 @@ public function __toString() {
       $update[] = "$field = EXCLUDED.$field";
     }
 
-    $query .= ' ON CONFLICT (' . $this->connection->escapeField($this->key) . ') DO UPDATE SET ' . implode(', ', $update);
+    $query .= ' ON CONFLICT (' . implode(', ', $keys) . ') DO UPDATE SET ' . implode(', ', $update);
 
     return $query;
   }
diff --git a/core/modules/pgsql/tests/src/Kernel/pgsql/NonPublicSchemaTest.php b/core/modules/pgsql/tests/src/Kernel/pgsql/NonPublicSchemaTest.php
index 4fb514c0a8b..9b82be97dc3 100644
--- a/core/modules/pgsql/tests/src/Kernel/pgsql/NonPublicSchemaTest.php
+++ b/core/modules/pgsql/tests/src/Kernel/pgsql/NonPublicSchemaTest.php
@@ -216,7 +216,7 @@ public function testUpsert(): void {
     $num_records_before = $this->testingFakeConnection->query('SELECT COUNT(*) FROM {faking_table}')->fetchField();
 
     $upsert = $this->testingFakeConnection->upsert('faking_table')
-      ->key('id')
+      ->keyColumns(['id'])
       ->fields(['id', 'test_field']);
 
     // Upserting a new row.
diff --git a/core/modules/sqlite/src/Driver/Database/sqlite/Upsert.php b/core/modules/sqlite/src/Driver/Database/sqlite/Upsert.php
index 59974272531..a8a4194d3e7 100644
--- a/core/modules/sqlite/src/Driver/Database/sqlite/Upsert.php
+++ b/core/modules/sqlite/src/Driver/Database/sqlite/Upsert.php
@@ -18,8 +18,13 @@ public function __toString() {
     // Create a sanitized comment string to prepend to the query.
     $comments = $this->connection->makeComment($this->comments);
 
+    $keys = array_map(function ($key) {
+      return $this->connection->escapeField($key);
+    }, $this->keyColumns);
+
     // Default fields are always placed first for consistency.
     $insert_fields = array_merge($this->defaultFields, $this->insertFields);
+    $insert_fields = array_combine($insert_fields, $insert_fields);
     $insert_fields = array_map(function ($field) {
       return $this->connection->escapeField($field);
     }, $insert_fields);
@@ -29,8 +34,10 @@ public function __toString() {
     $values = $this->getInsertPlaceholderFragment($this->insertValues, $this->defaultFields);
     $query .= implode(', ', $values);
 
-    // Updating the unique / primary key is not necessary.
-    unset($insert_fields[$this->key]);
+    // Updating the unique / primary key fields is not necessary.
+    foreach ($this->keyColumns as $key) {
+      unset($insert_fields[$key]);
+    }
 
     $update = [];
     foreach ($insert_fields as $field) {
@@ -39,7 +46,7 @@ public function __toString() {
       $update[] = "$field = EXCLUDED.$field";
     }
 
-    $query .= ' ON CONFLICT (' . $this->connection->escapeField($this->key) . ') DO UPDATE SET ' . implode(', ', $update);
+    $query .= ' ON CONFLICT (' . implode(', ', $keys) . ') DO UPDATE SET ' . implode(', ', $update);
 
     return $query;
   }
diff --git a/core/modules/system/tests/src/Functional/UpdateSystem/BrokenCacheUpdateTest.php b/core/modules/system/tests/src/Functional/UpdateSystem/BrokenCacheUpdateTest.php
index 08fc8f4e3ef..4f26f82db33 100644
--- a/core/modules/system/tests/src/Functional/UpdateSystem/BrokenCacheUpdateTest.php
+++ b/core/modules/system/tests/src/Functional/UpdateSystem/BrokenCacheUpdateTest.php
@@ -39,7 +39,7 @@ public function testUpdate(): void {
 
     // Create broken element info caches entries.
     $insert = $connection->upsert('cache_discovery');
-    $insert->key('cid');
+    $insert->keyColumns(['cid']);
     $fields = [
       'cid' => 'element_info',
       'data' => 'BROKEN',
diff --git a/core/tests/Drupal/KernelTests/Core/Database/DriverSpecificTransactionTestBase.php b/core/tests/Drupal/KernelTests/Core/Database/DriverSpecificTransactionTestBase.php
index 35a1549fe1a..79aaa19f82c 100644
--- a/core/tests/Drupal/KernelTests/Core/Database/DriverSpecificTransactionTestBase.php
+++ b/core/tests/Drupal/KernelTests/Core/Database/DriverSpecificTransactionTestBase.php
@@ -663,7 +663,7 @@ public function testQueryFailureInTransaction(): void {
     // Test a failed upsert query.
     try {
       $this->connection->upsert('test')
-        ->key('job')
+        ->keyColumns(['job'])
         ->fields(['job', 'age', 'name'])
         ->values([
           'job' => 'Presenter',
diff --git a/core/tests/Drupal/KernelTests/Core/Database/TransactionTest.php b/core/tests/Drupal/KernelTests/Core/Database/TransactionTest.php
index 1d8b78a2d52..2db265252c1 100644
--- a/core/tests/Drupal/KernelTests/Core/Database/TransactionTest.php
+++ b/core/tests/Drupal/KernelTests/Core/Database/TransactionTest.php
@@ -886,7 +886,7 @@ public function testQueryFailureInTransaction(): void {
     // Test a failed upsert query.
     try {
       $this->connection->upsert('test')
-        ->key('job')
+        ->keyColumns(['job'])
         ->fields(['job', 'age', 'name'])
         ->values([
           'job' => 'Presenter',
diff --git a/core/tests/Drupal/KernelTests/Core/Database/UpsertTest.php b/core/tests/Drupal/KernelTests/Core/Database/UpsertTest.php
index 14b7402ed7d..52eee6144a0 100644
--- a/core/tests/Drupal/KernelTests/Core/Database/UpsertTest.php
+++ b/core/tests/Drupal/KernelTests/Core/Database/UpsertTest.php
@@ -24,7 +24,7 @@ public function testUpsert(): void {
     $num_records_before = $connection->query('SELECT COUNT(*) FROM {test_people}')->fetchField();
 
     $upsert = $connection->upsert('test_people')
-      ->key('job')
+      ->keyColumns(['job'])
       ->fields(['job', 'age', 'name']);
 
     // Add a new row.
@@ -60,6 +60,88 @@ public function testUpsert(): void {
     $this->assertEquals('Meredith', $person->name, 'Name was not changed.');
   }
 
+  /**
+   * Confirms that we can upsert records with composite keys successfully.
+   */
+  public function testMultipleKeysUpsert(): void {
+    $connection = Database::getConnection();
+    $this->installSchema('database_test', ['test_composite_primary']);
+
+    // Add some initial test data.
+    $connection->insert('test_composite_primary')
+      ->fields(['name', 'age', 'job'])
+      ->values([
+        'name' => 'Tiffany',
+        'age' => 31,
+        'job' => 'Presenter',
+      ])
+      ->values([
+        'name' => 'Meredith',
+        'age' => 30,
+        'job' => 'Speaker',
+      ])
+      ->execute();
+
+    $num_records_before = $connection->query('SELECT COUNT(*) FROM {test_composite_primary}')->fetchField();
+
+    $upsert = $connection->upsert('test_composite_primary')
+      ->keyColumns(['name', 'age'])
+      // Add a new row directly from ::fields().
+      ->fields(['job'], [
+        'name' => 'Kate',
+        'age' => 25,
+        'job' => 'Volunteer',
+      ]);
+
+    // Add a new row.
+    $upsert->values([
+      'name' => 'Karen',
+      'age' => 35,
+      'job' => 'Manager',
+    ]);
+
+    // Update an existing row.
+    $upsert->values([
+      'name' => 'Meredith',
+      'age' => 30,
+      // The initial job was 'Speaker'.
+      'job' => 'Organizer',
+    ]);
+
+    // Add a new row by reusing a name but with a different age. This won't
+    // match the composite primary key constraint.
+    $upsert->values([
+      'name' => 'Meredith',
+      'age' => 40,
+      'job' => 'Supervisor',
+    ]);
+
+    $upsert->execute();
+
+    $num_records_after = $connection->query('SELECT COUNT(*) FROM {test_composite_primary}')->fetchField();
+    $this->assertEquals($num_records_before + 3, $num_records_after, 'Rows were inserted and updated properly.');
+
+    $person = $connection->query('SELECT * FROM {test_composite_primary} WHERE [job] = :job', [':job' => 'Volunteer'])->fetch();
+    $this->assertEquals('Volunteer', $person->job, 'Job set correctly.');
+    $this->assertEquals(25, $person->age, 'Age set correctly.');
+    $this->assertEquals('Kate', $person->name, 'Name set correctly.');
+
+    $person = $connection->query('SELECT * FROM {test_composite_primary} WHERE [job] = :job', [':job' => 'Manager'])->fetch();
+    $this->assertEquals('Manager', $person->job, 'Job set correctly.');
+    $this->assertEquals(35, $person->age, 'Age set correctly.');
+    $this->assertEquals('Karen', $person->name, 'Name set correctly.');
+
+    $person = $connection->query('SELECT * FROM {test_composite_primary} WHERE [job] = :job', [':job' => 'Organizer'])->fetch();
+    $this->assertEquals('Organizer', $person->job, 'Job set correctly.');
+    $this->assertEquals(30, $person->age, 'Age set correctly.');
+    $this->assertEquals('Meredith', $person->name, 'Name set correctly.');
+
+    $person = $connection->query('SELECT * FROM {test_composite_primary} WHERE [job] = :job', [':job' => 'Supervisor'])->fetch();
+    $this->assertEquals('Supervisor', $person->job, 'Job set correctly.');
+    $this->assertEquals(40, $person->age, 'Age set correctly.');
+    $this->assertEquals('Meredith', $person->name, 'Name set correctly.');
+  }
+
   /**
    * Confirms that we can upsert records with keywords successfully.
    */
@@ -67,7 +149,7 @@ public function testUpsertWithKeywords(): void {
     $num_records_before = $this->connection->query('SELECT COUNT(*) FROM {select}')->fetchField();
 
     $upsert = $this->connection->upsert('select')
-      ->key('id')
+      ->keyColumns(['id'])
       ->fields(['id', 'update']);
 
     // Add a new row.
@@ -117,7 +199,7 @@ public function testUpsertWithKeywords(): void {
   public function testUpsertNonExistingTable(): void {
     $this->expectException(DatabaseExceptionWrapper::class);
     $upsert = $this->connection->upsert('a-table-that-does-not-exist')
-      ->key('id')
+      ->keyColumns(['id'])
       ->fields(['id', 'update']);
     $upsert->values([
       'id' => 1,
@@ -137,7 +219,7 @@ public function testUpsertNullBlob(): void {
     $this->assertSame('test', $r['blob1']);
 
     $this->connection->upsert('test_one_blob')
-      ->key('id')
+      ->keyColumns(['id'])
       ->fields(['id', 'blob1'])
       ->values(['id' => $id, 'blob1' => NULL])
       ->values(['id' => $id + 1, 'blob1' => NULL])
