diff --git a/core/tests/Drupal/Tests/Listeners/DrupalListener.php b/core/tests/Drupal/Tests/Listeners/DrupalListener.php
index 663c2e52..54655a53 100644
--- a/core/tests/Drupal/Tests/Listeners/DrupalListener.php
+++ b/core/tests/Drupal/Tests/Listeners/DrupalListener.php
@@ -22,6 +22,8 @@ class DrupalListener implements TestListener {
    * {@inheritdoc}
    */
   public function startTest(Test $test): void {
+    $class = new \ReflectionClass($test);
+dump('return type', $class->getMethod('setUp')->getReturnType());
     $this->deprecationStartTest($test);
   }
 
