diff --git a/core/tests/Drupal/KernelTests/Core/Block/MultipleBlockFormTest.php b/core/tests/Drupal/KernelTests/Core/Block/MultipleBlockFormTest.php
index feecfd48ab..86ea0ab784 100644
--- a/core/tests/Drupal/KernelTests/Core/Block/MultipleBlockFormTest.php
+++ b/core/tests/Drupal/KernelTests/Core/Block/MultipleBlockFormTest.php
@@ -23,7 +23,7 @@ class MultipleBlockFormTest extends KernelTestBase {
   public function testMultipleForms() {
     $configuration = ['label' => 'A very cool block'];
     $block = \Drupal::service('plugin.manager.block')->createInstance('test_multiple_forms_block', $configuration);
-
+dump($block);
     $form_object1 = \Drupal::service('plugin_form.factory')->createInstance($block, 'configure');
     $form_object2 = \Drupal::service('plugin_form.factory')->createInstance($block, 'secondary');
 
