diff --git a/core/modules/views/src/Plugin/views/query/Sql.php b/core/modules/views/src/Plugin/views/query/Sql.php
index 14b335b63f..0f03b38254 100644
--- a/core/modules/views/src/Plugin/views/query/Sql.php
+++ b/core/modules/views/src/Plugin/views/query/Sql.php
@@ -1530,13 +1530,17 @@ public function execute(ViewExecutable $view) {
         }
 
         $result = $query->execute();
-        $result->setFetchMode(\PDO::FETCH_CLASS, 'Drupal\views\ResultRow');
+        $result->setFetchMode(\PDO::FETCH_OBJ);
 
         // Setup the result row objects.
-        $view->result = iterator_to_array($result);
-        array_walk($view->result, function (ResultRow $row, $index) {
-          $row->index = $index;
-        });
+        $i = 0;
+        $view->result = [];
+        foreach ($result as $rowObj) {
+          $view->result[] = new ResultRow([
+            'data' => $rowObj,
+            'index' => $i++,
+          ]);
+        }
 
         $view->pager->postExecute($view->result);
         $view->pager->updatePageInfo();
diff --git a/core/modules/views/src/ResultRow.php b/core/modules/views/src/ResultRow.php
index 2399f69983..d0255d5d36 100644
--- a/core/modules/views/src/ResultRow.php
+++ b/core/modules/views/src/ResultRow.php
@@ -5,9 +5,13 @@
 /**
  * A class representing a view result row.
  */
-#[\AllowDynamicProperties]
 class ResultRow {
 
+  /**
+   * Raw row data.
+   */
+  protected \stdClass $data;
+
   /**
    * The entity for this result.
    *
@@ -41,6 +45,7 @@ public function __construct(array $values = []) {
     foreach ($values as $key => $value) {
       $this->{$key} = $value;
     }
+dump([__METHOD__, $this]);
   }
 
   /**
