diff --git a/core/modules/contact/tests/src/Functional/ContactSitewideTest.php b/core/modules/contact/tests/src/Functional/ContactSitewideTest.php
index abbe71420c..e273d8c057 100644
--- a/core/modules/contact/tests/src/Functional/ContactSitewideTest.php
+++ b/core/modules/contact/tests/src/Functional/ContactSitewideTest.php
@@ -212,6 +212,15 @@ public function testSiteWideContact() {
     user_role_grant_permissions(RoleInterface::ANONYMOUS_ID, ['access site-wide contact form']);
     $this->drupalLogout();
     $this->drupalGet('contact');
+
+    $dom = new \DOMDocument();
+    // Suppress libxml warnings when loading HTML 5 content.
+    @$dom->loadHTML($this->getSession()->getPage()->getHtml());
+    $body = $dom->getElementsByTagName('body')[0];
+    foreach ($body->getElementsByTagName('script') as $item) {
+      $item->parentNode->removeChild($item);
+    }
+dump($body, $dom->saveHTML($body));
     $this->assertSession()->pageTextContains('Your email address');
     $this->assertNoText('Form');
     $this->drupalLogin($admin_user);
