diff --git a/composer.json b/composer.json
index 61c5d3735a..48a914793b 100644
--- a/composer.json
+++ b/composer.json
@@ -105,6 +105,11 @@
             "Drupal\\Composer\\": "composer"
         }
     },
+    "autoload-dev": {
+        "psr-4": {
+            "Drupal\\PHPStan\\": "core/tests/PHPStan"
+        }
+    },
     "scripts": {
         "pre-install-cmd": "Drupal\\Composer\\Composer::ensureComposerVersion",
         "pre-update-cmd": "Drupal\\Composer\\Composer::ensureComposerVersion",
diff --git a/core/misc/cspell/dictionary.txt b/core/misc/cspell/dictionary.txt
index 1a46815d4a..1289ee0829 100644
--- a/core/misc/cspell/dictionary.txt
+++ b/core/misc/cspell/dictionary.txt
@@ -10,6 +10,7 @@ ajaxing
 allwords
 alphadecimal
 ampm
+analyse
 anyall
 archiver
 archivers
diff --git a/core/phpstan.neon.dist b/core/phpstan.neon.dist
index 17c3970c21..073baf8af6 100644
--- a/core/phpstan.neon.dist
+++ b/core/phpstan.neon.dist
@@ -44,3 +44,6 @@ parameters:
     - "#Drupal calls should be avoided in classes, use dependency injection instead#"
     - "#^Plugin definitions cannot be altered.#"
     - "#^Class .* extends @internal class#"
+
+rules:
+  - Drupal\PHPStan\Rules\ComponentTestDoesNotExtendCoreTest
diff --git a/core/tests/Drupal/Tests/Listeners/DrupalComponentTestListenerTrait.php b/core/tests/Drupal/Tests/Listeners/DrupalComponentTestListenerTrait.php
deleted file mode 100644
index 90a5638e2d..0000000000
--- a/core/tests/Drupal/Tests/Listeners/DrupalComponentTestListenerTrait.php
+++ /dev/null
@@ -1,37 +0,0 @@
-<?php
-
-declare(strict_types=1);
-
-namespace Drupal\Tests\Listeners;
-
-use Drupal\KernelTests\KernelTestBase;
-use Drupal\Tests\BrowserTestBase;
-use Drupal\Tests\UnitTestCase;
-use PHPUnit\Framework\AssertionFailedError;
-
-/**
- * Ensures that no component tests are extending a core test base class.
- *
- * @internal
- */
-trait DrupalComponentTestListenerTrait {
-
-  /**
-   * Reacts to the end of a test.
-   *
-   * @param \PHPUnit\Framework\Test $test
-   *   The test object that has ended its test run.
-   * @param float $time
-   *   The time the test took.
-   */
-  protected function componentEndTest($test, $time) {
-    /** @var \PHPUnit\Framework\Test $test */
-    if (str_starts_with($test->toString(), 'Drupal\Tests\Component')) {
-      if ($test instanceof BrowserTestBase || $test instanceof KernelTestBase || $test instanceof UnitTestCase) {
-        $error = new AssertionFailedError('Component tests should not extend a core test base class.');
-        $test->getTestResultObject()->addFailure($test, $error, $time);
-      }
-    }
-  }
-
-}
diff --git a/core/tests/Drupal/Tests/Listeners/DrupalListener.php b/core/tests/Drupal/Tests/Listeners/DrupalListener.php
index 182e4fe4e7..26fa1e5001 100644
--- a/core/tests/Drupal/Tests/Listeners/DrupalListener.php
+++ b/core/tests/Drupal/Tests/Listeners/DrupalListener.php
@@ -18,7 +18,6 @@
 class DrupalListener implements TestListener {
 
   use TestListenerDefaultImplementation;
-  use DrupalComponentTestListenerTrait;
 
   /**
    * The wrapped Symfony test listener.
@@ -60,7 +59,6 @@ public function startTest(Test $test): void {
    */
   public function endTest(Test $test, float $time): void {
     $this->symfonyListener->endTest($test, $time);
-    $this->componentEndTest($test, $time);
   }
 
 }
diff --git a/core/tests/Drupal/Tests/PHPStan/Rules/ComponentTestDoesNotExtendCoreTestTest.php b/core/tests/Drupal/Tests/PHPStan/Rules/ComponentTestDoesNotExtendCoreTestTest.php
new file mode 100644
index 0000000000..be59eed4bb
--- /dev/null
+++ b/core/tests/Drupal/Tests/PHPStan/Rules/ComponentTestDoesNotExtendCoreTestTest.php
@@ -0,0 +1,34 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Drupal\Tests\PHPStan\Rules;
+
+use Drupal\PHPStan\Rules\ComponentTestDoesNotExtendCoreTest;
+use PHPStan\Rules\Rule;
+use PHPStan\Testing\RuleTestCase;
+
+/**
+ * Tests ComponentTestDoesNotExtendCoreTest rule.
+ *
+ * @group phpstan
+ */
+class ComponentTestDoesNotExtendCoreTestTest extends RuleTestCase {
+
+  protected function getRule(): Rule {
+    return new ComponentTestDoesNotExtendCoreTest();
+  }
+
+  public function testRule(): void {
+    $this->analyse(
+      [__DIR__ . '../../../../../fixtures/PHPStan/component-tests.php'],
+      [
+        [
+          'Component tests should not extend a core test base class.',
+          20,
+        ],
+      ]
+    );
+  }
+
+}
diff --git a/core/tests/PHPStan/Rules/ComponentTestDoesNotExtendCoreTest.php b/core/tests/PHPStan/Rules/ComponentTestDoesNotExtendCoreTest.php
new file mode 100644
index 0000000000..d1e4ffc580
--- /dev/null
+++ b/core/tests/PHPStan/Rules/ComponentTestDoesNotExtendCoreTest.php
@@ -0,0 +1,55 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Drupal\PHPStan\Rules;
+
+use PhpParser\Node;
+use PHPStan\Node\InClassNode;
+use PhpParser\Node\Stmt\Class_;
+use PHPStan\Analyser\Scope;
+use PHPStan\Rules\Rule;
+use PHPStan\Rules\RuleErrorBuilder;
+
+/**
+ * Ensures that no component tests are extending a core test base class.
+ *
+ * @implements Rule<\PhpParser\Node\Stmt\Class_>
+ *
+ * @internal
+ */
+final class ComponentTestDoesNotExtendCoreTest implements Rule {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getNodeType(): string {
+//    return Class_::class;
+    return InClassNode::class;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function processNode(Node $node, Scope $scope): array {
+dump([get_class($node), get_class($node->getClassReflection()), $node->getClassReflection()]);
+return [];
+        if ($node->extends === NULL || $node->namespacedName === NULL) {
+      return [];
+    }
+    if (!str_starts_with($node->namespacedName->toString(), 'Drupal\Tests\Component')) {
+      return [];
+    }
+dump([$node, $scope]);
+    if (str_starts_with($node->extends->toString(), 'Drupal\Tests\Core')) {
+      return [
+        RuleErrorBuilder::message('Component tests should not extend a core test base class.')
+          ->line($node->getStartLine())
+          ->build(),
+      ];
+    }
+
+    return [];
+  }
+
+}
diff --git a/core/tests/fixtures/PHPStan/component-tests.php b/core/tests/fixtures/PHPStan/component-tests.php
new file mode 100644
index 0000000000..79d2672037
--- /dev/null
+++ b/core/tests/fixtures/PHPStan/component-tests.php
@@ -0,0 +1,32 @@
+<?php
+
+// phpcs:ignoreFile
+
+declare(strict_types=1);
+
+namespace Drupal\Tests\Core\Foo;
+
+use PHPUnit\Framework\TestCase;
+
+abstract class FooTestBase extends TestCase {
+
+}
+
+namespace Drupal\Tests\Component\Foo;
+
+use Drupal\Example\Something\TestCase;
+use Drupal\Tests\Core\Foo\FooTestBase;
+
+final class FooTest extends FooTestBase {
+
+    public function testBeacon(): void {
+    }
+
+}
+
+final class BarTest extends TestCase {
+
+    public function testBeacon(): void {
+    }
+
+}
