diff --git a/core/modules/migrate_drupal_ui/src/Form/CredentialForm.php b/core/modules/migrate_drupal_ui/src/Form/CredentialForm.php
index 87e57604bb..e8274aa32f 100644
--- a/core/modules/migrate_drupal_ui/src/Form/CredentialForm.php
+++ b/core/modules/migrate_drupal_ui/src/Form/CredentialForm.php
@@ -250,7 +250,7 @@ public function buildForm(array $form, FormStateInterface $form_state) {
       ],
       '#element_validate' => ['::validatePaths'],
     ];
-
+dump($form);
     return $form;
   }
 
