diff --git a/core/modules/mysql/src/Driver/Database/mysql/Schema.php b/core/modules/mysql/src/Driver/Database/mysql/Schema.php
index 85e71af18d1..30e41cf9995 100644
--- a/core/modules/mysql/src/Driver/Database/mysql/Schema.php
+++ b/core/modules/mysql/src/Driver/Database/mysql/Schema.php
@@ -439,6 +439,7 @@ public function addField($table, $field, $spec, $keys_new = []) {
       $this->executeDdlStatement($query);
     }
     catch (DatabaseExceptionWrapper $e) {
+dump($e, $e->getPrevious());
       // MySQL error number 4111 (ER_DROP_PK_COLUMN_TO_DROP_GIPK) indicates that
       // when dropping and adding a primary key, the generated invisible primary
       // key (GIPK) column must also be dropped.
