diff --git a/core/modules/views/src/ResultRow.php b/core/modules/views/src/ResultRow.php
index 2399f69983..4d715ceaf1 100644
--- a/core/modules/views/src/ResultRow.php
+++ b/core/modules/views/src/ResultRow.php
@@ -38,6 +38,7 @@ class ResultRow {
    *   (optional) An array of values to add as properties on the object.
    */
   public function __construct(array $values = []) {
+dump([__METHOD__, $this, $values]);
     foreach ($values as $key => $value) {
       $this->{$key} = $value;
     }
