diff --git a/core/modules/file/tests/src/Functional/Views/RelationshipUserFileDataTest.php b/core/modules/file/tests/src/Functional/Views/RelationshipUserFileDataTest.php
index 0169e0a6d8..887b6384fa 100644
--- a/core/modules/file/tests/src/Functional/Views/RelationshipUserFileDataTest.php
+++ b/core/modules/file/tests/src/Functional/Views/RelationshipUserFileDataTest.php
@@ -67,7 +67,7 @@ public function testViewsHandlerRelationshipUserFileData() {
       'status' => FILE_STATUS_PERMANENT,
     ]);
     $file->enforceIsNew();
-    file_put_contents($file->getFileUri(), file_get_contents('core/modules/simpletest/files/image-1.png'));
+    file_put_contents($file->getFileUri(), file_get_contents('core/tests/fixtures/files/image-1.png'));
     $file->save();
 
     $account = $this->drupalCreateUser();
diff --git a/core/modules/file/tests/src/Kernel/Migrate/d6/MigrateFileTest.php b/core/modules/file/tests/src/Kernel/Migrate/d6/MigrateFileTest.php
index 6986dc1685..df713443a8 100644
--- a/core/modules/file/tests/src/Kernel/Migrate/d6/MigrateFileTest.php
+++ b/core/modules/file/tests/src/Kernel/Migrate/d6/MigrateFileTest.php
@@ -107,7 +107,7 @@ public function testFiles() {
     // File 2, when migrated for the second time, is treated as a different file
     // (due to having a different uri this time) and is given fid 6.
     $file = File::load(6);
-    $this->assertIdentical('public://core/modules/simpletest/files/image-2.jpg', $file->getFileUri());
+    $this->assertIdentical('public://core/tests/fixtures/files/image-2.jpg', $file->getFileUri());
 
     $map_table = $this->getMigration('d6_file')->getIdMap()->mapTableName();
     $map = \Drupal::database()
@@ -146,7 +146,7 @@ public static function migrateDumpAlter(KernelTestBase $test) {
       ->condition('fid', 3)
       ->fields([
         'filename' => 'image-3.jpg',
-        'filepath' => 'core/modules/simpletest/files/image-3.jpg',
+        'filepath' => 'core/tests/fixtures/files/image-3.jpg',
       ])
       ->execute();
 
diff --git a/core/modules/image/tests/src/Functional/ImageFieldValidateTest.php b/core/modules/image/tests/src/Functional/ImageFieldValidateTest.php
index d7d8c3c05e..1e531b2320 100644
--- a/core/modules/image/tests/src/Functional/ImageFieldValidateTest.php
+++ b/core/modules/image/tests/src/Functional/ImageFieldValidateTest.php
@@ -40,7 +40,7 @@ public function testValid() {
     $this->drupalPostForm(NULL, [], t('Save'));
 
     // Get invalid image test files from simpletest.
-    $dir = drupal_get_path('module', 'simpletest') . '/files';
+    $dir = 'core/tests/fixtures/files';
     $files = [];
     if (is_dir($dir)) {
       $files = $file_system->scanDirectory($dir, '/invalid-img-.*/');
diff --git a/core/modules/image/tests/src/Kernel/Views/RelationshipUserImageDataTest.php b/core/modules/image/tests/src/Kernel/Views/RelationshipUserImageDataTest.php
index e0f605e92d..36e12972ce 100644
--- a/core/modules/image/tests/src/Kernel/Views/RelationshipUserImageDataTest.php
+++ b/core/modules/image/tests/src/Kernel/Views/RelationshipUserImageDataTest.php
@@ -72,7 +72,7 @@ public function testViewsHandlerRelationshipUserImageData() {
       'status' => FILE_STATUS_PERMANENT,
     ]);
     $file->enforceIsNew();
-    file_put_contents($file->getFileUri(), file_get_contents('core/modules/simpletest/files/image-1.png'));
+    file_put_contents($file->getFileUri(), file_get_contents('core/tests/fixtures/files/image-1.png'));
     $file->save();
 
     $account = User::create([
diff --git a/core/modules/migrate/tests/src/Kernel/process/FileCopyTest.php b/core/modules/migrate/tests/src/Kernel/process/FileCopyTest.php
index 148d322f2b..38d92f90b4 100644
--- a/core/modules/migrate/tests/src/Kernel/process/FileCopyTest.php
+++ b/core/modules/migrate/tests/src/Kernel/process/FileCopyTest.php
@@ -50,7 +50,7 @@ public function testSuccessfulCopies() {
     $data_sets = [
       // Test a local to local copy.
       [
-        $this->root . '/core/modules/simpletest/files/image-test.jpg',
+        $this->root . '/core/tests/fixtures/files/image-test.jpg',
         'public://file1.jpg',
       ],
       // Test a temporary file using an absolute path.
@@ -61,7 +61,7 @@ public function testSuccessfulCopies() {
       // Test a temporary file using a relative path.
       [
         $file_absolute,
-        'temporary://core/modules/simpletest/files/test.jpg',
+        'temporary://core/tests/fixtures/files/test.jpg',
       ],
     ];
     foreach ($data_sets as $data) {
@@ -113,7 +113,7 @@ public function testSuccessfulReuse($source_path, $destination_path) {
   public function providerSuccessfulReuse() {
     return [
       [
-        'local_source_path' => static::getDrupalRoot() . '/core/modules/simpletest/files/image-test.jpg',
+        'local_source_path' => static::getDrupalRoot() . '/core/tests/fixtures/files/image-test.jpg',
         'local_destination_path' => 'public://file1.jpg',
       ],
       [
@@ -146,7 +146,7 @@ public function testSuccessfulMoves() {
       // Test a temporary file using a relative path.
       [
         $file_2_absolute,
-        'temporary://core/modules/simpletest/files/test.jpg',
+        'temporary://core/tests/fixtures/files/test.jpg',
       ],
     ];
     foreach ($data_sets as $data) {
diff --git a/core/modules/migrate_drupal/tests/fixtures/drupal6.php b/core/modules/migrate_drupal/tests/fixtures/drupal6.php
index caa55b0e4c..f597bfad16 100644
--- a/core/modules/migrate_drupal/tests/fixtures/drupal6.php
+++ b/core/modules/migrate_drupal/tests/fixtures/drupal6.php
@@ -8353,7 +8353,7 @@
   'fid' => '1',
   'uid' => '1',
   'filename' => 'Image1.png',
-  'filepath' => 'core/modules/simpletest/files/image-1.png',
+  'filepath' => 'core/tests/fixtures/files/image-1.png',
   'filemime' => 'image/png',
   'filesize' => '39325',
   'status' => '1',
@@ -8363,7 +8363,7 @@
   'fid' => '2',
   'uid' => '1',
   'filename' => 'Image2.jpg',
-  'filepath' => 'core/modules/simpletest/files/image-2.jpg',
+  'filepath' => 'core/tests/fixtures/files/image-2.jpg',
   'filemime' => 'image/jpeg',
   'filesize' => '1831',
   'status' => '1',
@@ -8373,7 +8373,7 @@
   'fid' => '3',
   'uid' => '1',
   'filename' => 'Image-test.gif',
-  'filepath' => 'core/modules/simpletest/files/image-test.gif',
+  'filepath' => 'core/tests/fixtures/files/image-test.gif',
   'filemime' => 'image/jpeg',
   'filesize' => '183',
   'status' => '1',
@@ -8383,7 +8383,7 @@
   'fid' => '5',
   'uid' => '1',
   'filename' => 'html-1.txt',
-  'filepath' => 'core/modules/simpletest/files/html-1.txt',
+  'filepath' => 'core/tests/fixtures/files/html-1.txt',
   'filemime' => 'text/plain',
   'filesize' => '24',
   'status' => '1',
@@ -33171,8 +33171,8 @@
   'menu_name' => 'navigation',
   'mlid' => '411',
   'plid' => '0',
-  'link_path' => 'core/modules/simpletest/files/imagecache',
-  'router_path' => 'core/modules/simpletest/files/imagecache',
+  'link_path' => 'core/tests/fixtures/files/imagecache',
+  'router_path' => 'core/tests/fixtures/files/imagecache',
   'link_title' => '',
   'options' => 'a:0:{}',
   'module' => 'system',
@@ -42845,7 +42845,7 @@
   'file' => 'sites/all/modules/cck/includes/content.node_form.inc',
 ))
 ->values(array(
-  'path' => 'core/modules/simpletest/files/imagecache',
+  'path' => 'core/tests/fixtures/files/imagecache',
   'load_functions' => '',
   'to_arg_functions' => '',
   'access_callback' => '_imagecache_menu_access_public_files',
@@ -42855,7 +42855,7 @@
   'fit' => '31',
   'number_parts' => '5',
   'tab_parent' => '',
-  'tab_root' => 'core/modules/simpletest/files/imagecache',
+  'tab_root' => 'core/tests/fixtures/files/imagecache',
   'title' => '',
   'title_callback' => 't',
   'title_arguments' => '',
@@ -48585,7 +48585,7 @@
   'status' => '1',
   'timezone' => '3600',
   'language' => 'fr',
-  'picture' => 'core/modules/simpletest/files/image-test.jpg',
+  'picture' => 'core/tests/fixtures/files/image-test.jpg',
   'init' => 'doe@example.com',
   'data' => 'a:2:{s:7:"contact";i:1;s:13:"form_build_id";s:48:"form-qu_DMjE-Vfg01arT5J4VbuBCkOgx_LeySJx4qrPOSuA";}',
   'timezone_name' => 'Europe/Berlin',
@@ -48610,7 +48610,7 @@
   'status' => '1',
   'timezone' => '7200',
   'language' => 'ro',
-  'picture' => 'core/modules/simpletest/files/image-test.png',
+  'picture' => 'core/tests/fixtures/files/image-test.png',
   'init' => 'roe@example.com',
   'data' => 'a:2:{s:7:"contact";i:0;s:13:"form_build_id";s:48:"form-1TxjbL2_1dEHIxEu2Db6OvEsSN1x9ILH1VCgnvsO6LE";}',
   'timezone_name' => 'Europe/Helsinki',
@@ -49567,7 +49567,7 @@
 ))
 ->values(array(
   'name' => 'file_directory_path',
-  'value' => 's:29:"core/modules/simpletest/files";',
+  'value' => 's:25:"core/tests/fixtures/files";',
 ))
 ->values(array(
   'name' => 'file_directory_temp',
diff --git a/core/modules/migrate_drupal_ui/tests/src/Functional/d6/files/core/modules/simpletest/files/html-1.txt b/core/modules/migrate_drupal_ui/tests/src/Functional/d6/files/core/tests/fixtures/files/html-1.txt
similarity index 100%
rename from core/modules/migrate_drupal_ui/tests/src/Functional/d6/files/core/modules/simpletest/files/html-1.txt
rename to core/modules/migrate_drupal_ui/tests/src/Functional/d6/files/core/tests/fixtures/files/html-1.txt
diff --git a/core/modules/migrate_drupal_ui/tests/src/Functional/d6/files/core/modules/simpletest/files/image-1.png b/core/modules/migrate_drupal_ui/tests/src/Functional/d6/files/core/tests/fixtures/files/image-1.png
similarity index 100%
rename from core/modules/migrate_drupal_ui/tests/src/Functional/d6/files/core/modules/simpletest/files/image-1.png
rename to core/modules/migrate_drupal_ui/tests/src/Functional/d6/files/core/tests/fixtures/files/image-1.png
diff --git a/core/modules/migrate_drupal_ui/tests/src/Functional/d6/files/core/modules/simpletest/files/image-2.jpg b/core/modules/migrate_drupal_ui/tests/src/Functional/d6/files/core/tests/fixtures/files/image-2.jpg
similarity index 100%
rename from core/modules/migrate_drupal_ui/tests/src/Functional/d6/files/core/modules/simpletest/files/image-2.jpg
rename to core/modules/migrate_drupal_ui/tests/src/Functional/d6/files/core/tests/fixtures/files/image-2.jpg
diff --git a/core/modules/migrate_drupal_ui/tests/src/Functional/d6/files/core/modules/simpletest/files/image-test.gif b/core/modules/migrate_drupal_ui/tests/src/Functional/d6/files/core/tests/fixtures/files/image-test.gif
similarity index 100%
rename from core/modules/migrate_drupal_ui/tests/src/Functional/d6/files/core/modules/simpletest/files/image-test.gif
rename to core/modules/migrate_drupal_ui/tests/src/Functional/d6/files/core/tests/fixtures/files/image-test.gif
diff --git a/core/modules/migrate_drupal_ui/tests/src/Functional/d6/files/core/modules/simpletest/files/image-test.jpg b/core/modules/migrate_drupal_ui/tests/src/Functional/d6/files/core/tests/fixtures/files/image-test.jpg
similarity index 100%
rename from core/modules/migrate_drupal_ui/tests/src/Functional/d6/files/core/modules/simpletest/files/image-test.jpg
rename to core/modules/migrate_drupal_ui/tests/src/Functional/d6/files/core/tests/fixtures/files/image-test.jpg
diff --git a/core/modules/migrate_drupal_ui/tests/src/Functional/d6/files/core/modules/simpletest/files/image-test.png b/core/modules/migrate_drupal_ui/tests/src/Functional/d6/files/core/tests/fixtures/files/image-test.png
similarity index 100%
rename from core/modules/migrate_drupal_ui/tests/src/Functional/d6/files/core/modules/simpletest/files/image-test.png
rename to core/modules/migrate_drupal_ui/tests/src/Functional/d6/files/core/tests/fixtures/files/image-test.png
diff --git a/core/modules/user/tests/src/Kernel/Migrate/d6/MigrateUserTest.php b/core/modules/user/tests/src/Kernel/Migrate/d6/MigrateUserTest.php
index d699f8b215..5e2c663428 100644
--- a/core/modules/user/tests/src/Kernel/Migrate/d6/MigrateUserTest.php
+++ b/core/modules/user/tests/src/Kernel/Migrate/d6/MigrateUserTest.php
@@ -48,7 +48,7 @@ protected function setUp() {
       'status' => FILE_STATUS_PERMANENT,
     ]);
     $file->enforceIsNew();
-    file_put_contents($file->getFileUri(), file_get_contents('core/modules/simpletest/files/image-1.png'));
+    file_put_contents($file->getFileUri(), file_get_contents('core/tests/fixtures/files/image-1.png'));
     $file->save();
 
     $file = File::create([
@@ -62,7 +62,7 @@ protected function setUp() {
       'status' => FILE_STATUS_PERMANENT,
     ]);
     $file->enforceIsNew();
-    file_put_contents($file->getFileUri(), file_get_contents('core/modules/simpletest/files/image-2.jpg'));
+    file_put_contents($file->getFileUri(), file_get_contents('core/tests/fixtures/files/image-2.jpg'));
     $file->save();
 
     $this->executeMigration('language');
diff --git a/core/modules/user/tests/src/Kernel/Plugin/migrate/source/d6/UserPictureFileTest.php b/core/modules/user/tests/src/Kernel/Plugin/migrate/source/d6/UserPictureFileTest.php
index 2724d35f59..d26f61cee5 100644
--- a/core/modules/user/tests/src/Kernel/Plugin/migrate/source/d6/UserPictureFileTest.php
+++ b/core/modules/user/tests/src/Kernel/Plugin/migrate/source/d6/UserPictureFileTest.php
@@ -27,7 +27,7 @@ public function providerSource() {
     $tests[0]['source_data']['users'] = [
       [
         'uid' => '2',
-        'picture' => 'core/modules/simpletest/files/image-test.jpg',
+        'picture' => 'core/tests/fixtures/files/image-test.jpg',
       ],
       [
         'uid' => '15',
@@ -39,7 +39,7 @@ public function providerSource() {
     $tests[0]['expected_data'] = [
       [
         'uid' => '2',
-        'picture' => 'core/modules/simpletest/files/image-test.jpg',
+        'picture' => 'core/tests/fixtures/files/image-test.jpg',
       ],
     ];
 
diff --git a/core/tests/Drupal/KernelTests/Core/File/PharWrapperTest.php b/core/tests/Drupal/KernelTests/Core/File/PharWrapperTest.php
index 535f30ffcc..e0dbddc467 100644
--- a/core/tests/Drupal/KernelTests/Core/File/PharWrapperTest.php
+++ b/core/tests/Drupal/KernelTests/Core/File/PharWrapperTest.php
@@ -15,7 +15,7 @@ class PharWrapperTest extends KernelTestBase {
    * Tests that only valid phar files can be used.
    */
   public function testPharFile() {
-    $base = $this->getDrupalRoot() . '/core/modules/simpletest/files';
+    $base = $this->getDrupalRoot() . '/core/tests/fixtures/files';
     // Ensure that file operations via the phar:// stream wrapper work for phar
     // files with the .phar extension.
     $this->assertFalse(file_exists("phar://$base/phar-1.phar/no-such-file.php"));
diff --git a/core/tests/Drupal/KernelTests/Core/File/ScanDirectoryTest.php b/core/tests/Drupal/KernelTests/Core/File/ScanDirectoryTest.php
index 0a0dfc1e15..79f5044769 100644
--- a/core/tests/Drupal/KernelTests/Core/File/ScanDirectoryTest.php
+++ b/core/tests/Drupal/KernelTests/Core/File/ScanDirectoryTest.php
@@ -34,11 +34,11 @@ class ScanDirectoryTest extends FileTestBase {
    */
   protected function setUp() {
     parent::setUp();
-    // Hardcode the location of the simpletest files as it is already known
+    // Hardcode the location of the fixtures files as it is already known
     // and shouldn't change, and we don't yet have a way to retrieve their
     // location from drupal_get_filename() in a cached way.
     // @todo Remove as part of https://www.drupal.org/node/2186491
-    $this->path = 'core/modules/simpletest/files';
+    $this->path = 'core/tests/fixtures/files';
     $this->fileSystem = $this->container->get('file_system');
   }
 
diff --git a/core/tests/Drupal/KernelTests/Core/Image/ToolkitGdTest.php b/core/tests/Drupal/KernelTests/Core/Image/ToolkitGdTest.php
index 0496973a7d..e8948d14df 100644
--- a/core/tests/Drupal/KernelTests/Core/Image/ToolkitGdTest.php
+++ b/core/tests/Drupal/KernelTests/Core/Image/ToolkitGdTest.php
@@ -280,7 +280,7 @@ public function testManipulations() {
     foreach ($files as $file) {
       foreach ($operations as $op => $values) {
         // Load up a fresh image.
-        $image = $this->imageFactory->get(drupal_get_path('module', 'simpletest') . '/files/' . $file);
+        $image = $this->imageFactory->get('core/tests/fixtures/files/' . $file);
         $toolkit = $image->getToolkit();
         if (!$image->isValid()) {
           $this->fail(new FormattableMarkup('Could not load image %file.', ['%file' => $file]));
@@ -426,14 +426,14 @@ public function testManipulations() {
    */
   public function testResourceDestruction() {
     // Test that an Image object going out of scope releases its GD resource.
-    $image = $this->imageFactory->get(drupal_get_path('module', 'simpletest') . '/files/image-test.png');
+    $image = $this->imageFactory->get('core/tests/fixtures/files/image-test.png');
     $res = $image->getToolkit()->getResource();
     $this->assertTrue(is_resource($res), 'Successfully loaded image resource.');
     $image = NULL;
     $this->assertFalse(is_resource($res), 'Image resource was destroyed after losing scope.');
 
     // Test that 'create_new' operation does not leave orphaned GD resources.
-    $image = $this->imageFactory->get(drupal_get_path('module', 'simpletest') . '/files/image-test.png');
+    $image = $this->imageFactory->get('core/tests/fixtures/files/image-test.png');
     $old_res = $image->getToolkit()->getResource();
     // Check if resource has been created successfully.
     $this->assertTrue(is_resource($old_res));
@@ -456,7 +456,7 @@ public function testGifTransparentImages() {
     // Test loading an indexed GIF image with transparent color set.
     // Color at top-right pixel should be fully transparent.
     $file = 'image-test-transparent-indexed.gif';
-    $image = $this->imageFactory->get(drupal_get_path('module', 'simpletest') . '/files/' . $file);
+    $image = $this->imageFactory->get('core/tests/fixtures/files/' . $file);
     $resource = $image->getToolkit()->getResource();
     $color_index = imagecolorat($resource, $image->getWidth() - 1, 0);
     $color = array_values(imagecolorsforindex($resource, $color_index));
@@ -490,14 +490,14 @@ public function testGifTransparentImages() {
     // of 6 colors, and setting the transparent color index to 6 (one higher
     // than the largest allowed index), as follows:
     // @code
-    // $image = imagecreatefromgif('core/modules/simpletest/files/image-test.gif');
+    // $image = imagecreatefromgif('core/tests/fixtures/files/image-test.gif');
     // imagecolortransparent($image, 6);
-    // imagegif($image, 'core/modules/simpletest/files/image-test-transparent-out-of-range.gif');
+    // imagegif($image, 'core/tests/fixtures/files/image-test-transparent-out-of-range.gif');
     // @endcode
     // This allows us to test that an image with an out-of-range color index
     // can be loaded correctly.
     $file = 'image-test-transparent-out-of-range.gif';
-    $image = $this->imageFactory->get(drupal_get_path('module', 'simpletest') . '/files/' . $file);
+    $image = $this->imageFactory->get('core/tests/fixtures/files/' . $file);
     $toolkit = $image->getToolkit();
 
     if (!$image->isValid()) {
@@ -522,7 +522,7 @@ public function testMissingOperation() {
     $file = 'image-test.png';
 
     // Load up a fresh image.
-    $image = $this->imageFactory->get(drupal_get_path('module', 'simpletest') . '/files/' . $file);
+    $image = $this->imageFactory->get('core/tests/fixtures/files/' . $file);
     if (!$image->isValid()) {
       $this->fail(new FormattableMarkup('Could not load image %file.', ['%file' => $file]));
     }
diff --git a/core/tests/Drupal/KernelTests/Core/Installer/InstallerLanguageTest.php b/core/tests/Drupal/KernelTests/Core/Installer/InstallerLanguageTest.php
index a265ed4ba0..0e74c35b4d 100644
--- a/core/tests/Drupal/KernelTests/Core/Installer/InstallerLanguageTest.php
+++ b/core/tests/Drupal/KernelTests/Core/Installer/InstallerLanguageTest.php
@@ -26,9 +26,9 @@ public function testInstallerTranslationFiles() {
       'it' => [],
     ];
 
-    // Hardcode the simpletest module location as we don't yet know where it is.
+    // Hardcode the fixtures location as we don't yet know where it is.
     // @todo Remove as part of https://www.drupal.org/node/2186491
-    $file_translation = new FileTranslation('core/modules/simpletest/files/translations', $this->container->get('file_system'));
+    $file_translation = new FileTranslation('core/tests/fixtures/files/translations', $this->container->get('file_system'));
     foreach ($expected_translation_files as $langcode => $files_expected) {
       $files_found = $file_translation->findTranslationFiles($langcode);
       $this->assertTrue(count($files_found) == count($files_expected), new FormattableMarkup('@count installer languages found.', ['@count' => count($files_expected)]));
diff --git a/core/tests/Drupal/Tests/TestFileCreationTrait.php b/core/tests/Drupal/Tests/TestFileCreationTrait.php
index ad664e4108..27ee641b50 100644
--- a/core/tests/Drupal/Tests/TestFileCreationTrait.php
+++ b/core/tests/Drupal/Tests/TestFileCreationTrait.php
@@ -24,7 +24,7 @@ trait TestFileCreationTrait {
    * The first time this method is called, it will call
    * $this->generateFile() to generate binary and ASCII text files in the
    * public:// directory. It will also copy all files in
-   * core/modules/simpletest/files to public://. These contain image, SQL, PHP,
+   * core/tests/fixtures/files to public://. These contain image, SQL, PHP,
    * JavaScript, and HTML files.
    *
    * All filenames are prefixed with their type and have appropriate extensions:
@@ -69,8 +69,8 @@ protected function getTestFiles($type, $size = NULL) {
         $this->generateFile('text-' . $count++, 64, $line, 'text');
       }
 
-      // Copy other test files from simpletest.
-      $original = drupal_get_path('module', 'simpletest') . '/files';
+      // Copy other test files from fixtures.
+      $original = \Drupal::service('app.root') . '/core/tests/fixtures/files';
       $files = $file_system->scanDirectory($original, '/(html|image|javascript|php|sql)-.*/');
       foreach ($files as $file) {
         $file_system->copy($file->uri, PublicStream::basePath());
diff --git a/core/modules/simpletest/files/README.txt b/core/tests/fixtures/files/README.txt
similarity index 100%
rename from core/modules/simpletest/files/README.txt
rename to core/tests/fixtures/files/README.txt
diff --git a/core/modules/simpletest/files/html-1.txt b/core/tests/fixtures/files/html-1.txt
similarity index 100%
rename from core/modules/simpletest/files/html-1.txt
rename to core/tests/fixtures/files/html-1.txt
diff --git a/core/modules/simpletest/files/html-2.html b/core/tests/fixtures/files/html-2.html
similarity index 100%
rename from core/modules/simpletest/files/html-2.html
rename to core/tests/fixtures/files/html-2.html
diff --git a/core/modules/simpletest/files/image-1.png b/core/tests/fixtures/files/image-1.png
similarity index 100%
rename from core/modules/simpletest/files/image-1.png
rename to core/tests/fixtures/files/image-1.png
diff --git a/core/modules/simpletest/files/image-2.jpg b/core/tests/fixtures/files/image-2.jpg
similarity index 100%
rename from core/modules/simpletest/files/image-2.jpg
rename to core/tests/fixtures/files/image-2.jpg
diff --git a/core/modules/simpletest/files/image-3.jpg b/core/tests/fixtures/files/image-3.jpg
similarity index 100%
rename from core/modules/simpletest/files/image-3.jpg
rename to core/tests/fixtures/files/image-3.jpg
diff --git a/core/modules/simpletest/files/image-test-no-transparency.gif b/core/tests/fixtures/files/image-test-no-transparency.gif
similarity index 100%
rename from core/modules/simpletest/files/image-test-no-transparency.gif
rename to core/tests/fixtures/files/image-test-no-transparency.gif
diff --git a/core/modules/simpletest/files/image-test-transparent-indexed.gif b/core/tests/fixtures/files/image-test-transparent-indexed.gif
similarity index 100%
rename from core/modules/simpletest/files/image-test-transparent-indexed.gif
rename to core/tests/fixtures/files/image-test-transparent-indexed.gif
diff --git a/core/modules/simpletest/files/image-test-transparent-out-of-range.gif b/core/tests/fixtures/files/image-test-transparent-out-of-range.gif
similarity index 100%
rename from core/modules/simpletest/files/image-test-transparent-out-of-range.gif
rename to core/tests/fixtures/files/image-test-transparent-out-of-range.gif
diff --git a/core/modules/simpletest/files/image-test.gif b/core/tests/fixtures/files/image-test.gif
similarity index 100%
rename from core/modules/simpletest/files/image-test.gif
rename to core/tests/fixtures/files/image-test.gif
diff --git a/core/modules/simpletest/files/image-test.jpg b/core/tests/fixtures/files/image-test.jpg
similarity index 100%
rename from core/modules/simpletest/files/image-test.jpg
rename to core/tests/fixtures/files/image-test.jpg
diff --git a/core/modules/simpletest/files/image-test.png b/core/tests/fixtures/files/image-test.png
similarity index 100%
rename from core/modules/simpletest/files/image-test.png
rename to core/tests/fixtures/files/image-test.png
diff --git a/core/modules/simpletest/files/invalid-img-test.png b/core/tests/fixtures/files/invalid-img-test.png
similarity index 100%
rename from core/modules/simpletest/files/invalid-img-test.png
rename to core/tests/fixtures/files/invalid-img-test.png
diff --git a/core/modules/simpletest/files/invalid-img-zero-size.png b/core/tests/fixtures/files/invalid-img-zero-size.png
similarity index 100%
rename from core/modules/simpletest/files/invalid-img-zero-size.png
rename to core/tests/fixtures/files/invalid-img-zero-size.png
diff --git a/core/modules/simpletest/files/javascript-1.txt b/core/tests/fixtures/files/javascript-1.txt
similarity index 100%
rename from core/modules/simpletest/files/javascript-1.txt
rename to core/tests/fixtures/files/javascript-1.txt
diff --git a/core/modules/simpletest/files/javascript-2.script b/core/tests/fixtures/files/javascript-2.script
similarity index 100%
rename from core/modules/simpletest/files/javascript-2.script
rename to core/tests/fixtures/files/javascript-2.script
diff --git a/core/modules/simpletest/files/phar-1.phar b/core/tests/fixtures/files/phar-1.phar
similarity index 100%
rename from core/modules/simpletest/files/phar-1.phar
rename to core/tests/fixtures/files/phar-1.phar
diff --git a/core/modules/simpletest/files/php-1.txt b/core/tests/fixtures/files/php-1.txt
similarity index 100%
rename from core/modules/simpletest/files/php-1.txt
rename to core/tests/fixtures/files/php-1.txt
diff --git a/core/modules/simpletest/files/php-2.php b/core/tests/fixtures/files/php-2.php
similarity index 100%
rename from core/modules/simpletest/files/php-2.php
rename to core/tests/fixtures/files/php-2.php
diff --git a/core/modules/simpletest/files/sql-1.txt b/core/tests/fixtures/files/sql-1.txt
similarity index 100%
rename from core/modules/simpletest/files/sql-1.txt
rename to core/tests/fixtures/files/sql-1.txt
diff --git a/core/modules/simpletest/files/sql-2.sql b/core/tests/fixtures/files/sql-2.sql
similarity index 100%
rename from core/modules/simpletest/files/sql-2.sql
rename to core/tests/fixtures/files/sql-2.sql
diff --git a/core/modules/simpletest/files/translations/drupal-8.0.0-beta2.hu.po b/core/tests/fixtures/files/translations/drupal-8.0.0-beta2.hu.po
similarity index 100%
rename from core/modules/simpletest/files/translations/drupal-8.0.0-beta2.hu.po
rename to core/tests/fixtures/files/translations/drupal-8.0.0-beta2.hu.po
diff --git a/core/modules/simpletest/files/translations/drupal-8.0.0.de.po b/core/tests/fixtures/files/translations/drupal-8.0.0.de.po
similarity index 100%
rename from core/modules/simpletest/files/translations/drupal-8.0.0.de.po
rename to core/tests/fixtures/files/translations/drupal-8.0.0.de.po
