diff --git a/core/phpstan.neon.dist b/core/phpstan.neon.dist
index 4ec2b3acea..a6f7a80ded 100644
--- a/core/phpstan.neon.dist
+++ b/core/phpstan.neon.dist
@@ -13,6 +13,9 @@ parameters:
 
   excludePaths:
     - '../sites/simpletest/*.php'
+    - tests/Drupal/Tests/Core/Form/FormStateTest.php
+    - lib/Drupal/Core/Extension/ModuleHandler.php
+    - modules/link/tests/src/Kernel/LinkItemTest.php
 
   reportUnmatchedIgnoredErrors: false
 
