diff --git a/core/tests/Drupal/KernelTests/Core/Image/ToolkitGdTest.php b/core/tests/Drupal/KernelTests/Core/Image/ToolkitGdTest.php
index 0496973a7d..6b3c1b9255 100644
--- a/core/tests/Drupal/KernelTests/Core/Image/ToolkitGdTest.php
+++ b/core/tests/Drupal/KernelTests/Core/Image/ToolkitGdTest.php
@@ -531,4 +531,13 @@ public function testMissingOperation() {
     $this->assertFalse($image->apply('missing_op', []), 'Calling a missing image toolkit operation plugin fails.');
   }
 
+  /**
+   * Tests creating an image that exceeds memory available.
+   */
+  public function testImageTooBig() {
+    $image = $this->imageFactory->get();
+    $image->createNew(50000, 50000);
+    $this->assertFalse($image->isValid());
+  }
+
 }
