diff --git a/core/modules/migrate/tests/src/Unit/MigrateSqlIdMapEnsureTablesTest.php b/core/modules/migrate/tests/src/Unit/MigrateSqlIdMapEnsureTablesTest.php
index f4a4dc2596..a66c1a706d 100644
--- a/core/modules/migrate/tests/src/Unit/MigrateSqlIdMapEnsureTablesTest.php
+++ b/core/modules/migrate/tests/src/Unit/MigrateSqlIdMapEnsureTablesTest.php
@@ -126,10 +126,15 @@ public function testEnsureTablesNotExist() {
       ]);
     $schema->expects($this->exactly(2))
       ->method('createTable')
-      ->withConsecutive(
-        ['migrate_map_sql_idmap_test', $map_table_schema],
-        ['migrate_message_sql_idmap_test', $table_schema],
-      );
+      ->willReturnCallback(function (...$args) use ($map_table_schema, $table_schema) {
+        static $series = [
+          ['migrate_map_sql_idmap_test', $map_table_schema],
+          ['migrate_message_sql_idmap_test', $table_schema],
+        ];
+
+        $expectedArgs = array_shift($series);
+        $this->assertSame($expectedArgs, $args);
+      });
 
     $this->runEnsureTablesTest($schema);
   }
