diff --git a/core/tests/Drupal/Tests/Listeners/DrupalListener.php b/core/tests/Drupal/Tests/Listeners/DrupalListener.php
index 663c2e52..2fa24af3 100644
--- a/core/tests/Drupal/Tests/Listeners/DrupalListener.php
+++ b/core/tests/Drupal/Tests/Listeners/DrupalListener.php
@@ -23,6 +23,17 @@ class DrupalListener implements TestListener {
    */
   public function startTest(Test $test): void {
     $this->deprecationStartTest($test);
+
+    // Check for missing void return typehints in concrete test classes'
+    // methods. If the method is inherited from a base test class, do
+    // nothing.
+    $class = new \ReflectionClass($test);
+    $method = $class->getMethod('setUp');
+    if ($method->getDeclaringClass()->getName() === get_class($test)) {
+      if ($method->getReturnType()->getName() !== 'void') {
+        @trigger_error('Declaring setUp without a void return typehint is deprecated in drupal:9.0.0. It will be required before drupal:10.0.0. See https://www.drupal.org/node/xxx', E_USER_DEPRECATED);
+      }
+    }
   }
 
   /**
