diff --git a/core/lib/Drupal/Component/PhpStorage/FileStorage.php b/core/lib/Drupal/Component/PhpStorage/FileStorage.php
index a694144b71..79989f7674 100644
--- a/core/lib/Drupal/Component/PhpStorage/FileStorage.php
+++ b/core/lib/Drupal/Component/PhpStorage/FileStorage.php
@@ -51,7 +51,12 @@ public function load($name) {
   public function save($name, $code) {
     $path = $this->getFullPath($name);
     $directory = dirname($path);
-    $this->ensureDirectory($directory);
+    try {
+      $this->ensureDirectory($directory);
+    }
+    catch (\RuntimeException $e) {
+      return FALSE;
+    }
     return (bool) file_put_contents($path, $code);
   }
 
@@ -68,6 +73,9 @@ public function save($name, $code) {
    *   The directory path.
    * @param int $mode
    *   The mode, permissions, the directory should have.
+   *
+   * @throws \RuntimeException
+   *   If directory creation failed.
    */
   protected function ensureDirectory($directory, $mode = 0777) {
     if ($this->createDirectory($directory, $mode)) {
@@ -90,7 +98,10 @@ protected function ensureDirectory($directory, $mode = 0777) {
    *   The mode, permissions, the directory should have.
    *
    * @return bool
-   *   TRUE if the directory exists or has been created, FALSE otherwise.
+   *   TRUE if the directory exists or has been created.
+   *
+   * @throws \RuntimeException
+   *   If directory creation failed.
    */
   protected function createDirectory($directory, $mode = 0777) {
     // If the directory exists already, there's nothing to do.
@@ -120,7 +131,7 @@ protected function createDirectory($directory, $mode = 0777) {
       }
       else {
         // Something failed and the directory doesn't exist.
-        trigger_error('mkdir(): Permission Denied', E_USER_WARNING);
+        throw new \RuntimeException('mkdir(): Permission Denied');
       }
     }
     return FALSE;
diff --git a/core/lib/Drupal/Component/PhpStorage/MTimeProtectedFastFileStorage.php b/core/lib/Drupal/Component/PhpStorage/MTimeProtectedFastFileStorage.php
index 415079376f..3fd3572a8e 100644
--- a/core/lib/Drupal/Component/PhpStorage/MTimeProtectedFastFileStorage.php
+++ b/core/lib/Drupal/Component/PhpStorage/MTimeProtectedFastFileStorage.php
@@ -63,7 +63,12 @@ public function __construct(array $configuration) {
    * {@inheritdoc}
    */
   public function save($name, $data) {
-    $this->ensureDirectory($this->directory);
+    try {
+      $this->ensureDirectory($this->directory);
+    }
+    catch (\RuntimeException $e) {
+      return FALSE;
+    }
 
     // Write the file out to a temporary location. Prepend with a '.' to keep it
     // hidden from listings and web servers.
@@ -83,7 +88,12 @@ public function save($name, $data) {
     // atomically, i.e. there is no point at which another process attempting to
     // access the new path will find it missing.
     $directory = $this->getContainingDirectoryFullPath($name);
-    $this->ensureDirectory($directory);
+    try {
+      $this->ensureDirectory($directory);
+    }
+    catch (\RuntimeException $e) {
+      return FALSE;
+    }
     $full_path = $this->getFullPath($name, $directory, $mtime);
     $result = rename($temporary_path, $full_path);
 
diff --git a/core/phpstan-baseline.neon b/core/phpstan-baseline.neon
index 15db53472c..6526c6e7e2 100644
--- a/core/phpstan-baseline.neon
+++ b/core/phpstan-baseline.neon
@@ -2261,22 +2261,6 @@ parameters:
 			count: 1
 			path: tests/Drupal/Tests/Component/DependencyInjection/Dumper/OptimizedPhpArrayDumperTest.php
 
-		-
-			message: """
-				#^Call to deprecated method expectWarning\\(\\) of class PHPUnit\\\\Framework\\\\TestCase\\:
-				https\\://github\\.com/sebastianbergmann/phpunit/issues/5062$#
-			"""
-			count: 1
-			path: tests/Drupal/Tests/Component/PhpStorage/FileStorageTest.php
-
-		-
-			message: """
-				#^Call to deprecated method expectWarningMessage\\(\\) of class PHPUnit\\\\Framework\\\\TestCase\\:
-				https\\://github\\.com/sebastianbergmann/phpunit/issues/5062$#
-			"""
-			count: 1
-			path: tests/Drupal/Tests/Component/PhpStorage/FileStorageTest.php
-
 		-
 			message: "#^Constructor of class Drupal\\\\Tests\\\\Component\\\\Plugin\\\\Factory\\\\ArgumentsAllNull has an unused parameter \\$charismatic\\.$#"
 			count: 1
diff --git a/core/tests/Drupal/Tests/Component/PhpStorage/FileStorageTest.php b/core/tests/Drupal/Tests/Component/PhpStorage/FileStorageTest.php
index d7ca82e28b..abc38a86b3 100644
--- a/core/tests/Drupal/Tests/Component/PhpStorage/FileStorageTest.php
+++ b/core/tests/Drupal/Tests/Component/PhpStorage/FileStorageTest.php
@@ -106,9 +106,12 @@ public function testCreateDirectoryFailWarning() {
       'bin' => 'test',
     ]);
     $code = "<?php\n echo 'here';";
-    $this->expectWarning();
-    $this->expectWarningMessage('mkdir(): Permission Denied');
-    $storage->save('subdirectory/foo.php', $code);
+    $this->expectException(\RuntimeException::class);
+    $this->expectExceptionMessage('mkdir(): Permission Denied');
+    $reflection = new \ReflectionClass($storage);
+    $reflectionMethod = $reflection->getMethod('createDirectory');
+dump([__METHOD__, $reflectionMethod]);
+    $reflectionMethod->invoke($storage, 'subdirectory');
   }
 
 }
