diff --git a/core/lib/Drupal/Core/Database/StatementWrapper.php b/core/lib/Drupal/Core/Database/StatementWrapper.php
index 8a3bc208c5..1e57a694a1 100644
--- a/core/lib/Drupal/Core/Database/StatementWrapper.php
+++ b/core/lib/Drupal/Core/Database/StatementWrapper.php
@@ -7,7 +7,7 @@
 /**
  * Implementation of StatementInterface encapsulating PDOStatement.
  */
-class StatementWrapper implements \IteratorAggregate, StatementInterface {
+class StatementWrapper implements \Iterator, StatementInterface {
 
   /**
    * The Drupal database connection object.
@@ -32,6 +32,20 @@ class StatementWrapper implements \IteratorAggregate, StatementInterface {
    */
   protected $rowCountEnabled = FALSE;
 
+  /**
+   * The current row, retrieved in the current fetch format.
+   *
+   * @var mixed
+   */
+  protected mixed $currentRow = NULL;
+
+  /**
+   * The key of the current row.
+   *
+   * @var int|null
+   */
+  protected ?int $currentKey = NULL;
+
   /**
    * Constructs a StatementWrapper object.
    *
@@ -76,6 +90,9 @@ public function getConnectionTarget(): string {
    * {@inheritdoc}
    */
   public function execute($args = [], $options = []) {
+    $this->currentRow = NULL;
+    $this->currentKey = NULL;
+
     if (isset($options['fetch'])) {
       if (is_string($options['fetch'])) {
         // \PDO::FETCH_PROPS_LATE tells __construct() to run before properties
@@ -154,8 +171,7 @@ public function fetchAllKeyed($key_index = 0, $value_index = 1) {
    * {@inheritdoc}
    */
   public function fetchField($index = 0) {
-    // Call \PDOStatement::fetchColumn to fetch the field.
-    return $this->clientStatement->fetchColumn($index);
+    return $this->fetch(\PDO::FETCH_NUM)[$index] ?? NULL;
   }
 
   /**
@@ -171,9 +187,19 @@ public function fetchAssoc() {
    */
   public function fetchObject(string $class_name = NULL, array $constructor_arguments = NULL) {
     if ($class_name) {
-      return $this->clientStatement->fetchObject($class_name, $constructor_arguments);
+      $this->currentRow = $this->clientStatement->fetchObject($class_name, $constructor_arguments);
+    }
+    $this->currentRow = $this->clientStatement->fetchObject();
+
+    if ($this->currentRow) {
+      $this->currentKey = $this->currentKey === NULL ? 0 : $this->currentKey++;
+    }
+    else {
+      $this->currentRow = NULL;
+      $this->currentKey = NULL;
     }
-    return $this->clientStatement->fetchObject();
+
+    return $this->currentRow;
   }
 
   /**
@@ -218,24 +244,37 @@ public function fetch($mode = NULL, $cursor_orientation = NULL, $cursor_offset =
     // need to be pass the exact number of arguments we where given.
     switch (func_num_args()) {
       case 0:
-        return $this->clientStatement->fetch();
+        $this->currentRow = $this->clientStatement->fetch();
 
       case 1:
-        return $this->clientStatement->fetch($mode);
+        $this->currentRow = $this->clientStatement->fetch($mode);
 
       case 2:
-        return $this->clientStatement->fetch($mode, $cursor_orientation);
+        $this->currentRow = $this->clientStatement->fetch($mode, $cursor_orientation);
 
       case 3:
       default:
-        return $this->clientStatement->fetch($mode, $cursor_orientation, $cursor_offset);
+        $this->currentRow = $this->clientStatement->fetch($mode, $cursor_orientation, $cursor_offset);
+    }
+
+    if ($this->currentRow) {
+      $this->currentKey = $this->currentKey === NULL ? 0 : $this->currentKey++;
+    }
+    else {
+      $this->currentRow = NULL;
+      $this->currentKey = NULL;
     }
+
+    return $this->currentRow;
   }
 
   /**
    * {@inheritdoc}
    */
   public function fetchAll($mode = NULL, $column_index = NULL, $constructor_arguments = NULL) {
+    $this->currentRow = NULL;
+    $this->currentKey = NULL;
+
     // Call \PDOStatement::fetchAll to fetch all rows.
     // \PDOStatement is picky about the number of arguments in some cases so we
     // need to be pass the exact number of arguments we where given.
@@ -258,9 +297,36 @@ public function fetchAll($mode = NULL, $column_index = NULL, $constructor_argume
   /**
    * {@inheritdoc}
    */
-  #[\ReturnTypeWillChange]
-  public function getIterator() {
-    return new \ArrayIterator($this->fetchAll());
+  public function current(): mixed {
+    return $this->currentRow;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function key(): mixed {
+    return $this->currentKey;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function rewind(): void {
+    // Nothing to do: our DatabaseStatement can't be rewound.
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function next(): void {
+    $this->fetch();
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function valid(): bool {
+    return (bool) $this->currentRow;
   }
 
 }
