diff --git a/core/modules/image/tests/src/FunctionalJavascript/ImageAdminStylesTest.php b/core/modules/image/tests/src/FunctionalJavascript/ImageAdminStylesTest.php
index faea9dd8d4b39c37d01976e65354304ce9f43739..2f04ae36e086eaebe00ce859bd021fa5734bc512 100644
--- a/core/modules/image/tests/src/FunctionalJavascript/ImageAdminStylesTest.php
+++ b/core/modules/image/tests/src/FunctionalJavascript/ImageAdminStylesTest.php
@@ -68,6 +68,26 @@ public function testAjaxEnabledEffectForm(): void {
       $page->pressButton('Update effect');
       $assert->statusMessageContains('The image effect was successfully applied.', 'status');
     }
+
+    // Edit the 1st effect, multiple AJAX calls before updating.
+    $style = ImageStyle::load($style_name);
+    $uuid = array_values($style->getEffects()->getInstanceIds())[0];
+    $this->drupalGet($admin_path . '/manage/' . $style_name . '/effects/' . $uuid);
+    $field = $page->findField('data[test_parameter]');
+    $field->setValue(200);
+    $page->pressButton('Ajax refresh');
+    $this->assertSession()->assertWaitOnAjaxRequest();
+    $field->setValue(300);
+    $page->pressButton('Ajax refresh');
+    $this->assertSession()->assertWaitOnAjaxRequest();
+    $field->setValue(400);
+    $page->pressButton('Ajax refresh');
+    $this->assertSession()->assertWaitOnAjaxRequest();
+    $page->pressButton('Update effect');
+    $this->assertSession()->statusMessageContains('The image effect was successfully applied.', 'status');
+    $style = ImageStyle::load($style_name);
+    $effectConfiguration = $style->getEffect($uuid)->getConfiguration();
+    $this->assertSame('bongo', var_export([$uuid, $effectConfiguration], TRUE));
   }
 
 }
