diff --git a/tests/src/Functional/ToolkitImagemagickTest.php b/tests/src/Functional/ToolkitImagemagickTest.php
index b91ce51..041dc26 100644
--- a/tests/src/Functional/ToolkitImagemagickTest.php
+++ b/tests/src/Functional/ToolkitImagemagickTest.php
@@ -95,6 +95,33 @@ class ToolkitImagemagickTest extends BrowserTestBase {
     $this->assertCount(0, $this->fileSystem->scanDirectory('temporary://', '/ima.*/'), 'No files left in the temporary directory after the Drupal shutdown.');
   }
 
+  /**
+   * Tests that double cropping returns an image of expected size.
+   *
+   * @param string $toolkit_id
+   *   The id of the toolkit to set up.
+   * @param string $toolkit_config
+   *   The config object of the toolkit to set up.
+   * @param array $toolkit_settings
+   *   The settings of the toolkit to set up.
+   *
+   * @dataProvider providerToolkitConfiguration
+   */
+  public function testDoubleCropping($toolkit_id, $toolkit_config, array $toolkit_settings): void {
+    $this->setUpToolkit($toolkit_id, $toolkit_config, $toolkit_settings);
+    $this->prepareImageFileHandling();
+
+    // Prepare a test image.
+    $image = $this->imageFactory->get('public://image-test.png');
+    $image->resize(5833, 3889);
+    $image->save('public://3204498-test.png');
+
+    // Process the image.
+    $test_image = $this->imageFactory->get('public://3204498-test.png');
+    dump($test_image);
+    $this->assertNotNull($test_image);
+  }
+
   /**
    * Test image toolkit operations.
    *
