diff --git a/core/tests/Drupal/Tests/Listeners/DrupalStandardsListenerTrait.php b/core/tests/Drupal/Tests/Listeners/DrupalStandardsListenerTrait.php
index a13af0e0b8..f97da2aeaa 100644
--- a/core/tests/Drupal/Tests/Listeners/DrupalStandardsListenerTrait.php
+++ b/core/tests/Drupal/Tests/Listeners/DrupalStandardsListenerTrait.php
@@ -60,10 +60,12 @@ private function checkValidCoversForTest(TestCase $test) {
       return;
     }
     // Gather our annotations.
+    [$method] = explode(' ', $test->getName());
     $annotations = Test::parseTestMethodAnnotations(
-      static::class,
-      $test->getName()
+      get_class($test),
+      $method
     );
+dump([get_class($test), $test->getName(), $method, $annotations]);
     // Glean the @coversDefaultClass annotation.
     $default_class = '';
     $valid_default_class = FALSE;
