diff --git a/core/phpstan.neon.dist b/core/phpstan.neon.dist
index ca2787a16bd..f2a1af4dd49 100644
--- a/core/phpstan.neon.dist
+++ b/core/phpstan.neon.dist
@@ -1,12 +1,11 @@
 # Configuration file for PHPStan static code checking, see https://phpstan.org .
 # PHPStan is triggered on Drupal CI in commit-code-check.sh.
 includes:
-  - phpstan-baseline.neon
   - phar://phpstan.phar/conf/bleedingEdge.neon
 
 parameters:
 
-  level: 1
+  level: 9
 
   fileExtensions:
     - sh
