diff --git a/core/modules/system/tests/src/Functional/UpdateSystem/RebuildScriptTest.php b/core/modules/system/tests/src/Functional/UpdateSystem/RebuildScriptTest.php
index 484f520da0b..56ff390329e 100644
--- a/core/modules/system/tests/src/Functional/UpdateSystem/RebuildScriptTest.php
+++ b/core/modules/system/tests/src/Functional/UpdateSystem/RebuildScriptTest.php
@@ -32,9 +32,9 @@ public function testRebuild(): void {
     $cache = $this->container->get('cache.default');
 
     $cache->set('rebuild_test', TRUE);
-    $this->drupalGet(Url::fromUri('base:core/rebuild.php'));
-    $this->assertSession()->addressEquals(new Url('<front>'));
-    $this->assertInstanceOf(\stdClass::class, $cache->get('rebuild_test'));
+//    $this->drupalGet(Url::fromUri('base:core/rebuild.php'));
+//    $this->assertSession()->addressEquals(new Url('<front>'));
+//    $this->assertInstanceOf(\stdClass::class, $cache->get('rebuild_test'));
 
     $settings['settings']['rebuild_access'] = (object) [
       'value' => TRUE,
diff --git a/core/rebuild.php b/core/rebuild.php
index 4a6bc87a6d5..7132708e0ec 100644
--- a/core/rebuild.php
+++ b/core/rebuild.php
@@ -23,10 +23,14 @@
 $autoloader = require_once __DIR__ . '/../autoload.php';
 require_once __DIR__ . '/includes/utility.inc';
 
+dump('A');
+
 $request = Request::createFromGlobals();
 // Manually resemble early bootstrap of DrupalKernel::boot().
 DrupalKernel::bootEnvironment();
 
+dump('B');
+
 try {
   Settings::initialize(dirname(__DIR__), DrupalKernel::findSitePath($request), $autoloader);
 }
@@ -36,6 +40,8 @@
   exit;
 }
 
+dump('C', Settings::getAll());
+
 if (Settings::get('rebuild_access', FALSE) ||
   ($request->query->get('token') && $request->query->get('timestamp') &&
     (($request->server->getInt('REQUEST_TIME') - $request->query->get('timestamp')) < 300) &&
@@ -48,8 +54,13 @@
   ];
   array_map('call_user_func', array_filter($user_caches, 'is_callable'));
 
+  dump('D');
   drupal_rebuild($autoloader, $request);
   \Drupal::messenger()->addStatus('Cache rebuild complete.');
+  dump('E');
 }
+
+dump('F');
+
 $base_path = dirname($request->getBaseUrl(), 2);
-header('Location: ' . $request->getSchemeAndHttpHost() . $base_path);
+//header('Location: ' . $request->getSchemeAndHttpHost() . $base_path);
