diff --git a/core/lib/Drupal/Core/Batch/BatchStorage.php b/core/lib/Drupal/Core/Batch/BatchStorage.php
index 46f9fb97c0d..e50a7330136 100644
--- a/core/lib/Drupal/Core/Batch/BatchStorage.php
+++ b/core/lib/Drupal/Core/Batch/BatchStorage.php
@@ -6,6 +6,7 @@
 use Drupal\Core\Access\CsrfTokenGenerator;
 use Drupal\Core\Database\Connection;
 use Drupal\Core\Database\DatabaseException;
+use Drupal\Core\Database\Event\ExecuteMethodEnsuringSchemaEvent;
 use Symfony\Component\HttpFoundation\Session\SessionInterface;
 
 class BatchStorage implements BatchStorageInterface {
@@ -126,7 +127,7 @@ public function create(array $batch) {
    *   A batch id.
    */
   public function getId(): int {
-    $try_again = FALSE;
+/*    $try_again = FALSE;
     try {
       // The batch table might not yet exist.
       return $this->doInsertBatchRecord();
@@ -142,7 +143,13 @@ public function getId(): int {
     // Now that the table has been created, try again if necessary.
     if ($try_again) {
       return $this->doInsertBatchRecord();
-    }
+    }*/
+    $event = new ExecuteMethodEnsuringSchemaEvent(
+      [$this, 'doInsertBatchRecord'],
+      [static::TABLE_NAME => $this->schemaDefinition()],
+    );
+    $this->connection->dispatchEvent($event);
+    return $event->getResult();
   }
 
   /**
diff --git a/core/lib/Drupal/Core/Database/Event/ExecuteMethodEnsuringSchemaEvent.php b/core/lib/Drupal/Core/Database/Event/ExecuteMethodEnsuringSchemaEvent.php
new file mode 100644
index 00000000000..8a031a819d6
--- /dev/null
+++ b/core/lib/Drupal/Core/Database/Event/ExecuteMethodEnsuringSchemaEvent.php
@@ -0,0 +1,27 @@
+<?php
+
+namespace Drupal\Core\Database\Event;
+
+/**
+ * Represents the execution of a method ensuring an indicated schema exists.
+ */
+class ExecuteMethodEnsuringSchemaEvent extends DatabaseEvent {
+
+  protected mixed $callMethodResult;
+
+  public function __construct(
+    callable $callMethod,
+    protected readonly array $schema,
+  ) {
+    parent::__construct();
+  }
+
+  public function setResult(mixed $result): void {
+    $this->callMethodResult = $result;
+  }
+
+  public function getResult(): mixed {
+    return $this->callMethodResult;
+  }
+
+}
