diff --git a/core/tests/Drupal/Tests/Listeners/DrupalListener.php b/core/tests/Drupal/Tests/Listeners/DrupalListener.php
index ec1eb30ec1..7bf8bfde93 100644
--- a/core/tests/Drupal/Tests/Listeners/DrupalListener.php
+++ b/core/tests/Drupal/Tests/Listeners/DrupalListener.php
@@ -6,6 +6,7 @@
 use PHPUnit\Framework\TestListenerDefaultImplementation;
 use PHPUnit\Framework\Test;
 use PHPUnit\Framework\TestSuite;
+use PHPUnit\Util\Test as TestUtil;
 use Symfony\Bridge\PhpUnit\SymfonyTestsListener;
 
 /**
@@ -51,6 +52,7 @@ public function addSkippedTest(Test $test, \Throwable $t, float $time): void {
    */
   public function startTest(Test $test): void {
     $this->symfonyListener->startTest($test);
+dump(TestUtil::parseTestMethodAnnotations(get_class($test)));
     // Check for incorrect visibility of the $modules property.
     $class = new \ReflectionClass($test);
     if ($class->hasProperty('modules') && !$class->getProperty('modules')->isProtected()) {
