diff --git a/core/modules/image/image.module b/core/modules/image/image.module
index 0d8bd4b1f26..06e92d3c778 100644
--- a/core/modules/image/image.module
+++ b/core/modules/image/image.module
@@ -15,6 +15,7 @@
 use Drupal\file\FileInterface;
 use Drupal\image\Controller\ImageStyleDownloadController;
 use Drupal\image\Entity\ImageStyle;
+use Drupal\image\ImageProcessor\TransformedDimensions;
 
 /**
  * The name of the query parameter for image derivative tokens.
@@ -280,11 +281,7 @@ function template_preprocess_image_style(&$variables) {
   $style = ImageStyle::load($variables['style_name']);
 
   // Determine the dimensions of the styled image.
-  $dimensions = [
-    'width' => $variables['width'],
-    'height' => $variables['height'],
-  ];
-
+  $dimensions = new TransformedDimensions($variables['width'], $variables['height'], $variables['uri']);
   $style->transformDimensions($dimensions, $variables['uri']);
 
   $variables['image'] = [
diff --git a/core/modules/image/src/Entity/ImageStyle.php b/core/modules/image/src/Entity/ImageStyle.php
index f975dc10932..2aa71420056 100644
--- a/core/modules/image/src/Entity/ImageStyle.php
+++ b/core/modules/image/src/Entity/ImageStyle.php
@@ -17,6 +17,7 @@
 use Drupal\Core\Url;
 use Drupal\image\ImageEffectPluginCollection;
 use Drupal\image\ImageEffectInterface;
+use Drupal\image\ImageProcessor\TransformedDimensions;
 use Drupal\image\ImageStyleInterface;
 use Drupal\Component\Utility\Crypt;
 use Drupal\Component\Utility\UrlHelper;
@@ -346,10 +347,19 @@ public function createDerivative($original_uri, $derivative_uri) {
   /**
    * {@inheritdoc}
    */
-  public function transformDimensions(array &$dimensions, $uri) {
+  public function transformDimensions(array|TransformedDimensions &$dimensions, $uri): void {
+    if (is_array($dimensions)) {
+      @trigger_error('one', E_USER_DEPRECATED);
+    }
+    if (isset(func_num_args()[1])) {
+      @trigger_error('two', E_USER_DEPRECATED);
+    }
+    $d = ['width' => $dimensions['width'], 'height' => $dimensions['height']];
     foreach ($this->getEffects() as $effect) {
-      $effect->transformDimensions($dimensions, $uri);
+      $effect->transformDimensions($d, $dimensions->uri);
     }
+    $dimensions['width'] = $d['width'];
+    $dimensions['height'] = $d['height'];
   }
 
   /**
diff --git a/core/modules/image/src/ImageProcessor/TransformedDimensions.php b/core/modules/image/src/ImageProcessor/TransformedDimensions.php
new file mode 100644
index 00000000000..1829828af28
--- /dev/null
+++ b/core/modules/image/src/ImageProcessor/TransformedDimensions.php
@@ -0,0 +1,69 @@
+<?php
+
+namespace Drupal\image\ImageProcessor;
+
+/**
+ * Keeps track of the image dimensions transformation by a style.
+ */
+class TransformedDimensions implements \ArrayAccess {
+
+  /**
+   * The current width.
+   */
+  protected ?int $width;
+
+  /**
+   * The current height.
+   */
+  protected ?int $height;
+
+  /**
+   * Any third party key added to legacy $dimensions array.
+   *
+   * @deprecated
+   */
+  protected array $thirdPartyKey = [];
+
+  public function __construct(
+    public readonly int $originalWidth,
+    public readonly int $originalHeight,
+    public readonly ?string $uri,
+  ) {
+    $this->width = $this->originalWidth;
+    $this->height = $this->originalHeight;
+  }
+
+  public function offsetExists(mixed $offset): bool {
+    if (in_array($offset, ['width', 'height'])) {
+      return TRUE;
+    }
+    return array_key_exists($offset, $this->thirdPartyKey);
+  }
+
+  public function offsetGet(mixed $offset): mixed {
+    return match ($offset) {
+      'width' => $this->width,
+      'height' => $this->height,
+      default => $this->thirdPartyKey[$offset],
+    };
+  }
+
+  public function offsetSet(mixed $offset, mixed $value): void {
+    if ($offset === 'width') {
+      $this->width = $value;
+    }
+    elseif ($offset === 'height') {
+      $this->height = $value;
+    }
+    else {
+      $this->thirdPartyKey[$offset] = $value;
+    }
+  }
+
+  public function offsetUnset(mixed $offset): void {
+    if (!in_array($offset, ['width', 'height'])) {
+      unset($offset);
+    }
+  }
+
+}
diff --git a/core/modules/image/src/ImageStyleInterface.php b/core/modules/image/src/ImageStyleInterface.php
index 2339858e761..5421cc7757b 100644
--- a/core/modules/image/src/ImageStyleInterface.php
+++ b/core/modules/image/src/ImageStyleInterface.php
@@ -111,7 +111,7 @@ public function createDerivative($original_uri, $derivative_uri);
    * - width: Integer with the derivative image width.
    * - height: Integer with the derivative image height.
    *
-   * @param array $dimensions
+   * @param array|\Drupal\image\ImageProcessor\TransformedDimensions $dimensions
    *   Associative array passed by reference. Implementations have to store the
    *   resulting width and height, in pixels.
    * @param string $uri
diff --git a/core/modules/image/tests/src/Kernel/ImageEffectsTest.php b/core/modules/image/tests/src/Kernel/ImageEffectsTest.php
index 6b7662e33e4..951f5b2f878 100644
--- a/core/modules/image/tests/src/Kernel/ImageEffectsTest.php
+++ b/core/modules/image/tests/src/Kernel/ImageEffectsTest.php
@@ -7,6 +7,7 @@
 use Drupal\Core\Form\FormState;
 use Drupal\image\Entity\ImageStyle;
 use Drupal\image\Form\ImageEffectEditForm;
+use Drupal\image\ImageProcessor\TransformedDimensions;
 use Drupal\Tests\Traits\Core\Image\ToolkitTestTrait;
 use Drupal\KernelTests\KernelTestBase;
 
@@ -44,6 +45,15 @@ protected function setUp(): void {
     $this->imageEffectPluginManager = $this->container->get('plugin.manager.image.effect');
   }
 
+  public function testTransformedDimensions(): void {
+    $d = new TransformedDimensions(40, 20, 'public://image.png');
+    $this->assertSame(40, $d->originalWidth);
+    $this->assertSame(20, $d->originalHeight);
+    $this->assertSame('public://image.png', $d->uri);
+    $this->assertSame(40, $d['width']);
+    $this->assertSame(20, $d['height']);
+  }
+
   /**
    * Tests the 'image_resize' effect.
    */
