diff --git a/core/modules/mysqli/src/Driver/Database/mysqli/Connection.php b/core/modules/mysqli/src/Driver/Database/mysqli/Connection.php
index e41df23075a..dc053a73d20 100644
--- a/core/modules/mysqli/src/Driver/Database/mysqli/Connection.php
+++ b/core/modules/mysqli/src/Driver/Database/mysqli/Connection.php
@@ -64,6 +64,7 @@ public static function open(array &$connection_options = []) {
     // Allow PDO options to be overridden.
     $connection_options += [
       'pdo' => [],
+      'use_prepared_statements' => FALSE,
     ];
 
     try {
diff --git a/core/modules/mysqli/src/Driver/Database/mysqli/Statement.php b/core/modules/mysqli/src/Driver/Database/mysqli/Statement.php
index f3b4346992d..f2378ed96ae 100644
--- a/core/modules/mysqli/src/Driver/Database/mysqli/Statement.php
+++ b/core/modules/mysqli/src/Driver/Database/mysqli/Statement.php
@@ -88,10 +88,28 @@ public function execute($args = [], $options = []) {
       if (!$this->hasClientStatement()) {
         // Replace named placeholders with positional ones if needed.
         $this->paramsPositions = array_flip(array_keys($args));
+dump(['a', $this->queryString, $args, $this->paramsPositions]);
         $converter = new NamedPlaceholderConverter();
         $converter->parse($this->queryString, $args);
         [$convertedQueryString, $args] = [$converter->getConvertedSQL(), $converter->getConvertedParameters()];
-        $this->clientStatement = $this->clientConnection->prepare($convertedQueryString);
+dump(['b', $convertedQueryString, $args]);
+        if ($this->connection->getConnectionOptions()['use_prepared_statements']) {
+          $this->clientStatement = $this->clientConnection->prepare($convertedQueryString);
+        }
+        else {
+          $args = array_map(fn (string $arg): string => "'". $this->clientConnection->real_escape_string($arg) . "'", $args);
+          preg_match_all('/(?<!\?)\?(?!\?)/', $convertedQueryString, $matches, \PREG_OFFSET_CAPTURE);
+          $strx = '';
+          $iStart = 0;
+          $iArg = 0;
+          foreach ($matches[0] as $x) {
+              $iLength = $x[1] - $iStart;
+              $strx .= substr($convertedQueryString, $iStart, $iLength) . $args[$iArg++];
+              $iStart = $iStart + $iLength + 1;
+          }
+          $strx .= substr($convertedQueryString, $iStart);
+dump(['c', $convertedQueryString, $args, $strx]);
+        }
       }
       else {
         // Transform the $args to positional.
@@ -104,13 +122,18 @@ public function execute($args = [], $options = []) {
 
       // In mysqli, the results of the statement execution are returned in a
       // different object than the statement itself.
-      $return = $this->getClientStatement()->execute($args);
-      $this->result = new Result(
-        $this->fetchMode,
-        $this->fetchOptions,
-        $this->getClientStatement()->get_result(),
-        $this->clientConnection,
-      );
+      if ($this->connection->getConnectionOptions()['use_prepared_statements']) {
+        $return = $this->getClientStatement()->execute($args);
+        $this->result = new Result(
+          $this->fetchMode,
+          $this->fetchOptions,
+          $this->getClientStatement()->get_result(),
+          $this->clientConnection,
+        );
+      }
+      else {
+
+      }
       $this->markResultsetIterable($return);
     }
     catch (\Exception $e) {
