diff --git a/core/lib/Drupal/Core/Extension/DatabaseDriverUninstallValidator.php b/core/lib/Drupal/Core/Extension/DatabaseDriverUninstallValidator.php
index 672eafd4336..5878fcf9f6e 100644
--- a/core/lib/Drupal/Core/Extension/DatabaseDriverUninstallValidator.php
+++ b/core/lib/Drupal/Core/Extension/DatabaseDriverUninstallValidator.php
@@ -3,7 +3,6 @@
 namespace Drupal\Core\Extension;
 
 use Drupal\Core\Database\Connection;
-use Drupal\Core\Database\Database;
 use Drupal\Core\StringTranslation\StringTranslationTrait;
 use Drupal\Core\StringTranslation\TranslationInterface;
 
@@ -14,34 +13,22 @@ class DatabaseDriverUninstallValidator implements ModuleUninstallValidatorInterf
 
   use StringTranslationTrait;
 
-  /**
-   * The module extension list.
-   *
-   * @var \Drupal\Core\Extension\ModuleExtensionList
-   */
-  protected $moduleExtensionList;
-
-  /**
-   * The database connection.
-   *
-   * @var \Drupal\Core\Database\Connection
-   */
-  protected $connection;
-
   /**
    * Constructs a new DatabaseDriverUninstallValidator.
    *
-   * @param \Drupal\Core\StringTranslation\TranslationInterface $string_translation
+   * @param \Drupal\Core\StringTranslation\TranslationInterface $stringTranslation
    *   The string translation service.
-   * @param \Drupal\Core\Extension\ModuleExtensionList $extension_list_module
+   * @param \Drupal\Core\Extension\ModuleExtensionList $moduleExtensionList
    *   The module extension list.
    * @param \Drupal\Core\Database\Connection $connection
    *   The database connection.
    */
-  public function __construct(TranslationInterface $string_translation, ModuleExtensionList $extension_list_module, Connection $connection) {
-    $this->stringTranslation = $string_translation;
-    $this->moduleExtensionList = $extension_list_module;
-    $this->connection = $connection;
+  public function __construct(
+    TranslationInterface $stringTranslation,
+    protected readonly ModuleExtensionList $moduleExtensionList,
+    protected readonly Connection $connection,
+  ) {
+    $this->stringTranslation = $stringTranslation;
   }
 
   /**
@@ -50,12 +37,6 @@ public function __construct(TranslationInterface $string_translation, ModuleExte
   public function validate($module) {
     $reasons = [];
 
-    // This is here to allow InstallerNonDefaultDatabaseDriverTest execute,
-    // it needs to get a new connection than the one passed in construction.
-    // @todo Remove the next line of code in
-    // https://www.drupal.org/project/drupal/issues/3433034.
-    $this->connection = Database::getConnection();
-
     // When the database driver is provided by a module, then that module
     // cannot be uninstalled.
     if ($module === $this->connection->getProvider()) {
diff --git a/core/tests/Drupal/FunctionalTests/Installer/InstallerNonDefaultDatabaseDriverTest.php b/core/tests/Drupal/FunctionalTests/Installer/InstallerNonDefaultDatabaseDriverTest.php
index 2f90be519ff..f8c267e5132 100644
--- a/core/tests/Drupal/FunctionalTests/Installer/InstallerNonDefaultDatabaseDriverTest.php
+++ b/core/tests/Drupal/FunctionalTests/Installer/InstallerNonDefaultDatabaseDriverTest.php
@@ -5,8 +5,6 @@
 namespace Drupal\FunctionalTests\Installer;
 
 use Drupal\Core\Database\Database;
-use Drupal\Core\Extension\Extension;
-use Drupal\Core\Extension\ModuleUninstallValidatorException;
 
 // cspell:ignore drupaldriver testdriverdatabasedrivertestmysql
 // cspell:ignore testdriverdatabasedrivertestpgsql
@@ -65,51 +63,48 @@ public function testInstalled(): void {
 
     // Assert that in the settings.php the database connection array has the
     // correct values set.
-    $contents = file_get_contents($this->container->getParameter('app.root') . '/' . $this->siteDirectory . '/settings.php');
-    $this->assertStringContainsString("'namespace' => 'Drupal\\\\driver_test\\\\Driver\\\\Database\\\\{$this->testDriverName}',", $contents);
-    $this->assertStringContainsString("'driver' => '{$this->testDriverName}',", $contents);
-    $this->assertStringContainsString("'autoload' => 'core/modules/system/tests/modules/driver_test/src/Driver/Database/{$this->testDriverName}/',", $contents);
-
-    $dependencies = "'dependencies' => " . PHP_EOL .
-      "  array (" . PHP_EOL .
-      "    'mysql' => " . PHP_EOL .
-      "    array (" . PHP_EOL .
-      "      'namespace' => 'Drupal\\\\mysql'," . PHP_EOL .
-      "      'autoload' => 'core/modules/mysql/src/'," . PHP_EOL .
-      "    )," . PHP_EOL .
-      "    'pgsql' => " . PHP_EOL .
-      "    array (" . PHP_EOL .
-      "      'namespace' => 'Drupal\\\\pgsql'," . PHP_EOL .
-      "      'autoload' => 'core/modules/pgsql/src/'," . PHP_EOL .
-      "    )," . PHP_EOL .
-      "  )," . PHP_EOL;
-    $this->assertStringContainsString($dependencies, $contents);
-
-    // Assert that the module "driver_test" has been installed.
-    $this->assertEquals(\Drupal::service('module_handler')->getModule('driver_test'), new Extension($this->root, 'module', 'core/modules/system/tests/modules/driver_test/driver_test.info.yml'));
-
-    // Change the default database connection to use the database driver from
-    // the module "driver_test".
-    $connection_info = Database::getConnectionInfo();
-    $driver_test_connection = $connection_info['default'];
-    $driver_test_connection['driver'] = $this->testDriverName;
-    $driver_test_connection['namespace'] = 'Drupal\\driver_test\\Driver\\Database\\' . $this->testDriverName;
-    $driver_test_connection['autoload'] = "core/modules/system/tests/modules/driver_test/src/Driver/Database/{$this->testDriverName}/";
-    Database::renameConnection('default', 'original_database_connection');
-    Database::addConnectionInfo('default', 'default', $driver_test_connection);
-
-    // The module "driver_test" should not be uninstallable, because it is
-    // providing the database driver.
-    try {
-      $this->container->get('module_installer')->uninstall(['driver_test']);
-      $this->fail('Uninstalled driver_test module.');
-    }
-    catch (ModuleUninstallValidatorException $e) {
-      $this->assertStringContainsString("The module 'Contrib database driver test' is providing the database driver '{$this->testDriverName}'.", $e->getMessage());
-    }
+    $installedDatabaseSettings = $this->getInstalledDatabaseSettings();
+    $this->assertEquals("Drupal\\driver_test\\Driver\\Database\\{$this->testDriverName}", $installedDatabaseSettings['default']['default']['namespace']);
+    $this->assertEquals($this->testDriverName, $installedDatabaseSettings['default']['default']['driver']);
+    $this->assertEquals("core/modules/system/tests/modules/driver_test/src/Driver/Database/{$this->testDriverName}/", $installedDatabaseSettings['default']['default']['autoload']);
+    $this->assertEquals([
+      'mysql' => [
+        'namespace' => 'Drupal\\mysql',
+        'autoload' => 'core/modules/mysql/src/',
+      ],
+      'pgsql' => [
+        'namespace' => 'Drupal\\pgsql',
+        'autoload' => 'core/modules/pgsql/src/',
+      ],
+    ], $installedDatabaseSettings['default']['default']['dependencies']);
+
+    // Assert that the module "driver_test" and its dependencies have been
+    // installed.
+    $this->drupalGet('admin/modules');
+    $this->assertSession()->checkboxChecked('modules[driver_test][enable]');
+    $this->assertSession()->checkboxChecked('modules[mysql][enable]');
+    $this->assertSession()->checkboxChecked('modules[pgsql][enable]');
+
+    // The module "driver_test" can not be uninstalled, because it is providing
+    // the database driver. Also, the "mysql" and "pgsql" modules can not be
+    // uninstalled being dependencies of the "driver_test" module.
+    $this->drupalGet('admin/modules/uninstall');
+    $this->assertSession()->elementTextContains('xpath', '//tr[@data-drupal-selector="edit-driver-test"]', "The following reason prevents Contrib database driver test from being uninstalled: The module 'Contrib database driver test' is providing the database driver 'DriverTestMysql'.");
+    $this->assertSession()->elementTextContains('xpath', '//tr[@data-drupal-selector="edit-mysql"]', "The following reason prevents MySQL from being uninstalled: Required by: driver_test");
+    $this->assertSession()->elementTextContains('xpath', '//tr[@data-drupal-selector="edit-pgsql"]', "The following reason prevents PostgreSQL from being uninstalled: Required by: driver_test");
+  }
 
-    // Restore the old database connection.
-    Database::addConnectionInfo('default', 'default', $connection_info['default']);
+  /**
+   * Returns the databases setup from the SUT's settings.php.
+   *
+   * @return array<string,mixed>
+   *   The value of the $databases variable.
+   */
+  protected function getInstalledDatabaseSettings(): array {
+    $app_root = $this->container->getParameter('app.root');
+    $site_path = $this->siteDirectory;
+    include $app_root . '/' . $site_path . '/settings.php';
+    return $databases;
   }
 
 }
