diff --git a/core/tests/Drupal/KernelTests/Core/Image/ToolkitGdTest.php b/core/tests/Drupal/KernelTests/Core/Image/ToolkitGdTest.php
index b5dddf395..24a7a5751 100644
--- a/core/tests/Drupal/KernelTests/Core/Image/ToolkitGdTest.php
+++ b/core/tests/Drupal/KernelTests/Core/Image/ToolkitGdTest.php
@@ -459,6 +459,33 @@ public function testResourceDestruction() {
     $this->assertIsResource($new_res);
   }
 
+  /**
+   * Tests creation of image that will exceed memory limit.
+   */
+  public function testInsufficientMemory() {
+    $image = $this->imageFactory->get('core/tests/fixtures/files/image-test.png');
+
+    // Test creation of an image that will exceed the memory available.
+    $old_res = $image->getToolkit()->getResource();
+    $this->assertFalse($image->createNew(200000, 200000));
+    $new_res = $image->getToolkit()->getResource();
+
+    // Check that a new resource has not been created, and the old one is still
+    // valid.
+    $this->assertEquals($old_res, $new_res);
+  }
+
+  /**
+   * Tests creation of image that will exceed memory available.
+   */
+  public function testInsufficientMemory() {
+    $image = $this->imageFactory->get('core/tests/fixtures/files/image-test.png');
+
+    $memory_in_use = memory_get_usage(TRUE);
+    ini_set('memory_limit', $memory_in_use + 2048);
+    $this->assertFalse($image->resize(20000, 20000));
+  }
+
   /**
    * Tests for GIF images with transparency.
    */
