diff --git a/composer.json b/composer.json
index c137ad2ab4..6e577cd2e2 100644
--- a/composer.json
+++ b/composer.json
@@ -22,8 +22,7 @@
         "drupal/coder": "^8.3.10",
         "instaclick/php-webdriver": "^1.4.1",
         "mikey179/vfsstream": "^1.6.8",
-        "phpunit/phpunit": "^8.5.14 || ^9",
-        "phpspec/prophecy": "^1.12",
+        "phpunit/phpunit": "^8.5.14 || ^9 || dev-master",
         "symfony/css-selector": "^4.4",
         "symfony/phpunit-bridge": "^5.3.0",
         "symfony/var-dumper": "^5.3.0",
@@ -48,7 +47,7 @@
         "preferred-install": "dist",
         "autoloader-suffix": "Drupal9",
         "platform": {
-            "php": "7.3.0"
+            "php": "8.0.0"
         }
     },
     "extra": {
@@ -98,8 +97,10 @@
         "pre-autoload-dump": "Drupal\\Core\\Composer\\Composer::preAutoloadDump",
         "drupal-phpunit-upgrade-check": "Drupal\\Core\\Composer\\Composer::upgradePHPUnit",
         "drupal-phpunit-upgrade": [
-            "@composer update phpunit/phpunit --with-dependencies --no-progress",
-            "@composer require phpspec/prophecy-phpunit:^2 --no-progress --no-suggest"
+            "@composer remove phpunit/phpunit phpspec/prophecy --dev",
+            "@composer require phpspec/prophecy:^1.3 --dev --no-progress  --update-with-all-dependencies --ignore-platform-req php",
+            "@composer require phpspec/prophecy-phpunit:^2 --dev --no-progress  --update-with-all-dependencies --ignore-platform-req php",
+            "@composer require phpunit/phpunit:\"dev-master as 9.10.99\" --dev --no-progress --update-with-all-dependencies --ignore-platform-req php"
         ],
         "post-update-cmd": [
             "Drupal\\Composer\\Composer::generateMetapackages"
diff --git a/core/drupalci.yml b/core/drupalci.yml
index 94aa2ad22f..7faa3cd961 100644
--- a/core/drupalci.yml
+++ b/core/drupalci.yml
@@ -5,42 +5,43 @@ build:
   assessment:
     testing:
       # Run code quality checks.
-      container_command.commit-checks:
-        commands:
-          - "core/scripts/dev/commit-code-check.sh --drupalci"
-        halt-on-fail: true
+#      container_command.commit-checks:
+#        commands:
+#          - "core/scripts/dev/commit-code-check.sh --drupalci"
+#        halt-on-fail: true
       # run_tests task is executed several times in order of performance speeds.
       # halt-on-fail can be set on the run_tests tasks in order to fail fast.
       # suppress-deprecations is false in order to be alerted to usages of
       # deprecated code.
       run_tests.phpunit:
         types: 'PHPUnit-Unit'
-        testgroups: '--all'
-        suppress-deprecations: false
+#        testgroups: '--all'
+        testgroups: 'Serialization,Render,Session'
+        suppress-deprecations: true
         halt-on-fail: false
-      run_tests.kernel:
-        types: 'PHPUnit-Kernel'
-        testgroups: '--all'
-        suppress-deprecations: false
-        halt-on-fail: false
-      run_tests.build:
-        # Limit concurrency due to disk space concerns.
-        concurrency: 15
-        types: 'PHPUnit-Build'
-        testgroups: '--all'
-        suppress-deprecations: false
-        halt-on-fail: false
-      run_tests.functional:
-        types: 'PHPUnit-Functional'
-        testgroups: '--all'
-        suppress-deprecations: false
-        halt-on-fail: false
-      run_tests.javascript:
-        concurrency: 15
-        types: 'PHPUnit-FunctionalJavascript'
-        testgroups: '--all'
-        suppress-deprecations: false
-        halt-on-fail: false
-      # Run nightwatch testing.
-      # @see https://www.drupal.org/project/drupal/issues/2869825
-      nightwatchjs: {}
+#      run_tests.kernel:
+#        types: 'PHPUnit-Kernel'
+#        testgroups: '--all'
+#        suppress-deprecations: false
+#        halt-on-fail: false
+#      run_tests.build:
+#        # Limit concurrency due to disk space concerns.
+#        concurrency: 15
+#        types: 'PHPUnit-Build'
+#        testgroups: '--all'
+#        suppress-deprecations: false
+#        halt-on-fail: false
+#      run_tests.functional:
+#        types: 'PHPUnit-Functional'
+#        testgroups: '--all'
+#        suppress-deprecations: false
+#        halt-on-fail: false
+#      run_tests.javascript:
+#        concurrency: 15
+#        types: 'PHPUnit-FunctionalJavascript'
+#        testgroups: '--all'
+#        suppress-deprecations: false
+#        halt-on-fail: false
+#      # Run nightwatch testing.
+#      # @see https://www.drupal.org/project/drupal/issues/2869825
+#      nightwatchjs: {}
diff --git a/core/lib/Drupal/Core/Test/PhpUnitTestRunner.php b/core/lib/Drupal/Core/Test/PhpUnitTestRunner.php
index 84171a17b4..6832b668ac 100644
--- a/core/lib/Drupal/Core/Test/PhpUnitTestRunner.php
+++ b/core/lib/Drupal/Core/Test/PhpUnitTestRunner.php
@@ -145,8 +145,8 @@ public function runCommand(array $unescaped_test_classnames, $phpunit_file, &$st
       $phpunit_bin,
       '--log-junit',
       escapeshellarg($phpunit_file),
-      '--printer',
-      escapeshellarg(SimpletestUiPrinter::class),
+//      '--printer',
+//      escapeshellarg(SimpletestUiPrinter::class),
     ];
 
     // Optimized for running a single test.
diff --git a/core/phpunit.xml.dist b/core/phpunit.xml.dist
index 18e1f3b3e9..3570012d7b 100644
--- a/core/phpunit.xml.dist
+++ b/core/phpunit.xml.dist
@@ -6,13 +6,15 @@
  or your current system user. See core/tests/README.md and
  https://www.drupal.org/node/2116263 for details.
 -->
-<phpunit bootstrap="tests/bootstrap.php" colors="true"
+<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         bootstrap="tests/bootstrap.php" colors="true"
          beStrictAboutTestsThatDoNotTestAnything="true"
          beStrictAboutOutputDuringTests="true"
          beStrictAboutChangesToGlobalState="true"
          failOnWarning="true"
          printerClass="\Drupal\Tests\Listeners\HtmlOutputPrinter"
-         cacheResult="false">
+         cacheResult="false"
+         xsi:noNamespaceSchemaLocation="https://schema.phpunit.de/9.3/phpunit.xsd">
   <php>
     <!-- Set error reporting to E_ALL. -->
     <ini name="error_reporting" value="32767"/>
@@ -55,29 +57,22 @@
       <file>./tests/TestSuites/BuildTestSuite.php</file>
     </testsuite>
   </testsuites>
-  <listeners>
-    <listener class="\Drupal\Tests\Listeners\DrupalListener">
-    </listener>
-  </listeners>
-  <!-- Filter for coverage reports. -->
-  <filter>
-    <whitelist>
+  <!-- Settings for coverage reports. -->
+  <coverage>
+    <include>
       <directory>./includes</directory>
       <directory>./lib</directory>
-      <!-- Extensions can have their own test directories, so exclude those. -->
       <directory>./modules</directory>
-      <exclude>
-        <directory>./modules/*/src/Tests</directory>
-        <directory>./modules/*/tests</directory>
-      </exclude>
       <directory>../modules</directory>
-      <exclude>
-        <directory>../modules/*/src/Tests</directory>
-        <directory>../modules/*/tests</directory>
-        <directory>../modules/*/*/src/Tests</directory>
-        <directory>../modules/*/*/tests</directory>
-      </exclude>
       <directory>../sites</directory>
-     </whitelist>
-  </filter>
+    </include>
+    <exclude>
+      <directory>./modules/*/src/Tests</directory>
+      <directory>./modules/*/tests</directory>
+      <directory>../modules/*/src/Tests</directory>
+      <directory>../modules/*/tests</directory>
+      <directory>../modules/*/*/src/Tests</directory>
+      <directory>../modules/*/*/tests</directory>
+    </exclude>
+  </coverage>
 </phpunit>
diff --git a/core/tests/Drupal/TestTools/PhpUnitCompatibility/PhpUnit10/TestCompatibilityTrait.php b/core/tests/Drupal/TestTools/PhpUnitCompatibility/PhpUnit10/TestCompatibilityTrait.php
new file mode 100644
index 0000000000..fc1e828758
--- /dev/null
+++ b/core/tests/Drupal/TestTools/PhpUnitCompatibility/PhpUnit10/TestCompatibilityTrait.php
@@ -0,0 +1,20 @@
+<?php
+
+namespace Drupal\TestTools\PhpUnitCompatibility\PhpUnit10;
+
+//use Prophecy\PhpUnit\ProphecyTrait;
+
+// @todo Replace with a proper dependency when we stop supporting PHPUnit 8.
+//if (!trait_exists(ProphecyTrait::class)) {
+//  print "Drupal requires Prophecy PhpUnit when using PHPUnit 9 or greater. Please use 'composer require --dev phpspec/prophecy-phpunit:^2' to ensure that it is present.\n";
+//  exit(1);
+//}
+
+/**
+ * Drupal's forward compatibility layer with multiple versions of PHPUnit.
+ */
+trait TestCompatibilityTrait {
+
+//  use ProphecyTrait;
+
+}
