diff --git a/core/modules/contact/tests/src/Functional/ContactSitewideTest.php b/core/modules/contact/tests/src/Functional/ContactSitewideTest.php
index abbe71420c..223d6a1609 100644
--- a/core/modules/contact/tests/src/Functional/ContactSitewideTest.php
+++ b/core/modules/contact/tests/src/Functional/ContactSitewideTest.php
@@ -213,7 +213,7 @@ public function testSiteWideContact() {
     $this->drupalLogout();
     $this->drupalGet('contact');
     $this->assertSession()->pageTextContains('Your email address');
-    $this->assertNoText('Form');
+    $this->assertSession()->pageTextNotContains('Form');
     $this->drupalLogin($admin_user);
 
     // Add more forms.
diff --git a/core/tests/Drupal/Tests/WebAssert.php b/core/tests/Drupal/Tests/WebAssert.php
index 47a5c283e5..14ed01c604 100644
--- a/core/tests/Drupal/Tests/WebAssert.php
+++ b/core/tests/Drupal/Tests/WebAssert.php
@@ -894,7 +894,23 @@ public function pageTextNotContains($text) {
     if (func_num_args() > 1) {
       @trigger_error('Calling ' . __METHOD__ . ' with more than one argument is deprecated in drupal:9.1.0 and will throw an \InvalidArgumentException in drupal:10.0.0. See https://www.drupal.org/node/3162537', E_USER_DEPRECATED);
     }
-    return parent::pageTextNotContains($text);
+    $dom = new \DOMDocument();
+    // Suppress libxml warnings when loading HTML 5 content.
+    @$dom->loadHTML($this->session->getPage()->getHtml());
+    $body = $dom->getElementsByTagName('body')[0];
+dump(['000', $body->textContent]);
+    foreach ($body->getElementsByTagName('script') as $item) {
+      $item->parentNode->removeChild($item);
+    }
+    $actual = preg_replace('/\s+/u', ' ', $body->textContent);
+dump(['aaa', $body->textContent, $actual]);
+    $regex = '/' . preg_quote($text, '/') . '/ui';
+    if (!preg_match($regex, $actual)) {
+      return;
+    }
+
+    $message = sprintf('The text "%s" appears in the text of this page, but it should not.', $text);
+    throw new ResponseTextException($message, $this->session->getDriver());
   }
 
   /**
