diff --git a/core/modules/contact/tests/src/Functional/ContactSitewideTest.php b/core/modules/contact/tests/src/Functional/ContactSitewideTest.php
index abbe71420c..9f2536e5d8 100644
--- a/core/modules/contact/tests/src/Functional/ContactSitewideTest.php
+++ b/core/modules/contact/tests/src/Functional/ContactSitewideTest.php
@@ -212,6 +212,7 @@ public function testSiteWideContact() {
     user_role_grant_permissions(RoleInterface::ANONYMOUS_ID, ['access site-wide contact form']);
     $this->drupalLogout();
     $this->drupalGet('contact');
+dump($this->getSession()->getPage()->getHtml());
     $this->assertSession()->pageTextContains('Your email address');
     $this->assertNoText('Form');
     $this->drupalLogin($admin_user);
