diff --git a/core/lib/Drupal/Core/Database/StatementWrapper.php b/core/lib/Drupal/Core/Database/StatementWrapper.php
index 4381fd1af6..ea65ea3681 100644
--- a/core/lib/Drupal/Core/Database/StatementWrapper.php
+++ b/core/lib/Drupal/Core/Database/StatementWrapper.php
@@ -9,6 +9,22 @@
  */
 class StatementWrapper implements \IteratorAggregate, StatementInterface {
 
+  protected array $fetchModes = [
+    \PDO::FETCH_DEFAULT => 'FETCH_DEFAULT',
+    \PDO::FETCH_LAZY => 'FETCH_LAZY',
+    \PDO::FETCH_ASSOC => 'FETCH_ASSOC',
+    \PDO::FETCH_NUM => 'FETCH_NUM',
+    \PDO::FETCH_BOTH => 'FETCH_BOTH',
+    \PDO::FETCH_OBJ => 'FETCH_OBJ',
+    \PDO::FETCH_BOUND => 'FETCH_BOUND',
+    \PDO::FETCH_COLUMN => 'FETCH_COLUMN',
+    \PDO::FETCH_CLASS => 'FETCH_CLASS',
+    \PDO::FETCH_INTO => 'FETCH_INTO',
+    \PDO::FETCH_FUNC => 'FETCH_FUNC',
+    \PDO::FETCH_NAMED => 'FETCH_NAMED',
+    \PDO::FETCH_KEY_PAIR => 'FETCH_KEY_PAIR',
+  ];
+
   /**
    * The Drupal database connection object.
    *
@@ -193,6 +209,9 @@ public function rowCount() {
    * {@inheritdoc}
    */
   public function setFetchMode($mode, $a1 = NULL, $a2 = []) {
+    if (!in_array($mode, [\PDO::FETCH_ASSOC, \PDO::FETCH_NUM, \PDO::FETCH_OBJ])) {
+      @trigger_error('Fetch mode ' . ($this->fetchModes[$mode] ?? $mode) . ' is deprecated in drupal:10.1.0 and is removed from drupal:11.0.0. Use supported modes only. See https://www.drupal.org/node/3244583', E_USER_DEPRECATED);
+    }
     // Call \PDOStatement::setFetchMode to set fetch mode.
     // \PDOStatement is picky about the number of arguments in some cases so we
     // need to be pass the exact number of arguments we where given.
@@ -213,6 +232,9 @@ public function setFetchMode($mode, $a1 = NULL, $a2 = []) {
    * {@inheritdoc}
    */
   public function fetch($mode = NULL, $cursor_orientation = NULL, $cursor_offset = NULL) {
+    if (!in_array($mode, [\PDO::FETCH_ASSOC, \PDO::FETCH_NUM, \PDO::FETCH_OBJ])) {
+      @trigger_error('Fetch mode ' . ($this->fetchModes[$mode] ?? $mode) . ' is deprecated in drupal:10.1.0 and is removed from drupal:11.0.0. Use supported modes only. See https://www.drupal.org/node/3244583', E_USER_DEPRECATED);
+    }
     // Call \PDOStatement::fetchAll to fetch all rows.
     // \PDOStatement is picky about the number of arguments in some cases so we
     // need to be pass the exact number of arguments we where given.
@@ -236,6 +258,9 @@ public function fetch($mode = NULL, $cursor_orientation = NULL, $cursor_offset =
    * {@inheritdoc}
    */
   public function fetchAll($mode = NULL, $column_index = NULL, $constructor_arguments = NULL) {
+    if (!in_array($mode, [\PDO::FETCH_ASSOC, \PDO::FETCH_NUM, \PDO::FETCH_OBJ])) {
+      @trigger_error('Fetch mode ' . ($this->fetchModes[$mode] ?? $mode) . ' is deprecated in drupal:10.1.0 and is removed from drupal:11.0.0. Use supported modes only. See https://www.drupal.org/node/3244583', E_USER_DEPRECATED);
+    }
     // Call \PDOStatement::fetchAll to fetch all rows.
     // \PDOStatement is picky about the number of arguments in some cases so we
     // need to be pass the exact number of arguments we where given.
