diff --git a/src/ImagemagickExecManager.php b/src/ImagemagickExecManager.php
index b292da3..e3ae71a 100644
--- a/src/ImagemagickExecManager.php
+++ b/src/ImagemagickExecManager.php
@@ -256,7 +256,9 @@ if ($return_code) dump(['command' => implode(' ', $cmdline), 'code' => $return_c
     if (($pre = $arguments->toArray(ArgumentMode::PreSource)) !== []) {
       array_push($cmdline, ...$pre);
     }
-    $cmdline[] = $sourcePath;
+    if ($sourcePath) {
+      $cmdline[] = $sourcePath;
+    }
     array_push($cmdline, ...$arguments->toArray(ArgumentMode::PostSource));
     $cmdline[] = $destinationPath;
     return $cmdline;
@@ -285,7 +287,9 @@ if ($return_code) dump(['command' => implode(' ', $cmdline), 'code' => $return_c
       array_push($cmdline, ...$pre);
     }
     array_push($cmdline, ...$arguments->toArray(ArgumentMode::PostSource));
-    $cmdline[] = $sourcePath;
+    if ($sourcePath) {
+      $cmdline[] = $sourcePath;
+    }
     $cmdline[] = $destinationPath;
     return $cmdline;
   }
