diff --git a/core/tests/Drupal/Tests/WebAssert.php b/core/tests/Drupal/Tests/WebAssert.php
index 47a5c283e5..074380ae62 100644
--- a/core/tests/Drupal/Tests/WebAssert.php
+++ b/core/tests/Drupal/Tests/WebAssert.php
@@ -109,8 +109,7 @@ public function responseHeaderDoesNotExist(string $name, string $message = ''):
    *   (Optional) the failure message.
    */
   public function pageTextMatchesCount(int $count, string $regex, string $message = ''): void {
-    $actual = preg_replace('/\s+/u', ' ', $this->session->getPage()->getText());
-    $matches = preg_match_all($regex, $actual);
+    $matches = preg_match_all($regex, $this->getCleanedUpPageText(TRUE));
     if ($message === '') {
       $message = "Failed asserting that the page matches the pattern '$regex' $count time(s), $matches found.";
     }
@@ -877,6 +876,29 @@ public function responseHeaderNotMatches($name, $regex) {
     return parent::responseHeaderNotMatches($name, $regex);
   }
 
+  /**
+   * Returns a Drupal's page text cleaned up of script tags.
+   *
+   * @param bool $pack
+   *   (Optional) If TRUE, packs contiguous spaces and newlines into a single
+   *   space. Defaults to FALSE.
+   *
+   * @return string
+   *   The cleaned up page text.
+   */
+  private function getCleanedUpPageText(bool $pack = FALSE): string {
+    $dom = new \DOMDocument();
+    // Suppress libxml warnings when loading HTML 5 content.
+    @$dom->loadHTML($this->session->getPage()->getHtml());
+    $body = $dom->getElementsByTagName('body')[0];
+    dump(['pre-', $body->textContent]);
+    while ($items = $body->getElementsByTagName('script')) {
+      $items[0]->parentNode->removeChild($items[0]);
+    }
+    dump(['post', $body->textContent, ($pack ? preg_replace('/\s+/u', ' ', $body->textContent) : $body->textContent)]);
+    return $pack ? preg_replace('/\s+/u', ' ', $body->textContent) : $body->textContent;
+  }
+
   /**
    * {@inheritdoc}
    */
@@ -884,7 +906,12 @@ public function pageTextContains($text) {
     if (func_num_args() > 1) {
       @trigger_error('Calling ' . __METHOD__ . ' with more than one argument is deprecated in drupal:9.1.0 and will throw an \InvalidArgumentException in drupal:10.0.0. See https://www.drupal.org/node/3162537', E_USER_DEPRECATED);
     }
-    return parent::pageTextContains($text);
+    $regex = '/' . preg_quote($text, '/') . '/ui';
+    if ((bool) preg_match($regex, $this->getCleanedUpPageText(TRUE))) {
+      return;
+    }
+    $message = sprintf('The text "%s" was not found anywhere in the text of the current page.', $text);
+    throw new ResponseTextException($message, $this->session->getDriver());
   }
 
   /**
@@ -894,7 +921,12 @@ public function pageTextNotContains($text) {
     if (func_num_args() > 1) {
       @trigger_error('Calling ' . __METHOD__ . ' with more than one argument is deprecated in drupal:9.1.0 and will throw an \InvalidArgumentException in drupal:10.0.0. See https://www.drupal.org/node/3162537', E_USER_DEPRECATED);
     }
-    return parent::pageTextNotContains($text);
+    $regex = '/' . preg_quote($text, '/') . '/ui';
+    if (!preg_match($regex, $this->getCleanedUpPageText(TRUE))) {
+      return;
+    }
+    $message = sprintf('The text "%s" appears in the text of this page, but it should not.', $text);
+    throw new ResponseTextException($message, $this->session->getDriver());
   }
 
   /**
@@ -904,7 +936,11 @@ public function pageTextMatches($regex) {
     if (func_num_args() > 1) {
       @trigger_error('Calling ' . __METHOD__ . ' with more than one argument is deprecated in drupal:9.1.0 and will throw an \InvalidArgumentException in drupal:10.0.0. See https://www.drupal.org/node/3162537', E_USER_DEPRECATED);
     }
-    return parent::pageTextMatches($regex);
+    if ((bool) preg_match($regex, $this->getCleanedUpPageText())) {
+      return;
+    }
+    $message = sprintf('The pattern %s was not found anywhere in the text of the current page.', $regex);
+    throw new ResponseTextException($message, $this->session->getDriver());
   }
 
   /**
@@ -914,7 +950,11 @@ public function pageTextNotMatches($regex) {
     if (func_num_args() > 1) {
       @trigger_error('Calling ' . __METHOD__ . ' with more than one argument is deprecated in drupal:9.1.0 and will throw an \InvalidArgumentException in drupal:10.0.0. See https://www.drupal.org/node/3162537', E_USER_DEPRECATED);
     }
-    return parent::pageTextNotMatches($regex);
+    if (!preg_match($regex, $this->getCleanedUpPageText())) {
+      return;
+    }
+    $message = sprintf('The pattern %s was found in the text of the current page, but it should not.', $regex);
+    throw new ResponseTextException($message, $this->session->getDriver());
   }
 
   /**
