diff --git a/core/lib/Drupal/Core/Test/TestDatabase.php b/core/lib/Drupal/Core/Test/TestDatabase.php
index 20cc8e6687..57c9a94c7e 100644
--- a/core/lib/Drupal/Core/Test/TestDatabase.php
+++ b/core/lib/Drupal/Core/Test/TestDatabase.php
@@ -261,6 +261,36 @@ public static function lastTestGet($test_id) {
         ':test_id' => $test_id,
       ])
       ->fetchField();
+dump([$last_prefix, $last_test_class]);
+    $res = $connection
+      ->select('simpletest', 'st')
+      ->fields('st')
+      ->condition('test_id', $test_id)
+      ->execute()
+      ->fetchAll();
+dump($res);
+    $res = $connection
+      ->select('simpletest', 'st')
+      ->fields('st', ['test_class'])
+      ->condition('test_id', $test_id)
+      ->groupBy('message_id')
+      ->having('MAX(message_id)')
+      ->execute()
+      ->fetchAll();
+dump($res);
+    $max_message_id_subquery = $connection
+      ->select('simpletest', 'sts')
+      ->condition('test_id', $test_id)
+      ->addExpression('MAX(message_id)', 'max_message_id');
+    $select = $connection->select($max_message_id_subquery, 'stb');
+    $select->join('simpletest', 'sta', 'sta.message_id = stb.max_message_id');
+    $select->fields('sta');
+dump((string) $select);
+    $res = $select
+      ->execute()
+      ->fetchAll();
+dump($res);
+      
     return [$last_prefix, $last_test_class];
   }
 
diff --git a/core/scripts/run-tests.sh b/core/scripts/run-tests.sh
index 3039fab006..9d252e5f99 100755
--- a/core/scripts/run-tests.sh
+++ b/core/scripts/run-tests.sh
@@ -740,6 +740,7 @@ function simpletest_script_execute_batch($test_classes) {
       if (empty($status['running'])) {
         // The child exited, unregister it.
         proc_close($child['process']);
+TestDatabase::lastTestGet($test_id);
         if ($status['exitcode'] === SIMPLETEST_SCRIPT_EXIT_FAILURE) {
           $total_status = max($status['exitcode'], $total_status);
         }
