diff --git a/core/tests/Drupal/FunctionalTests/Routing/PathEncodedTest.php b/core/tests/Drupal/FunctionalTests/Routing/PathEncodedTest.php
index fd5e94a6f9..f1ccd9ba61 100644
--- a/core/tests/Drupal/FunctionalTests/Routing/PathEncodedTest.php
+++ b/core/tests/Drupal/FunctionalTests/Routing/PathEncodedTest.php
@@ -35,6 +35,7 @@ public function testGetEncoded() {
       'path_encoded_test.parentheses' => '/cat(box)',
     ];
     foreach ($route_paths as $route_name => $path) {
+dump($path);
       $this->drupalGet(Url::fromRoute($route_name));
       $this->assertSession()->pageTextContains('PathEncodedTestController works');
     }
