diff --git a/src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler/Configuration.php b/src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler/Configuration.php
index 04d24883a9..253aa6ea59 100644
--- a/src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler/Configuration.php
+++ b/src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler/Configuration.php
@@ -52,7 +52,7 @@ class Configuration
     private $baselineFile = '';
 
     /**
-     * @var array
+     * @var array<string,array<string,array<string,array<string,int>>>>
      */
     private $baselineDeprecations = [];
 
@@ -144,7 +144,7 @@ class Configuration
             if (is_file($this->baselineFile)) {
                 $map = json_decode(file_get_contents($this->baselineFile));
                 foreach ($map as $baseline_deprecation) {
-                    $this->baselineDeprecations[$baseline_deprecation->location][$baseline_deprecation->message] = $baseline_deprecation->count;
+                    $this->baselineDeprecations[$baseline_deprecation->location][$baseline_deprecation->type ?? 'any'][$baseline_deprecation->message] = $baseline_deprecation->count;
                 }
             } else {
                 throw new \InvalidArgumentException(sprintf('The baselineFile "%s" does not exist.', $this->baselineFile));
@@ -214,17 +214,18 @@ class Configuration
             $location = 'procedural code';
         }
 
+        $type = $deprecation->getType();
         $message = $deprecation->getMessage();
-        $result = isset($this->baselineDeprecations[$location][$message]) && $this->baselineDeprecations[$location][$message] > 0;
+        $result = isset($this->baselineDeprecations[$location][$type][$message]) && $this->baselineDeprecations[$location][$type][$message] > 0;
         if ($this->generateBaseline) {
             if ($result) {
-                ++$this->baselineDeprecations[$location][$message];
+                ++$this->baselineDeprecations[$location][$type][$message];
             } else {
-                $this->baselineDeprecations[$location][$message] = 1;
+                $this->baselineDeprecations[$location][$type][$message] = 1;
                 $result = true;
             }
         } elseif ($result) {
-            --$this->baselineDeprecations[$location][$message];
+            --$this->baselineDeprecations[$location][$type][$message];
         }
 
         return $result;
@@ -246,13 +247,16 @@ class Configuration
     public function writeBaseline()
     {
         $map = [];
-        foreach ($this->baselineDeprecations as $location => $messages) {
-            foreach ($messages as $message => $count) {
-                $map[] = [
-                    'location' => $location,
-                    'message' => $message,
-                    'count' => $count,
-                ];
+        foreach ($this->baselineDeprecations as $location => $types) {
+            foreach ($types as $type => $messages) {
+                foreach ($messages as $message => $count) {
+                    $map[] = [
+                        'location' => $location,
+                        'type' => $type,
+                        'message' => $message,
+                        'count' => $count,
+                    ];
+                }
             }
         }
         file_put_contents($this->baselineFile, json_encode($map, \JSON_PRETTY_PRINT | \JSON_UNESCAPED_SLASHES));
