diff --git a/core/modules/comment/tests/src/Unit/CommentLinkBuilderTest.php b/core/modules/comment/tests/src/Unit/CommentLinkBuilderTest.php
index 1b00a4ff02..b55852106e 100644
--- a/core/modules/comment/tests/src/Unit/CommentLinkBuilderTest.php
+++ b/core/modules/comment/tests/src/Unit/CommentLinkBuilderTest.php
@@ -132,6 +132,7 @@ public function testCommentLinkBuilder(NodeInterface $node, $context, $has_acces
       ->method('isAnonymous')
       ->willReturn($is_anonymous);
     $links = $this->commentLinkBuilder->buildCommentedEntityLinks($node, $context);
+dump([get_class($node), $context, $has_access_comments, $history_exists, $has_post_comments, $is_anonymous, $expected, $links]);
     if (!empty($expected)) {
       if (!empty($links)) {
         foreach ($expected as $link => $detail) {
