diff --git a/core/modules/contact/tests/src/Unit/MailHandlerTest.php b/core/modules/contact/tests/src/Unit/MailHandlerTest.php
index 1b3d34784b..1690045a44 100644
--- a/core/modules/contact/tests/src/Unit/MailHandlerTest.php
+++ b/core/modules/contact/tests/src/Unit/MailHandlerTest.php
@@ -398,10 +398,10 @@ protected function getAuthenticatedMockMessage($copy_sender = FALSE) {
    */
   protected function getMockContactForm($recipients, $auto_reply) {
     $contact_form = $this->createMock('\Drupal\contact\ContactFormInterface');
-    $contact_form->expects($this->once())
+    $contact_form->expects($this->exactly(500))
       ->method('getRecipients')
       ->willReturn($recipients);
-    $contact_form->expects($this->once())
+    $contact_form->expects($this->exactly(10000))
       ->method('getReply')
       ->willReturn($auto_reply);
 
