diff --git a/src/ImagemagickExecManager.php b/src/ImagemagickExecManager.php
index b292da3..e3ae71a 100644
--- a/src/ImagemagickExecManager.php
+++ b/src/ImagemagickExecManager.php
@@ -256,7 +256,9 @@ if ($return_code) dump(['command' => implode(' ', $cmdline), 'code' => $return_c
     if (($pre = $arguments->toArray(ArgumentMode::PreSource)) !== []) {
       array_push($cmdline, ...$pre);
     }
-    $cmdline[] = $sourcePath;
+    if ($sourcePath) {
+      $cmdline[] = $sourcePath;
+    }
     array_push($cmdline, ...$arguments->toArray(ArgumentMode::PostSource));
     $cmdline[] = $destinationPath;
     return $cmdline;
@@ -285,7 +287,9 @@ if ($return_code) dump(['command' => implode(' ', $cmdline), 'code' => $return_c
       array_push($cmdline, ...$pre);
     }
     array_push($cmdline, ...$arguments->toArray(ArgumentMode::PostSource));
-    $cmdline[] = $sourcePath;
+    if ($sourcePath) {
+      $cmdline[] = $sourcePath;
+    }
     $cmdline[] = $destinationPath;
     return $cmdline;
   }
diff --git a/tests/src/Functional/ToolkitImagemagickTest.php b/tests/src/Functional/ToolkitImagemagickTest.php
index ddd7f5f..87aa243 100644
--- a/tests/src/Functional/ToolkitImagemagickTest.php
+++ b/tests/src/Functional/ToolkitImagemagickTest.php
@@ -452,7 +452,6 @@ class ToolkitImagemagickTest extends BrowserTestBase {
     foreach ([IMAGETYPE_PNG, IMAGETYPE_GIF, IMAGETYPE_JPEG] as $type) {
       $image = $this->imageFactory->get();
       $image->createNew(50, 20, image_type_to_extension($type, FALSE), '#ffff00');
-      $image->getToolkit()->arguments()->setDestinationFormatFromExtension(image_type_to_extension($type, FALSE));
       $file = 'from_null' . image_type_to_extension($type);
       $file_path = $this->testDirectory . '/' . $file;
       $this->assertEquals(50, $image->getWidth(), "Image file '$file' has the correct width.");
@@ -597,9 +596,9 @@ class ToolkitImagemagickTest extends BrowserTestBase {
     $file_names = [
       'ascii_test.png',
       'εικόναδοκιμής.png',
-      '\'greek εικόνα δοκιμής.png\'',
-      '"russian Тестовое изображение.png"',
-      '"simplified chinese 测试图片.png"',
+      'greek εικόνα δοκιμής.png',
+      'russian Тестовое изображение.png',
+      'simplified chinese 测试图片.png',
       '"japanese 試験画像.png"',
       '"arabic صورة الاختبار.png"',
       'armenian փորձարկման պատկերը.png',
