diff --git a/core/tests/Drupal/BuildTests/Composer/Component/ComponentsIsolatedBuildTest.php b/core/tests/Drupal/BuildTests/Composer/Component/ComponentsIsolatedBuildTest.php
index fdad24db7d..85b5570fba 100644
--- a/core/tests/Drupal/BuildTests/Composer/Component/ComponentsIsolatedBuildTest.php
+++ b/core/tests/Drupal/BuildTests/Composer/Component/ComponentsIsolatedBuildTest.php
@@ -13,8 +13,6 @@
  * @group Component
  *
  * @coversNothing
- *
- * @requires externalCommand composer
  */
 class ComponentsIsolatedBuildTest extends ComposerBuildTestBase {
 
diff --git a/core/tests/Drupal/BuildTests/Composer/Component/ComponentsTaggedReleaseTest.php b/core/tests/Drupal/BuildTests/Composer/Component/ComponentsTaggedReleaseTest.php
index 8fd0fbbfad..df12ea6b76 100644
--- a/core/tests/Drupal/BuildTests/Composer/Component/ComponentsTaggedReleaseTest.php
+++ b/core/tests/Drupal/BuildTests/Composer/Component/ComponentsTaggedReleaseTest.php
@@ -13,8 +13,6 @@
  * @group Component
  *
  * @coversNothing
- *
- * @requires externalCommand composer
  */
 class ComponentsTaggedReleaseTest extends ComposerBuildTestBase {
 
diff --git a/core/tests/Drupal/BuildTests/Composer/ComposerValidateTest.php b/core/tests/Drupal/BuildTests/Composer/ComposerValidateTest.php
index ab105f0110..13d7260564 100644
--- a/core/tests/Drupal/BuildTests/Composer/ComposerValidateTest.php
+++ b/core/tests/Drupal/BuildTests/Composer/ComposerValidateTest.php
@@ -7,7 +7,6 @@
 
 /**
  * @group Composer
- * @requires externalCommand composer
  */
 class ComposerValidateTest extends BuildTestBase {
 
diff --git a/core/tests/Drupal/BuildTests/Composer/Template/ComposerProjectTemplatesTest.php b/core/tests/Drupal/BuildTests/Composer/Template/ComposerProjectTemplatesTest.php
index 413b4d25cd..4246826df2 100644
--- a/core/tests/Drupal/BuildTests/Composer/Template/ComposerProjectTemplatesTest.php
+++ b/core/tests/Drupal/BuildTests/Composer/Template/ComposerProjectTemplatesTest.php
@@ -22,8 +22,6 @@
  *
  * @group #slow
  * @group Template
- *
- * @requires externalCommand composer
  */
 class ComposerProjectTemplatesTest extends ComposerBuildTestBase {
 
diff --git a/core/tests/Drupal/BuildTests/Framework/BuildTestBase.php b/core/tests/Drupal/BuildTests/Framework/BuildTestBase.php
index d700a13f4b..7d10abb863 100644
--- a/core/tests/Drupal/BuildTests/Framework/BuildTestBase.php
+++ b/core/tests/Drupal/BuildTests/Framework/BuildTestBase.php
@@ -9,6 +9,7 @@
 use Drupal\Tests\DrupalTestBrowser;
 use Drupal\Tests\PhpUnitCompatibilityTrait;
 use Drupal\Tests\Traits\PhpUnitWarnings;
+use Drupal\TestTools\Extension\RequiresComposerTrait;
 use PHPUnit\Framework\TestCase;
 use Symfony\Component\Filesystem\Filesystem as SymfonyFilesystem;
 use Symfony\Component\Finder\Finder;
@@ -52,7 +53,7 @@
  */
 abstract class BuildTestBase extends TestCase {
 
-  use ExternalCommandRequirementsTrait;
+  use RequiresComposerTrait;
   use PhpUnitWarnings;
   use PhpUnitCompatibilityTrait;
 
@@ -149,20 +150,11 @@ abstract class BuildTestBase extends TestCase {
    */
   private PhpExecutableFinder $phpFinder;
 
-  /**
-   * {@inheritdoc}
-   */
-  public static function setUpBeforeClass(): void {
-    parent::setUpBeforeClass();
-    static::checkClassCommandRequirements();
-  }
-
   /**
    * {@inheritdoc}
    */
   protected function setUp(): void {
     parent::setUp();
-    static::checkMethodCommandRequirements($this->getName());
     $this->phpFinder = new PhpExecutableFinder();
     // Set up the workspace directory.
     // @todo Glean working directory from env vars, etc.
diff --git a/core/tests/Drupal/BuildTests/Framework/ExternalCommandRequirementsTrait.php b/core/tests/Drupal/BuildTests/Framework/ExternalCommandRequirementsTrait.php
index 0faf6d5f7d..7b69a18cfb 100644
--- a/core/tests/Drupal/BuildTests/Framework/ExternalCommandRequirementsTrait.php
+++ b/core/tests/Drupal/BuildTests/Framework/ExternalCommandRequirementsTrait.php
@@ -28,6 +28,7 @@ trait ExternalCommandRequirementsTrait {
    *   skipped. Callers should not catch this exception.
    */
   private static function checkClassCommandRequirements() {
+    @trigger_error(__METHOD__ . "() is deprecated in drupal:10.2.0 and is removed from drupal:11.0.0. Use the ::checkCommandRequirements() method instead. See https://www.drupal.org/node/3362239", E_USER_DEPRECATED);
     $annotations = Test::parseTestMethodAnnotations(static::class);
     if (!empty($annotations['class']['requires'])) {
       static::checkExternalCommandRequirements($annotations['class']['requires']);
@@ -42,6 +43,7 @@ private static function checkClassCommandRequirements() {
    *   skipped. Callers should not catch this exception.
    */
   private static function checkMethodCommandRequirements($name) {
+    @trigger_error(__METHOD__ . "() is deprecated in drupal:10.2.0 and is removed from drupal:11.0.0. Use the ::checkCommandRequirements() method instead. See https://www.drupal.org/node/3362239", E_USER_DEPRECATED);
     $annotations = Test::parseTestMethodAnnotations(static::class, $name);
     if (!empty($annotations['method']['requires'])) {
       static::checkExternalCommandRequirements($annotations['method']['requires']);
@@ -59,10 +61,12 @@ private static function checkMethodCommandRequirements($name) {
    *   skipped. Callers should not catch this exception.
    */
   private static function checkExternalCommandRequirements(array $annotations) {
+    @trigger_error(__METHOD__ . "() is deprecated in drupal:10.2.0 and is removed from drupal:11.0.0. Use the ::checkCommandRequirements() method instead. See https://www.drupal.org/node/3362239", E_USER_DEPRECATED);
     // Make a list of required commands.
     $required_commands = [];
     foreach ($annotations as $requirement) {
       if (str_starts_with($requirement, 'externalCommand ')) {
+        @trigger_error("The '@require externalCommand' annotation for tests is deprecated in drupal:10.2.0 and is removed from drupal:11.0.0. Use the ::checkCommandRequirements() method instead. See https://www.drupal.org/node/3362239", E_USER_DEPRECATED);
         $command = trim(str_replace('externalCommand ', '', $requirement));
         // Use named keys to avoid duplicates.
         $required_commands[$command] = $command;
diff --git a/core/tests/Drupal/BuildTests/Framework/Tests/ExternalCommandRequirementTest.php b/core/tests/Drupal/BuildTests/Framework/Tests/ExternalCommandRequirementTest.php
index b4f6170fc1..1c4c0452cb 100644
--- a/core/tests/Drupal/BuildTests/Framework/Tests/ExternalCommandRequirementTest.php
+++ b/core/tests/Drupal/BuildTests/Framework/Tests/ExternalCommandRequirementTest.php
@@ -5,6 +5,7 @@
 use Drupal\BuildTests\Framework\ExternalCommandRequirementsTrait;
 use PHPUnit\Framework\SkippedTestError;
 use PHPUnit\Framework\TestCase;
+use Symfony\Bridge\PhpUnit\ExpectDeprecationTrait;
 
 /**
  * @coversDefaultClass \Drupal\BuildTests\Framework\ExternalCommandRequirementsTrait
@@ -12,13 +13,17 @@
  */
 class ExternalCommandRequirementTest extends TestCase {
 
+  use ExpectDeprecationTrait;
+
   /**
    * @covers ::checkExternalCommandRequirements
+   * @group legacy
    */
   public function testCheckExternalCommandRequirementsNotAvailable() {
+    $this->expectDeprecation('Drupal\BuildTests\Framework\ExternalCommandRequirementsTrait::checkExternalCommandRequirements() is deprecated in drupal:10.2.0 and is removed from drupal:11.0.0. Use the ::checkCommandRequirements() method instead. See https://www.drupal.org/node/3362239');
+    $this->expectDeprecation('The \'@require externalCommand\' annotation for tests is deprecated in drupal:10.2.0 and is removed from drupal:11.0.0. Use the ::checkCommandRequirements() method instead. See https://www.drupal.org/node/3362239');
     $requires = new UsesCommandRequirements();
     $ref_check_requirements = new \ReflectionMethod($requires, 'checkExternalCommandRequirements');
-    $ref_check_requirements->setAccessible(TRUE);
 
     // Use a try/catch block because otherwise PHPUnit might think this test is
     // legitimately skipped.
@@ -35,11 +40,13 @@ public function testCheckExternalCommandRequirementsNotAvailable() {
 
   /**
    * @covers ::checkExternalCommandRequirements
+   * @group legacy
    */
   public function testCheckExternalCommandRequirementsAvailable() {
+    $this->expectDeprecation('Drupal\BuildTests\Framework\ExternalCommandRequirementsTrait::checkExternalCommandRequirements() is deprecated in drupal:10.2.0 and is removed from drupal:11.0.0. Use the ::checkCommandRequirements() method instead. See https://www.drupal.org/node/3362239');
+    $this->expectDeprecation('The \'@require externalCommand\' annotation for tests is deprecated in drupal:10.2.0 and is removed from drupal:11.0.0. Use the ::checkCommandRequirements() method instead. See https://www.drupal.org/node/3362239');
     $requires = new UsesCommandRequirements();
     $ref_check_requirements = new \ReflectionMethod($requires, 'checkExternalCommandRequirements');
-    $ref_check_requirements->setAccessible(TRUE);
 
     // Use a try/catch block because otherwise PHPUnit might think this test is
     // legitimately skipped.
@@ -55,11 +62,12 @@ public function testCheckExternalCommandRequirementsAvailable() {
 
   /**
    * @covers ::checkClassCommandRequirements
+   * @group legacy
    */
   public function testClassRequiresAvailable() {
+    $this->expectDeprecation('Drupal\BuildTests\Framework\ExternalCommandRequirementsTrait::checkClassCommandRequirements() is deprecated in drupal:10.2.0 and is removed from drupal:11.0.0. Use the ::checkCommandRequirements() method instead. See https://www.drupal.org/node/3362239');
     $requires = new ClassRequiresAvailable();
     $ref_check = new \ReflectionMethod($requires, 'checkClassCommandRequirements');
-    $ref_check->setAccessible(TRUE);
     // Use a try/catch block because otherwise PHPUnit might think this test is
     // legitimately skipped.
     try {
@@ -72,11 +80,12 @@ public function testClassRequiresAvailable() {
 
   /**
    * @covers ::checkClassCommandRequirements
+   * @group legacy
    */
   public function testClassRequiresUnavailable() {
+    $this->expectDeprecation('Drupal\BuildTests\Framework\ExternalCommandRequirementsTrait::checkClassCommandRequirements() is deprecated in drupal:10.2.0 and is removed from drupal:11.0.0. Use the ::checkCommandRequirements() method instead. See https://www.drupal.org/node/3362239');
     $requires = new ClassRequiresUnavailable();
     $ref_check = new \ReflectionMethod($requires, 'checkClassCommandRequirements');
-    $ref_check->setAccessible(TRUE);
     // Use a try/catch block because otherwise PHPUnit might think this test is
     // legitimately skipped.
     try {
@@ -90,11 +99,12 @@ public function testClassRequiresUnavailable() {
 
   /**
    * @covers ::checkMethodCommandRequirements
+   * @group legacy
    */
   public function testMethodRequiresAvailable() {
+    $this->expectDeprecation('Drupal\BuildTests\Framework\ExternalCommandRequirementsTrait::checkMethodCommandRequirements() is deprecated in drupal:10.2.0 and is removed from drupal:11.0.0. Use the ::checkCommandRequirements() method instead. See https://www.drupal.org/node/3362239');
     $requires = new MethodRequires();
     $ref_check = new \ReflectionMethod($requires, 'checkMethodCommandRequirements');
-    $ref_check->setAccessible(TRUE);
     // Use a try/catch block because otherwise PHPUnit might think this test is
     // legitimately skipped.
     try {
@@ -107,11 +117,12 @@ public function testMethodRequiresAvailable() {
 
   /**
    * @covers ::checkMethodCommandRequirements
+   * @group legacy
    */
   public function testMethodRequiresUnavailable() {
+    $this->expectDeprecation('Drupal\BuildTests\Framework\ExternalCommandRequirementsTrait::checkMethodCommandRequirements() is deprecated in drupal:10.2.0 and is removed from drupal:11.0.0. Use the ::checkCommandRequirements() method instead. See https://www.drupal.org/node/3362239');
     $requires = new MethodRequires();
     $ref_check = new \ReflectionMethod($requires, 'checkMethodCommandRequirements');
-    $ref_check->setAccessible(TRUE);
     // Use a try/catch block because otherwise PHPUnit might think this test is
     // legitimately skipped.
     try {
diff --git a/core/tests/Drupal/TestTools/Extension/RequiresComposerTrait.php b/core/tests/Drupal/TestTools/Extension/RequiresComposerTrait.php
new file mode 100644
index 0000000000..a0bde6be90
--- /dev/null
+++ b/core/tests/Drupal/TestTools/Extension/RequiresComposerTrait.php
@@ -0,0 +1,21 @@
+<?php
+
+namespace Drupal\TestTools\Extension;
+
+use Symfony\Component\Process\ExecutableFinder;
+
+/**
+ * Ensures Composer executable is available, skips test otherwise.
+ */
+trait RequiresComposerTrait {
+
+  /**
+   * @beforeClass
+   */
+  public static function requiresComposer(): void {
+    if (!((new ExecutableFinder())->find('composxxer'))) {
+      static::markTestSkipped('This test requires the Composer executable to be accessible.');
+    }
+  }
+
+}
