diff --git a/.gitlab-ci.yml b/.gitlab-ci.yml
index 106b7eb756e..34e64e2d264 100644
--- a/.gitlab-ci.yml
+++ b/.gitlab-ci.yml
@@ -68,7 +68,7 @@ variables:
   PARENT_PIPELINE_ID: $CI_PIPELINE_ID
   _TARGET_PHP: "8.4-ubuntu"
   PHPUNIT_CONFIGURATION_FILE_PATH: $CI_PROJECT_DIR/core
-  PHPUNIT_FAIL_ON_PHPUNIT_DEPRECATION: false
+  PHPUNIT_FAIL_ON_PHPUNIT_DEPRECATION: true
 
 #############
 # Stages    #
@@ -583,7 +583,6 @@ default:
     KUBERNETES_CPU_REQUEST: "4"
     CONCURRENCY: 24
     _TARGET_PHP: "8.3-ubuntu"
-    PHPUNIT_FAIL_ON_PHPUNIT_DEPRECATION: true
 
 '⚡️ PHPUnit Unit tests on PHP 8.4':
   <<: [ *default-job-settings-lint, *default-phpunit-job-settings ]
@@ -591,7 +590,6 @@ default:
     TESTSUITE: PHPUnit-Unit
     KUBERNETES_CPU_REQUEST: "4"
     CONCURRENCY: 24
-    PHPUNIT_FAIL_ON_PHPUNIT_DEPRECATION: true
 
 '⚡️ PHPUnit Unit tests on PHP 8.5':
   <<: [ *default-job-settings-lint, *default-phpunit-job-settings ]
@@ -601,20 +599,17 @@ default:
     KUBERNETES_CPU_REQUEST: "4"
     CONCURRENCY: 24
     _TARGET_PHP: "8.5-ubuntu"
-    PHPUNIT_FAIL_ON_PHPUNIT_DEPRECATION: true
 
 '⚡️ Component unit tests on PHP 8.3':
   <<: [ *default-job-settings-lint, *default-component-unit-test-job-settings ]
   variables:
     KUBERNETES_CPU_REQUEST: "2"
     _TARGET_PHP: "8.3-ubuntu"
-    PHPUNIT_FAIL_ON_PHPUNIT_DEPRECATION: true
 
 '⚡️ Component unit tests on PHP 8.4':
   <<: [ *default-job-settings-lint, *default-component-unit-test-job-settings ]
   variables:
     KUBERNETES_CPU_REQUEST: "2"
-    PHPUNIT_FAIL_ON_PHPUNIT_DEPRECATION: true
 
 '⚡️ Component unit tests on PHP 8.5':
   <<: [ *default-job-settings-lint, *default-component-unit-test-job-settings ]
@@ -622,7 +617,6 @@ default:
   variables:
     KUBERNETES_CPU_REQUEST: "2"
     _TARGET_PHP: "8.5-ubuntu"
-    PHPUNIT_FAIL_ON_PHPUNIT_DEPRECATION: true
 
 '✅️ PHPStan rules tests':
   <<: [ *default-job-settings-lint ]
diff --git a/.gitlab-ci/pipeline.yml b/.gitlab-ci/pipeline.yml
index 981afc15ebe..bf3f378f1d1 100644
--- a/.gitlab-ci/pipeline.yml
+++ b/.gitlab-ci/pipeline.yml
@@ -27,10 +27,6 @@ default:
       [[ $_TARGET_DB == pgsql* ]] && export SIMPLETEST_DB=pgsql://$POSTGRES_USER:$POSTGRES_PASSWORD@database/$POSTGRES_DB?module=pgsql
     - echo "SIMPLETEST_DB = $SIMPLETEST_DB"
     - export PHPUNIT_CONFIGURATION_FILE_PATH=$CI_PROJECT_DIR/core
-    - |
-      [[ $TESTSUITE == PHPUnit-Build ]] && export PHPUNIT_FAIL_ON_PHPUNIT_DEPRECATION=true
-      [[ $TESTSUITE == PHPUnit-Functional ]] && export PHPUNIT_FAIL_ON_PHPUNIT_DEPRECATION=true
-      [[ $TESTSUITE == PHPUnit-FunctionalJavascript ]] && export PHPUNIT_FAIL_ON_PHPUNIT_DEPRECATION=true
     - $CI_PROJECT_DIR/.gitlab-ci/scripts/server-setup.sh
   after_script:
     - sed -i "s#$CI_PROJECT_DIR/##" ./sites/default/files/simpletest/phpunit-*.xml || true
@@ -77,7 +73,7 @@ variables:
   MINK_DRIVER_ARGS_WEBDRIVER_CHROMEDRIVER_NON_W3C: '["chrome", {"browserName":"chrome","goog:chromeOptions":{"args":["--disable-dev-shm-usage","--disable-gpu","--headless","--dns-prefetch-disable"]}}, "http://localhost:9515"]'
   CI_PARALLEL_NODE_INDEX: $CI_NODE_INDEX
   CI_PARALLEL_NODE_TOTAL: $CI_NODE_TOTAL
-  PHPUNIT_FAIL_ON_PHPUNIT_DEPRECATION: false
+  PHPUNIT_FAIL_ON_PHPUNIT_DEPRECATION: true
 
 .with-database: &with-database
   name: $_CONFIG_DOCKERHUB_ROOT/$_TARGET_DB:production
diff --git a/core/.deprecation-ignore.txt b/core/.deprecation-ignore.txt
index 3d220395485..b3bb0e6616e 100644
--- a/core/.deprecation-ignore.txt
+++ b/core/.deprecation-ignore.txt
@@ -33,6 +33,7 @@
 %Since symfony/http-foundation 7.2: NativeSessionStorage's "sid_bits_per_character" option is deprecated and will be ignored in Symfony 8.0.%
 
 # Drupal 12.
+%expectDeprecation\(\) is deprecated in drupal:11\.3\.0%
 %The "Drupal\\Core\\Database\\Query\\Select::hasAllTags\(\)" method will require a new "string \.\.\. \$tags" argument in the next major version of its interface%
 %The "Drupal\\Core\\Database\\Query\\Select::hasAnyTag\(\)" method will require a new "string \.\.\. \$tags" argument in the next major version of its interface%
 %The "Drupal\\Core\\Database\\Query\\SelectExtender::hasAllTags\(\)" method will require a new "string \.\.\. \$tags" argument in the next major version of its interface%
diff --git a/core/.phpstan-baseline.php b/core/.phpstan-baseline.php
index 8689c24a62b..c76c3da0cda 100644
--- a/core/.phpstan-baseline.php
+++ b/core/.phpstan-baseline.php
@@ -25856,6 +25856,15 @@
 	'count' => 1,
 	'path' => __DIR__ . '/modules/node/tests/src/Functional/NodeBlockFunctionalTest.php',
 ];
+$ignoreErrors[] = [
+	'message' => '#^Call to deprecated method expectDeprecation\\(\\) of class Drupal\\\\Tests\\\\BrowserTestBase\\:
+in drupal\\:11\\.3\\.0 and is removed from drupal\\:12\\.0\\.0\\. Use
+  \\$this\\-\\>expectUserDeprecationMessage\\(\\) or
+  \\$this\\-\\>expectUserDeprecationMessageMatches\\(\\) instead\\.$#',
+	'identifier' => 'method.deprecated',
+	'count' => 1,
+	'path' => __DIR__ . '/modules/node/tests/src/Functional/NodePageVariableDeprecationTest.php',
+];
 $ignoreErrors[] = [
 	'message' => '#^Method Drupal\\\\Tests\\\\node\\\\Functional\\\\NodeTestBase\\:\\:assertNodeAccess\\(\\) has no return type specified\\.$#',
 	'identifier' => 'missingType.return',
@@ -33672,6 +33681,15 @@
 	'count' => 1,
 	'path' => __DIR__ . '/modules/taxonomy/tests/src/Functional/Rest/VocabularyXmlBasicAuthTest.php',
 ];
+$ignoreErrors[] = [
+	'message' => '#^Call to deprecated method expectDeprecation\\(\\) of class Drupal\\\\Tests\\\\BrowserTestBase\\:
+in drupal\\:11\\.3\\.0 and is removed from drupal\\:12\\.0\\.0\\. Use
+  \\$this\\-\\>expectUserDeprecationMessage\\(\\) or
+  \\$this\\-\\>expectUserDeprecationMessageMatches\\(\\) instead\\.$#',
+	'identifier' => 'method.deprecated',
+	'count' => 1,
+	'path' => __DIR__ . '/modules/taxonomy/tests/src/Functional/TaxonomyTermPageVariableDeprecationTest.php',
+];
 $ignoreErrors[] = [
 	'message' => '#^Method Drupal\\\\Tests\\\\taxonomy\\\\Functional\\\\TaxonomyTestBase\\:\\:createEntityReferenceField\\(\\) has no return type specified\\.$#',
 	'identifier' => 'missingType.return',
@@ -46544,6 +46562,15 @@
 	'count' => 1,
 	'path' => __DIR__ . '/tests/Drupal/KernelTests/KernelTestBase.php',
 ];
+$ignoreErrors[] = [
+	'message' => '#^Usage of deprecated trait Drupal\\\\TestTools\\\\Extension\\\\DeprecationBridge\\\\ExpectDeprecationTrait in class Drupal\\\\KernelTests\\\\KernelTestBase\\:
+in drupal\\:11\\.3\\.0 and is removed from drupal\\:12\\.0\\.0\\. Use
+  \\$this\\-\\>expectUserDeprecationMessage\\(\\) or
+  \\$this\\-\\>expectUserDeprecationMessageMatches\\(\\) instead\\.$#',
+	'identifier' => 'traitUse.deprecatedTrait',
+	'count' => 1,
+	'path' => __DIR__ . '/tests/Drupal/KernelTests/KernelTestBase.php',
+];
 $ignoreErrors[] = [
 	'message' => '#^Variable \\$value in isset\\(\\) always exists and is not nullable\\.$#',
 	'identifier' => 'isset.variable',
@@ -46971,6 +46998,15 @@
 	'count' => 1,
 	'path' => __DIR__ . '/tests/Drupal/Tests/BrowserTestBase.php',
 ];
+$ignoreErrors[] = [
+	'message' => '#^Usage of deprecated trait Drupal\\\\TestTools\\\\Extension\\\\DeprecationBridge\\\\ExpectDeprecationTrait in class Drupal\\\\Tests\\\\BrowserTestBase\\:
+in drupal\\:11\\.3\\.0 and is removed from drupal\\:12\\.0\\.0\\. Use
+  \\$this\\-\\>expectUserDeprecationMessage\\(\\) or
+  \\$this\\-\\>expectUserDeprecationMessageMatches\\(\\) instead\\.$#',
+	'identifier' => 'traitUse.deprecatedTrait',
+	'count' => 1,
+	'path' => __DIR__ . '/tests/Drupal/Tests/BrowserTestBase.php',
+];
 $ignoreErrors[] = [
 	'message' => '#^Method Drupal\\\\Tests\\\\Component\\\\Annotation\\\\AnnotationBaseStub\\:\\:get\\(\\) has no return type specified\\.$#',
 	'identifier' => 'missingType.return',
@@ -49905,6 +49941,15 @@
 	'count' => 1,
 	'path' => __DIR__ . '/tests/Drupal/Tests/PerformanceData.php',
 ];
+$ignoreErrors[] = [
+	'message' => '#^Usage of deprecated trait Drupal\\\\TestTools\\\\Extension\\\\DeprecationBridge\\\\ExpectDeprecationTrait in class Drupal\\\\Tests\\\\UnitTestCase\\:
+in drupal\\:11\\.3\\.0 and is removed from drupal\\:12\\.0\\.0\\. Use
+  \\$this\\-\\>expectUserDeprecationMessage\\(\\) or
+  \\$this\\-\\>expectUserDeprecationMessageMatches\\(\\) instead\\.$#',
+	'identifier' => 'traitUse.deprecatedTrait',
+	'count' => 1,
+	'path' => __DIR__ . '/tests/Drupal/Tests/UnitTestCase.php',
+];
 $ignoreErrors[] = [
 	'message' => '#^Method Drupal\\\\Tests\\\\WebAssert\\:\\:assert\\(\\) has no return type specified\\.$#',
 	'identifier' => 'missingType.return',
diff --git a/core/assets/schemas/v1/icon_pack.schema.json b/core/assets/schemas/v1/icon_pack.schema.json
index 548341717ea..18e0a70bc62 100644
--- a/core/assets/schemas/v1/icon_pack.schema.json
+++ b/core/assets/schemas/v1/icon_pack.schema.json
@@ -93,7 +93,7 @@
       "title": "Asset library",
       "description": "The ID of an asset library",
       "type": "string",
-      "pattern": "^\\w+/[A-Za-z]+[\\w-]*$"
+      "pattern": "^\\w+/[A-Za-z]+\\w*$"
     }
   }
 }
\ No newline at end of file
diff --git a/core/modules/migrate/tests/src/Unit/Exception/MigrateSkipProcessExceptionTest.php b/core/modules/migrate/tests/src/Unit/Exception/MigrateSkipProcessExceptionTest.php
index eddae890f98..7970c3eaa9f 100644
--- a/core/modules/migrate/tests/src/Unit/Exception/MigrateSkipProcessExceptionTest.php
+++ b/core/modules/migrate/tests/src/Unit/Exception/MigrateSkipProcessExceptionTest.php
@@ -21,7 +21,7 @@ class MigrateSkipProcessExceptionTest extends UnitTestCase {
    */
   public function testDeprecation(): void {
     $this->expectException(MigrateSkipProcessException::class);
-    $this->expectDeprecation("Unsilenced deprecation: " . MigrateSkipProcessException::class . " is deprecated in drupal:10.3.0 and is removed from drupal:12.0.0. Return TRUE from a process plugin's isPipelineStopped() method to halt further processing on a pipeline. See https://www.drupal.org/node/3414511");
+    $this->expectDeprecation(MigrateSkipProcessException::class . " is deprecated in drupal:10.3.0 and is removed from drupal:12.0.0. Return TRUE from a process plugin's isPipelineStopped() method to halt further processing on a pipeline. See https://www.drupal.org/node/3414511");
     throw new MigrateSkipProcessException();
   }
 
diff --git a/core/tests/Drupal/FunctionalJavascriptTests/DrupalSelenium2Driver.php b/core/tests/Drupal/FunctionalJavascriptTests/DrupalSelenium2Driver.php
index 78814d59f9b..caaf506af79 100644
--- a/core/tests/Drupal/FunctionalJavascriptTests/DrupalSelenium2Driver.php
+++ b/core/tests/Drupal/FunctionalJavascriptTests/DrupalSelenium2Driver.php
@@ -230,4 +230,25 @@ private function executeJsOnElement(Element $element, string $script) {
     return $this->getWebDriverSession()->execute($options);
   }
 
+  /**
+   * {@inheritdoc}
+   */
+  public function executeScript(string $script) {
+    parent::executeScript($script);
+    $this->processJavascriptDeprecations();
+  }
+
+  /**
+   * Re-triggers within PHPUnit Javascript deprecations from the front-end.
+   */
+  public function processJavascriptDeprecations(): void {
+    $warnings = $this->evaluateScript("JSON.parse(sessionStorage.getItem('js_testing_log_test.warnings') || JSON.stringify([]))");
+    foreach ($warnings as $warning) {
+      if (str_starts_with($warning, '[Deprecation]')) {
+        // phpcs:ignore Drupal.Semantics.FunctionTriggerError
+        @trigger_error('Javascript Deprecation:' . substr($warning, 13), E_USER_DEPRECATED);
+      }
+    }
+  }
+
 }
diff --git a/core/tests/Drupal/FunctionalJavascriptTests/WebDriverTestBase.php b/core/tests/Drupal/FunctionalJavascriptTests/WebDriverTestBase.php
index d378208f03b..ba584ec0048 100644
--- a/core/tests/Drupal/FunctionalJavascriptTests/WebDriverTestBase.php
+++ b/core/tests/Drupal/FunctionalJavascriptTests/WebDriverTestBase.php
@@ -109,14 +109,6 @@ protected function tearDown(): void {
         // explaining what the problem is.
         throw new \RuntimeException('Unfinished AJAX requests while tearing down a test');
       }
-
-      $warnings = $this->getSession()->evaluateScript("JSON.parse(sessionStorage.getItem('js_testing_log_test.warnings') || JSON.stringify([]))");
-      foreach ($warnings as $warning) {
-        if (str_starts_with($warning, '[Deprecation]')) {
-          // phpcs:ignore Drupal.Semantics.FunctionTriggerError
-          @trigger_error('Javascript Deprecation:' . substr($warning, 13), E_USER_DEPRECATED);
-        }
-      }
     }
     parent::tearDown();
   }
@@ -214,6 +206,23 @@ public function assertSession($name = NULL) {
     return new WebDriverWebAssert($this->getSession($name), $this->baseUrl);
   }
 
+  /**
+   * {@inheritdoc}
+   */
+  protected function drupalGet($path, array $options = [], array $headers = []) {
+    $result = parent::drupalGet($path, $options, $headers);
+
+    // Wait for all requests to finish.
+    $this->getSession()->wait(5000, 'window.drupalActiveXhrCount === 0 || typeof window.drupalActiveXhrCount === "undefined"');
+
+    // Process Javascript deprecations.
+    $driver = $this->getSession()->getDriver();
+    assert($driver instanceof DrupalSelenium2Driver);
+    $driver->processJavascriptDeprecations();
+
+    return $result;
+  }
+
   /**
    * Gets the current Drupal javascript settings and parses into an array.
    *
diff --git a/core/tests/Drupal/KernelTests/KernelTestBase.php b/core/tests/Drupal/KernelTests/KernelTestBase.php
index 4396f2cd3cd..e75935d4120 100644
--- a/core/tests/Drupal/KernelTests/KernelTestBase.php
+++ b/core/tests/Drupal/KernelTests/KernelTestBase.php
@@ -25,6 +25,8 @@
 use Drupal\Tests\PhpUnitCompatibilityTrait;
 use Drupal\Tests\TestRequirementsTrait;
 use Drupal\TestTools\Comparator\MarkupInterfaceComparator;
+use Drupal\TestTools\ErrorHandler\BootstrapErrorHandler;
+use Drupal\TestTools\Extension\DeprecationBridge\DeprecationHandler;
 use Drupal\TestTools\Extension\DeprecationBridge\ExpectDeprecationTrait;
 use Drupal\TestTools\Extension\Dump\DebugDump;
 use Drupal\TestTools\Extension\SchemaInspector;
@@ -699,6 +701,13 @@ protected function tearDown(): void {
     $this->container = NULL;
     new Settings([]);
 
+    // We expect that the current error handler is the one set during the
+    // PHPUnit bootstrap. If not, the error handler was changed during the test
+    // execution but not properly restored during ::tearDown().
+    if (DeprecationHandler::isEnabled() && !get_error_handler() instanceof BootstrapErrorHandler) {
+      throw new \RuntimeException(sprintf('%s registered its own error handler without restoring the previous one before or during tear down. This can cause unpredictable test results. Ensure the test cleans up after itself.', $this->name()));
+    }
+
     parent::tearDown();
   }
 
diff --git a/core/tests/Drupal/TestTools/ErrorHandler/BootstrapErrorHandler.php b/core/tests/Drupal/TestTools/ErrorHandler/BootstrapErrorHandler.php
index 71079770a1a..1665a9b4fee 100644
--- a/core/tests/Drupal/TestTools/ErrorHandler/BootstrapErrorHandler.php
+++ b/core/tests/Drupal/TestTools/ErrorHandler/BootstrapErrorHandler.php
@@ -13,10 +13,8 @@
  *
  * This code works in coordination with DeprecationHandler.
  *
- * This error handler is registered during PHPUnit's runner bootstrap, and is
- * essentially used to capture deprecations occurring before tests are run (for
- * example, deprecations triggered by the DebugClassloader). When test runs
- * are prepared, a test specific TestErrorHandler is activated instead.
+ * This error handler is registered during PHPUnit's runner bootstrap, and
+ * overrides PHPUnit's own error handler.
  *
  * @see \Drupal\TestTools\Extension\DeprecationBridge\DeprecationHandler
  *
@@ -55,12 +53,6 @@ public function __invoke(int $errorNumber, string $errorString, string $errorFil
       throw new \RuntimeException(__METHOD__ . '() must not be called if the deprecation handler is not enabled.');
     }
 
-    // We collect a deprecation no matter what.
-    if (E_USER_DEPRECATED === $errorNumber || E_DEPRECATED === $errorNumber) {
-      $prefix = (error_reporting() & $errorNumber) ? 'Unsilenced deprecation: ' : '';
-      DeprecationHandler::collectActualDeprecation($prefix . $errorString);
-    }
-
     // If the deprecation handled is one of those in the ignore list, we keep
     // running.
     if ((E_USER_DEPRECATED === $errorNumber || E_DEPRECATED === $errorNumber) && DeprecationHandler::isIgnoredDeprecation($errorString)) {
diff --git a/core/tests/Drupal/TestTools/ErrorHandler/TestErrorHandler.php b/core/tests/Drupal/TestTools/ErrorHandler/TestErrorHandler.php
deleted file mode 100644
index 8ad1de3f634..00000000000
--- a/core/tests/Drupal/TestTools/ErrorHandler/TestErrorHandler.php
+++ /dev/null
@@ -1,75 +0,0 @@
-<?php
-
-declare(strict_types=1);
-
-namespace Drupal\TestTools\ErrorHandler;
-
-use Drupal\TestTools\Extension\DeprecationBridge\DeprecationHandler;
-use PHPUnit\Framework\TestCase;
-
-/**
- * Drupal's PHPUnit test level error handler.
- *
- * This code works in coordination with DeprecationHandler.
- *
- * This error handler is registered during the preparation of a PHPUnit's test,
- * and is essentially used to capture deprecations occurring during test
- * executions. When test runs are torn down, the more generic
- * BootstrapErrorHandler is restored.
- *
- * @see \Drupal\TestTools\Extension\DeprecationBridge\DeprecationHandler
- *
- * @internal
- */
-final class TestErrorHandler {
-
-  /**
-   * @param \Drupal\TestTools\ErrorHandler\BootstrapErrorHandler $parentHandler
-   *   The parent error handler.
-   * @param \PHPUnit\Framework\TestCase $testCase
-   *   The test case being executed.
-   */
-  public function __construct(
-    private readonly BootstrapErrorHandler $parentHandler,
-    private readonly TestCase $testCase,
-  ) {
-  }
-
-  /**
-   * Executes when the object is called as a function.
-   *
-   * @param int $errorNumber
-   *   The level of the error raised.
-   * @param string $errorString
-   *   The error message.
-   * @param string $errorFile
-   *   The filename that the error was raised in.
-   * @param int $errorLine
-   *   The line number the error was raised at.
-   *
-   * @return bool
-   *   TRUE to stop error handling, FALSE to let the normal error handler
-   *   continue.
-   */
-  public function __invoke(int $errorNumber, string $errorString, string $errorFile, int $errorLine): bool {
-    if (!DeprecationHandler::isEnabled()) {
-      throw new \RuntimeException(__METHOD__ . '() must not be called if the deprecation handler is not enabled.');
-    }
-
-    // We are within a test execution. If we have a deprecation and the test is
-    // a deprecation test, than we just collect the deprecation and return to
-    // execution, since deprecations are expected.
-    if ((E_USER_DEPRECATED === $errorNumber || E_DEPRECATED === $errorNumber) && DeprecationHandler::isDeprecationTest($this->testCase)) {
-      $prefix = (error_reporting() & $errorNumber) ? 'Unsilenced deprecation: ' : '';
-      DeprecationHandler::collectActualDeprecation($prefix . $errorString);
-      return TRUE;
-    }
-
-    // In all other cases (errors, warnings, deprecations in normal tests), we
-    // fall back to the parent error handler, which is the one that was
-    // registered in the test runner bootstrap (BootstrapErrorHandler).
-    call_user_func($this->parentHandler, $errorNumber, $errorString, $errorFile, $errorLine);
-    return TRUE;
-  }
-
-}
diff --git a/core/tests/Drupal/TestTools/Extension/DeprecationBridge/DeprecationHandler.php b/core/tests/Drupal/TestTools/Extension/DeprecationBridge/DeprecationHandler.php
index f66ef7c8c94..59d7f1e4d3a 100644
--- a/core/tests/Drupal/TestTools/Extension/DeprecationBridge/DeprecationHandler.php
+++ b/core/tests/Drupal/TestTools/Extension/DeprecationBridge/DeprecationHandler.php
@@ -4,14 +4,12 @@
 
 namespace Drupal\TestTools\Extension\DeprecationBridge;
 
-use PHPUnit\Framework\TestCase;
-
 /**
  * Drupal's PHPUnit extension to manage code deprecation.
  *
- * This class is a replacement for symfony/phpunit-bridge that does not
- * support PHPUnit 10. In the future this extension might be dropped if
- * PHPUnit adds support for all deprecation management needs.
+ * This class is a replacement for symfony/phpunit-bridge that is not
+ * supporting PHPUnit 10 and above. In the future this extension might be
+ * dropped if PHPUnit will support all deprecation management needs.
  *
  * @internal
  */
@@ -29,20 +27,6 @@ final class DeprecationHandler {
    */
   private static array $deprecationIgnorePatterns = [];
 
-  /**
-   * A list of expected deprecation messages.
-   *
-   * @var list<string>
-   */
-  private static array $expectedDeprecations = [];
-
-  /**
-   * A list of deprecation messages collected during test run.
-   *
-   * @var list<string>
-   */
-  private static array $collectedDeprecations = [];
-
   /**
    * This class should not be instantiated.
    */
@@ -130,80 +114,6 @@ public static function init(?string $ignoreFile = NULL): void {
     self::$enabled = TRUE;
   }
 
-  /**
-   * Resets the extension.
-   *
-   * The extension should be reset at the beginning of each test run to ensure
-   * matching of expected and actual deprecations.
-   */
-  public static function reset(): void {
-    if (!self::isEnabled()) {
-      return;
-    }
-    self::$expectedDeprecations = [];
-    self::$collectedDeprecations = [];
-  }
-
-  /**
-   * Adds an expected deprecation.
-   *
-   * Tests will expect deprecations during the test execution; at the end of
-   * each test run, collected deprecations are checked against the expected
-   * ones.
-   *
-   * @param string $message
-   *   The expected deprecation message.
-   */
-  public static function expectDeprecation(string $message): void {
-    if (!self::isEnabled()) {
-      return;
-    }
-    self::$expectedDeprecations[] = $message;
-  }
-
-  /**
-   * Returns all expected deprecations.
-   *
-   * @return list<string>
-   *   The expected deprecation messages.
-   */
-  public static function getExpectedDeprecations(): array {
-    if (!self::isEnabled()) {
-      throw new \LogicException(__CLASS__ . ' is not initialized');
-    }
-    return self::$expectedDeprecations;
-  }
-
-  /**
-   * Collects an actual deprecation.
-   *
-   * Tests will expect deprecations during the test execution; at the end of
-   * each test run, collected deprecations are checked against the expected
-   * ones.
-   *
-   * @param string $message
-   *   The actual deprecation message triggered via trigger_error().
-   */
-  public static function collectActualDeprecation(string $message): void {
-    if (!self::isEnabled()) {
-      return;
-    }
-    self::$collectedDeprecations[] = $message;
-  }
-
-  /**
-   * Returns all collected deprecations.
-   *
-   * @return list<string>
-   *   The collected deprecation messages.
-   */
-  public static function getCollectedDeprecations(): array {
-    if (!self::isEnabled()) {
-      throw new \LogicException(__CLASS__ . ' is not initialized');
-    }
-    return self::$collectedDeprecations;
-  }
-
   /**
    * Determines if an actual deprecation should be ignored.
    *
@@ -224,31 +134,4 @@ public static function isIgnoredDeprecation(string $deprecationMessage): bool {
     return (bool) $result;
   }
 
-  /**
-   * Determines if a test case is a deprecation test.
-   *
-   * Deprecation tests are those that are annotated with '@group legacy' or
-   * that have a '#[IgnoreDeprecations]' attribute.
-   *
-   * @param \PHPUnit\Framework\TestCase $testCase
-   *   The test case being executed.
-   */
-  public static function isDeprecationTest(TestCase $testCase): bool {
-    return $testCase->valueObjectForEvents()->metadata()->isIgnoreDeprecations()->isNotEmpty() || self::isTestInLegacyGroup($testCase);
-  }
-
-  /**
-   * Determines if a test case is part of the 'legacy' group.
-   *
-   * @param \PHPUnit\Framework\TestCase $testCase
-   *   The test case being executed.
-   */
-  private static function isTestInLegacyGroup(TestCase $testCase): bool {
-    $groups = [];
-    foreach ($testCase->valueObjectForEvents()->metadata()->isGroup() as $metadata) {
-      $groups[] = $metadata->groupName();
-    }
-    return in_array('legacy', $groups, TRUE);
-  }
-
 }
diff --git a/core/tests/Drupal/TestTools/Extension/DeprecationBridge/ExpectDeprecationTrait.php b/core/tests/Drupal/TestTools/Extension/DeprecationBridge/ExpectDeprecationTrait.php
index ed73ca8fd93..e93793cfb31 100644
--- a/core/tests/Drupal/TestTools/Extension/DeprecationBridge/ExpectDeprecationTrait.php
+++ b/core/tests/Drupal/TestTools/Extension/DeprecationBridge/ExpectDeprecationTrait.php
@@ -4,9 +4,7 @@
 
 namespace Drupal\TestTools\Extension\DeprecationBridge;
 
-use Drupal\TestTools\ErrorHandler\TestErrorHandler;
-use PHPUnit\Framework\Attributes\After;
-use PHPUnit\Framework\Attributes\Before;
+@trigger_error('\Drupal\TestTools\Extension\DeprecationBridge\ExpectDeprecationTrait is deprecated in drupal:11.3.0 and is removed from drupal:12.0.0. Use $this->expectUserDeprecationMessage() or $this->expectUserDeprecationMessageMatches() instead. See https://www.drupal.org/node/7654312', E_USER_DEPRECATED);
 
 /**
  * A trait to include in Drupal tests to manage expected deprecations.
@@ -14,82 +12,53 @@
  * This code works in coordination with DeprecationHandler.
  *
  * This trait is a replacement for symfony/phpunit-bridge that is not
- * supporting PHPUnit 10. In the future this extension might be dropped if
- * PHPUnit will support all deprecation management needs.
+ * supporting PHPUnit 10 and above. In the future this extension might be
+ * dropped if PHPUnit will support all deprecation management needs.
  *
- * @see \Drupal\TestTools\Extension\DeprecationBridge\DeprecationHandler
+ * @deprecated in drupal:11.3.0 and is removed from drupal:12.0.0. Use
+ *   $this->expectUserDeprecationMessage() or
+ *   $this->expectUserDeprecationMessageMatches() instead.
+ *
+ * @see https://www.drupal.org/node/7654312
  *
  * @internal
  */
 trait ExpectDeprecationTrait {
 
-  /**
-   * Sets up the test error handler.
-   *
-   * This method is run before each test's ::setUp() method, and when the
-   * DeprecationHandler is active, resets the extension to be able to collect
-   * the test's deprecations, and sets TestErrorHandler as the current error
-   * handler.
-   *
-   * @see \Drupal\TestTools\ErrorHandler\TestErrorHandler
-   */
-  #[Before]
-  public function setUpErrorHandler(): void {
-    if (!DeprecationHandler::isEnabled()) {
-      return;
-    }
-
-    DeprecationHandler::reset();
-    set_error_handler(new TestErrorHandler(get_error_handler(), $this));
-  }
-
-  /**
-   * Tears down the test error handler.
-   *
-   * This method is run after each test's ::tearDown() method, and checks if
-   * collected deprecations match the expectations; it also resets the error
-   * handler to the one set prior of the change made by ::setUpErrorHandler().
-   */
-  #[After]
-  public function tearDownErrorHandler(): void {
-    if (!DeprecationHandler::isEnabled()) {
-      return;
-    }
-
-    // We expect that the current error handler is the one set by
-    // ::setUpErrorHandler() prior to the start of the test execution. If not,
-    // the error handler was changed during the test execution but not properly
-    // restored during ::tearDown().
-    if (!get_error_handler() instanceof TestErrorHandler) {
-      throw new \RuntimeException(sprintf('%s registered its own error handler without restoring the previous one before or during tear down. This can cause unpredictable test results. Ensure the test cleans up after itself.', $this->name()));
-    }
-    restore_error_handler();
-
-    // Checks if collected deprecations match the expectations.
-    if (DeprecationHandler::getExpectedDeprecations()) {
-      $prefix = "@expectedDeprecation:\n";
-      $expDep = $prefix . '%A  ' . implode("\n%A  ", DeprecationHandler::getExpectedDeprecations()) . "\n%A";
-      $actDep = $prefix . '  ' . implode("\n  ", DeprecationHandler::getCollectedDeprecations()) . "\n";
-      $this->assertStringMatchesFormat($expDep, $actDep);
-    }
-  }
-
   /**
    * Adds an expected deprecation.
    *
    * @param string $message
    *   The expected deprecation message.
+   *
+   * @deprecated in drupal:11.3.0 and is removed from drupal:12.0.0. Use
+   *   $this->expectUserDeprecationMessage() or
+   *   $this->expectUserDeprecationMessageMatches() instead.
+   *
+   * @see https://www.drupal.org/node/7654312
    */
   public function expectDeprecation(string $message): void {
-    if (!DeprecationHandler::isDeprecationTest($this)) {
-      throw new \RuntimeException('expectDeprecation() can only be called from tests marked with #[IgnoreDeprecations] or \'@group legacy\'');
-    }
-
-    if (!DeprecationHandler::isEnabled()) {
-      return;
-    }
+    @trigger_error('expectDeprecation() is deprecated in drupal:11.3.0 and is removed from drupal:12.0.0. Use $this->expectUserDeprecationMessage() or $this->expectUserDeprecationMessageMatches() instead. See https://www.drupal.org/node/7654312', E_USER_DEPRECATED);
+    $this->expectUserDeprecationMessageMatches($this->regularExpressionForFormatDescription('%A' . $message . '%A'));
+  }
 
-    DeprecationHandler::expectDeprecation($message);
+  private function regularExpressionForFormatDescription(string $string): string {
+    $string = strtr(preg_quote($string, '/'), [
+      '%%' => '%',
+      '%e' => preg_quote(\DIRECTORY_SEPARATOR, '/'),
+      '%s' => '[^\r\n]+',
+      '%S' => '[^\r\n]*',
+      '%a' => '.+?',
+      '%A' => '.*?',
+      '%w' => '\s*',
+      '%i' => '[+-]?\d+',
+      '%d' => '\d+',
+      '%x' => '[0-9a-fA-F]+',
+      '%f' => '[+-]?(?:\d+|(?=\.\d))(?:\.\d+)?(?:[Ee][+-]?\d+)?',
+      '%c' => '.',
+      '%0' => '\x00',
+    ]);
+    return '/^' . $string . '$/s';
   }
 
 }
diff --git a/core/tests/Drupal/Tests/BrowserTestBase.php b/core/tests/Drupal/Tests/BrowserTestBase.php
index 98aafad7df1..c26011b692e 100644
--- a/core/tests/Drupal/Tests/BrowserTestBase.php
+++ b/core/tests/Drupal/Tests/BrowserTestBase.php
@@ -21,6 +21,8 @@
 use Drupal\Tests\node\Traits\NodeCreationTrait;
 use Drupal\Tests\user\Traits\UserCreationTrait;
 use Drupal\TestTools\Comparator\MarkupInterfaceComparator;
+use Drupal\TestTools\ErrorHandler\BootstrapErrorHandler;
+use Drupal\TestTools\Extension\DeprecationBridge\DeprecationHandler;
 use Drupal\TestTools\Extension\DeprecationBridge\ExpectDeprecationTrait;
 use Drupal\TestTools\Extension\Dump\DebugDump;
 use GuzzleHttp\Cookie\CookieJar;
@@ -436,6 +438,13 @@ protected function tearDown(): void {
       $callbacks = &drupal_register_shutdown_function();
       $callbacks = $this->originalShutdownCallbacks;
     }
+
+    // We expect that the current error handler is the one set during the
+    // PHPUnit bootstrap. If not, the error handler was changed during the test
+    // execution but not properly restored during ::tearDown().
+    if (DeprecationHandler::isEnabled() && !get_error_handler() instanceof BootstrapErrorHandler) {
+      throw new \RuntimeException(sprintf('%s registered its own error handler without restoring the previous one before or during tear down. This can cause unpredictable test results. Ensure the test cleans up after itself.', $this->name()));
+    }
   }
 
   /**
diff --git a/core/tests/Drupal/Tests/Component/Assertion/InspectorTest.php b/core/tests/Drupal/Tests/Component/Assertion/InspectorTest.php
index 2d02f97c571..86bbc4c18fe 100644
--- a/core/tests/Drupal/Tests/Component/Assertion/InspectorTest.php
+++ b/core/tests/Drupal/Tests/Component/Assertion/InspectorTest.php
@@ -5,7 +5,6 @@
 namespace Drupal\Tests\Component\Assertion;
 
 use Drupal\Component\Assertion\Inspector;
-use Drupal\TestTools\Extension\DeprecationBridge\ExpectDeprecationTrait;
 use PHPUnit\Framework\Attributes\CoversClass;
 use PHPUnit\Framework\Attributes\DataProvider;
 use PHPUnit\Framework\Attributes\Group;
@@ -18,8 +17,6 @@
 #[Group('Assertion')]
 class InspectorTest extends TestCase {
 
-  use ExpectDeprecationTrait;
-
   /**
    * Tests asserting all members are strings.
    *
diff --git a/core/tests/Drupal/Tests/Component/DependencyInjection/ContainerTest.php b/core/tests/Drupal/Tests/Component/DependencyInjection/ContainerTest.php
index 87cd766c288..792da8c6a96 100644
--- a/core/tests/Drupal/Tests/Component/DependencyInjection/ContainerTest.php
+++ b/core/tests/Drupal/Tests/Component/DependencyInjection/ContainerTest.php
@@ -6,7 +6,6 @@
 
 use Drupal\Component\DependencyInjection\Container;
 use Drupal\Component\Utility\Crypt;
-use Drupal\TestTools\Extension\DeprecationBridge\ExpectDeprecationTrait;
 use PHPUnit\Framework\Attributes\CoversClass;
 use PHPUnit\Framework\Attributes\Group;
 use PHPUnit\Framework\TestCase;
@@ -26,7 +25,7 @@
 #[CoversClass(Container::class)]
 #[Group('DependencyInjection')]
 class ContainerTest extends TestCase {
-  use ExpectDeprecationTrait;
+
   use ProphecyTrait;
 
   /**
diff --git a/core/tests/Drupal/Tests/Component/DependencyInjection/Dumper/OptimizedPhpArrayDumperTest.php b/core/tests/Drupal/Tests/Component/DependencyInjection/Dumper/OptimizedPhpArrayDumperTest.php
index 0465d11b247..4d907872bdd 100644
--- a/core/tests/Drupal/Tests/Component/DependencyInjection/Dumper/OptimizedPhpArrayDumperTest.php
+++ b/core/tests/Drupal/Tests/Component/DependencyInjection/Dumper/OptimizedPhpArrayDumperTest.php
@@ -6,7 +6,6 @@
 
 use Drupal\Component\DependencyInjection\Dumper\OptimizedPhpArrayDumper;
 use Drupal\Component\Utility\Crypt;
-use Drupal\TestTools\Extension\DeprecationBridge\ExpectDeprecationTrait;
 use PHPUnit\Framework\Attributes\CoversClass;
 use PHPUnit\Framework\Attributes\DataProvider;
 use PHPUnit\Framework\Attributes\Group;
@@ -31,7 +30,6 @@
 #[Group('DependencyInjection')]
 class OptimizedPhpArrayDumperTest extends TestCase {
 
-  use ExpectDeprecationTrait;
   use ProphecyTrait;
 
   /**
diff --git a/core/tests/Drupal/Tests/Component/PhpStorage/FileStorageReadOnlyTest.php b/core/tests/Drupal/Tests/Component/PhpStorage/FileStorageReadOnlyTest.php
index e1c05c7111c..66f6bab4081 100644
--- a/core/tests/Drupal/Tests/Component/PhpStorage/FileStorageReadOnlyTest.php
+++ b/core/tests/Drupal/Tests/Component/PhpStorage/FileStorageReadOnlyTest.php
@@ -7,7 +7,6 @@
 use Drupal\Component\PhpStorage\FileReadOnlyStorage;
 use Drupal\Component\PhpStorage\FileStorage;
 use Drupal\Component\Utility\Random;
-use Drupal\TestTools\Extension\DeprecationBridge\ExpectDeprecationTrait;
 use PHPUnit\Framework\Attributes\CoversClass;
 use PHPUnit\Framework\Attributes\Group;
 
@@ -19,8 +18,6 @@
 #[Group('PhpStorage')]
 class FileStorageReadOnlyTest extends PhpStorageTestBase {
 
-  use ExpectDeprecationTrait;
-
   /**
    * Standard test settings to pass to storage instances.
    *
diff --git a/core/tests/Drupal/Tests/Component/PhpStorage/FileStorageTest.php b/core/tests/Drupal/Tests/Component/PhpStorage/FileStorageTest.php
index e24013cfcc7..429dfd03c62 100644
--- a/core/tests/Drupal/Tests/Component/PhpStorage/FileStorageTest.php
+++ b/core/tests/Drupal/Tests/Component/PhpStorage/FileStorageTest.php
@@ -6,7 +6,6 @@
 
 use Drupal\Component\PhpStorage\FileStorage;
 use Drupal\Component\Utility\Random;
-use Drupal\TestTools\Extension\DeprecationBridge\ExpectDeprecationTrait;
 use org\bovigo\vfs\vfsStreamDirectory;
 use PHPUnit\Framework\Attributes\CoversClass;
 use PHPUnit\Framework\Attributes\Group;
@@ -19,8 +18,6 @@
 #[Group('PhpStorage')]
 class FileStorageTest extends PhpStorageTestBase {
 
-  use ExpectDeprecationTrait;
-
   /**
    * Standard test settings to pass to storage instances.
    *
diff --git a/core/tests/Drupal/Tests/Component/Render/FormattableMarkupTest.php b/core/tests/Drupal/Tests/Component/Render/FormattableMarkupTest.php
index 4547de6b9d3..a93c603801c 100644
--- a/core/tests/Drupal/Tests/Component/Render/FormattableMarkupTest.php
+++ b/core/tests/Drupal/Tests/Component/Render/FormattableMarkupTest.php
@@ -5,7 +5,6 @@
 namespace Drupal\Tests\Component\Render;
 
 use Drupal\Component\Render\FormattableMarkup;
-use Drupal\TestTools\Extension\DeprecationBridge\ExpectDeprecationTrait;
 use PHPUnit\Framework\Attributes\CoversClass;
 use PHPUnit\Framework\Attributes\DataProvider;
 use PHPUnit\Framework\Attributes\Group;
@@ -18,8 +17,6 @@
 #[Group('utility')]
 class FormattableMarkupTest extends TestCase {
 
-  use ExpectDeprecationTrait;
-
   /**
    * The error message of the last error in the error handler.
    *
diff --git a/core/tests/Drupal/Tests/Component/Utility/BytesTest.php b/core/tests/Drupal/Tests/Component/Utility/BytesTest.php
index 3a7d57caa16..75ab2d28a16 100644
--- a/core/tests/Drupal/Tests/Component/Utility/BytesTest.php
+++ b/core/tests/Drupal/Tests/Component/Utility/BytesTest.php
@@ -5,7 +5,6 @@
 namespace Drupal\Tests\Component\Utility;
 
 use Drupal\Component\Utility\Bytes;
-use Drupal\TestTools\Extension\DeprecationBridge\ExpectDeprecationTrait;
 use PHPUnit\Framework\Attributes\CoversClass;
 use PHPUnit\Framework\Attributes\DataProvider;
 use PHPUnit\Framework\Attributes\Group;
@@ -21,7 +20,6 @@
 #[Group('Utility')]
 class BytesTest extends TestCase {
 
-  use ExpectDeprecationTrait;
   use ProphecyTrait;
 
   /**
diff --git a/core/tests/Drupal/Tests/Component/Utility/NumberTest.php b/core/tests/Drupal/Tests/Component/Utility/NumberTest.php
index fdac9baf796..bb41f644692 100644
--- a/core/tests/Drupal/Tests/Component/Utility/NumberTest.php
+++ b/core/tests/Drupal/Tests/Component/Utility/NumberTest.php
@@ -5,7 +5,6 @@
 namespace Drupal\Tests\Component\Utility;
 
 use Drupal\Component\Utility\Number;
-use Drupal\TestTools\Extension\DeprecationBridge\ExpectDeprecationTrait;
 use PHPUnit\Framework\Attributes\CoversClass;
 use PHPUnit\Framework\Attributes\DataProvider;
 use PHPUnit\Framework\Attributes\Group;
@@ -21,8 +20,6 @@
 #[Group('Utility')]
 class NumberTest extends TestCase {
 
-  use ExpectDeprecationTrait;
-
   /**
    * Tests Number::validStep() without offset.
    *
@@ -187,9 +184,9 @@ public function testAlphadecimalToIntThrowsExceptionWithMalformedStrings(): void
   #[IgnoreDeprecations]
   public function testAlphadecimalToIntReturnsZeroWithNullAndEmptyString(): void {
     $deprecationMessage = 'Passing NULL or an empty string to Drupal\Component\Utility\Number::alphadecimalToInt() is deprecated in drupal:11.2.0 and will be removed in drupal:12.0.0. See https://www.drupal.org/node/3494472';
-    $this->expectDeprecation($deprecationMessage);
+    $this->expectUserDeprecationMessage($deprecationMessage);
     $this->assertSame(0, Number::alphadecimalToInt(NULL));
-    $this->expectDeprecation($deprecationMessage);
+    $this->expectUserDeprecationMessage($deprecationMessage);
     $this->assertSame(0, Number::alphadecimalToInt(''));
   }
 
diff --git a/core/tests/Drupal/Tests/Component/Utility/UnicodeTest.php b/core/tests/Drupal/Tests/Component/Utility/UnicodeTest.php
index 06e6e1ef7d4..116a1d297fb 100644
--- a/core/tests/Drupal/Tests/Component/Utility/UnicodeTest.php
+++ b/core/tests/Drupal/Tests/Component/Utility/UnicodeTest.php
@@ -5,7 +5,6 @@
 namespace Drupal\Tests\Component\Utility;
 
 use Drupal\Component\Utility\Unicode;
-use Drupal\TestTools\Extension\DeprecationBridge\ExpectDeprecationTrait;
 use PHPUnit\Framework\Attributes\CoversClass;
 use PHPUnit\Framework\Attributes\DataProvider;
 use PHPUnit\Framework\Attributes\Group;
@@ -17,7 +16,6 @@
 #[CoversClass(Unicode::class)]
 #[Group('Utility')]
 class UnicodeTest extends TestCase {
-  use ExpectDeprecationTrait;
 
   /**
    * Tests multibyte ucfirst.
diff --git a/core/tests/Drupal/Tests/Core/EventSubscriber/CustomPageExceptionHtmlSubscriberTest.php b/core/tests/Drupal/Tests/Core/EventSubscriber/CustomPageExceptionHtmlSubscriberTest.php
index 0e4d108f0fc..91efc74602a 100644
--- a/core/tests/Drupal/Tests/Core/EventSubscriber/CustomPageExceptionHtmlSubscriberTest.php
+++ b/core/tests/Drupal/Tests/Core/EventSubscriber/CustomPageExceptionHtmlSubscriberTest.php
@@ -129,6 +129,7 @@ protected function setUp(): void {
    */
   protected function tearDown(): void {
     ini_set('error_log', $this->errorLog);
+    parent::tearDown();
   }
 
   /**
diff --git a/core/tests/Drupal/Tests/Core/Form/FormTestBase.php b/core/tests/Drupal/Tests/Core/Form/FormTestBase.php
index ffbff566382..2a594a700ad 100644
--- a/core/tests/Drupal/Tests/Core/Form/FormTestBase.php
+++ b/core/tests/Drupal/Tests/Core/Form/FormTestBase.php
@@ -205,6 +205,7 @@ protected function setUp(): void {
   protected function tearDown(): void {
     Html::resetSeenIds();
     (new FormState())->clearErrors();
+    parent::tearDown();
   }
 
   /**
diff --git a/core/tests/Drupal/Tests/Core/Test/PhpUnitBridgeIsolatedTest.php b/core/tests/Drupal/Tests/Core/Test/PhpUnitBridgeIsolatedTest.php
index 384f2a2d549..a04e726ca49 100644
--- a/core/tests/Drupal/Tests/Core/Test/PhpUnitBridgeIsolatedTest.php
+++ b/core/tests/Drupal/Tests/Core/Test/PhpUnitBridgeIsolatedTest.php
@@ -21,7 +21,7 @@
 class PhpUnitBridgeIsolatedTest extends UnitTestCase {
 
   public function testDeprecatedClass(): void {
-    $this->expectDeprecation('Drupal\deprecation_test\Deprecation\FixtureDeprecatedClass is deprecated.');
+    $this->expectUserDeprecationMessage('Drupal\deprecation_test\Deprecation\FixtureDeprecatedClass is deprecated.');
     $deprecated = new FixtureDeprecatedClass();
     $this->assertEquals('test', $deprecated->testFunction());
   }
diff --git a/core/tests/Drupal/Tests/Core/Test/PhpUnitBridgeTest.php b/core/tests/Drupal/Tests/Core/Test/PhpUnitBridgeTest.php
index 46bf8e00853..dc741d8ba12 100644
--- a/core/tests/Drupal/Tests/Core/Test/PhpUnitBridgeTest.php
+++ b/core/tests/Drupal/Tests/Core/Test/PhpUnitBridgeTest.php
@@ -20,7 +20,7 @@ class PhpUnitBridgeTest extends UnitTestCase {
    * Tests class-level deprecation.
    */
   public function testDeprecatedClass(): void {
-    $this->expectDeprecation('Drupal\deprecation_test\Deprecation\FixtureDeprecatedClass is deprecated.');
+    $this->expectUserDeprecationMessage('Drupal\deprecation_test\Deprecation\FixtureDeprecatedClass is deprecated.');
     $deprecated = new FixtureDeprecatedClass();
     $this->assertEquals('test', $deprecated->testFunction());
   }
diff --git a/core/tests/Drupal/Tests/Core/Theme/Icon/IconPackManagerTest.php b/core/tests/Drupal/Tests/Core/Theme/Icon/IconPackManagerTest.php
deleted file mode 100644
index 79ecd9ae3df..00000000000
--- a/core/tests/Drupal/Tests/Core/Theme/Icon/IconPackManagerTest.php
+++ /dev/null
@@ -1,108 +0,0 @@
-<?php
-
-declare(strict_types=1);
-
-namespace Drupal\Tests\Core\Theme\Icon;
-
-use Drupal\Core\Theme\Icon\IconExtractorPluginManager;
-use Drupal\Core\Cache\CacheBackendInterface;
-use Drupal\Core\Extension\ModuleHandlerInterface;
-use Drupal\Core\Extension\ThemeHandlerInterface;
-use Drupal\Core\Theme\Icon\Plugin\IconPackManager;
-use Drupal\Core\Theme\Icon\IconCollector;
-use Drupal\Core\Theme\Icon\Exception\IconPackConfigErrorException;
-use Drupal\Tests\UnitTestCase;
-use PHPUnit\Framework\Attributes\CoversClass;
-use PHPUnit\Framework\Attributes\DataProvider;
-use PHPUnit\Framework\Attributes\Group;
-
-/**
- * Tests Drupal\Core\Theme\Icon\Plugin\IconPackManager.
- *
- * @group icon
- */
-#[CoversClass(IconPackManager::class)]
-#[Group('icon')]
-class IconPackManagerTest extends UnitTestCase {
-
-  /**
-   * Data provider for ::testLibraryName().
-   *
-   * @return array
-   *   Provide test data as:
-   *   - (string) Library name to test
-   *   - (bool) Flag for validity
-   */
-  public static function providerIconPackLibraryName(): array {
-    return [
-      'valid name with underscore' => ['my_theme/my_library', TRUE],
-      'valid name with hyphen' => ['my_theme/my-library', TRUE],
-      'valid name with hyphen and numbers' => ['my_theme/my-long-library_with_mixed-hyphens123', TRUE],
-      'invalid theme name case 1' => ['my-theme/my-library', FALSE],
-      'invalid theme name case 2' => ['my-theme/my_library', FALSE],
-      'Invalid library name case 1' => ['my_theme/my library', FALSE],
-      'Invalid library name special characters' => ['my-theme/$#random!', FALSE],
-    ];
-  }
-
-  /**
-   * Test the library names.
-   *
-   * @param string $name
-   *   The icon data.
-   * @param bool $is_valid
-   *   The result expected is valid or not.
-   */
-  #[DataProvider('providerIconPackLibraryName')]
-  public function testIconPackLibraryName(string $name, bool $is_valid): void {
-    // Minimal required values.
-    $definition = [
-      'id' => 'foo',
-      'provider' => 'icon_test',
-      'extractor' => 'bar',
-      'template' => '',
-      'library' => $name,
-    ];
-
-    // Mock all dependencies for IconPackManager.
-    $module_handler = $this->createMock(ModuleHandlerInterface::class);
-    $theme_handler = $this->createMock(ThemeHandlerInterface::class);
-    $cache_backend = $this->createMock(CacheBackendInterface::class);
-
-    // Add the IconExtractorPluginManager mock here.
-    $icon_extractor_manager = $this->createMock(IconExtractorPluginManager::class);
-
-    $icon_collector = $this->createMock(IconCollector::class);
-
-    // Create the IconPackManager instance and pass the manager with definition.
-    $iconPackManager = new IconPackManager(
-      $module_handler,
-      $theme_handler,
-      $cache_backend,
-      $icon_extractor_manager,
-      $icon_collector,
-      $this->root,
-    );
-
-    $iconPackManager->setValidator();
-
-    $reflection = new \ReflectionClass($iconPackManager);
-    $method = $reflection->getMethod('validateDefinition');
-    $method->setAccessible(TRUE);
-
-    try {
-      $result = $method->invoke($iconPackManager, $definition);
-      if ($is_valid) {
-        $this->assertTrue($result);
-      }
-    }
-    catch (\Exception $e) {
-      if (!$is_valid) {
-        $this->assertSame('icon_test:foo Error in definition `foo`:[library] Does not match the regex pattern ^\w+/[A-Za-z]+[\w-]*$', $e->getMessage());
-        $this->assertInstanceOf(IconPackConfigErrorException::class, $e);
-      }
-    }
-
-  }
-
-}
diff --git a/core/tests/Drupal/Tests/ExpectDeprecationTest.php b/core/tests/Drupal/Tests/ExpectDeprecationTest.php
index f377fa12cfa..dc05454d4c5 100644
--- a/core/tests/Drupal/Tests/ExpectDeprecationTest.php
+++ b/core/tests/Drupal/Tests/ExpectDeprecationTest.php
@@ -12,11 +12,12 @@
 use PHPUnit\Framework\TestCase;
 
 /**
- * Ensures Drupal has test coverage of Symfony's deprecation testing.
+ * Ensures test coverage for deprecation testing.
  */
 #[Group('Test')]
 #[IgnoreDeprecations]
 class ExpectDeprecationTest extends TestCase {
+
   use ExpectDeprecationTrait;
 
   /**
diff --git a/core/tests/Drupal/Tests/UnitTestCase.php b/core/tests/Drupal/Tests/UnitTestCase.php
index b279bd1ed25..d584547d730 100644
--- a/core/tests/Drupal/Tests/UnitTestCase.php
+++ b/core/tests/Drupal/Tests/UnitTestCase.php
@@ -10,6 +10,8 @@
 use Drupal\Core\DependencyInjection\ContainerBuilder;
 use Drupal\Core\StringTranslation\TranslatableMarkup;
 use Drupal\Core\StringTranslation\PluralTranslatableMarkup;
+use Drupal\TestTools\ErrorHandler\BootstrapErrorHandler;
+use Drupal\TestTools\Extension\DeprecationBridge\DeprecationHandler;
 use Drupal\TestTools\Extension\DeprecationBridge\ExpectDeprecationTrait;
 use Drupal\TestTools\Extension\Dump\DebugDump;
 use PHPUnit\Framework\Attributes\BeforeClass;
@@ -77,6 +79,20 @@ protected function setUp(): void {
     chdir($this->root);
   }
 
+  /**
+   * {@inheritdoc}
+   */
+  protected function tearDown(): void {
+    // We expect that the current error handler is the one set during the
+    // PHPUnit bootstrap. If not, the error handler was changed during the test
+    // execution but not properly restored during ::tearDown().
+    if (DeprecationHandler::isEnabled() && !get_error_handler() instanceof BootstrapErrorHandler) {
+      throw new \RuntimeException(sprintf('%s registered its own error handler without restoring the previous one before or during tear down. This can cause unpredictable test results. Ensure the test cleans up after itself.', $this->name()));
+    }
+
+    parent::tearDown();
+  }
+
   /**
    * Returns a stub config factory that behaves according to the passed array.
    *
