diff --git a/core/lib/Drupal/Core/Test/TestDatabase.php b/core/lib/Drupal/Core/Test/TestDatabase.php
index 20cc8e6687..fcc1e42747 100644
--- a/core/lib/Drupal/Core/Test/TestDatabase.php
+++ b/core/lib/Drupal/Core/Test/TestDatabase.php
@@ -261,6 +261,9 @@ public static function lastTestGet($test_id) {
         ':test_id' => $test_id,
       ])
       ->fetchField();
+dump([$last_prefix, $last_test_class]);
+    $res = $connection->select('simpletest', 'st')->fields('st', 'test_class')->execute()->fetchAll();
+dump($res);
     return [$last_prefix, $last_test_class];
   }
 
diff --git a/core/modules/image/tests/src/Unit/ImageStyleTest.php b/core/modules/image/tests/src/Unit/ImageStyleTest.php
index 27acd46e0e..7ae7f3049f 100644
--- a/core/modules/image/tests/src/Unit/ImageStyleTest.php
+++ b/core/modules/image/tests/src/Unit/ImageStyleTest.php
@@ -10,7 +10,7 @@
  *
  * @group Image
  */
-class ImageStyleTest extends UnitTestCase {
+class ImageStyleTest extends UnitTestCase 
 
   /**
    * The entity type used for testing.
