diff --git a/core/lib/Drupal/Core/TypedData/Validation/RecursiveContextualValidator.php b/core/lib/Drupal/Core/TypedData/Validation/RecursiveContextualValidator.php
index f0b885a10c..792ed88097 100644
--- a/core/lib/Drupal/Core/TypedData/Validation/RecursiveContextualValidator.php
+++ b/core/lib/Drupal/Core/TypedData/Validation/RecursiveContextualValidator.php
@@ -177,6 +177,7 @@ protected function validateNode(TypedDataInterface $data, $constraints = NULL, $
    *   The constraints which should be ensured for the given value.
    */
   protected function validateConstraints($value, $cache_key, $constraints) {
+dump([__METHOD__, $constraints]);
     foreach ($constraints as $constraint) {
       // Prevent duplicate validation of constraints, in the case
       // that constraints belong to multiple validated groups
diff --git a/core/lib/Drupal/Core/Validation/Plugin/Validation/Constraint/RangeConstraint.php b/core/lib/Drupal/Core/Validation/Plugin/Validation/Constraint/RangeConstraint.php
index 161aa10db2..dbf3c79e3c 100644
--- a/core/lib/Drupal/Core/Validation/Plugin/Validation/Constraint/RangeConstraint.php
+++ b/core/lib/Drupal/Core/Validation/Plugin/Validation/Constraint/RangeConstraint.php
@@ -19,7 +19,19 @@
  */
 class RangeConstraint extends Range {
 
-  public $minMessage = 'This value should be %limit or more.';
-  public $maxMessage = 'This value should be %limit or less.';
+  /**
+   * {@inheritdoc}
+   */
+  public function __construct(array $options = NULL) {
+dump($options);
+    if (isset($options['min']) && isset($options['max'])) {
+      $options['notInRangeMessage'] = $options['notInRangeMessage'] ?? 'This value should be between %min and %max.';
+    }
+    else {
+      $options['minMessage'] = $options['minMessage'] ?? 'This value should be %limit or more.';
+      $options['maxMessage'] = $options['maxMessage'] ?? 'This value should be %limit or less.';
+    }
+    parent::__construct($options);
+  }
 
 }
