diff --git a/core/modules/views/src/Plugin/views/query/Sql.php b/core/modules/views/src/Plugin/views/query/Sql.php
index 14b335b63f..8a33d3559d 100644
--- a/core/modules/views/src/Plugin/views/query/Sql.php
+++ b/core/modules/views/src/Plugin/views/query/Sql.php
@@ -1534,9 +1534,11 @@ public function execute(ViewExecutable $view) {
 
         // Setup the result row objects.
         $view->result = iterator_to_array($result);
+dump([__METHOD__, 'a', $view->result]);
         array_walk($view->result, function (ResultRow $row, $index) {
           $row->index = $index;
         });
+dump([__METHOD__, 'b', $view->result]);
 
         $view->pager->postExecute($view->result);
         $view->pager->updatePageInfo();
diff --git a/core/modules/views/src/ResultRow.php b/core/modules/views/src/ResultRow.php
index 2399f69983..4d715ceaf1 100644
--- a/core/modules/views/src/ResultRow.php
+++ b/core/modules/views/src/ResultRow.php
@@ -38,6 +38,7 @@ class ResultRow {
    *   (optional) An array of values to add as properties on the object.
    */
   public function __construct(array $values = []) {
+dump([__METHOD__, $this, $values]);
     foreach ($values as $key => $value) {
       $this->{$key} = $value;
     }
