diff --git a/core/modules/mysqli/src/Driver/Database/mysqli/Connection.php b/core/modules/mysqli/src/Driver/Database/mysqli/Connection.php
index e41df23075a..dc053a73d20 100644
--- a/core/modules/mysqli/src/Driver/Database/mysqli/Connection.php
+++ b/core/modules/mysqli/src/Driver/Database/mysqli/Connection.php
@@ -64,6 +64,7 @@ public static function open(array &$connection_options = []) {
     // Allow PDO options to be overridden.
     $connection_options += [
       'pdo' => [],
+      'use_prepared_statements' => FALSE,
     ];
 
     try {
diff --git a/core/modules/mysqli/src/Driver/Database/mysqli/Result.php b/core/modules/mysqli/src/Driver/Database/mysqli/Result.php
index 2c5e57c3aa8..2713236baa6 100644
--- a/core/modules/mysqli/src/Driver/Database/mysqli/Result.php
+++ b/core/modules/mysqli/src/Driver/Database/mysqli/Result.php
@@ -23,7 +23,7 @@ class Result extends ResultBase {
    *   The fetch mode.
    * @param array{class: class-string, constructor_args: list<mixed>, column: int, cursor_orientation?: int, cursor_offset?: int} $fetchOptions
    *   The fetch options.
-   * @param \mysqli_result|false $mysqliResult
+   * @param \mysqli_result|bool $mysqliResult
    *   The MySQLi result object.
    * @param \mysqli $mysqliConnection
    *   Client database connection object.
@@ -31,7 +31,7 @@ class Result extends ResultBase {
   public function __construct(
     FetchAs $fetchMode,
     array $fetchOptions,
-    protected readonly \mysqli_result|false $mysqliResult,
+    protected readonly \mysqli_result|bool $mysqliResult,
     protected readonly \mysqli $mysqliConnection,
   ) {
     parent::__construct($fetchMode, $fetchOptions);
diff --git a/core/modules/mysqli/src/Driver/Database/mysqli/Statement.php b/core/modules/mysqli/src/Driver/Database/mysqli/Statement.php
index f3b4346992d..ce7d349d4b5 100644
--- a/core/modules/mysqli/src/Driver/Database/mysqli/Statement.php
+++ b/core/modules/mysqli/src/Driver/Database/mysqli/Statement.php
@@ -91,7 +91,25 @@ public function execute($args = [], $options = []) {
         $converter = new NamedPlaceholderConverter();
         $converter->parse($this->queryString, $args);
         [$convertedQueryString, $args] = [$converter->getConvertedSQL(), $converter->getConvertedParameters()];
-        $this->clientStatement = $this->clientConnection->prepare($convertedQueryString);
+        if ($this->connection->getConnectionOptions()['use_prepared_statements']) {
+          $this->clientStatement = $this->clientConnection->prepare($convertedQueryString);
+        }
+        else {
+          $args = array_map(
+            fn (?string $arg): string => $arg === NULL ? 'NULL' : "'" . $this->clientConnection->real_escape_string($arg) . "'",
+            $args,
+          );
+          preg_match_all('/(?<!\?)\?(?!\?)/', $convertedQueryString, $matches, \PREG_OFFSET_CAPTURE);
+          $executableQueryString = '';
+          $iStart = 0;
+          $iArg = 0;
+          foreach ($matches[0] as $match) {
+            $iLength = $match[1] - $iStart;
+            $executableQueryString .= substr($convertedQueryString, $iStart, $iLength) . $args[$iArg++];
+            $iStart = $iStart + $iLength + 1;
+          }
+          $executableQueryString .= substr($convertedQueryString, $iStart);
+        }
       }
       else {
         // Transform the $args to positional.
@@ -104,11 +122,19 @@ public function execute($args = [], $options = []) {
 
       // In mysqli, the results of the statement execution are returned in a
       // different object than the statement itself.
-      $return = $this->getClientStatement()->execute($args);
+      if ($this->connection->getConnectionOptions()['use_prepared_statements']) {
+        $return = $this->getClientStatement()->execute($args);
+        $result = $this->getClientStatement()->get_result();
+      }
+      else {
+        assert(isset($executableQueryString));
+        $result = $this->clientConnection->query($executableQueryString);
+        $return = (bool) $result;
+      }
       $this->result = new Result(
         $this->fetchMode,
         $this->fetchOptions,
-        $this->getClientStatement()->get_result(),
+        $result,
         $this->clientConnection,
       );
       $this->markResultsetIterable($return);
diff --git a/core/modules/mysqli/tests/src/Kernel/mysqli/PreparedStatementOptionTest.php b/core/modules/mysqli/tests/src/Kernel/mysqli/PreparedStatementOptionTest.php
new file mode 100644
index 00000000000..64ee2723d79
--- /dev/null
+++ b/core/modules/mysqli/tests/src/Kernel/mysqli/PreparedStatementOptionTest.php
@@ -0,0 +1,37 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Drupal\Tests\mysqli\Kernel\mysqli;
+
+use Drupal\KernelTests\Core\Database\DriverSpecificTransactionTestBase;
+use PHPUnit\Framework\Attributes\Group;
+use PHPUnit\Framework\Attributes\RunTestsInSeparateProcesses;
+use PHPUnit\Framework\Attributes\TestWith;
+
+/**
+ * Tests for the use of prepared statements in mysqli.
+ */
+#[Group('Database')]
+#[RunTestsInSeparateProcesses]
+class PreparedStatementOptionTest extends DriverSpecificTransactionTestBase {
+
+  /**
+   * Tests for the use of prepared statements in mysqli.
+   *
+   * @param bool $usePreparedStatements
+   *   Whether or not to use prepared statements when querying.
+   */
+  #[TestWith([TRUE])]
+  #[TestWith([FALSE])]
+  public function testNoBlocksCreatedDuringConfigSync(bool $usePreparedStatements): void {
+    // Create an additional connection with the 'use_prepared_statements'
+    // option set to our testing.
+    $connectionInfo = Database::getConnectionInfo();
+    $connectionInfo['default']['use_prepared_statements'] = $usePreparedStatements;
+    Database::addConnectionInfo('default', 'test', $connectionInfo['default']);
+    $testConnection = Database::getConnection('test');
+    $this->assertSame($usePreparedStatements, $testConnection->getConnectionOptions()['use_prepared_statements']);
+  }
+
+}
diff --git a/core/tests/Drupal/FunctionalTests/Installer/InstallerTestBase.php b/core/tests/Drupal/FunctionalTests/Installer/InstallerTestBase.php
index 82c5d8e6b69..be346ca35b2 100644
--- a/core/tests/Drupal/FunctionalTests/Installer/InstallerTestBase.php
+++ b/core/tests/Drupal/FunctionalTests/Installer/InstallerTestBase.php
@@ -259,7 +259,10 @@ protected function setUpProfile() {
   protected function setUpSettings() {
     $parameters = $this->parameters['forms']['install_settings_form'];
     $driver = $parameters['driver'];
-    unset($parameters[$driver]['dependencies']);
+    unset(
+      $parameters[$driver]['dependencies'],
+      $parameters[$driver]['use_prepared_statements'],
+    );
     $edit = $this->translatePostValues($parameters);
     $this->submitForm($edit, $this->translations['Save and continue']);
   }
