diff --git a/core/lib/Drupal/Core/Database/Connection.php b/core/lib/Drupal/Core/Database/Connection.php
index a0d9fc67d1a..8abfdac57d5 100644
--- a/core/lib/Drupal/Core/Database/Connection.php
+++ b/core/lib/Drupal/Core/Database/Connection.php
@@ -1495,7 +1495,7 @@ public function isEventEnabled(string $eventName): bool {
   /**
    * Enables database API events dispatching.
    *
-   * @param string[] $eventNames
+   * @param list<class-string<\Drupal\Core\Database\Event\DatabaseEvent>> $eventNames
    *   A list of database events to be enabled.
    *
    * @return static
@@ -1503,6 +1503,7 @@ public function isEventEnabled(string $eventName): bool {
   public function enableEvents(array $eventNames): static {
     foreach ($eventNames as $eventName) {
       assert(class_exists($eventName), "Event class {$eventName} does not exist");
+      assert(is_a($eventName, DatabaseEvent::class, TRUE), "Event class {$eventName} is not a valid database event class");
       $this->enabledEvents[$eventName] = TRUE;
     }
     return $this;
@@ -1511,7 +1512,7 @@ public function enableEvents(array $eventNames): static {
   /**
    * Disables database API events dispatching.
    *
-   * @param string[] $eventNames
+   * @param list<class-string<\Drupal\Core\Database\Event\DatabaseEvent>> $eventNames
    *   A list of database events to be disabled.
    *
    * @return static
@@ -1519,6 +1520,7 @@ public function enableEvents(array $eventNames): static {
   public function disableEvents(array $eventNames): static {
     foreach ($eventNames as $eventName) {
       assert(class_exists($eventName), "Event class {$eventName} does not exist");
+      assert(is_a($eventName, DatabaseEvent::class, TRUE), "Event class {$eventName} is not a valid database event class");
       $this->enabledEvents[$eventName] = FALSE;
     }
     return $this;
diff --git a/core/lib/Drupal/Core/Database/Event/StatementExecutionEndEvent.php b/core/lib/Drupal/Core/Database/Event/StatementExecutionEndEvent.php
index 30718843d6c..4114ccbbc7d 100644
--- a/core/lib/Drupal/Core/Database/Event/StatementExecutionEndEvent.php
+++ b/core/lib/Drupal/Core/Database/Event/StatementExecutionEndEvent.php
@@ -28,6 +28,10 @@ class StatementExecutionEndEvent extends DatabaseEvent {
    *   called.
    * @param float $startTime
    *   The time of the statement execution start.
+   * @param string|null $transactionId
+   *   The current transaction id if active.
+   * @param string|null $transactionName
+   *   The current transaction name if active.
    */
   public function __construct(
     public readonly int $statementObjectId,
@@ -37,6 +41,8 @@ public function __construct(
     public readonly array $args,
     public readonly array $caller,
     public readonly float $startTime,
+    public readonly ?string $transactionId = NULL,
+    public readonly ?string $transactionName = NULL,
   ) {
     parent::__construct();
   }
diff --git a/core/lib/Drupal/Core/Database/Event/StatementExecutionFailureEvent.php b/core/lib/Drupal/Core/Database/Event/StatementExecutionFailureEvent.php
index 682e59f2389..b13f87fc1a3 100644
--- a/core/lib/Drupal/Core/Database/Event/StatementExecutionFailureEvent.php
+++ b/core/lib/Drupal/Core/Database/Event/StatementExecutionFailureEvent.php
@@ -34,6 +34,10 @@ class StatementExecutionFailureEvent extends StatementExecutionEndEvent {
    *   The code of the exception that was thrown.
    * @param string $exceptionMessage
    *   The message of the exception that was thrown.
+   * @param string|null $transactionId
+   *   The current transaction id if active.
+   * @param string|null $transactionName
+   *   The current transaction name if active.
    */
   public function __construct(
     int $statementObjectId,
@@ -46,8 +50,10 @@ public function __construct(
     public readonly string $exceptionClass,
     public readonly int|string $exceptionCode,
     public readonly string $exceptionMessage,
+    ?string $transactionId = NULL,
+    ?string $transactionName = NULL,
   ) {
-    parent::__construct($statementObjectId, $key, $target, $queryString, $args, $caller, $startTime);
+    parent::__construct($statementObjectId, $key, $target, $queryString, $args, $caller, $startTime, $transactionId, $transactionName);
   }
 
 }
diff --git a/core/lib/Drupal/Core/Database/Event/StatementExecutionStartEvent.php b/core/lib/Drupal/Core/Database/Event/StatementExecutionStartEvent.php
index 010f810966a..124142556ed 100644
--- a/core/lib/Drupal/Core/Database/Event/StatementExecutionStartEvent.php
+++ b/core/lib/Drupal/Core/Database/Event/StatementExecutionStartEvent.php
@@ -26,6 +26,10 @@ class StatementExecutionStartEvent extends DatabaseEvent {
    * @param array $caller
    *   A normalized debug backtrace entry representing the last non-db method
    *   called.
+   * @param string|null $transactionId
+   *   The current transaction id if active.
+   * @param string|null $transactionName
+   *   The current transaction name if active.
    */
   public function __construct(
     public readonly int $statementObjectId,
@@ -34,6 +38,8 @@ public function __construct(
     public readonly string $queryString,
     public readonly array $args,
     public readonly array $caller,
+    public readonly ?string $transactionId = NULL,
+    public readonly ?string $transactionName = NULL,
   ) {
     parent::__construct();
   }
diff --git a/core/lib/Drupal/Core/Database/Event/TransactionAutoCommitEvent.php b/core/lib/Drupal/Core/Database/Event/TransactionAutoCommitEvent.php
new file mode 100644
index 00000000000..922e579077c
--- /dev/null
+++ b/core/lib/Drupal/Core/Database/Event/TransactionAutoCommitEvent.php
@@ -0,0 +1,18 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Drupal\Core\Database\Event;
+
+/**
+ * Emitted when an automatic commit is detected.
+ *
+ * In some cases the active transaction is automatically committed by the
+ * database server when a DDL statement is executed during a transaction
+ * (for example, MySql). In such cases we emit this event to notify that a
+ * commit occurred while the Drupal transaction is still active.
+ * A normal TransactionCommitEvent is anyway emitted later once the Drupal
+ * transaction is committed.
+ */
+final class TransactionAutoCommitEvent extends TransactionEventBase {
+}
diff --git a/core/lib/Drupal/Core/Database/Event/TransactionBeginEvent.php b/core/lib/Drupal/Core/Database/Event/TransactionBeginEvent.php
new file mode 100644
index 00000000000..fd8801bdbd4
--- /dev/null
+++ b/core/lib/Drupal/Core/Database/Event/TransactionBeginEvent.php
@@ -0,0 +1,11 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Drupal\Core\Database\Event;
+
+/**
+ * Represents the beginning of a transaction.
+ */
+final class TransactionBeginEvent extends TransactionEventBase {
+}
diff --git a/core/lib/Drupal/Core/Database/Event/TransactionCommitEvent.php b/core/lib/Drupal/Core/Database/Event/TransactionCommitEvent.php
new file mode 100644
index 00000000000..e95cf6c57f5
--- /dev/null
+++ b/core/lib/Drupal/Core/Database/Event/TransactionCommitEvent.php
@@ -0,0 +1,11 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Drupal\Core\Database\Event;
+
+/**
+ * Represents the commit of a transaction.
+ */
+final class TransactionCommitEvent extends TransactionCommitOrReleaseEventBase {
+}
diff --git a/core/lib/Drupal/Core/Database/Event/TransactionCommitFailureEvent.php b/core/lib/Drupal/Core/Database/Event/TransactionCommitFailureEvent.php
new file mode 100644
index 00000000000..801f0d0bc4e
--- /dev/null
+++ b/core/lib/Drupal/Core/Database/Event/TransactionCommitFailureEvent.php
@@ -0,0 +1,11 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Drupal\Core\Database\Event;
+
+/**
+ * Represents the failed commit of a transaction.
+ */
+final class TransactionCommitFailureEvent extends TransactionCommitOrReleaseFailureEventBase {
+}
diff --git a/core/lib/Drupal/Core/Database/Event/TransactionCommitOrReleaseEventBase.php b/core/lib/Drupal/Core/Database/Event/TransactionCommitOrReleaseEventBase.php
new file mode 100644
index 00000000000..90c1147ca32
--- /dev/null
+++ b/core/lib/Drupal/Core/Database/Event/TransactionCommitOrReleaseEventBase.php
@@ -0,0 +1,48 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Drupal\Core\Database\Event;
+
+use Drupal\Core\Database\Transaction\ClientConnectionTransactionState;
+use Drupal\Core\Database\Transaction\CommitType;
+
+/**
+ * Represents the commit of a transaction.
+ */
+abstract class TransactionCommitOrReleaseEventBase extends TransactionEventBase {
+
+  /**
+   * Constructor.
+   *
+   * See 'Customizing database settings' in settings.php for an explanation of
+   * the $key and $target connection values.
+   *
+   * @param string $key
+   *   The database connection key.
+   * @param string $target
+   *   The database connection target.
+   * @param string $id
+   *   The id of the transaction.
+   * @param string $name
+   *   The name of the transaction.
+   * @param \Drupal\Core\Database\Transaction\ClientConnectionTransactionState $transactionState
+   *   The state of the client transaction.
+   * @param array $stackItems
+   *   The current transaction stack items.
+   * @param \Drupal\Core\Database\Transaction\CommitType $type
+   *   The type of commit.
+   */
+  public function __construct(
+    string $key,
+    string $target,
+    string $id,
+    string $name,
+    ClientConnectionTransactionState $transactionState,
+    array $stackItems,
+    public readonly CommitType $type,
+  ) {
+    parent::__construct($key, $target, $id, $name, $transactionState, $stackItems);
+  }
+
+}
diff --git a/core/lib/Drupal/Core/Database/Event/TransactionCommitOrReleaseFailureEventBase.php b/core/lib/Drupal/Core/Database/Event/TransactionCommitOrReleaseFailureEventBase.php
new file mode 100644
index 00000000000..6b140596d15
--- /dev/null
+++ b/core/lib/Drupal/Core/Database/Event/TransactionCommitOrReleaseFailureEventBase.php
@@ -0,0 +1,54 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Drupal\Core\Database\Event;
+
+use Drupal\Core\Database\Transaction\ClientConnectionTransactionState;
+use Drupal\Core\Database\Transaction\CommitType;
+
+/**
+ * Base class for database transaction commit failure events.
+ */
+abstract class TransactionCommitOrReleaseFailureEventBase extends TransactionFailureEventBase {
+
+  /**
+   * Constructor.
+   *
+   * See 'Customizing database settings' in settings.php for an explanation of
+   * the $key and $target connection values.
+   *
+   * @param string $key
+   *   The database connection key.
+   * @param string $target
+   *   The database connection target.
+   * @param string $id
+   *   The id of the transaction.
+   * @param string $name
+   *   The name of the transaction.
+   * @param \Drupal\Core\Database\Transaction\ClientConnectionTransactionState $transactionState
+   *   The state of the client transaction.
+   * @param array $stackItems
+   *   The current transaction stack items.
+   * @param class-string $className
+   *   The FQCN of the exception.
+   * @param string $message
+   *   The message of the exception.
+   * @param \Drupal\Core\Database\Transaction\CommitType $type
+   *   The type of commit.
+   */
+  public function __construct(
+    string $key,
+    string $target,
+    string $id,
+    string $name,
+    ClientConnectionTransactionState $transactionState,
+    array $stackItems,
+    string $className,
+    string $message,
+    public readonly CommitType $type,
+  ) {
+    parent::__construct($key, $target, $id, $name, $transactionState, $stackItems, $className, $message);
+  }
+
+}
diff --git a/core/lib/Drupal/Core/Database/Event/TransactionEvent.php b/core/lib/Drupal/Core/Database/Event/TransactionEvent.php
new file mode 100644
index 00000000000..b810b886f71
--- /dev/null
+++ b/core/lib/Drupal/Core/Database/Event/TransactionEvent.php
@@ -0,0 +1,35 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Drupal\Core\Database\Event;
+
+/**
+ * Enumeration of the transaction related database events.
+ */
+enum TransactionEvent: string {
+
+  case Begin = TransactionBeginEvent::class;
+  case Commit = TransactionCommitEvent::class;
+  case CommitFailure = TransactionCommitFailureEvent::class;
+  case ReleaseSavepoint = TransactionReleaseSavepointEvent::class;
+  case ReleaseSavepointFailure = TransactionReleaseSavepointFailureEvent::class;
+  case Rollback = TransactionRollbackEvent::class;
+  case RollbackFailure = TransactionRollbackFailureEvent::class;
+  case RollbackToSavepoint = TransactionRollbackToSavepointEvent::class;
+  case RollbackToSavepointFailure = TransactionRollbackToSavepointFailureEvent::class;
+  case Savepoint = TransactionSavepointEvent::class;
+
+  case AutoCommit = TransactionAutoCommitEvent::class;
+
+  /**
+   * Returns an array with all transaction related event classes.
+   *
+   * @return list<class-string<\Drupal\Core\Database\Event\TransactionEventBase>>
+   *   An array with all transaction related event classes.
+   */
+  public static function all(): array {
+    return array_map(fn(self $case) => $case->value, self::cases());
+  }
+
+}
diff --git a/core/lib/Drupal/Core/Database/Event/TransactionEventBase.php b/core/lib/Drupal/Core/Database/Event/TransactionEventBase.php
new file mode 100644
index 00000000000..0e41b9d95a8
--- /dev/null
+++ b/core/lib/Drupal/Core/Database/Event/TransactionEventBase.php
@@ -0,0 +1,44 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Drupal\Core\Database\Event;
+
+use Drupal\Core\Database\Transaction\ClientConnectionTransactionState;
+
+/**
+ * Base class for database transaction events.
+ */
+abstract class TransactionEventBase extends DatabaseEvent {
+
+  /**
+   * Constructor.
+   *
+   * See 'Customizing database settings' in settings.php for an explanation of
+   * the $key and $target connection values.
+   *
+   * @param string $key
+   *   The database connection key.
+   * @param string $target
+   *   The database connection target.
+   * @param string $id
+   *   The id of the transaction.
+   * @param string $name
+   *   The name of the transaction.
+   * @param \Drupal\Core\Database\Transaction\ClientConnectionTransactionState $transactionState
+   *   The state of the client transaction.
+   * @param array $stackItems
+   *   The current transaction stack items.
+   */
+  public function __construct(
+    public readonly string $key,
+    public readonly string $target,
+    public readonly string $id,
+    public readonly string $name,
+    public readonly ClientConnectionTransactionState $transactionState,
+    public readonly array $stackItems,
+  ) {
+    parent::__construct();
+  }
+
+}
diff --git a/core/lib/Drupal/Core/Database/Event/TransactionFailureEventBase.php b/core/lib/Drupal/Core/Database/Event/TransactionFailureEventBase.php
new file mode 100644
index 00000000000..bf04371a498
--- /dev/null
+++ b/core/lib/Drupal/Core/Database/Event/TransactionFailureEventBase.php
@@ -0,0 +1,50 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Drupal\Core\Database\Event;
+
+use Drupal\Core\Database\Transaction\ClientConnectionTransactionState;
+
+/**
+ * Base class for database transaction failure events.
+ */
+abstract class TransactionFailureEventBase extends TransactionEventBase {
+
+  /**
+   * Constructor.
+   *
+   * See 'Customizing database settings' in settings.php for an explanation of
+   * the $key and $target connection values.
+   *
+   * @param string $key
+   *   The database connection key.
+   * @param string $target
+   *   The database connection target.
+   * @param string $id
+   *   The id of the transaction.
+   * @param string $name
+   *   The name of the transaction.
+   * @param \Drupal\Core\Database\Transaction\ClientConnectionTransactionState $transactionState
+   *   The state of the client transaction.
+   * @param array $stackItems
+   *   The current transaction stack items.
+   * @param class-string $className
+   *   The FQCN of the exception.
+   * @param string $message
+   *   The message of the exception.
+   */
+  public function __construct(
+    string $key,
+    string $target,
+    string $id,
+    string $name,
+    ClientConnectionTransactionState $transactionState,
+    array $stackItems,
+    public readonly string $className,
+    public readonly string $message,
+  ) {
+    parent::__construct($key, $target, $id, $name, $transactionState, $stackItems);
+  }
+
+}
diff --git a/core/lib/Drupal/Core/Database/Event/TransactionReleaseSavepointEvent.php b/core/lib/Drupal/Core/Database/Event/TransactionReleaseSavepointEvent.php
new file mode 100644
index 00000000000..61d61710da2
--- /dev/null
+++ b/core/lib/Drupal/Core/Database/Event/TransactionReleaseSavepointEvent.php
@@ -0,0 +1,11 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Drupal\Core\Database\Event;
+
+/**
+ * Represents a savepoint release.
+ */
+final class TransactionReleaseSavepointEvent extends TransactionCommitOrReleaseEventBase {
+}
diff --git a/core/lib/Drupal/Core/Database/Event/TransactionReleaseSavepointFailureEvent.php b/core/lib/Drupal/Core/Database/Event/TransactionReleaseSavepointFailureEvent.php
new file mode 100644
index 00000000000..df0367bcfdc
--- /dev/null
+++ b/core/lib/Drupal/Core/Database/Event/TransactionReleaseSavepointFailureEvent.php
@@ -0,0 +1,11 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Drupal\Core\Database\Event;
+
+/**
+ * Represents a failed savepoint release.
+ */
+final class TransactionReleaseSavepointFailureEvent extends TransactionCommitOrReleaseFailureEventBase {
+}
diff --git a/core/lib/Drupal/Core/Database/Event/TransactionRollbackEvent.php b/core/lib/Drupal/Core/Database/Event/TransactionRollbackEvent.php
new file mode 100644
index 00000000000..3f64c0d0e89
--- /dev/null
+++ b/core/lib/Drupal/Core/Database/Event/TransactionRollbackEvent.php
@@ -0,0 +1,11 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Drupal\Core\Database\Event;
+
+/**
+ * Represents the rollback of a transaction.
+ */
+final class TransactionRollbackEvent extends TransactionEventBase {
+}
diff --git a/core/lib/Drupal/Core/Database/Event/TransactionRollbackFailureEvent.php b/core/lib/Drupal/Core/Database/Event/TransactionRollbackFailureEvent.php
new file mode 100644
index 00000000000..6b1c7d59cbd
--- /dev/null
+++ b/core/lib/Drupal/Core/Database/Event/TransactionRollbackFailureEvent.php
@@ -0,0 +1,11 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Drupal\Core\Database\Event;
+
+/**
+ * Represents the failed rollback of a transaction.
+ */
+final class TransactionRollbackFailureEvent extends TransactionFailureEventBase {
+}
diff --git a/core/lib/Drupal/Core/Database/Event/TransactionRollbackToSavepointEvent.php b/core/lib/Drupal/Core/Database/Event/TransactionRollbackToSavepointEvent.php
new file mode 100644
index 00000000000..2c5be72e257
--- /dev/null
+++ b/core/lib/Drupal/Core/Database/Event/TransactionRollbackToSavepointEvent.php
@@ -0,0 +1,11 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Drupal\Core\Database\Event;
+
+/**
+ * Represents the rollback of a transaction to a savepoint.
+ */
+final class TransactionRollbackToSavepointEvent extends TransactionEventBase {
+}
diff --git a/core/lib/Drupal/Core/Database/Event/TransactionRollbackToSavepointFailureEvent.php b/core/lib/Drupal/Core/Database/Event/TransactionRollbackToSavepointFailureEvent.php
new file mode 100644
index 00000000000..9c17ee84d68
--- /dev/null
+++ b/core/lib/Drupal/Core/Database/Event/TransactionRollbackToSavepointFailureEvent.php
@@ -0,0 +1,11 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Drupal\Core\Database\Event;
+
+/**
+ * Represents the failed rollback of a transaction to a savepoint.
+ */
+final class TransactionRollbackToSavepointFailureEvent extends TransactionFailureEventBase {
+}
diff --git a/core/lib/Drupal/Core/Database/Event/TransactionSavepointEvent.php b/core/lib/Drupal/Core/Database/Event/TransactionSavepointEvent.php
new file mode 100644
index 00000000000..b3109980fbb
--- /dev/null
+++ b/core/lib/Drupal/Core/Database/Event/TransactionSavepointEvent.php
@@ -0,0 +1,11 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Drupal\Core\Database\Event;
+
+/**
+ * Represents the activation of a transaction savepoint.
+ */
+final class TransactionSavepointEvent extends TransactionEventBase {
+}
diff --git a/core/lib/Drupal/Core/Database/Schema.php b/core/lib/Drupal/Core/Database/Schema.php
index 2f548ffb1d1..b07585813aa 100644
--- a/core/lib/Drupal/Core/Database/Schema.php
+++ b/core/lib/Drupal/Core/Database/Schema.php
@@ -138,7 +138,7 @@ protected function executeDdlStatement(string $sql, array $arguments = [], array
     // DDL statements when in a transaction force a commit in some databases.
     // Void the transaction in that case.
     if (!$this->connection->supportsTransactionalDDL() && $this->connection->transactionManager()->inTransaction()) {
-      $this->connection->transactionManager()->voidClientTransaction();
+      $this->connection->transactionManager()->handleAutomaticCommitUponDDL();
     }
   }
 
diff --git a/core/lib/Drupal/Core/Database/Statement/StatementBase.php b/core/lib/Drupal/Core/Database/Statement/StatementBase.php
index e266616975f..a7374310443 100644
--- a/core/lib/Drupal/Core/Database/Statement/StatementBase.php
+++ b/core/lib/Drupal/Core/Database/Statement/StatementBase.php
@@ -145,7 +145,9 @@ protected function dispatchStatementExecutionStartEvent(array $args): ?Statement
         $this->connection->getTarget(),
         $this->getQueryString(),
         $args,
-        $this->connection->findCallerFromDebugBacktrace()
+        $this->connection->findCallerFromDebugBacktrace(),
+        $this->connection->transactionManager()->currentTransactionId(),
+        $this->connection->transactionManager()->currentTransactionName(),
       );
       $this->connection->dispatchEvent($startEvent);
       return $startEvent;
@@ -168,7 +170,9 @@ protected function dispatchStatementExecutionEndEvent(?StatementExecutionStartEv
         $startEvent->queryString,
         $startEvent->args,
         $startEvent->caller,
-        $startEvent->time
+        $startEvent->time,
+        $startEvent->transactionId,
+        $startEvent->transactionName,
       ));
     }
   }
@@ -194,6 +198,8 @@ protected function dispatchStatementExecutionFailureEvent(?StatementExecutionSta
         get_class($e),
         $e->getCode(),
         $e->getMessage(),
+        $startEvent->transactionId,
+        $startEvent->transactionName,
       ));
     }
   }
diff --git a/core/lib/Drupal/Core/Database/Transaction/CommitType.php b/core/lib/Drupal/Core/Database/Transaction/CommitType.php
new file mode 100644
index 00000000000..cff631275cc
--- /dev/null
+++ b/core/lib/Drupal/Core/Database/Transaction/CommitType.php
@@ -0,0 +1,16 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Drupal\Core\Database\Transaction;
+
+/**
+ * Enumeration of the types of commit for Drupal transactions.
+ */
+enum CommitType {
+
+  case Explicit;
+  case ImplicitOnDestruct;
+  case AutomaticUponDDL;
+
+}
diff --git a/core/lib/Drupal/Core/Database/Transaction/TransactionManagerBase.php b/core/lib/Drupal/Core/Database/Transaction/TransactionManagerBase.php
index f379e770c15..e6530f1e04e 100644
--- a/core/lib/Drupal/Core/Database/Transaction/TransactionManagerBase.php
+++ b/core/lib/Drupal/Core/Database/Transaction/TransactionManagerBase.php
@@ -5,8 +5,21 @@
 namespace Drupal\Core\Database\Transaction;
 
 use Drupal\Core\Database\Connection;
+use Drupal\Core\Database\Event\TransactionAutoCommitEvent;
+use Drupal\Core\Database\Event\TransactionBeginEvent;
+use Drupal\Core\Database\Event\TransactionCommitEvent;
+use Drupal\Core\Database\Event\TransactionCommitFailureEvent;
+use Drupal\Core\Database\Event\TransactionEvent;
+use Drupal\Core\Database\Event\TransactionReleaseSavepointEvent;
+use Drupal\Core\Database\Event\TransactionReleaseSavepointFailureEvent;
+use Drupal\Core\Database\Event\TransactionRollbackEvent;
+use Drupal\Core\Database\Event\TransactionRollbackFailureEvent;
+use Drupal\Core\Database\Event\TransactionRollbackToSavepointEvent;
+use Drupal\Core\Database\Event\TransactionRollbackToSavepointFailureEvent;
+use Drupal\Core\Database\Event\TransactionSavepointEvent;
 use Drupal\Core\Database\Transaction;
 use Drupal\Core\Database\TransactionCommitFailedException;
+use Drupal\Core\Database\TransactionException;
 use Drupal\Core\Database\TransactionNameNonUniqueException;
 use Drupal\Core\Database\TransactionOutOfOrderException;
 
@@ -167,6 +180,7 @@ protected function stack(): array {
    *   changes the order of object destruction when the develop mode is enabled.
    */
   public function commitAll(): void {
+    $this->connection->disableEvents(TransactionEvent::all());
     foreach (array_reverse($this->stack()) as $id => $item) {
       $this->unpile($item->name, $id);
     }
@@ -233,12 +247,23 @@ protected function dumpStackItemsAsString(): string {
     if ($this->stack() === []) {
       return '*** empty ***';
     }
+    return implode(' > ', $this->stackItemsAsArray());
+  }
 
-    $temp = [];
-    foreach ($this->stack() as $id => $item) {
-      $temp[] = $id . '\\' . $item->name;
-    }
-    return implode(' > ', $temp);
+  /**
+   * Produces an array representation of the stack items.
+   *
+   * Drivers should not override this method unless they also override the
+   * $stack property.
+   *
+   * @return list<string>
+   *   The array of stack items represented like id\name.
+   */
+  protected function stackItemsAsArray(): array {
+    return array_map(fn(string $id, StackItem $item): string => $id . '\\' . $item->name,
+      array_keys($this->stack()),
+      array_values($this->stack()),
+    );
   }
 
   /**
@@ -248,6 +273,32 @@ public function inTransaction(): bool {
     return (bool) $this->stackDepth() && $this->getConnectionTransactionState() === ClientConnectionTransactionState::Active;
   }
 
+  /**
+   * Gets the current transaction id.
+   *
+   * @return string|null
+   *   The current transaction id.
+   */
+  public function currentTransactionId(): ?string {
+    if (!$this->inTransaction()) {
+      return NULL;
+    }
+    return array_key_last($this->stack);
+  }
+
+  /**
+   * Gets the current transaction name.
+   *
+   * @return string|null
+   *   The current transaction name.
+   */
+  public function currentTransactionName(): ?string {
+    if (!$this->inTransaction()) {
+      return NULL;
+    }
+    return end($this->stack)->name;
+  }
+
   /**
    * {@inheritdoc}
    */
@@ -273,25 +324,42 @@ public function push(string $name = ''): Transaction {
     // Do the client-level processing.
     if ($this->stackDepth() === 0) {
       $this->beginClientTransaction();
-      $type = StackItemType::Root;
       $this->setConnectionTransactionState(ClientConnectionTransactionState::Active);
+      $this->addStackItem($id, new StackItem($name, StackItemType::Root));
       // Only set ::rootId if there's not one set already, which may happen in
       // case of broken transactions.
       if ($this->rootId === NULL) {
         $this->rootId = $id;
       }
+      if ($this->connection->isEventEnabled(TransactionBeginEvent::class)) {
+        $this->connection->dispatchEvent(new TransactionBeginEvent(
+          $this->connection->getKey(),
+          $this->connection->getTarget(),
+          $id,
+          $name,
+          $this->getConnectionTransactionState(),
+          $this->stackItemsAsArray(),
+        ));
+      }
     }
     else {
       // If we're already in a Drupal transaction then we want to create a
       // database savepoint, rather than try to begin another database
       // transaction.
       $this->addClientSavepoint($name);
-      $type = StackItemType::Savepoint;
+      $this->addStackItem($id, new StackItem($name, StackItemType::Savepoint));
+      if ($this->connection->isEventEnabled(TransactionSavepointEvent::class)) {
+        $this->connection->dispatchEvent(new TransactionSavepointEvent(
+          $this->connection->getKey(),
+          $this->connection->getTarget(),
+          $id,
+          $name,
+          $this->getConnectionTransactionState(),
+          $this->stackItemsAsArray(),
+        ));
+      }
     }
 
-    // Add an item on the stack, increasing its depth.
-    $this->addStackItem($id, new StackItem($name, $type));
-
     // Actually return a new Transaction object.
     return new Transaction($this->connection, $name, $id);
   }
@@ -324,7 +392,7 @@ public function purge(string $name, string $id): void {
     // When we get here, the transaction (or savepoint) is still active on the
     // database. We can unpile it, and if we are left with no more items in the
     // stack, we can also process the post transaction callbacks.
-    $this->commit($name, $id);
+    $this->commit($name, $id, CommitType::ImplicitOnDestruct);
     $this->removeStackItem($id);
     if ($this->rootId === $id) {
       $this->processPostTransactionCallbacks();
@@ -352,7 +420,7 @@ public function unpile(string $name, string $id): void {
     }
 
     // Commit the transaction.
-    $this->commit($name, $id);
+    $this->commit($name, $id, CommitType::Explicit);
 
     // Void the transaction stack item.
     $this->voidStackItem($id);
@@ -365,13 +433,15 @@ public function unpile(string $name, string $id): void {
    *   The name of the transaction.
    * @param string $id
    *   The id of the transaction.
+   * @param \Drupal\Core\Database\Transaction\CommitType|null $commitType
+   *   The type of commit.
    *
    * @throws \Drupal\Core\Database\TransactionOutOfOrderException
    *   If a Drupal Transaction with the specified name does not exist.
    * @throws \Drupal\Core\Database\TransactionCommitFailedException
    *   If the commit of the root transaction failed.
    */
-  protected function commit(string $name, string $id): void {
+  protected function commit(string $name, string $id, ?CommitType $commitType = NULL): void {
     if ($this->getConnectionTransactionState() !== ClientConnectionTransactionState::Active) {
       // The stack got corrupted.
       throw new TransactionOutOfOrderException("Transaction {$id}\\{$name} is out of order. Active stack: " . $this->dumpStackItemsAsString());
@@ -388,12 +458,76 @@ protected function commit(string $name, string $id): void {
     if ($this->stackDepth() > 1 && $this->stack()[$id]->type === StackItemType::Savepoint) {
       // Release the client transaction savepoint in case the Drupal
       // transaction is not a root one.
-      $this->releaseClientSavepoint($name);
+      try {
+        if (!$this->releaseClientSavepoint($name)) {
+          throw new TransactionCommitFailedException("Release savepoint $name failed");
+        }
+        $this->removeStackItem($id);
+        if ($this->connection->isEventEnabled(TransactionReleaseSavepointEvent::class)) {
+          $this->connection->dispatchEvent(new TransactionReleaseSavepointEvent(
+            $this->connection->getKey(),
+            $this->connection->getTarget(),
+            $id,
+            $name,
+            $this->getConnectionTransactionState(),
+            $this->stackItemsAsArray(),
+            $commitType,
+          ));
+        }
+      }
+      catch (\Exception $e) {
+        $this->setConnectionTransactionState(ClientConnectionTransactionState::CommitFailed);
+        $this->voidClientTransaction();
+        if ($this->connection->isEventEnabled(TransactionReleaseSavepointFailureEvent::class)) {
+          $this->connection->dispatchEvent(new TransactionReleaseSavepointFailureEvent(
+            $this->connection->getKey(),
+            $this->connection->getTarget(),
+            $id,
+            $name,
+            $this->getConnectionTransactionState(),
+            $this->stackItemsAsArray(),
+            get_class($e),
+            $e->getMessage(),
+            $commitType,
+          ));
+        }
+      }
     }
     elseif ($this->stackDepth() === 1 && $this->stack()[$id]->type === StackItemType::Root) {
       // If this was the root Drupal transaction, we can commit the client
       // transaction.
-      $this->processRootCommit();
+      try {
+        $this->processRootCommit();
+        $this->voidStackItem($id);
+        if ($this->connection->isEventEnabled(TransactionCommitEvent::class)) {
+          $this->connection->dispatchEvent(new TransactionCommitEvent(
+            $this->connection->getKey(),
+            $this->connection->getTarget(),
+            $id,
+            $name,
+            $this->getConnectionTransactionState(),
+            $this->stackItemsAsArray(),
+            $commitType,
+          ));
+        }
+      }
+      catch (\Exception $e) {
+        $this->voidClientTransaction();
+        if ($this->connection->isEventEnabled(TransactionCommitFailureEvent::class)) {
+          $this->connection->dispatchEvent(new TransactionCommitFailureEvent(
+            $this->connection->getKey(),
+            $this->connection->getTarget(),
+            $id,
+            $name,
+            $this->getConnectionTransactionState(),
+            $this->stackItemsAsArray(),
+            get_class($e),
+            $e->getMessage(),
+            $commitType,
+          ));
+        }
+        throw $e;
+      }
     }
     else {
       // The stack got corrupted.
@@ -423,20 +557,77 @@ public function rollback(string $name, string $id): void {
         // Rollback the client transaction to the savepoint when the Drupal
         // transaction is not a root one. Then, release the savepoint too. The
         // client connection remains active.
-        $this->rollbackClientSavepoint($name);
-        $this->releaseClientSavepoint($name);
-        // The Transaction object remains open, and when it will get destructed
-        // no commit should happen. Void the stack item.
-        $this->voidStackItem($id);
+        try {
+          if (!$this->rollbackClientSavepoint($name)) {
+            throw new TransactionException("Rollback to savepoint $name failed");
+          }
+          if (!$this->releaseClientSavepoint($name)) {
+            throw new TransactionException("Release savepoint $name failed");
+          }
+          // The Transaction object remains open, and when it will get
+          // destructed no commit should happen. Void the stack item.
+          $this->voidStackItem($id);
+          if ($this->connection->isEventEnabled(TransactionRollbackToSavepointEvent::class)) {
+            $this->connection->dispatchEvent(new TransactionRollbackToSavepointEvent(
+              $this->connection->getKey(),
+              $this->connection->getTarget(),
+              $id,
+              $name,
+              $this->getConnectionTransactionState(),
+              $this->stackItemsAsArray(),
+            ));
+          }
+        }
+        catch (\Exception $e) {
+          $this->setConnectionTransactionState(ClientConnectionTransactionState::RollbackFailed);
+          $this->voidClientTransaction();
+          if ($this->connection->isEventEnabled(TransactionRollbackToSavepointFailureEvent::class)) {
+            $this->connection->dispatchEvent(new TransactionRollbackToSavepointFailureEvent(
+              $this->connection->getKey(),
+              $this->connection->getTarget(),
+              $id,
+              $name,
+              $this->getConnectionTransactionState(),
+              $this->stackItemsAsArray(),
+              get_class($e),
+              $e->getMessage(),
+            ));
+          }
+        }
       }
       elseif ($this->stackDepth() === 1 && $this->stack()[$id]->type === StackItemType::Root) {
         // If this was the root Drupal transaction, we can rollback the client
         // transaction. The transaction is closed.
-        $this->processRootRollback();
-        if ($this->getConnectionTransactionState() === ClientConnectionTransactionState::RolledBack) {
+        try {
+          $this->processRootRollback();
           // The Transaction object remains open, and when it will get
           // destructed no commit should happen. Void the stack item.
           $this->voidStackItem($id);
+          if ($this->connection->isEventEnabled(TransactionRollbackEvent::class)) {
+            $this->connection->dispatchEvent(new TransactionRollbackEvent(
+              $this->connection->getKey(),
+              $this->connection->getTarget(),
+              $id,
+              $name,
+              $this->getConnectionTransactionState(),
+              $this->stackItemsAsArray(),
+            ));
+          }
+        }
+        catch (\Exception $e) {
+          $this->voidClientTransaction();
+          if ($this->connection->isEventEnabled(TransactionRollbackFailureEvent::class)) {
+            $this->connection->dispatchEvent(new TransactionRollbackFailureEvent(
+              $this->connection->getKey(),
+              $this->connection->getTarget(),
+              $id,
+              $name,
+              $this->getConnectionTransactionState(),
+              $this->stackItemsAsArray(),
+              get_class($e),
+              $e->getMessage(),
+            ));
+          }
         }
       }
       else {
@@ -510,7 +701,12 @@ protected function getConnectionTransactionState(): ClientConnectionTransactionS
    * Processes the root transaction rollback.
    */
   protected function processRootRollback(): void {
-    $this->rollbackClientTransaction();
+    $clientRollback = $this->rollbackClientTransaction();
+    if (!$clientRollback) {
+      $this->setConnectionTransactionState(ClientConnectionTransactionState::RollbackFailed);
+      throw new TransactionException('Transaction rollback failed');
+    }
+    $this->setConnectionTransactionState(ClientConnectionTransactionState::RolledBack);
   }
 
   /**
@@ -522,8 +718,10 @@ protected function processRootRollback(): void {
   protected function processRootCommit(): void {
     $clientCommit = $this->commitClientTransaction();
     if (!$clientCommit) {
-      throw new TransactionCommitFailedException();
+      $this->setConnectionTransactionState(ClientConnectionTransactionState::CommitFailed);
+      throw new TransactionCommitFailedException('Transaction commit failed');
     }
+    $this->setConnectionTransactionState(ClientConnectionTransactionState::Committed);
   }
 
   /**
@@ -621,7 +819,35 @@ public function voidClientTransaction(): void {
     while ($i = array_key_last($this->stack())) {
       $this->voidStackItem((string) $i);
     }
-    $this->setConnectionTransactionState(ClientConnectionTransactionState::Voided);
+    if ($this->getConnectionTransactionState() === ClientConnectionTransactionState::Active) {
+      $this->setConnectionTransactionState(ClientConnectionTransactionState::Voided);
+    }
+  }
+
+  /**
+   * Handles automatic commit occurred upon execution of a DDL statement.
+   *
+   * In some cases the active transaction is automatically committed by the
+   * database server when a DDL statement is executed during a transaction
+   * (for example, MySql). In such cases we need to trigger a
+   * TransactionCommitEvent and void the remaining items on the stack.
+   *
+   * This method should only be called internally by a database driver.
+   */
+  public function handleAutomaticCommitUponDDL(): void {
+    if ($this->connection->isEventEnabled(TransactionAutoCommitEvent::class)) {
+      $i = array_key_first($this->stack);
+      assert($this->stack[$i]->type === StackItemType::Root);
+      $this->connection->dispatchEvent(new TransactionAutoCommitEvent(
+        $this->connection->getKey(),
+        $this->connection->getTarget(),
+        $i,
+        $this->stack[$i]->name,
+        $this->getConnectionTransactionState(),
+        $this->stackItemsAsArray(),
+      ));
+    }
+    $this->voidClientTransaction();
   }
 
 }
diff --git a/core/lib/Drupal/Core/Database/Transaction/TransactionManagerInterface.php b/core/lib/Drupal/Core/Database/Transaction/TransactionManagerInterface.php
index 0f7744da95d..3c3d620edce 100644
--- a/core/lib/Drupal/Core/Database/Transaction/TransactionManagerInterface.php
+++ b/core/lib/Drupal/Core/Database/Transaction/TransactionManagerInterface.php
@@ -19,6 +19,22 @@ interface TransactionManagerInterface {
    */
   public function inTransaction(): bool;
 
+  /**
+   * Gets the current transaction id.
+   *
+   * @return string|null
+   *   The current transaction id.
+   */
+  public function currentTransactionId(): ?string;
+
+  /**
+   * Gets the current transaction name.
+   *
+   * @return string|null
+   *   The current transaction name.
+   */
+  public function currentTransactionName(): ?string;
+
   /**
    * Checks if a named Drupal transaction is active.
    *
@@ -122,16 +138,25 @@ public function rollback(string $name, string $id): void;
   /**
    * Voids the client connection.
    *
-   * In some cases the active transaction can be automatically committed by the
-   * database server (for example, MySql when a DDL statement is executed
-   * during a transaction). In such cases we need to void the remaining items
-   * on the stack so that when outliving Transaction object get out of scope
-   * they will not try operations on the database.
+   * Void any existing items on the stack so that when outliving Transaction
+   * objects get out of scope they will not try operations on the database.
    *
    * This method should only be called internally by a database driver.
    */
   public function voidClientTransaction(): void;
 
+  /**
+   * Handles automatic commit occurred upon execution of a DDL statement.
+   *
+   * In some cases the active transaction is automatically committed by the
+   * database server when a DDL statement is executed during a transaction
+   * (for example, MySql). In such cases we need to trigger a
+   * TransactionCommitEvent and void the remaining items on the stack.
+   *
+   * This method should only be called internally by a database driver.
+   */
+  public function handleAutomaticCommitUponDDL(): void;
+
   /**
    * Adds a root transaction end callback.
    *
diff --git a/core/lib/Drupal/Core/Database/TransactionCommitFailedException.php b/core/lib/Drupal/Core/Database/TransactionCommitFailedException.php
index f1dbf956543..4cb99dfd4d4 100644
--- a/core/lib/Drupal/Core/Database/TransactionCommitFailedException.php
+++ b/core/lib/Drupal/Core/Database/TransactionCommitFailedException.php
@@ -3,6 +3,6 @@
 namespace Drupal\Core\Database;
 
 /**
- * Exception thrown when a commit() function fails.
+ * Exception thrown when a commit or release savepoint function fails.
  */
 class TransactionCommitFailedException extends TransactionException implements DatabaseException {}
diff --git a/core/modules/mysql/src/Driver/Database/mysql/TransactionManager.php b/core/modules/mysql/src/Driver/Database/mysql/TransactionManager.php
index a79749ff4bc..f0084969f35 100644
--- a/core/modules/mysql/src/Driver/Database/mysql/TransactionManager.php
+++ b/core/modules/mysql/src/Driver/Database/mysql/TransactionManager.php
@@ -4,8 +4,8 @@
 
 namespace Drupal\mysql\Driver\Database\mysql;
 
-use Drupal\Core\Database\Transaction\ClientConnectionTransactionState;
 use Drupal\Core\Database\Transaction\TransactionManagerBase;
+use Drupal\Core\Database\TransactionNoActiveException;
 
 /**
  * MySql implementation of TransactionManagerInterface.
@@ -24,24 +24,12 @@ protected function beginClientTransaction(): bool {
     return $this->connection->getClientConnection()->beginTransaction();
   }
 
-  /**
-   * {@inheritdoc}
-   */
-  protected function processRootCommit(): void {
-    if (!$this->connection->getClientConnection()->inTransaction()) {
-      $this->voidClientTransaction();
-      return;
-    }
-    parent::processRootCommit();
-  }
-
   /**
    * {@inheritdoc}
    */
   protected function rollbackClientSavepoint(string $name): bool {
     if (!$this->connection->getClientConnection()->inTransaction()) {
-      $this->voidClientTransaction();
-      return TRUE;
+      throw new TransactionNoActiveException();
     }
     return parent::rollbackClientSavepoint($name);
   }
@@ -51,8 +39,7 @@ protected function rollbackClientSavepoint(string $name): bool {
    */
   protected function releaseClientSavepoint(string $name): bool {
     if (!$this->connection->getClientConnection()->inTransaction()) {
-      $this->voidClientTransaction();
-      return TRUE;
+      throw new TransactionNoActiveException();
     }
     return parent::releaseClientSavepoint($name);
   }
@@ -61,12 +48,7 @@ protected function releaseClientSavepoint(string $name): bool {
    * {@inheritdoc}
    */
   protected function commitClientTransaction(): bool {
-    $clientCommit = $this->connection->getClientConnection()->commit();
-    $this->setConnectionTransactionState($clientCommit ?
-      ClientConnectionTransactionState::Committed :
-      ClientConnectionTransactionState::CommitFailed
-    );
-    return $clientCommit;
+    return $this->connection->getClientConnection()->commit();
   }
 
   /**
@@ -74,15 +56,9 @@ protected function commitClientTransaction(): bool {
    */
   protected function rollbackClientTransaction(): bool {
     if (!$this->connection->getClientConnection()->inTransaction()) {
-      $this->voidClientTransaction();
-      return FALSE;
+      throw new TransactionNoActiveException();
     }
-    $clientRollback = $this->connection->getClientConnection()->rollBack();
-    $this->setConnectionTransactionState($clientRollback ?
-      ClientConnectionTransactionState::RolledBack :
-      ClientConnectionTransactionState::RollbackFailed
-    );
-    return $clientRollback;
+    return $this->connection->getClientConnection()->rollBack();
   }
 
 }
diff --git a/core/modules/mysqli/src/Driver/Database/mysqli/TransactionManager.php b/core/modules/mysqli/src/Driver/Database/mysqli/TransactionManager.php
index 90237fd6a43..7327c98061e 100644
--- a/core/modules/mysqli/src/Driver/Database/mysqli/TransactionManager.php
+++ b/core/modules/mysqli/src/Driver/Database/mysqli/TransactionManager.php
@@ -4,7 +4,6 @@
 
 namespace Drupal\mysqli\Driver\Database\mysqli;
 
-use Drupal\Core\Database\Transaction\ClientConnectionTransactionState;
 use Drupal\Core\Database\Transaction\TransactionManagerBase;
 
 /**
@@ -33,16 +32,7 @@ protected function rollbackClientSavepoint(string $name): bool {
     // Mysqli does not have a rollback_to_savepoint method, and it does not
     // allow a prepared statement for 'ROLLBACK TO SAVEPOINT', so we need to
     // fallback to querying on the client connection directly.
-    try {
-      return (bool) $this->connection->getClientConnection()->query('ROLLBACK TO SAVEPOINT ' . $name);
-    }
-    catch (\mysqli_sql_exception) {
-      // If the rollback failed, most likely the savepoint was not there
-      // because the transaction is no longer active. In this case we void the
-      // transaction stack.
-      $this->voidClientTransaction();
-      return TRUE;
-    }
+    return (bool) $this->connection->getClientConnection()->query('ROLLBACK TO SAVEPOINT ' . $name);
   }
 
   /**
@@ -59,24 +49,14 @@ protected function rollbackClientTransaction(): bool {
     // Note: mysqli::rollback() returns TRUE if there's no active transaction.
     // This is diverging from PDO MySql. A PHP bug report exists.
     // @see https://bugs.php.net/bug.php?id=81533.
-    $clientRollback = $this->connection->getClientConnection()->rollBack();
-    $this->setConnectionTransactionState($clientRollback ?
-      ClientConnectionTransactionState::RolledBack :
-      ClientConnectionTransactionState::RollbackFailed
-    );
-    return $clientRollback;
+    return $this->connection->getClientConnection()->rollBack();
   }
 
   /**
    * {@inheritdoc}
    */
   protected function commitClientTransaction(): bool {
-    $clientCommit = $this->connection->getClientConnection()->commit();
-    $this->setConnectionTransactionState($clientCommit ?
-      ClientConnectionTransactionState::Committed :
-      ClientConnectionTransactionState::CommitFailed
-    );
-    return $clientCommit;
+    return $this->connection->getClientConnection()->commit();
   }
 
 }
diff --git a/core/modules/pgsql/src/Driver/Database/pgsql/TransactionManager.php b/core/modules/pgsql/src/Driver/Database/pgsql/TransactionManager.php
index 4374ac993c3..86020c3f9b1 100644
--- a/core/modules/pgsql/src/Driver/Database/pgsql/TransactionManager.php
+++ b/core/modules/pgsql/src/Driver/Database/pgsql/TransactionManager.php
@@ -4,7 +4,6 @@
 
 namespace Drupal\pgsql\Driver\Database\pgsql;
 
-use Drupal\Core\Database\Transaction\ClientConnectionTransactionState;
 use Drupal\Core\Database\Transaction\TransactionManagerBase;
 
 /**
@@ -23,24 +22,14 @@ protected function beginClientTransaction(): bool {
    * {@inheritdoc}
    */
   protected function rollbackClientTransaction(): bool {
-    $clientRollback = $this->connection->getClientConnection()->rollBack();
-    $this->setConnectionTransactionState($clientRollback ?
-      ClientConnectionTransactionState::RolledBack :
-      ClientConnectionTransactionState::RollbackFailed
-    );
-    return $clientRollback;
+    return $this->connection->getClientConnection()->rollBack();
   }
 
   /**
    * {@inheritdoc}
    */
   protected function commitClientTransaction(): bool {
-    $clientCommit = $this->connection->getClientConnection()->commit();
-    $this->setConnectionTransactionState($clientCommit ?
-      ClientConnectionTransactionState::Committed :
-      ClientConnectionTransactionState::CommitFailed
-    );
-    return $clientCommit;
+    return $this->connection->getClientConnection()->commit();
   }
 
 }
diff --git a/core/modules/sqlite/src/Driver/Database/sqlite/TransactionManager.php b/core/modules/sqlite/src/Driver/Database/sqlite/TransactionManager.php
index 7b422767d63..7c25879d1f7 100644
--- a/core/modules/sqlite/src/Driver/Database/sqlite/TransactionManager.php
+++ b/core/modules/sqlite/src/Driver/Database/sqlite/TransactionManager.php
@@ -4,7 +4,6 @@
 
 namespace Drupal\sqlite\Driver\Database\sqlite;
 
-use Drupal\Core\Database\Transaction\ClientConnectionTransactionState;
 use Drupal\Core\Database\Transaction\TransactionManagerBase;
 
 /**
@@ -23,24 +22,14 @@ protected function beginClientTransaction(): bool {
    * {@inheritdoc}
    */
   protected function rollbackClientTransaction(): bool {
-    $clientRollback = $this->connection->getClientConnection()->rollBack();
-    $this->setConnectionTransactionState($clientRollback ?
-      ClientConnectionTransactionState::RolledBack :
-      ClientConnectionTransactionState::RollbackFailed
-    );
-    return $clientRollback;
+    return $this->connection->getClientConnection()->rollBack();
   }
 
   /**
    * {@inheritdoc}
    */
   protected function commitClientTransaction(): bool {
-    $clientCommit = $this->connection->getClientConnection()->commit();
-    $this->setConnectionTransactionState($clientCommit ?
-      ClientConnectionTransactionState::Committed :
-      ClientConnectionTransactionState::CommitFailed
-    );
-    return $clientCommit;
+    return $this->connection->getClientConnection()->commit();
   }
 
 }
diff --git a/core/modules/system/tests/modules/database_test/database_test.services.yml b/core/modules/system/tests/modules/database_test/database_test.services.yml
index a4211a5bec1..ea8a1ebe872 100644
--- a/core/modules/system/tests/modules/database_test/database_test.services.yml
+++ b/core/modules/system/tests/modules/database_test/database_test.services.yml
@@ -1,5 +1,6 @@
 services:
   _defaults:
+    autowire: true
     autoconfigure: true
   # A test subscriber for the database API events.
   Drupal\database_test\EventSubscriber\DatabaseEventSubscriber: ~
diff --git a/core/modules/system/tests/modules/database_test/src/EventSubscriber/DatabaseEventSubscriber.php b/core/modules/system/tests/modules/database_test/src/EventSubscriber/DatabaseEventSubscriber.php
index 029d6ee5a31..6ea778c15e6 100644
--- a/core/modules/system/tests/modules/database_test/src/EventSubscriber/DatabaseEventSubscriber.php
+++ b/core/modules/system/tests/modules/database_test/src/EventSubscriber/DatabaseEventSubscriber.php
@@ -7,6 +7,15 @@
 use Drupal\Core\Database\Event\StatementExecutionEndEvent;
 use Drupal\Core\Database\Event\StatementExecutionFailureEvent;
 use Drupal\Core\Database\Event\StatementExecutionStartEvent;
+use Drupal\Core\Database\Event\TransactionAutoCommitEvent;
+use Drupal\Core\Database\Event\TransactionBeginEvent;
+use Drupal\Core\Database\Event\TransactionCommitEvent;
+use Drupal\Core\Database\Event\TransactionReleaseSavepointEvent;
+use Drupal\Core\Database\Event\TransactionRollbackEvent;
+use Drupal\Core\Database\Event\TransactionRollbackToSavepointEvent;
+use Drupal\Core\Database\Event\TransactionSavepointEvent;
+use Drupal\Core\KeyValueStore\KeyValueFactoryInterface;
+use Symfony\Component\DependencyInjection\Attribute\Autowire;
 use Symfony\Component\EventDispatcher\EventSubscriberInterface;
 
 /**
@@ -34,6 +43,12 @@ class DatabaseEventSubscriber implements EventSubscriberInterface {
    */
   public array $statementIdsInExecution = [];
 
+  public function __construct(
+    #[Autowire(service: 'keyvalue')]
+    protected KeyValueFactoryInterface $keyValue,
+  ) {
+  }
+
   /**
    * {@inheritdoc}
    */
@@ -42,6 +57,13 @@ public static function getSubscribedEvents(): array {
       StatementExecutionStartEvent::class => 'onStatementExecutionStart',
       StatementExecutionEndEvent::class => 'onStatementExecutionEnd',
       StatementExecutionFailureEvent::class => 'onStatementExecutionFailure',
+      TransactionBeginEvent::class => 'onTransactionBegin',
+      TransactionSavepointEvent::class => 'onTransactionSavepoint',
+      TransactionAutoCommitEvent::class => 'onTransactionAutoCommit',
+      TransactionCommitEvent::class => 'onTransactionCommit',
+      TransactionReleaseSavepointEvent::class => 'onTransactionReleaseSavepoint',
+      TransactionRollbackEvent::class => 'onTransactionRollback',
+      TransactionRollbackToSavepointEvent::class => 'onTransactionRollbackToSavepoint',
     ];
   }
 
@@ -78,4 +100,74 @@ public function onStatementExecutionFailure(StatementExecutionFailureEvent $even
     $this->countStatementFailures++;
   }
 
+  /**
+   * Subscribes to a TransactionBeginEvent.
+   *
+   * @param \Drupal\Core\Database\Event\TransactionBeginEvent $event
+   *   The transaction event.
+   */
+  public function onTransactionBegin(TransactionBeginEvent $event): void {
+    $this->keyValue->get('TransactionEventTest')->set('event_data', get_class($event) . ' ' . $event->key . ' ' . $event->target . ' ' . $event->id . '\\' . $event->name);
+  }
+
+  /**
+   * Subscribes to a TransactionSavepointEvent.
+   *
+   * @param \Drupal\Core\Database\Event\TransactionSavepointEvent $event
+   *   The transaction event.
+   */
+  public function onTransactionSavepoint(TransactionSavepointEvent $event): void {
+    $this->keyValue->get('TransactionEventTest')->set('event_data', get_class($event) . ' ' . $event->key . ' ' . $event->target . ' ' . $event->id . '\\' . $event->name . ' stack: ' . implode(' > ', $event->stackItems));
+  }
+
+  /**
+   * Subscribes to a TransactionAutoCommitEvent.
+   *
+   * @param \Drupal\Core\Database\Event\TransactionAutoCommitEvent $event
+   *   The transaction event.
+   */
+  public function onTransactionCommit(TransactionAutoCommitEvent $event): void {
+    $this->keyValue->get('TransactionEventTest')->set('event_data', get_class($event) . ' ' . $event->key . ' ' . $event->target . ' ' . $event->id . '\\' . $event->name);
+  }
+
+  /**
+   * Subscribes to a TransactionCommitEvent.
+   *
+   * @param \Drupal\Core\Database\Event\TransactionCommitEvent $event
+   *   The transaction event.
+   */
+  public function onTransactionCommit(TransactionCommitEvent $event): void {
+    $this->keyValue->get('TransactionEventTest')->set('event_data', get_class($event) . ' ' . $event->key . ' ' . $event->target . ' ' . $event->id . '\\' . $event->name . ' commit-type: ' . $event->type->name);
+  }
+
+  /**
+   * Subscribes to a TransactionReleaseSavepointEvent.
+   *
+   * @param \Drupal\Core\Database\Event\TransactionReleaseSavepointEvent $event
+   *   The transaction event.
+   */
+  public function onTransactionReleaseSavepoint(TransactionReleaseSavepointEvent $event): void {
+    $this->keyValue->get('TransactionEventTest')->set('event_data', get_class($event) . ' ' . $event->key . ' ' . $event->target . ' ' . $event->id . '\\' . $event->name . ' stack: ' . implode(' > ', $event->stackItems));
+  }
+
+  /**
+   * Subscribes to a TransactionRollbackEvent.
+   *
+   * @param \Drupal\Core\Database\Event\TransactionRollbackEvent $event
+   *   The transaction event.
+   */
+  public function onTransactionRollback(TransactionRollbackEvent $event): void {
+    $this->keyValue->get('TransactionEventTest')->set('event_data', get_class($event) . ' ' . $event->key . ' ' . $event->target . ' ' . $event->id . '\\' . $event->name);
+  }
+
+  /**
+   * Subscribes to a TransactionRollbackToSavepointEvent.
+   *
+   * @param \Drupal\Core\Database\Event\TransactionRollbackToSavepointEvent $event
+   *   The transaction event.
+   */
+  public function onTransactionRollbackToSavepoint(TransactionRollbackToSavepointEvent $event): void {
+    $this->keyValue->get('TransactionEventTest')->set('event_data', get_class($event) . ' ' . $event->key . ' ' . $event->target . ' ' . $event->id . '\\' . $event->name . ' stack: ' . implode(' > ', $event->stackItems));
+  }
+
 }
diff --git a/core/tests/Drupal/KernelTests/Core/Database/TransactionEventTest.php b/core/tests/Drupal/KernelTests/Core/Database/TransactionEventTest.php
new file mode 100644
index 00000000000..a1c732c41ac
--- /dev/null
+++ b/core/tests/Drupal/KernelTests/Core/Database/TransactionEventTest.php
@@ -0,0 +1,130 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Drupal\KernelTests\Core\Database;
+
+use Drupal\Core\Database\Event\TransactionEvent;
+use PHPUnit\Framework\Attributes\Group;
+
+/**
+ * Tests the Transaction events.
+ */
+#[Group('Database')]
+class TransactionEventTest extends DatabaseTestBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function setUp(): void {
+    parent::setUp();
+    $this->connection->enableEvents(TransactionEvent::all());
+  }
+
+  /**
+   * Tests transaction beginning.
+   */
+  public function testTransactionBegin(): void {
+    $tx = $this->connection->startTransaction();
+    $this->assertMatchesRegularExpression("/^.*TransactionBeginEvent default default .*\\\\drupal_transaction$/", \Drupal::keyValue('TransactionEventTest')->get('event_data'));
+    $tx->commitOrRelease();
+  }
+
+  /**
+   * Tests adding a savepoint.
+   */
+  public function testTransactionSavepoint(): void {
+    $tx = $this->connection->startTransaction();
+    $savepoint = $this->connection->startTransaction();
+    $this->assertMatchesRegularExpression("/^.*TransactionSavepointEvent default default .*\\\\savepoint_1 stack: .*\\\\drupal_transaction > .*\\\\savepoint_1$/", \Drupal::keyValue('TransactionEventTest')->get('event_data'));
+    $savepoint->commitOrRelease();
+    $tx->commitOrRelease();
+  }
+
+  /**
+   * Tests committing a transaction.
+   */
+  public function testTransactionCommit(): void {
+    $tx = $this->connection->startTransaction();
+    $tx->commitOrRelease();
+    $this->assertMatchesRegularExpression("/^.*TransactionCommitEvent default default .*\\\\drupal_transaction commit-type: Explicit$/", \Drupal::keyValue('TransactionEventTest')->get('event_data'));
+  }
+
+  /**
+   * Tests implicitly committing a transaction via object destruction.
+   */
+  public function testTransactionImplicitCommit(): void {
+    $tx = $this->connection->startTransaction();
+    unset($tx);
+    $this->assertMatchesRegularExpression("/^.*TransactionCommitEvent default default .*\\\\drupal_transaction commit-type: ImplicitOnDestruct$/", \Drupal::keyValue('TransactionEventTest')->get('event_data'));
+  }
+
+  /**
+   * Tests committing a transaction after a savepoint was opened.
+   */
+  public function testTransactionCommitNested(): void {
+    $tx = $this->connection->startTransaction();
+    $savepoint = $this->connection->startTransaction('foo');
+    $tx->commitOrRelease();
+    $this->assertMatchesRegularExpression("/^.*TransactionCommitEvent default default .*\\\\drupal_transaction commit-type: Explicit$/", \Drupal::keyValue('TransactionEventTest')->get('event_data'));
+    unset($savepoint);
+  }
+
+  /**
+   * Tests DDL auto-commit.
+   */
+  public function testAutoCommitOnDataDefinitionLanguageStatement(): void {
+    if ($this->connection->supportsTransactionalDDL()) {
+      $this->markTestSkipped('This test is only relevant for databases that do not support transactional DDL.');
+    }
+
+    $table = [
+      'fields' => [
+        'id' => [
+          'type' => 'serial',
+          'unsigned' => TRUE,
+          'not null' => TRUE,
+        ],
+      ],
+      'primary key' => ['id'],
+    ];
+
+    $tx = $this->connection->startTransaction();
+    $this->connection->schema()->createTable('database_test_00', $table);
+    $this->assertMatchesRegularExpression("/^.*TransactionAutoCommitEvent default default .*\\\\drupal_transaction$/", \Drupal::keyValue('TransactionEventTest')->get('event_data'));
+    unset($tx);
+    $this->assertMatchesRegularExpression("/^.*TransactionCommitEvent default default .*\\\\drupal_transaction commit-type: AutomaticUponDDL$/", \Drupal::keyValue('TransactionEventTest')->get('event_data'));
+  }
+
+  /**
+   * Tests releasing a savepoint.
+   */
+  public function testTransactionReleaseSavepoint(): void {
+    $tx = $this->connection->startTransaction();
+    $savepoint = $this->connection->startTransaction('foo');
+    $savepoint->commitOrRelease();
+    $this->assertMatchesRegularExpression("/^.*TransactionReleaseSavepointEvent default default .*\\\\foo stack: .*\\\\drupal_transaction$/", \Drupal::keyValue('TransactionEventTest')->get('event_data'));
+    unset($tx);
+  }
+
+  /**
+   * Tests rolling back a transaction.
+   */
+  public function testTransactionRollback(): void {
+    $tx = $this->connection->startTransaction();
+    $tx->rollback();
+    $this->assertMatchesRegularExpression("/^.*TransactionRollbackEvent default default .*\\\\drupal_transaction$/", \Drupal::keyValue('TransactionEventTest')->get('event_data'));
+  }
+
+  /**
+   * Tests rolling back a transaction to a savepoint.
+   */
+  public function testTransactionRollbackToSavepoint(): void {
+    $tx = $this->connection->startTransaction();
+    $savepoint = $this->connection->startTransaction('foo');
+    $savepoint->rollback();
+    $this->assertMatchesRegularExpression("/^.*TransactionRollbackToSavepointEvent default default .*\\\\foo stack: .*\\\\drupal_transaction$/", \Drupal::keyValue('TransactionEventTest')->get('event_data'));
+    unset($tx);
+  }
+
+}
diff --git a/core/tests/Drupal/Tests/Core/Database/Stub/EventSubscriber/StubTransactionEventSubscriber.php b/core/tests/Drupal/Tests/Core/Database/Stub/EventSubscriber/StubTransactionEventSubscriber.php
new file mode 100644
index 00000000000..b76a04c9d3b
--- /dev/null
+++ b/core/tests/Drupal/Tests/Core/Database/Stub/EventSubscriber/StubTransactionEventSubscriber.php
@@ -0,0 +1,58 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Drupal\Tests\Core\Database\Stub\EventSubscriber;
+
+use Drupal\Core\Database\Event\TransactionBeginEvent;
+use Drupal\Core\Database\Event\TransactionCommitEvent;
+use Drupal\Core\Database\Event\TransactionCommitFailureEvent;
+use Drupal\Core\Database\Event\TransactionEventBase;
+use Drupal\Core\Database\Event\TransactionReleaseSavepointEvent;
+use Drupal\Core\Database\Event\TransactionReleaseSavepointFailureEvent;
+use Drupal\Core\Database\Event\TransactionRollbackEvent;
+use Drupal\Core\Database\Event\TransactionRollbackFailureEvent;
+use Drupal\Core\Database\Event\TransactionRollbackToSavepointEvent;
+use Drupal\Core\Database\Event\TransactionRollbackToSavepointFailureEvent;
+use Drupal\Core\Database\Event\TransactionSavepointEvent;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+/**
+ * Responds to database API transaction events.
+ */
+class StubTransactionEventSubscriber implements EventSubscriberInterface {
+
+  /**
+   * The last event dispatched.
+   */
+  public TransactionEventBase $lastEvent;
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function getSubscribedEvents(): array {
+    return [
+      TransactionBeginEvent::class => 'onTransactionEvent',
+      TransactionCommitEvent::class => 'onTransactionEvent',
+      TransactionCommitFailureEvent::class => 'onTransactionEvent',
+      TransactionReleaseSavepointEvent::class => 'onTransactionEvent',
+      TransactionReleaseSavepointFailureEvent::class => 'onTransactionEvent',
+      TransactionRollbackEvent::class => 'onTransactionEvent',
+      TransactionRollbackFailureEvent::class => 'onTransactionEvent',
+      TransactionRollbackToSavepointEvent::class => 'onTransactionEvent',
+      TransactionRollbackToSavepointFailureEvent::class => 'onTransactionEvent',
+      TransactionSavepointEvent::class => 'onTransactionEvent',
+    ];
+  }
+
+  /**
+   * Subscribes to a transaction event.
+   *
+   * @param \Drupal\Core\Database\Event\TransactionEventBase $event
+   *   The transaction event.
+   */
+  public function onTransactionEvent(TransactionEventBase $event): void {
+    $this->lastEvent = $event;
+  }
+
+}
diff --git a/core/tests/Drupal/Tests/Core/Database/Stub/StubConnection.php b/core/tests/Drupal/Tests/Core/Database/Stub/StubConnection.php
index 4b31c06c2e6..9c1e9ece3f4 100644
--- a/core/tests/Drupal/Tests/Core/Database/Stub/StubConnection.php
+++ b/core/tests/Drupal/Tests/Core/Database/Stub/StubConnection.php
@@ -8,6 +8,7 @@
 use Drupal\Core\Database\ExceptionHandler;
 use Drupal\Core\Database\Log;
 use Drupal\Core\Database\StatementWrapperIterator;
+use Drupal\Core\Database\Transaction\TransactionManagerInterface;
 use Drupal\Tests\Core\Database\Stub\Driver\Schema;
 
 /**
@@ -52,6 +53,20 @@ public static function open(array &$connection_options = []) {
     return new \stdClass();
   }
 
+  /**
+   * {@inheritdoc}
+   */
+  public function getKey() {
+    return 'default';
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getTarget() {
+    return 'default';
+  }
+
   /**
    * {@inheritdoc}
    */
@@ -126,4 +141,11 @@ public function condition($conjunction) {
     return new StubCondition($conjunction);
   }
 
+  /**
+   * {@inheritdoc}
+   */
+  protected function driverTransactionManager(): TransactionManagerInterface {
+    return new StubTransactionManager($this);
+  }
+
 }
diff --git a/core/tests/Drupal/Tests/Core/Database/Stub/StubTransactionManager.php b/core/tests/Drupal/Tests/Core/Database/Stub/StubTransactionManager.php
new file mode 100644
index 00000000000..1457e3b2c8c
--- /dev/null
+++ b/core/tests/Drupal/Tests/Core/Database/Stub/StubTransactionManager.php
@@ -0,0 +1,61 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Drupal\Tests\Core\Database\Stub;
+
+use Drupal\Core\Database\Transaction\TransactionManagerBase;
+
+/**
+ * A stub of the TransactionManagerBase class for testing purposes.
+ */
+class StubTransactionManager extends TransactionManagerBase {
+
+  /**
+   * Should methods succeed or not.
+   */
+  public bool $success = TRUE;
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function beginClientTransaction(): bool {
+    return $this->success;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function addClientSavepoint(string $name): bool {
+    return $this->success;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function rollbackClientSavepoint(string $name): bool {
+    return $this->success;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function releaseClientSavepoint(string $name): bool {
+    return $this->success;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function commitClientTransaction(): bool {
+    return $this->success;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function rollbackClientTransaction(): bool {
+    return $this->success;
+  }
+
+}
diff --git a/core/tests/Drupal/Tests/Core/Database/TransactionEventTest.php b/core/tests/Drupal/Tests/Core/Database/TransactionEventTest.php
new file mode 100644
index 00000000000..562807d50e6
--- /dev/null
+++ b/core/tests/Drupal/Tests/Core/Database/TransactionEventTest.php
@@ -0,0 +1,190 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Drupal\Tests\Core\Database;
+
+use Drupal\Core\Database\Connection;
+use Drupal\Core\Database\Event\TransactionCommitEvent;
+use Drupal\Core\Database\Event\TransactionCommitFailureEvent;
+use Drupal\Core\Database\Event\TransactionEvent;
+use Drupal\Core\Database\Event\TransactionReleaseSavepointEvent;
+use Drupal\Core\Database\Event\TransactionReleaseSavepointFailureEvent;
+use Drupal\Core\Database\Event\TransactionRollbackEvent;
+use Drupal\Core\Database\Event\TransactionRollbackFailureEvent;
+use Drupal\Core\Database\Event\TransactionRollbackToSavepointEvent;
+use Drupal\Core\Database\Event\TransactionRollbackToSavepointFailureEvent;
+use Drupal\Core\Database\Transaction\ClientConnectionTransactionState;
+use Drupal\Core\Database\Transaction\CommitType;
+use Drupal\Core\Database\TransactionCommitFailedException;
+use Drupal\Core\DependencyInjection\Container;
+use Drupal\Tests\Core\Database\Stub\EventSubscriber\StubTransactionEventSubscriber;
+use Drupal\Tests\Core\Database\Stub\StubConnection;
+use Drupal\Tests\Core\Database\Stub\StubPDO;
+use Drupal\Tests\UnitTestCase;
+use PHPUnit\Framework\Attributes\Group;
+use Symfony\Component\EventDispatcher\EventDispatcher;
+
+/**
+ * Tests transaction events.
+ */
+#[Group('Database')]
+class TransactionEventTest extends UnitTestCase {
+
+  /**
+   * A database connection.
+   */
+  protected Connection $connection;
+
+  /**
+   * A subscriber for transaction events.
+   */
+  protected StubTransactionEventSubscriber $subscriber;
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function setUp(): void {
+    parent::setUp();
+    $this->connection = new StubConnection($this->createMock(StubPDO::class), []);
+    $this->connection->enableEvents(TransactionEvent::all());
+
+    $container = new Container();
+    $dispatcher = new EventDispatcher();
+    $this->subscriber = new StubTransactionEventSubscriber();
+    $dispatcher->addSubscriber($this->subscriber);
+    $container->set('event_dispatcher', $dispatcher);
+    \Drupal::setContainer($container);
+  }
+
+  /**
+   * Tests committing a transaction.
+   */
+  public function testTransactionCommit(): void {
+    $tx = $this->connection->startTransaction();
+    $tx->commitOrRelease();
+    $this->assertSame(TransactionCommitEvent::class, get_class($this->subscriber->lastEvent));
+    $this->assertSame('drupal_transaction', $this->subscriber->lastEvent->name);
+    $this->assertSame(ClientConnectionTransactionState::Committed, $this->subscriber->lastEvent->transactionState);
+    $this->assertSame(CommitType::Explicit, $this->subscriber->lastEvent->type);
+    $this->assertCount(0, $this->subscriber->lastEvent->stackItems);
+  }
+
+  /**
+   * Tests committing a transaction via object destruction.
+   */
+  public function testTransactionCommitOnDestruct(): void {
+    $tx = $this->connection->startTransaction();
+    unset($tx);
+    $this->assertSame(TransactionCommitEvent::class, get_class($this->subscriber->lastEvent));
+    $this->assertSame('drupal_transaction', $this->subscriber->lastEvent->name);
+    $this->assertSame(ClientConnectionTransactionState::Committed, $this->subscriber->lastEvent->transactionState);
+    $this->assertSame(CommitType::ImplicitOnDestruct, $this->subscriber->lastEvent->type);
+    $this->assertCount(0, $this->subscriber->lastEvent->stackItems);
+  }
+
+  /**
+   * Tests a transaction commit failure.
+   */
+  public function testTransactionCommitFailure(): void {
+    $tx = $this->connection->startTransaction();
+    $this->connection->transactionManager()->success = FALSE;
+    try {
+      $tx->commitOrRelease();
+    }
+    catch (TransactionCommitFailedException) {
+      // Expected, do nothing.
+    }
+    $this->assertSame(TransactionCommitFailureEvent::class, get_class($this->subscriber->lastEvent));
+    $this->assertSame('drupal_transaction', $this->subscriber->lastEvent->name);
+    $this->assertSame(ClientConnectionTransactionState::CommitFailed, $this->subscriber->lastEvent->transactionState);
+    $this->assertSame(CommitType::Explicit, $this->subscriber->lastEvent->type);
+    $this->assertCount(0, $this->subscriber->lastEvent->stackItems);
+  }
+
+  /**
+   * Tests releasing a savepoint.
+   */
+  public function testTransactionReleaseSavepoint(): void {
+    $tx = $this->connection->startTransaction();
+    $savepoint = $this->connection->startTransaction('foo');
+    $savepoint->commitOrRelease();
+    $this->assertSame(TransactionReleaseSavepointEvent::class, get_class($this->subscriber->lastEvent));
+    $this->assertSame('foo', $this->subscriber->lastEvent->name);
+    $this->assertSame(ClientConnectionTransactionState::Active, $this->subscriber->lastEvent->transactionState);
+    $this->assertSame(CommitType::Explicit, $this->subscriber->lastEvent->type);
+    $this->assertCount(1, $this->subscriber->lastEvent->stackItems);
+    unset($tx);
+  }
+
+  /**
+   * Tests a transaction release savepoint failure.
+   */
+  public function testTransactionReleaseSavepointFailure(): void {
+    $tx = $this->connection->startTransaction();
+    $savepoint = $this->connection->startTransaction('foo');
+    $this->connection->transactionManager()->success = FALSE;
+    $savepoint->commitOrRelease();
+    $this->assertSame(TransactionReleaseSavepointFailureEvent::class, get_class($this->subscriber->lastEvent));
+    $this->assertSame('foo', $this->subscriber->lastEvent->name);
+    $this->assertSame(ClientConnectionTransactionState::CommitFailed, $this->subscriber->lastEvent->transactionState);
+    $this->assertSame(CommitType::Explicit, $this->subscriber->lastEvent->type);
+    $this->assertCount(0, $this->subscriber->lastEvent->stackItems);
+    unset($tx);
+  }
+
+  /**
+   * Tests rolling back a transaction.
+   */
+  public function testTransactionRollback(): void {
+    $tx = $this->connection->startTransaction();
+    $tx->rollback();
+    $this->assertSame(TransactionRollbackEvent::class, get_class($this->subscriber->lastEvent));
+    $this->assertSame('drupal_transaction', $this->subscriber->lastEvent->name);
+    $this->assertSame(ClientConnectionTransactionState::RolledBack, $this->subscriber->lastEvent->transactionState);
+    $this->assertCount(0, $this->subscriber->lastEvent->stackItems);
+  }
+
+  /**
+   * Tests a transaction rollback failure.
+   */
+  public function testTransactionRollbackFailure(): void {
+    $tx = $this->connection->startTransaction();
+    $this->connection->transactionManager()->success = FALSE;
+    $tx->rollback();
+    $this->assertSame(TransactionRollbackFailureEvent::class, get_class($this->subscriber->lastEvent));
+    $this->assertSame('drupal_transaction', $this->subscriber->lastEvent->name);
+    $this->assertSame(ClientConnectionTransactionState::RollbackFailed, $this->subscriber->lastEvent->transactionState);
+    $this->assertCount(0, $this->subscriber->lastEvent->stackItems);
+  }
+
+  /**
+   * Tests rolling back a transaction to a savepoint.
+   */
+  public function testTransactionRollbackToSavepoint(): void {
+    $tx = $this->connection->startTransaction();
+    $savepoint = $this->connection->startTransaction('foo');
+    $savepoint->rollback();
+    $this->assertSame(TransactionRollbackToSavepointEvent::class, get_class($this->subscriber->lastEvent));
+    $this->assertSame('foo', $this->subscriber->lastEvent->name);
+    $this->assertSame(ClientConnectionTransactionState::Active, $this->subscriber->lastEvent->transactionState);
+    $this->assertCount(1, $this->subscriber->lastEvent->stackItems);
+    unset($tx);
+  }
+
+  /**
+   * Tests failure when rolling back a transaction to a savepoint.
+   */
+  public function testTransactionRollbackToSavepointFailure(): void {
+    $tx = $this->connection->startTransaction();
+    $savepoint = $this->connection->startTransaction('foo');
+    $this->connection->transactionManager()->success = FALSE;
+    $savepoint->rollback();
+    $this->assertSame(TransactionRollbackToSavepointFailureEvent::class, get_class($this->subscriber->lastEvent));
+    $this->assertSame('foo', $this->subscriber->lastEvent->name);
+    $this->assertSame(ClientConnectionTransactionState::RollbackFailed, $this->subscriber->lastEvent->transactionState);
+    $this->assertCount(0, $this->subscriber->lastEvent->stackItems);
+    unset($tx);
+  }
+
+}
