$ git diff
diff --git a/core/modules/comment/src/CommentLinkBuilder.php b/core/modules/comment/src/CommentLinkBuilder.php
index 4725f88237..a220399bc8 100644
--- a/core/modules/comment/src/CommentLinkBuilder.php
+++ b/core/modules/comment/src/CommentLinkBuilder.php
@@ -75,6 +75,7 @@ public function __construct(AccountInterface $current_user, CommentManagerInterf
    * {@inheritdoc}
    */
   public function buildCommentedEntityLinks(FieldableEntityInterface $entity, array &$context) {
+dump(__METHOD__);
     $entity_links = [];
     $view_mode = $context['view_mode'];
     if ($view_mode == 'search_index' || $view_mode == 'search_result' || $view_mode == 'print' || $view_mode == 'rss') {
@@ -87,6 +88,7 @@ public function buildCommentedEntityLinks(FieldableEntityInterface $entity, arra
     }

     $fields = $this->commentManager->getFields($entity->getEntityTypeId());
+dump($view_mode, array_keys($fields));
     foreach ($fields as $field_name => $detail) {
       // Skip fields that the entity does not have.
       if (!$entity->hasField($field_name)) {
@@ -94,6 +96,7 @@ public function buildCommentedEntityLinks(FieldableEntityInterface $entity, arra
       }
       $links = [];
       $commenting_status = $entity->get($field_name)->status;
+dump($commenting_status);
       if ($commenting_status != CommentItemInterface::HIDDEN) {
         // Entity has commenting status open or closed.
         $field_definition = $entity->getFieldDefinition($field_name);
@@ -186,6 +189,7 @@ public function buildCommentedEntityLinks(FieldableEntityInterface $entity, arra
           }
         }
       }
+dump($links);

       if (!empty($links)) {
         $entity_links['comment__' . $field_name] = [

