diff --git a/core/lib/Drupal/Core/Test/FunctionalTestSetupTrait.php b/core/lib/Drupal/Core/Test/FunctionalTestSetupTrait.php
index fc4eed64cc..3cc4d2a0c2 100644
--- a/core/lib/Drupal/Core/Test/FunctionalTestSetupTrait.php
+++ b/core/lib/Drupal/Core/Test/FunctionalTestSetupTrait.php
@@ -530,7 +530,7 @@ protected function installParameters() {
       unset($formInput['host']);
       unset($formInput['port']);
     }
-
+dump([$driverNamespace, $formInput]);
     $parameters = [
       'interactive' => FALSE,
       'parameters' => [
