diff --git a/tests/src/Functional/Effect/RelativeCropTest.php b/tests/src/Functional/Effect/RelativeCropTest.php
index 1287279..5305e90 100644
--- a/tests/src/Functional/Effect/RelativeCropTest.php
+++ b/tests/src/Functional/Effect/RelativeCropTest.php
@@ -157,7 +157,7 @@ class RelativeCropTest extends ImageEffectsTestBase {
         '#width' => $image->getWidth(),
         '#height' => $image->getHeight(),
       ];
-      $this->assertEquals('<img src="' . $derivative_url . '" width="' . $test['expected_width'] . '" height="' . $test['expected_height'] . '" alt="" class="image-style-image-effects-test" />', $this->getImageTag($variables), $test_description);
+      $this->assertRegExp("/\<img src=\"" . preg_quote($derivative_url) . "\" width=\"{$test['expected_width']}\" height=\"{$test['expected_height']}\" alt=\"\" .*class=\"image-style-image-effects-test\" \/\>/", $this->getImageTag($variables), $test_description);
 
       // Check that ::applyEffect generates an image with the expected
       // dimensions.
diff --git a/tests/src/Functional/Effect/SmartCropTest.php b/tests/src/Functional/Effect/SmartCropTest.php
index d037b50..f3aefb7 100644
--- a/tests/src/Functional/Effect/SmartCropTest.php
+++ b/tests/src/Functional/Effect/SmartCropTest.php
@@ -75,7 +75,9 @@ class SmartCropTest extends ImageEffectsTestBase {
         '#width' => $image->getWidth(),
         '#height' => $image->getHeight(),
       ];
-      $this->assertEquals('<img src="' . $derivative_url . '" width="' . $test['expected_width'] . '" height="' . $test['expected_height'] . '" alt="" class="image-style-image-effects-test" />', $this->getImageTag($variables));
+//      $xx = "/\<img src=\"" . preg_quote($derivative_url) . "\" width=\"{$test['expected_width']}\" height=\"{$test['expected_height']}\" alt=\"\" .*class=\"image-style-image-effects-test\" \/\>/";
+//      $this->assertEquals($xx, $this->getImageTag($variables));
+      $this->assertRegExp("/src=\"" . preg_quote($derivative_url) . "\" width=\"{$test['expected_width']}\" height=\"{$test['expected_height']}\" alt=\"\" .*class=\"image\-style\-image\-effects\-test\" /", $this->getImageTag($variables));
 
       // Check that ::applyEffect generates image with expected dimensions.
       $this->testImageStyle->createDerivative($original_uri, $derivative_uri);
