diff --git a/core/tests/Drupal/Tests/Listeners/DrupalListener.php b/core/tests/Drupal/Tests/Listeners/DrupalListener.php
index ec1eb30ec1..738512ce0b 100644
--- a/core/tests/Drupal/Tests/Listeners/DrupalListener.php
+++ b/core/tests/Drupal/Tests/Listeners/DrupalListener.php
@@ -51,6 +51,7 @@ public function addSkippedTest(Test $test, \Throwable $t, float $time): void {
    */
   public function startTest(Test $test): void {
     $this->symfonyListener->startTest($test);
+dump($test);
     // Check for incorrect visibility of the $modules property.
     $class = new \ReflectionClass($test);
     if ($class->hasProperty('modules') && !$class->getProperty('modules')->isProtected()) {
