diff --git a/core/tests/Drupal/KernelTests/Core/Test/TestRunTest.php b/core/tests/Drupal/KernelTests/Core/Test/TestRunTest.php
index 4db17e0e..843a5f03 100644
--- a/core/tests/Drupal/KernelTests/Core/Test/TestRunTest.php
+++ b/core/tests/Drupal/KernelTests/Core/Test/TestRunTest.php
@@ -4,6 +4,7 @@
 
 use Drupal\Core\Database\Database;
 use Drupal\Core\Test\JUnitConverter;
+use Drupal\Core\Test\PhpUnitTestRunner;
 use Drupal\Core\Test\TestRun;
 use Drupal\Core\Test\SimpletestTestRunResultsStorage;
 use Drupal\KernelTests\KernelTestBase;
