diff --git a/core/modules/migrate/tests/src/Kernel/Plugin/MigrationPluginListTest.php b/core/modules/migrate/tests/src/Kernel/Plugin/MigrationPluginListTest.php
index 9dc4f88eae..d3ed0f9a11 100644
--- a/core/modules/migrate/tests/src/Kernel/Plugin/MigrationPluginListTest.php
+++ b/core/modules/migrate/tests/src/Kernel/Plugin/MigrationPluginListTest.php
@@ -129,6 +129,7 @@ public function testGetDefinitions() {
 
     // Set up a migrate database connection so that plugin discovery works.
     // Clone the current connection and replace the current prefix.
+dump(['before', Database::getAllConnectionInfo()]);
     $connection_info = Database::getConnectionInfo('migrate');
     if ($connection_info) {
       Database::renameConnection('migrate', 'simpletest_original_migrate');
@@ -145,6 +146,7 @@ public function testGetDefinitions() {
       $connection_info[$target]['prefix'][$value['prefix']['default']] = $value['prefix']['default'];
     }
     Database::addConnectionInfo('migrate', 'default', $connection_info['default']);
+dump(['after', Database::getAllConnectionInfo()]);
 
     $migration_plugins = $this->container->get('plugin.manager.migration')->getDefinitions();
     // All the plugins provided by core depend on migrate_drupal.
