diff --git a/core/lib/Drupal/Core/TypedData/Validation/RecursiveContextualValidator.php b/core/lib/Drupal/Core/TypedData/Validation/RecursiveContextualValidator.php
index f0b885a10c..792ed88097 100644
--- a/core/lib/Drupal/Core/TypedData/Validation/RecursiveContextualValidator.php
+++ b/core/lib/Drupal/Core/TypedData/Validation/RecursiveContextualValidator.php
@@ -177,6 +177,7 @@ protected function validateNode(TypedDataInterface $data, $constraints = NULL, $
    *   The constraints which should be ensured for the given value.
    */
   protected function validateConstraints($value, $cache_key, $constraints) {
+dump([__METHOD__, $constraints]);
     foreach ($constraints as $constraint) {
       // Prevent duplicate validation of constraints, in the case
       // that constraints belong to multiple validated groups
diff --git a/core/lib/Drupal/Core/Validation/Plugin/Validation/Constraint/RangeConstraint.php b/core/lib/Drupal/Core/Validation/Plugin/Validation/Constraint/RangeConstraint.php
index 161aa10db2..dbf3c79e3c 100644
--- a/core/lib/Drupal/Core/Validation/Plugin/Validation/Constraint/RangeConstraint.php
+++ b/core/lib/Drupal/Core/Validation/Plugin/Validation/Constraint/RangeConstraint.php
@@ -19,7 +19,19 @@
  */
 class RangeConstraint extends Range {
 
-  public $minMessage = 'This value should be %limit or more.';
-  public $maxMessage = 'This value should be %limit or less.';
+  /**
+   * {@inheritdoc}
+   */
+  public function __construct(array $options = NULL) {
+dump($options);
+    if (isset($options['min']) && isset($options['max'])) {
+      $options['notInRangeMessage'] = $options['notInRangeMessage'] ?? 'This value should be between %min and %max.';
+    }
+    else {
+      $options['minMessage'] = $options['minMessage'] ?? 'This value should be %limit or more.';
+      $options['maxMessage'] = $options['maxMessage'] ?? 'This value should be %limit or less.';
+    }
+    parent::__construct($options);
+  }
 
 }
diff --git a/core/lib/Drupal/Core/Validation/Plugin/Validation/Constraint/RangeConstraintValidator.php b/core/lib/Drupal/Core/Validation/Plugin/Validation/Constraint/RangeConstraintValidator.php
index b17a57d91a..2d1f956e05 100644
--- a/core/lib/Drupal/Core/Validation/Plugin/Validation/Constraint/RangeConstraintValidator.php
+++ b/core/lib/Drupal/Core/Validation/Plugin/Validation/Constraint/RangeConstraintValidator.php
@@ -50,7 +50,21 @@ public function validate($value, Constraint $constraint) {
       }
     }
 
-    if (NULL !== $constraint->max && $value > $max) {
+    $hasLowerLimit = NULL !== $constraint->min;
+    $hasUpperLimit = NULL !== $constraint->max;
+
+    if ($hasLowerLimit && $hasUpperLimit && ($value < $min || $value > $max)) {
+      $this->context->buildViolation($constraint->notInRangeMessage)
+        ->setParameter('{{ value }}', $this->formatValue($value, self::PRETTY_DATE))
+        ->setParameter('{{ min }}', $this->formatValue($min, self::PRETTY_DATE))
+        ->setParameter('{{ max }}', $this->formatValue($max, self::PRETTY_DATE))
+        ->setCode(Range::NOT_IN_RANGE_ERROR)
+        ->addViolation();
+
+        return;
+    }
+
+    if ($hasUpperLimit && $value > $max) {
       $this->context->buildViolation($constraint->maxMessage)
         ->setParameter('{{ value }}', $this->formatValue($value, self::PRETTY_DATE))
         ->setParameter('{{ limit }}', $this->formatValue($max, self::PRETTY_DATE))
@@ -60,7 +74,7 @@ public function validate($value, Constraint $constraint) {
       return;
     }
 
-    if (NULL !== $constraint->min && $value < $min) {
+    if ($hasLowerLimit && $value < $min) {
       $this->context->buildViolation($constraint->minMessage)
         ->setParameter('{{ value }}', $this->formatValue($value, self::PRETTY_DATE))
         ->setParameter('{{ limit }}', $this->formatValue($min, self::PRETTY_DATE))
