diff --git a/core/modules/views/src/Plugin/views/query/Sql.php b/core/modules/views/src/Plugin/views/query/Sql.php
index 035df8b593..683a4aa8c1 100644
--- a/core/modules/views/src/Plugin/views/query/Sql.php
+++ b/core/modules/views/src/Plugin/views/query/Sql.php
@@ -1380,9 +1380,9 @@ public function query($get_count = FALSE) {
         // Handle group by of field without table alias to avoid ambiguous
         // column error.
         if ($field == $this->view->storage->get('base_field')) {
-          $field = $this->view->storage->get('base_table') . '.' . $field;
+          $field = $this->view->storage->get('base_table') . '].[' . $field;
         }
-        $query->groupBy($field);
+        $query->groupBy('[' . $field . ']');
       }
       if (!empty($this->having) && $condition = $this->buildCondition('having')) {
         $query->havingCondition($condition);
@@ -1422,6 +1422,7 @@ public function query($get_count = FALSE) {
     // Add all query substitutions as metadata.
     $query->addMetaData('views_substitutions', \Drupal::moduleHandler()->invokeAll('views_query_substitutions', [$this->view]));
 
+dump((string) $query);
     return $query;
   }
 
diff --git a/core/modules/views/tests/src/Kernel/QueryGroupByTest.php b/core/modules/views/tests/src/Kernel/QueryGroupByTest.php
index 171688321e..f5fb1af4b0 100644
--- a/core/modules/views/tests/src/Kernel/QueryGroupByTest.php
+++ b/core/modules/views/tests/src/Kernel/QueryGroupByTest.php
@@ -230,7 +230,7 @@ public function testGroupByBaseField() {
     $view->displayHandlers->get('default')->options['fields']['name']['group_type'] = 'min';
     unset($view->displayHandlers->get('default')->options['fields']['id']['group_type']);
     $this->executeView($view);
-    $this->assertStringContainsString('GROUP BY entity_test.id', (string) $view->build_info['query'], 'GROUP BY field includes the base table name when grouping on the base field.');
+    $this->assertMatchesRegularExpression('/GROUP BY .entity_test.\..id./', (string) $view->build_info['query'], 'GROUP BY field includes the base table name when grouping on the base field.');
   }
 
   /**
