diff --git a/core/tests/Drupal/Tests/Listeners/DrupalListener.php b/core/tests/Drupal/Tests/Listeners/DrupalListener.php
index 663c2e52..2704e5bd 100644
--- a/core/tests/Drupal/Tests/Listeners/DrupalListener.php
+++ b/core/tests/Drupal/Tests/Listeners/DrupalListener.php
@@ -22,6 +22,11 @@ class DrupalListener implements TestListener {
    * {@inheritdoc}
    */
   public function startTest(Test $test): void {
+    $class = new \ReflectionClass($test);
+    if ($class->getMethod('setUp')->getDeclaringClass()->getName() !== get_class($test)) {
+dump(['declaring class', $class->getMethod('setUp')->getDeclaringClass()->getName()]);
+dump(['return type', $class->getMethod('setUp')->getReturnType()]);
+    }
     $this->deprecationStartTest($test);
   }
 
