diff --git a/core/tests/Drupal/Tests/Listeners/DrupalStandardsListenerTrait.php b/core/tests/Drupal/Tests/Listeners/DrupalStandardsListenerTrait.php
index a13af0e0b8..7577dad005 100644
--- a/core/tests/Drupal/Tests/Listeners/DrupalStandardsListenerTrait.php
+++ b/core/tests/Drupal/Tests/Listeners/DrupalStandardsListenerTrait.php
@@ -61,9 +61,11 @@ private function checkValidCoversForTest(TestCase $test) {
     }
     // Gather our annotations.
     $annotations = Test::parseTestMethodAnnotations(
-      static::class,
+//      static::class,
+      get_class($test),
       $test->getName()
     );
+dump([get_class($test), $test->getName(), $annotations]);
     // Glean the @coversDefaultClass annotation.
     $default_class = '';
     $valid_default_class = FALSE;
