diff --git a/core/modules/comment/tests/src/Unit/CommentLinkBuilderTest.php b/core/modules/comment/tests/src/Unit/CommentLinkBuilderTest.php
index 1b00a4ff02..b4c3ee07d2 100644
--- a/core/modules/comment/tests/src/Unit/CommentLinkBuilderTest.php
+++ b/core/modules/comment/tests/src/Unit/CommentLinkBuilderTest.php
@@ -132,6 +132,7 @@ public function testCommentLinkBuilder(NodeInterface $node, $context, $has_acces
       ->method('isAnonymous')
       ->willReturn($is_anonymous);
     $links = $this->commentLinkBuilder->buildCommentedEntityLinks($node, $context);
+dump([__METHOD__, $expected, $links]);
     if (!empty($expected)) {
       if (!empty($links)) {
         foreach ($expected as $link => $detail) {
@@ -312,6 +313,7 @@ protected function getMockNode($has_field, $comment_status, $form_location, $com
       ->willReturn(1);

     $url = Url::fromRoute('node.view');
+dump([__METHOD__, $url]);
     $node->expects($this->any())
       ->method('toUrl')
       ->willReturn($url);

