diff --git a/core/lib/Drupal/Core/Database/Schema.php b/core/lib/Drupal/Core/Database/Schema.php
index 190f8816fa3..b87f92bd2f0 100644
--- a/core/lib/Drupal/Core/Database/Schema.php
+++ b/core/lib/Drupal/Core/Database/Schema.php
@@ -134,6 +134,7 @@ public function prefixNonTable($table) {
    *   run. The given options will be merged with self::defaultOptions().
    */
   protected function executeDdlStatement(string $sql, array $arguments = [], array $options = []): void {
+dump([__METHOD__, $sql, $arguments, $options]);
     $this->connection->query($sql, $arguments, $options);
     // DDL statements when in a transaction force a commit in some databases.
     // Void the transaction in that case.
diff --git a/core/modules/pgsql/src/Driver/Database/pgsql/Schema.php b/core/modules/pgsql/src/Driver/Database/pgsql/Schema.php
index a4585e15da7..7cf11cccd90 100644
--- a/core/modules/pgsql/src/Driver/Database/pgsql/Schema.php
+++ b/core/modules/pgsql/src/Driver/Database/pgsql/Schema.php
@@ -166,6 +166,7 @@ public function queryTableInformation($table) {
 AND (format_type(pg_attribute.atttypid, pg_attribute.atttypmod) = 'bytea'
 OR pg_get_expr(pg_attrdef.adbin, pg_attribute.attrelid) LIKE 'nextval%')
 EOD;
+dump([__METHOD__, $sql, $quoted_key]);
         $result = $this->connection->query($sql, [
           ':key' => $quoted_key,
         ]);
