diff --git a/core/modules/migrate/tests/src/Kernel/Plugin/MigrationPluginListTest.php b/core/modules/migrate/tests/src/Kernel/Plugin/MigrationPluginListTest.php
index 9dc4f88eae..c855dfb2b2 100644
--- a/core/modules/migrate/tests/src/Kernel/Plugin/MigrationPluginListTest.php
+++ b/core/modules/migrate/tests/src/Kernel/Plugin/MigrationPluginListTest.php
@@ -129,6 +129,7 @@ public function testGetDefinitions() {
 
     // Set up a migrate database connection so that plugin discovery works.
     // Clone the current connection and replace the current prefix.
+//dump(['before', Database::getAllConnectionInfo()]);
     $connection_info = Database::getConnectionInfo('migrate');
     if ($connection_info) {
       Database::renameConnection('migrate', 'simpletest_original_migrate');
@@ -142,9 +143,10 @@ public function testGetDefinitions() {
 
       // Add the original simpletest prefix so SQLite can attach its database.
       // @see \Drupal\Core\Database\Driver\sqlite\Connection::init()
-      $connection_info[$target]['prefix'][$value['prefix']['default']] = $value['prefix']['default'];
+//      $connection_info[$target]['prefix'][$value['prefix']['default']] = $value['prefix']['default'];
     }
     Database::addConnectionInfo('migrate', 'default', $connection_info['default']);
+//dump(['after', Database::getAllConnectionInfo()]);
 
     $migration_plugins = $this->container->get('plugin.manager.migration')->getDefinitions();
     // All the plugins provided by core depend on migrate_drupal.
