diff --git a/composer.json b/composer.json
index 61c5d3735a..48a914793b 100644
--- a/composer.json
+++ b/composer.json
@@ -105,6 +105,11 @@
             "Drupal\\Composer\\": "composer"
         }
     },
+    "autoload-dev": {
+        "psr-4": {
+            "Drupal\\PHPStan\\": "core/tests/PHPStan"
+        }
+    },
     "scripts": {
         "pre-install-cmd": "Drupal\\Composer\\Composer::ensureComposerVersion",
         "pre-update-cmd": "Drupal\\Composer\\Composer::ensureComposerVersion",
diff --git a/core/misc/cspell/dictionary.txt b/core/misc/cspell/dictionary.txt
index 1a46815d4a..1289ee0829 100644
--- a/core/misc/cspell/dictionary.txt
+++ b/core/misc/cspell/dictionary.txt
@@ -10,6 +10,7 @@ ajaxing
 allwords
 alphadecimal
 ampm
+analyse
 anyall
 archiver
 archivers
diff --git a/core/phpstan.neon.dist b/core/phpstan.neon.dist
index 17c3970c21..073baf8af6 100644
--- a/core/phpstan.neon.dist
+++ b/core/phpstan.neon.dist
@@ -44,3 +44,6 @@ parameters:
     - "#Drupal calls should be avoided in classes, use dependency injection instead#"
     - "#^Plugin definitions cannot be altered.#"
     - "#^Class .* extends @internal class#"
+
+rules:
+  - Drupal\PHPStan\Rules\ComponentTestDoesNotExtendCoreTest
diff --git a/core/tests/Drupal/Tests/Component/Assertion/InspectorTest.php b/core/tests/Drupal/Tests/Component/Assertion/InspectorTest.php
index ef3dff30f8..2368fbb4e8 100644
--- a/core/tests/Drupal/Tests/Component/Assertion/InspectorTest.php
+++ b/core/tests/Drupal/Tests/Component/Assertion/InspectorTest.php
@@ -4,7 +4,7 @@
 
 namespace Drupal\Tests\Component\Assertion;
 
-use PHPUnit\Framework\TestCase;
+use Drupal\Tests\UnitTestCase;
 use Drupal\Component\Assertion\Inspector;
 use Symfony\Bridge\PhpUnit\ExpectDeprecationTrait;
 
@@ -12,7 +12,7 @@
  * @coversDefaultClass \Drupal\Component\Assertion\Inspector
  * @group Assertion
  */
-class InspectorTest extends TestCase {
+class InspectorTest extends UnitTestCase {
 
   use ExpectDeprecationTrait;
 
diff --git a/core/tests/Drupal/Tests/Listeners/DrupalComponentTestListenerTrait.php b/core/tests/Drupal/Tests/Listeners/DrupalComponentTestListenerTrait.php
deleted file mode 100644
index 90a5638e2d..0000000000
--- a/core/tests/Drupal/Tests/Listeners/DrupalComponentTestListenerTrait.php
+++ /dev/null
@@ -1,37 +0,0 @@
-<?php
-
-declare(strict_types=1);
-
-namespace Drupal\Tests\Listeners;
-
-use Drupal\KernelTests\KernelTestBase;
-use Drupal\Tests\BrowserTestBase;
-use Drupal\Tests\UnitTestCase;
-use PHPUnit\Framework\AssertionFailedError;
-
-/**
- * Ensures that no component tests are extending a core test base class.
- *
- * @internal
- */
-trait DrupalComponentTestListenerTrait {
-
-  /**
-   * Reacts to the end of a test.
-   *
-   * @param \PHPUnit\Framework\Test $test
-   *   The test object that has ended its test run.
-   * @param float $time
-   *   The time the test took.
-   */
-  protected function componentEndTest($test, $time) {
-    /** @var \PHPUnit\Framework\Test $test */
-    if (str_starts_with($test->toString(), 'Drupal\Tests\Component')) {
-      if ($test instanceof BrowserTestBase || $test instanceof KernelTestBase || $test instanceof UnitTestCase) {
-        $error = new AssertionFailedError('Component tests should not extend a core test base class.');
-        $test->getTestResultObject()->addFailure($test, $error, $time);
-      }
-    }
-  }
-
-}
diff --git a/core/tests/Drupal/Tests/Listeners/DrupalListener.php b/core/tests/Drupal/Tests/Listeners/DrupalListener.php
index 182e4fe4e7..26fa1e5001 100644
--- a/core/tests/Drupal/Tests/Listeners/DrupalListener.php
+++ b/core/tests/Drupal/Tests/Listeners/DrupalListener.php
@@ -18,7 +18,6 @@
 class DrupalListener implements TestListener {
 
   use TestListenerDefaultImplementation;
-  use DrupalComponentTestListenerTrait;
 
   /**
    * The wrapped Symfony test listener.
@@ -60,7 +59,6 @@ public function startTest(Test $test): void {
    */
   public function endTest(Test $test, float $time): void {
     $this->symfonyListener->endTest($test, $time);
-    $this->componentEndTest($test, $time);
   }
 
 }
diff --git a/core/tests/Drupal/Tests/PHPStan/Rules/ComponentTestDoesNotExtendCoreTestTest.php b/core/tests/Drupal/Tests/PHPStan/Rules/ComponentTestDoesNotExtendCoreTestTest.php
new file mode 100644
index 0000000000..6f7e5e124f
--- /dev/null
+++ b/core/tests/Drupal/Tests/PHPStan/Rules/ComponentTestDoesNotExtendCoreTestTest.php
@@ -0,0 +1,73 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Drupal\Tests\PHPStan\Rules;
+
+use Drupal\PHPStan\Rules\ComponentTestDoesNotExtendCoreTest;
+use PHPStan\Rules\Rule;
+use PHPStan\Testing\RuleTestCase;
+use Symfony\Component\ErrorHandler\DebugClassLoader;
+
+/**
+ * Tests ComponentTestDoesNotExtendCoreTest rule.
+ *
+ * @group phpstan
+ */
+class ComponentTestDoesNotExtendCoreTestTest extends RuleTestCase {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setUp(): void {
+    parent::setUp();
+    DebugClassLoader::disable();
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function tearDown(): void {
+    DebugClassLoader::enable();
+    parent::tearDown();
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function getRule(): Rule {
+    return new ComponentTestDoesNotExtendCoreTest();
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function testRule(): void {
+    $this->analyse(
+      [__DIR__ . '../../../../../fixtures/PHPStan/component-tests.php'],
+      [
+        [
+          'Component tests should not extend Drupal\Tests\UnitTestCase.',
+          19,
+        ],
+        [
+          'Component tests should not extend Drupal\BuildTests\Framework\BuildTestBase.',
+          22,
+        ],
+        [
+          'Component tests should not extend Drupal\KernelTests\KernelTestBase.',
+          25,
+        ],
+        [
+          'Component tests should not extend Drupal\Tests\BrowserTestBase.',
+          28,
+        ],
+        [
+          'Component tests should not extend Drupal\Tests\BrowserTestBase.',
+          31,
+        ],
+      ]
+    );
+  }
+
+}
diff --git a/core/tests/PHPStan/Rules/ComponentTestDoesNotExtendCoreTest.php b/core/tests/PHPStan/Rules/ComponentTestDoesNotExtendCoreTest.php
new file mode 100644
index 0000000000..169608ebe8
--- /dev/null
+++ b/core/tests/PHPStan/Rules/ComponentTestDoesNotExtendCoreTest.php
@@ -0,0 +1,63 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Drupal\PHPStan\Rules;
+
+use Drupal\BuildTests\Framework\BuildTestBase;
+use Drupal\KernelTests\KernelTestBase;
+use Drupal\Tests\BrowserTestBase;
+use Drupal\Tests\UnitTestCase;
+use PhpParser\Node;
+use PHPStan\Analyser\Scope;
+use PHPStan\Node\InClassNode;
+use PHPStan\Rules\Rule;
+use PHPStan\Rules\RuleErrorBuilder;
+
+/**
+ * Ensures that no component tests are extending a core test base class.
+ *
+ * @implements Rule<\PHPStan\Node\InClassNode>
+ *
+ * @internal
+ */
+final class ComponentTestDoesNotExtendCoreTest implements Rule {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getNodeType(): string {
+    return InClassNode::class;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function processNode(Node $node, Scope $scope): array {
+    $class = $node->getClassReflection();
+
+    if (!str_starts_with($class->getName(), 'Drupal\Tests\Component')) {
+      return [];
+    }
+
+    $invalidParents = [
+      UnitTestCase::class,
+      BuildTestBase::class,
+      KernelTestBase::class,
+      BrowserTestBase::class,
+    ];
+
+    foreach ($invalidParents as $invalidParent) {
+      if ($class->isSubclassOf($invalidParent)) {
+        return [
+          RuleErrorBuilder::message("Component tests should not extend {$invalidParent}.")
+            ->line($node->getStartLine())
+            ->build(),
+        ];
+      }
+    }
+
+    return [];
+  }
+
+}
diff --git a/core/tests/fixtures/PHPStan/component-tests.php b/core/tests/fixtures/PHPStan/component-tests.php
new file mode 100644
index 0000000000..6ece93459e
--- /dev/null
+++ b/core/tests/fixtures/PHPStan/component-tests.php
@@ -0,0 +1,32 @@
+<?php
+
+// phpcs:ignoreFile
+
+declare(strict_types=1);
+
+namespace Drupal\Tests\Component\Foo;
+
+use Drupal\BuildTests\Framework\BuildTestBase;
+use Drupal\FunctionalJavascriptTests\WebDriverTestBase;
+use Drupal\KernelTests\KernelTestBase;
+use Drupal\Tests\BrowserTestBase;
+use Drupal\Tests\UnitTestCase;
+use PHPUnit\Framework\TestCase;
+
+final class FooTest extends TestCase {
+}
+
+final class UnitTest extends UnitTestCase {
+}
+
+final class BuildTest extends BuildTestBase {
+}
+
+final class KernelTest extends KernelTestBase {
+}
+
+final class FunctionalTest extends BrowserTestBase {
+}
+
+final class FunctionalJavascriptTest extends WebDriverTestBase {
+}
