diff --git a/core/lib/Drupal/Core/Database/StatementWrapper.php b/core/lib/Drupal/Core/Database/StatementWrapper.php
index 4381fd1af6..1a84bf7831 100644
--- a/core/lib/Drupal/Core/Database/StatementWrapper.php
+++ b/core/lib/Drupal/Core/Database/StatementWrapper.php
@@ -260,7 +260,9 @@ public function fetchAll($mode = NULL, $column_index = NULL, $constructor_argume
    */
   #[\ReturnTypeWillChange]
   public function getIterator() {
-    return new \ArrayIterator($this->fetchAll());
+    while (($row = $this->fetch()) !== FALSE) {
+      yield $row;
+    }
   }
 
 }
