diff --git a/.gitlab-ci/pipeline.yml b/.gitlab-ci/pipeline.yml
index b094571428a..e22ea839859 100644
--- a/.gitlab-ci/pipeline.yml
+++ b/.gitlab-ci/pipeline.yml
@@ -215,6 +215,7 @@ variables:
     - if: $CI_PIPELINE_SOURCE == "parent_pipeline" && $PERFORMANCE_TEST != "1"
       changes:
         - core/tests/PHPStan/*
+        - core/tests/PHPStan/**/*
         - composer/Metapackage/PinnedDevDependencies/composer.json
     - when: manual
       allow_failure: true
@@ -224,7 +225,7 @@ variables:
     - docker-php-ext-enable pcov
     - cd core/tests/PHPStan
     - composer install
-    - vendor/bin/phpunit tests --testdox --coverage-text --colors=never --coverage-cobertura=coverage.cobertura.xml --log-junit junit.xml
+    - vendor/bin/phpunit tests --testdox --coverage-text --colors=always --coverage-cobertura=coverage.cobertura.xml --log-junit junit.xml
   # Default job settings runs a script that junit files in a specific location..
   after_script: []
   artifacts:
diff --git a/core/phpstan.neon.dist b/core/phpstan.neon.dist
index 99f249eb2a5..e981e74b08a 100644
--- a/core/phpstan.neon.dist
+++ b/core/phpstan.neon.dist
@@ -54,5 +54,6 @@ parameters:
 
 rules:
   - Drupal\PHPStan\Rules\ComponentTestDoesNotExtendCoreTest
+  - Drupal\PHPStan\Rules\ReturnTypeWillBeAdded
   - PHPStan\Rules\Functions\MissingFunctionReturnTypehintRule
   - PHPStan\Rules\Methods\MissingMethodReturnTypehintRule
diff --git a/core/tests/PHPStan/Rules/ReturnTypeWillBeAdded.php b/core/tests/PHPStan/Rules/ReturnTypeWillBeAdded.php
new file mode 100644
index 00000000000..114c7f9d658
--- /dev/null
+++ b/core/tests/PHPStan/Rules/ReturnTypeWillBeAdded.php
@@ -0,0 +1,224 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Drupal\PHPStan\Rules;
+
+use PhpParser\Node;
+use PHPStan\Analyser\Scope;
+use PHPStan\Node\InClassMethodNode;
+use PHPStan\PhpDoc\ResolvedPhpDocBlock;
+use PHPStan\PhpDocParser\Ast\PhpDoc\PhpDocTagNode;
+use PHPStan\Rules\Rule;
+use PHPStan\Rules\RuleErrorBuilder;
+use PHPStan\Type\FileTypeMapper;
+use PHPStan\Type\VerbosityLevel;
+use PHPStan\Type\ParserNodeTypeToPHPStanType;
+
+/**
+ * Check methods for future return type addition.
+ *
+ * @implements Rule<\PHPStan\Node\InClassMethodNode>
+ *
+ * @internal
+ */
+final class ReturnTypeWillBeAdded implements Rule {
+
+  public function __construct(
+    private readonly FileTypeMapper $fileTypeMapper,
+  ) {
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getNodeType(): string {
+    return InClassMethodNode::class;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function processNode(Node $node, Scope $scope): array {
+
+    $method = $node->getMethodReflection();
+    $methodName = $method->getName();
+
+    // Skip magic methods.
+    if ($node->getOriginalNode()->isMagic()) {
+      return [];
+    }
+
+    // Check that there are no duplicate '@return-type-will-be-added'
+    // annotations on the method.
+    $resolvedPhpDoc = $this->fileTypeMapper->getResolvedPhpDoc(
+      $scope->getFile(),
+      $scope->isInClass() ? $scope->getClassReflection()->getName() : NULL,
+      $scope->isInTrait() ? $scope->getTraitReflection()->getName() : NULL,
+      $methodName,
+      $method->getDocComment() ?? '',
+    );
+    try {
+      $returnTypeWillBeAddedTag = $this->getReturnTypeWillBeAddedTag($resolvedPhpDoc);
+    }
+    catch (\LogicException $e) {
+      $message = sprintf(
+        "%s::%s() %s",
+        $method->getDeclaringClass()->getName(),
+        $method->getName(),
+        $e->getMessage(),
+      );
+      return [
+        RuleErrorBuilder::message($message)
+          ->line($node->getStartLine())
+          ->identifier('drupal.returnTypeWillBeAdded')
+          ->build(),
+      ];
+    }
+
+    // Determine the native return type of the method.
+    $methodNativeParserReturnType = $node->getOriginalNode()->getReturnType();
+    // @phpstan-ignore phpstanApi.method
+    $methodNativeReturnType = $methodNativeParserReturnType ? ParserNodeTypeToPHPStanType::resolve($methodNativeParserReturnType, $scope->getClassReflection()) : NULL;
+
+    // Determine the PhpDoc return type of the method.
+    $methodPhpDocReturnTag = $resolvedPhpDoc->getReturnTag();
+    $methodPhpDocReturnType = $methodPhpDocReturnTag ? $methodPhpDocReturnTag->getType() : NULL;
+
+    // Annotation @return-type-will-be-added exists, but method already has a
+    // native return type.
+    if ($returnTypeWillBeAddedTag && $methodNativeReturnType) {
+      $message = sprintf(
+        "%s::%s() specifies a '@return-type-will-be-added' annotation, but already has a '%s' native return type.",
+        $method->getDeclaringClass()->getName(),
+        $method->getName(),
+        $methodNativeReturnType->describe(VerbosityLevel::value()),
+      );
+      return [
+        RuleErrorBuilder::message($message)
+          ->line($node->getStartLine())
+          ->identifier('drupal.returnTypeWillBeAdded')
+          ->build(),
+      ];
+    }
+
+    // Annotation @return-type-will-be-added exists, but no @return.
+    if ($returnTypeWillBeAddedTag && $methodPhpDocReturnType === NULL) {
+      $message = sprintf(
+        "%s::%s() specifies a '@return-type-will-be-added' annotation, but a '@return' annotation is missing.",
+        $method->getDeclaringClass()->getName(),
+        $method->getName(),
+      );
+      return [
+        RuleErrorBuilder::message($message)
+          ->line($node->getStartLine())
+          ->identifier('drupal.returnTypeWillBeAdded')
+          ->build(),
+      ];
+    }
+
+    // If method is declared in current class, we are done. We only continue
+    // checks for methods prototyped elsewhere.
+    $prototypeClass = $method->getPrototype()->getDeclaringClass();
+    if ($prototypeClass == $method->getDeclaringClass()) {
+      return [];
+    }
+    $prototypeMethod = $prototypeClass->getMethod($methodName, $scope);
+    $prototypeTrait = method_exists($prototypeMethod, 'getDeclaringTrait') ? $prototypeMethod->getDeclaringTrait() : NULL;
+
+    // Get PHPDoc for the prototype.
+    $resolvedPrototypePhpDoc = $this->fileTypeMapper->getResolvedPhpDoc(
+      $prototypeClass->getFileName(),
+      $prototypeClass->getName(),
+      $prototypeTrait ? $prototypeTrait->getName() : NULL,
+      $methodName,
+      $prototypeMethod->getDocComment() ?? '',
+    );
+
+    // Check for '@return-type-will-be-added' annotation on the prototype.
+    try {
+      $prototypeReturnTypeWillBeAddedTag = $this->getReturnTypeWillBeAddedTag($resolvedPrototypePhpDoc);
+    }
+    catch (\LogicException) {
+      $prototypeReturnTypeWillBeAddedTag = NULL;
+      // Keep going, duplicates were reported on the prototype parsing already.
+    }
+
+    // No annotation in prototype, return.
+    if ($prototypeReturnTypeWillBeAddedTag == NULL) {
+      return [];
+    }
+
+    // Determine the PhpDoc return type of the method.
+    $prototypeMethodPhpDocReturnTag = $resolvedPrototypePhpDoc->getReturnTag();
+    $prototypeMethodPhpDocReturnType = $prototypeMethodPhpDocReturnTag ? $prototypeMethodPhpDocReturnTag->getType() : NULL;
+
+    // Annotation in prototype, method has no native return type.
+    if ($prototypeReturnTypeWillBeAddedTag && $methodNativeReturnType === NULL && $prototypeMethodPhpDocReturnType) {
+      $message = sprintf(
+        "%s::%s() will add '%s' as a native return type declaration %s. Add the return type to the implementation now.",
+        $prototypeTrait ? $prototypeTrait->getName() : $prototypeClass->getName(),
+        $methodName,
+        $prototypeMethodPhpDocReturnType->describe(VerbosityLevel::value()),
+        (string) $prototypeReturnTypeWillBeAddedTag->value ?: 'in the future',
+      );
+      return [
+        RuleErrorBuilder::message($message)
+          ->line($node->getStartLine())
+          ->identifier('drupal.returnTypeWillBeAdded')
+          ->build(),
+      ];
+    }
+
+    // Annotation in prototype, method has a diverging native return type.
+    if ($prototypeReturnTypeWillBeAddedTag &&
+      $methodNativeReturnType &&
+      $prototypeMethodPhpDocReturnType &&
+      !$prototypeMethodPhpDocReturnType->isSuperTypeOf($methodNativeReturnType)->yes()
+    ) {
+      $message = sprintf(
+        "Declaration of %s::%s(): %s must be compatible with %s::%s(): %s that will be added as a native return type declaration %s. Change the return type to the implementation now.",
+        $method->getDeclaringClass()->getName(),
+        $methodName,
+        $methodNativeReturnType->describe(VerbosityLevel::value()),
+        $prototypeTrait ? $prototypeTrait->getName() : $prototypeClass->getName(),
+        $methodName,
+        $prototypeMethodPhpDocReturnType->describe(VerbosityLevel::value()),
+        (string) $prototypeReturnTypeWillBeAddedTag->value ?: 'in the future',
+      );
+      return [
+        RuleErrorBuilder::message($message)
+          ->line($node->getStartLine())
+          ->identifier('drupal.returnTypeWillBeAdded')
+          ->build(),
+      ];
+    }
+
+    return [];
+  }
+
+  /**
+   * Returns the @return-type-will-be-added tag if existing.
+   *
+   * @param \PHPStan\PhpDoc\ResolvedPhpDocBlock $resolvedPhpDoc
+   *   The resolved PHPDoc block.
+   *
+   * @return \PHPStan\PhpDocParser\Ast\PhpDoc\PhpDocTagNode|null
+   *   A PHPDoc tag node containing the @return-type-will-be-added annotation.
+   *
+   * @throws \LogicException
+   *   If multiple @return-type-will-be-added annotations are found in a single
+   *   PHPDoc block.
+   */
+  private function getReturnTypeWillBeAddedTag(ResolvedPhpDocBlock $resolvedPhpDoc): ?PhpDocTagNode {
+    $tags = [];
+    foreach ($resolvedPhpDoc->getPhpDocNodes() as $node) {
+      $tags = array_merge($tags, $node->getTagsByName('@return-type-will-be-added'));
+    }
+    if (($tagsCount = count($tags)) > 1) {
+      throw new \LogicException("'@return-type-will-be-added' annotation should only be defined once. Found {$tagsCount} instances.");
+    }
+    return $tags[0] ?? NULL;
+  }
+
+}
diff --git a/core/tests/PHPStan/fixtures/return-type-will-be-added.php b/core/tests/PHPStan/fixtures/return-type-will-be-added.php
new file mode 100644
index 00000000000..e9b486e09a0
--- /dev/null
+++ b/core/tests/PHPStan/fixtures/return-type-will-be-added.php
@@ -0,0 +1,172 @@
+<?php
+
+// phpcs:ignoreFile
+
+declare(strict_types=1);
+
+namespace Drupal\PHPStanTestFixture\ReturnTypeWillBeAdded;
+
+interface FooInterface {
+
+  /**
+   * @return-type-will-be-added in version 42.0.0
+   * @return bool
+   */
+  public function fooInterfaceValidAnnotationForAbstractClassImplementation();
+
+  /**
+   * @return-type-will-be-added in version 42.0.0
+   * @return bool
+   */
+  public function fooInterfaceValidAnnotationForConcreteClassImplementation();
+
+  /**
+   * @return-type-will-be-added in version 42.0.0
+   * @return bool
+   */
+  public function fooInterfaceValidAnnotationForConcreteClassImplementationOfWrongType();
+
+  /**
+   * @return-type-will-be-added in version 42.0.0
+   * @return-type-will-be-added in version 43.0.0
+   * @return bool
+   */
+  public function fooInterfaceDuplicateAnnotation();
+
+  /**
+   * @return-type-will-be-added in version 42.0.0
+   * @return bool
+   */
+  public function fooInterfaceNativeReturnTypeExistsAlready(): bool;
+
+  /**
+   * @return-type-will-be-added in version 42.0.0
+   */
+  public function fooInterfaceMissingPhpDocReturnType();
+
+}
+
+trait FooTrait {
+
+  /**
+   * @return-type-will-be-added
+   * @return bool
+   */
+  protected function fooTraitValidAnnotationForUseInAbstractClassOverriddenInConcreteClass() {
+    return TRUE;
+  }
+
+  /**
+   * @return-type-will-be-added
+   * @return bool
+   */
+  protected function fooTraitValidAnnotationForUseInAbstractClassOverriddenInConcreteClassWithNativeTypeSpecified() {
+    return TRUE;
+  }
+
+  /**
+   * @return-type-will-be-added in version 42.0.0
+   * @return bool
+   */
+  protected function fooTraitValidAnnotation() {
+    return TRUE;
+  }
+
+}
+
+abstract class FooAbstract implements FooInterface {
+
+  use FooTrait;
+
+  public function fooInterfaceValidAnnotationForAbstractClassImplementation() {
+    return TRUE;
+  }
+
+  /**
+   * Only implemented to respect inheritance.
+   */
+  public function fooInterfaceDuplicateAnnotation() {
+    return TRUE;
+  }
+
+  /**
+   * Only implemented to respect inheritance.
+   */
+  public function fooInterfaceNativeReturnTypeExistsAlready(): bool {
+    return TRUE;
+  }
+
+  /**
+   * Only implemented to respect inheritance.
+   */
+  public function fooInterfaceMissingPhpDocReturnType() {
+    return TRUE;
+  }
+
+  /**
+   * @return-type-will-be-added in version 42.0.0
+   * @return bool
+   */
+  abstract public function fooAbstractValidAnnotation();
+
+  /**
+   * @return-type-will-be-added in version 42.0.0
+   * @return-type-will-be-added in version 43.0.0
+   * @return bool
+   */
+  public function fooAbstractDuplicateAnnotation() {
+    return TRUE;
+  }
+
+  /**
+   * @return-type-will-be-added in version 42.0.0
+   * @return bool
+   */
+  public function fooAbstractNativeReturnTypeExistsAlready(): bool {
+    return TRUE;
+  }
+
+  /**
+   * @return-type-will-be-added in version 42.0.0
+   */
+  public function fooAbstractMissingPhpDocReturnType() {
+    return TRUE;
+  }
+
+}
+
+class Foo extends FooAbstract {
+
+  public function __construct(
+    private readonly ?int $baz,
+  ) {
+  }
+
+  public function fooInterfaceValidAnnotationForConcreteClassImplementation() {
+    return TRUE;
+  }
+
+  public function fooInterfaceValidAnnotationForConcreteClassImplementationOfWrongType(): string {
+    return 'bar';
+  }
+
+  protected function fooTraitValidAnnotationForUseInAbstractClassOverriddenInConcreteClass() {
+    return TRUE;
+  }
+
+  protected function fooTraitValidAnnotationForUseInAbstractClassOverriddenInConcreteClassWithNativeTypeSpecified(): string {
+    return 'baz';
+  }
+
+  public function fooAbstractValidAnnotation() {
+    return TRUE;
+  }
+
+  /**
+   * @return-type-will-be-added in version 42.0.0
+   */
+  public function fooMissingPhpDocReturnType() {
+    return TRUE;
+  }
+
+}
diff --git a/core/tests/PHPStan/tests/ReturnTypeWillBeAddedTest.php b/core/tests/PHPStan/tests/ReturnTypeWillBeAddedTest.php
new file mode 100644
index 00000000000..8ff5207eb8f
--- /dev/null
+++ b/core/tests/PHPStan/tests/ReturnTypeWillBeAddedTest.php
@@ -0,0 +1,92 @@
+<?php
+
+declare(strict_types=1);
+
+// cspell:ignore analyse
+
+namespace Drupal\PHPStan\Tests;
+
+use Drupal\PHPStan\Rules\ReturnTypeWillBeAdded;
+use PHPStan\Rules\Rule;
+use PHPStan\Testing\RuleTestCase;
+use PHPStan\Type\FileTypeMapper;
+
+/**
+ * Tests ReturnTypeWillBeAdded rule.
+ */
+class ReturnTypeWillBeAddedTest extends RuleTestCase {
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function getRule(): Rule {
+    return new ReturnTypeWillBeAdded(
+      self::getContainer()->getByType(FileTypeMapper::class),
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function testRule(): void {
+    include_once __DIR__ . '/../fixtures/return-type-will-be-added.php';
+    $this->analyse(
+      [__DIR__ . '/../fixtures/return-type-will-be-added.php'],
+      [
+        [
+          "Drupal\\PHPStanTestFixture\\ReturnTypeWillBeAdded\\FooInterface::fooInterfaceDuplicateAnnotation() '@return-type-will-be-added' annotation should only be defined once. Found 2 instances.",
+          34,
+        ],
+        [
+          "Drupal\\PHPStanTestFixture\\ReturnTypeWillBeAdded\\FooInterface::fooInterfaceNativeReturnTypeExistsAlready() specifies a '@return-type-will-be-added' annotation, but already has a 'bool' native return type.",
+          40,
+        ],
+        [
+          "Drupal\\PHPStanTestFixture\\ReturnTypeWillBeAdded\\FooInterface::fooInterfaceMissingPhpDocReturnType() specifies a '@return-type-will-be-added' annotation, but a '@return' annotation is missing.",
+          45,
+        ],
+        [
+          "Drupal\\PHPStanTestFixture\\ReturnTypeWillBeAdded\\FooInterface::fooInterfaceValidAnnotationForAbstractClassImplementation() will add 'bool' as a native return type declaration in version 42.0.0. Add the return type to the implementation now.",
+          81,
+        ],
+        [
+          "Drupal\\PHPStanTestFixture\\ReturnTypeWillBeAdded\\FooAbstract::fooAbstractDuplicateAnnotation() '@return-type-will-be-added' annotation should only be defined once. Found 2 instances.",
+          117,
+        ],
+        [
+          "Drupal\\PHPStanTestFixture\\ReturnTypeWillBeAdded\\FooAbstract::fooAbstractNativeReturnTypeExistsAlready() specifies a '@return-type-will-be-added' annotation, but already has a 'bool' native return type.",
+          125,
+        ],
+        [
+          "Drupal\\PHPStanTestFixture\\ReturnTypeWillBeAdded\\FooAbstract::fooAbstractMissingPhpDocReturnType() specifies a '@return-type-will-be-added' annotation, but a '@return' annotation is missing.",
+          132,
+        ],
+        [
+          "Drupal\\PHPStanTestFixture\\ReturnTypeWillBeAdded\\FooInterface::fooInterfaceValidAnnotationForConcreteClassImplementation() will add 'bool' as a native return type declaration in version 42.0.0. Add the return type to the implementation now.",
+          145,
+        ],
+        [
+          "Declaration of Drupal\\PHPStanTestFixture\\ReturnTypeWillBeAdded\\Foo::fooInterfaceValidAnnotationForConcreteClassImplementationOfWrongType(): string must be compatible with Drupal\\PHPStanTestFixture\\ReturnTypeWillBeAdded\\FooInterface::fooInterfaceValidAnnotationForConcreteClassImplementationOfWrongType(): bool that will be added as a native return type declaration in version 42.0.0. Change the return type to the implementation now.",
+          149,
+        ],
+        [
+          "Drupal\\PHPStanTestFixture\\ReturnTypeWillBeAdded\\FooTrait::fooTraitValidAnnotationForUseInAbstractClassOverriddenInConcreteClass() will add 'bool' as a native return type declaration in the future. Add the return type to the implementation now.",
+          153,
+        ],
+        [
+          "Declaration of Drupal\\PHPStanTestFixture\\ReturnTypeWillBeAdded\\Foo::fooTraitValidAnnotationForUseInAbstractClassOverriddenInConcreteClassWithNativeTypeSpecified(): string must be compatible with Drupal\\PHPStanTestFixture\\ReturnTypeWillBeAdded\\FooTrait::fooTraitValidAnnotationForUseInAbstractClassOverriddenInConcreteClassWithNativeTypeSpecified(): bool that will be added as a native return type declaration in the future. Change the return type to the implementation now.",
+          157,
+        ],
+        [
+          "Drupal\\PHPStanTestFixture\\ReturnTypeWillBeAdded\\FooAbstract::fooAbstractValidAnnotation() will add 'bool' as a native return type declaration in version 42.0.0. Add the return type to the implementation now.",
+          161,
+        ],
+        [
+          "Drupal\\PHPStanTestFixture\\ReturnTypeWillBeAdded\\Foo::fooMissingPhpDocReturnType() specifies a '@return-type-will-be-added' annotation, but a '@return' annotation is missing.",
+          168,
+        ],
+      ],
+    );
+  }
+
+}
