diff --git a/composer.json b/composer.json
index c137ad2ab4..6704d4b913 100644
--- a/composer.json
+++ b/composer.json
@@ -22,8 +22,7 @@
         "drupal/coder": "^8.3.10",
         "instaclick/php-webdriver": "^1.4.1",
         "mikey179/vfsstream": "^1.6.8",
-        "phpunit/phpunit": "^8.5.14 || ^9",
-        "phpspec/prophecy": "^1.12",
+        "phpunit/phpunit": "^8.5.14 || ^9 || dev-master",
         "symfony/css-selector": "^4.4",
         "symfony/phpunit-bridge": "^5.3.0",
         "symfony/var-dumper": "^5.3.0",
@@ -48,7 +47,7 @@
         "preferred-install": "dist",
         "autoloader-suffix": "Drupal9",
         "platform": {
-            "php": "7.3.0"
+            "php": "8.0.0"
         }
     },
     "extra": {
diff --git a/core/tests/Drupal/TestTools/PhpUnitCompatibility/PhpUnit10/TestCompatibilityTrait.php b/core/tests/Drupal/TestTools/PhpUnitCompatibility/PhpUnit10/TestCompatibilityTrait.php
new file mode 100644
index 0000000000..fc1e828758
--- /dev/null
+++ b/core/tests/Drupal/TestTools/PhpUnitCompatibility/PhpUnit10/TestCompatibilityTrait.php
@@ -0,0 +1,20 @@
+<?php
+
+namespace Drupal\TestTools\PhpUnitCompatibility\PhpUnit10;
+
+//use Prophecy\PhpUnit\ProphecyTrait;
+
+// @todo Replace with a proper dependency when we stop supporting PHPUnit 8.
+//if (!trait_exists(ProphecyTrait::class)) {
+//  print "Drupal requires Prophecy PhpUnit when using PHPUnit 9 or greater. Please use 'composer require --dev phpspec/prophecy-phpunit:^2' to ensure that it is present.\n";
+//  exit(1);
+//}
+
+/**
+ * Drupal's forward compatibility layer with multiple versions of PHPUnit.
+ */
+trait TestCompatibilityTrait {
+
+//  use ProphecyTrait;
+
+}
