diff --git a/core/tests/Drupal/KernelTests/Core/Test/Comparator/MarkupInterfaceComparatorTest.php b/core/tests/Drupal/KernelTests/Core/Test/Comparator/MarkupInterfaceComparatorTest.php
new file mode 100644
index 0000000000..c02582c6f7
--- /dev/null
+++ b/core/tests/Drupal/KernelTests/Core/Test/Comparator/MarkupInterfaceComparatorTest.php
@@ -0,0 +1,155 @@
+<?php
+
+namespace Drupal\KernelTests\Core\Test\Comparator;
+
+use Drupal\Component\Render\FormattableMarkup;
+use Drupal\Core\StringTranslation\TranslatableMarkup;
+use Drupal\KernelTests\KernelTestBase;
+use Drupal\TestTools\Comparator\MarkupInterfaceComparator;
+use PHPUnit\Framework\Error\Error;
+use PHPUnit\Framework\Error\Notice;
+use SebastianBergmann\Comparator\Factory;
+use SebastianBergmann\Comparator\ComparisonFailure;
+
+/**
+ * Tests \Drupal\TestTools\Comparator\MarkupInterfaceComparator.
+ *
+ * We need to test the class with a kernel test since casting MarkupInterface
+ * objects to strings can require an initialised container.
+ *
+ * @group Test
+ *
+ * @coversDefaultClass \Drupal\TestTools\Comparator\MarkupInterfaceComparator
+ */
+class MarkupInterfaceComparatorTest extends KernelTestBase {
+
+  /**
+   * @var \Drupal\TestTools\Comparator\MarkupInterfaceComparator
+   */
+  protected $comparator;
+
+  /**
+   * @var \SebastianBergmann\Comparator\Factory
+   */
+  protected $factory;
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function setUp() {
+    parent::setUp();
+    $this->factory = new Factory();
+    $this->comparator = new MarkupInterfaceComparator();
+    $this->comparator->setFactory($this->factory);
+  }
+
+  /**
+   * Provides test data for the comparator.
+   *
+   * @return array
+   *   Each array entry has:
+   *   - test expected value,
+   *   - test actual value,
+   *   - a bool indicating the expected return value of ::accepts,
+   *   - a value indicating the expected result of ::assertEquals, either TRUE
+   *     or a class name of an object thrown.
+   */
+  public function dataSetProvider() {
+    return [
+      'FormattableMarkup vs FormattableMarkup, equal' => [
+        new FormattableMarkup('goldfinger', []),
+        new FormattableMarkup('goldfinger', []),
+        TRUE,
+        TRUE,
+      ],
+      'FormattableMarkup vs FormattableMarkup, not equal' => [
+        new FormattableMarkup('goldfinger', []),
+        new FormattableMarkup('moonraker', []),
+        TRUE,
+        ComparisonFailure::class,
+      ],
+      'FormattableMarkup vs string, equal' => [
+        new FormattableMarkup('goldfinger', []),
+        'goldfinger',
+        TRUE,
+        TRUE,
+      ],
+      'string vs FormattableMarkup, equal' => [
+        'goldfinger',
+        new FormattableMarkup('goldfinger', []),
+        TRUE,
+        TRUE,
+      ],
+      'TranslatableMarkup vs FormattableMarkup, equal' => [
+        new TranslatableMarkup('goldfinger'),
+        new FormattableMarkup('goldfinger', []),
+        TRUE,
+        TRUE,
+      ],
+      'TranslatableMarkup vs string, not equal' => [
+        new TranslatableMarkup('goldfinger'),
+        'moonraker',
+        TRUE,
+        ComparisonFailure::class,
+      ],
+      'TranslatableMarkup vs int, equal' => [
+        new TranslatableMarkup('1234'),
+        1234,
+        TRUE,
+        TRUE,
+      ],
+      'int vs TranslatableMarkup, equal' => [
+        1234,
+        new TranslatableMarkup('1234'),
+        TRUE,
+        TRUE,
+      ],
+      'FormattableMarkup vs array' => [
+        new FormattableMarkup('goldfinger', []),
+        ['goldfinger'],
+        FALSE,
+        Notice::class,
+      ],
+      'stdClass vs TranslatableMarkup' => [
+        (object) ['goldfinger'],
+        new TranslatableMarkup('goldfinger'),
+        FALSE,
+        Error::class,
+      ],
+      'string vs string, equal' => [
+        'goldfinger',
+        'goldfinger',
+        FALSE,
+        TRUE,
+      ],
+    ];
+  }
+
+  /**
+   * @covers ::accepts
+   * @dataProvider dataSetProvider
+   */
+  public function testAccepts($expected, $actual, $accepts_result, $equals_result) {
+    if ($accepts_result) {
+      $this->assertTrue($this->comparator->accepts($expected, $actual));
+    }
+    else {
+      $this->assertFalse($this->comparator->accepts($expected, $actual));
+    }
+  }
+
+  /**
+   * @covers ::assertEquals
+   * @dataProvider dataSetProvider
+   */
+  public function testAssertEquals($expected, $actual, $accepts_result, $equals_result) {
+    try {
+      $this->assertNull($this->comparator->assertEquals($expected, $actual));
+      $this->assertTrue($equals_result);
+    }
+    catch (\Throwable $e) {
+      $this->assertInstanceOf($equals_result, $e);
+    }
+  }
+
+}
diff --git a/core/tests/Drupal/KernelTests/KernelTestBase.php b/core/tests/Drupal/KernelTests/KernelTestBase.php
index 957ed79582..a9999936a0 100644
--- a/core/tests/Drupal/KernelTests/KernelTestBase.php
+++ b/core/tests/Drupal/KernelTests/KernelTestBase.php
@@ -22,6 +22,7 @@
 use Drupal\Tests\PhpunitCompatibilityTrait;
 use Drupal\Tests\RandomGeneratorTrait;
 use Drupal\Tests\TestRequirementsTrait;
+use Drupal\TestTools\Comparator\MarkupInterfaceComparator;
 use PHPUnit\Framework\Exception;
 use PHPUnit\Framework\TestCase;
 use Symfony\Component\DependencyInjection\Reference;
@@ -237,6 +238,9 @@ public static function setUpBeforeClass() {
   protected function setUp() {
     parent::setUp();
 
+    // Allow tests to compare MarkupInterface objects via assertEquals().
+    $this->registerComparator(new MarkupInterfaceComparator());
+
     $this->root = static::getDrupalRoot();
     $this->initFileCache();
     $this->bootEnvironment();
diff --git a/core/tests/Drupal/TestTools/Comparator/MarkupInterfaceComparator.php b/core/tests/Drupal/TestTools/Comparator/MarkupInterfaceComparator.php
new file mode 100644
index 0000000000..71ea80b82d
--- /dev/null
+++ b/core/tests/Drupal/TestTools/Comparator/MarkupInterfaceComparator.php
@@ -0,0 +1,34 @@
+<?php
+
+namespace Drupal\TestTools\Comparator;
+
+use Drupal\Component\Render\MarkupInterface;
+use SebastianBergmann\Comparator\Comparator;
+
+/**
+ * Compares MarkupInterface objects for equality.
+ */
+class MarkupInterfaceComparator extends Comparator {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function accepts($expected, $actual) {
+    // If at least one argument is a MarkupInterface object, we take over and
+    // convert to strings before comparing.
+    return ($expected instanceof MarkupInterface && $actual instanceof MarkupInterface) ||
+      ($expected instanceof MarkupInterface && is_scalar($actual)) ||
+      (is_scalar($expected) && $actual instanceof MarkupInterface);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function assertEquals($expected, $actual, $delta = 0.0, $canonicalize = FALSE, $ignoreCase = FALSE) {
+    $expected_safe = (string) $expected;
+    $actual_safe = (string) $actual;
+    $comparator = $this->factory->getComparatorFor($expected_safe, $actual_safe);
+    $comparator->assertEquals($expected_safe, $actual_safe, $delta, $canonicalize, $ignoreCase);
+  }
+
+}
diff --git a/core/tests/Drupal/TestTools/PhpUnitCompatibility/PhpUnit6/TestCompatibilityTrait.php b/core/tests/Drupal/TestTools/PhpUnitCompatibility/PhpUnit6/TestCompatibilityTrait.php
index 92b1b9ebd7..e6d99e402a 100644
--- a/core/tests/Drupal/TestTools/PhpUnitCompatibility/PhpUnit6/TestCompatibilityTrait.php
+++ b/core/tests/Drupal/TestTools/PhpUnitCompatibility/PhpUnit6/TestCompatibilityTrait.php
@@ -39,18 +39,4 @@ public static function assertFalse($actual, $message = '') {
     }
   }
 
-  /**
-   * {@inheritdoc}
-   */
-  public static function assertEquals($expected, $actual, $message = '', $delta = 0, $maxDepth = 10, $canonicalize = FALSE, $ignoreCase = FALSE) {
-    // Cast objects implementing MarkupInterface to string instead of
-    // relying on PHP casting them to string depending on what they are being
-    // comparing with.
-    if (method_exists(self::class, 'castSafeStrings')) {
-      $expected = self::castSafeStrings($expected);
-      $actual = self::castSafeStrings($actual);
-    }
-    parent::assertEquals($expected, $actual, $message, $delta, $maxDepth, $canonicalize, $ignoreCase);
-  }
-
 }
diff --git a/core/tests/Drupal/TestTools/PhpUnitCompatibility/PhpUnit7/TestCompatibilityTrait.php b/core/tests/Drupal/TestTools/PhpUnitCompatibility/PhpUnit7/TestCompatibilityTrait.php
index 3d10e151a9..5db4beb372 100644
--- a/core/tests/Drupal/TestTools/PhpUnitCompatibility/PhpUnit7/TestCompatibilityTrait.php
+++ b/core/tests/Drupal/TestTools/PhpUnitCompatibility/PhpUnit7/TestCompatibilityTrait.php
@@ -39,18 +39,4 @@ public static function assertFalse($actual, string $message = ''): void {
     }
   }
 
-  /**
-   * {@inheritdoc}
-   */
-  public static function assertEquals($expected, $actual, string $message = '', float $delta = 0, int $maxDepth = 10, bool $canonicalize = FALSE, bool $ignoreCase = FALSE): void {
-    // Cast objects implementing MarkupInterface to string instead of
-    // relying on PHP casting them to string depending on what they are being
-    // comparing with.
-    if (method_exists(self::class, 'castSafeStrings')) {
-      $expected = self::castSafeStrings($expected);
-      $actual = self::castSafeStrings($actual);
-    }
-    parent::assertEquals($expected, $actual, $message, $delta, $maxDepth, $canonicalize, $ignoreCase);
-  }
-
 }
diff --git a/core/tests/Drupal/Tests/BrowserTestBase.php b/core/tests/Drupal/Tests/BrowserTestBase.php
index 96657a93e7..4845bf86bd 100644
--- a/core/tests/Drupal/Tests/BrowserTestBase.php
+++ b/core/tests/Drupal/Tests/BrowserTestBase.php
@@ -17,6 +17,7 @@
 use Drupal\Tests\node\Traits\ContentTypeCreationTrait;
 use Drupal\Tests\node\Traits\NodeCreationTrait;
 use Drupal\Tests\user\Traits\UserCreationTrait;
+use Drupal\TestTools\Comparator\MarkupInterfaceComparator;
 use GuzzleHttp\Cookie\CookieJar;
 use PHPUnit\Framework\TestCase;
 use Psr\Http\Message\RequestInterface;
@@ -391,6 +392,9 @@ protected function registerSessions() {}
   protected function setUp() {
     parent::setUp();
 
+    // Allow tests to compare MarkupInterface objects via assertEquals().
+    $this->registerComparator(new MarkupInterfaceComparator());
+
     $this->setupBaseUrl();
 
     // Install Drupal test site.
