diff --git a/core/tests/Drupal/Tests/Listeners/DrupalListener.php b/core/tests/Drupal/Tests/Listeners/DrupalListener.php
index ec1eb30ec1..7298ed4322 100644
--- a/core/tests/Drupal/Tests/Listeners/DrupalListener.php
+++ b/core/tests/Drupal/Tests/Listeners/DrupalListener.php
@@ -6,6 +6,7 @@
 use PHPUnit\Framework\TestListenerDefaultImplementation;
 use PHPUnit\Framework\Test;
 use PHPUnit\Framework\TestSuite;
+use PHPUnit\Util\Test as TestUtil;
 use Symfony\Bridge\PhpUnit\SymfonyTestsListener;
 
 /**
@@ -56,6 +57,20 @@ public function startTest(Test $test): void {
     if ($class->hasProperty('modules') && !$class->getProperty('modules')->isProtected()) {
       @trigger_error('The ' . get_class($test) . '::$modules property must be declared protected. See https://www.drupal.org/node/2909426', E_USER_DEPRECATED);
     }
+    // Check data provider visibility and scope.
+    [$method] = explode(' ', $test->getName());
+    $annotations = TestUtil::parseTestMethodAnnotations(get_class($test), $method);
+    if (isset($annotations['method']['dataProvider'])) {
+      foreach ($annotations['method']['dataProvider'] as $provider) {
+        $providerMethod = $class->getMethod($provider);
+        if (!$providerMethod->isPublic()) {
+          @trigger_error(get_class($test) . '::' . $provider . '() must be declared public.', E_USER_DEPRECATED);
+        }
+        if (!$providerMethod->isStatic()) {
+          @trigger_error(get_class($test) . '::' . $provider . '() must be declared static.', E_USER_DEPRECATED);
+        }
+      }
+    }
   }
 
   /**
