diff --git a/core/lib/Drupal/Component/Diff/Diff.php b/core/lib/Drupal/Component/Diff/Diff.php
index 289bab5f0e..5bc187965e 100644
--- a/core/lib/Drupal/Component/Diff/Diff.php
+++ b/core/lib/Drupal/Component/Diff/Diff.php
@@ -3,6 +3,8 @@
 namespace Drupal\Component\Diff;
 
 use Drupal\Component\Diff\Engine\DiffEngine;
+use SebastianBergmann\Diff\Differ;
+use SebastianBergmann\Diff\Output\UnifiedDiffOutputBuilder;
 
 /**
  * Class representing a 'diff' between two sequences of strings.
@@ -36,6 +38,9 @@ class Diff {
   public function __construct($from_lines, $to_lines) {
     $eng = new DiffEngine();
     $this->edits = $eng->diff($from_lines, $to_lines);
+dump([$from_lines, $to_lines, $this->edits]);
+    $differ = new Differ(new UnifiedDiffOutputBuilder("--- Removed\n+++ Added\n"));
+dump($differ->diffToArray($from_lines, $to_lines));
     //$this->_check($from_lines, $to_lines);
   }
 
