diff --git a/core/tests/Drupal/Tests/Core/Utility/TokenTest.php b/core/tests/Drupal/Tests/Core/Utility/TokenTest.php
index d9c102f372..5cd8afdced 100644
--- a/core/tests/Drupal/Tests/Core/Utility/TokenTest.php
+++ b/core/tests/Drupal/Tests/Core/Utility/TokenTest.php
@@ -339,6 +339,32 @@ public function testScanDeprecation() {
     $this->assertSame([], $this->token->scan(NULL));
   }
 
+  /**
+   * Tests passing a non-string value to Token::scan().
+   */
+  public function testScanDeprecation0() {
+    $this->expectDeprecation('Calling Drupal\Core\Utility\Token::scan() with a $text parameter of type other than string is deprecated in drupal:10.1.0, a typehint will be added in drupal:11.0.0. See https://www.drupal.org/node/3334317');
+    $this->assertSame([], $this->token->scan(NULL));
+  }
+
+  /**
+   * Tests passing a non-string value to Token::scan().
+   */
+  public function testScanDeprecation1() {
+    $this->assertSame([], $this->token->scan(NULL));
+  }
+
+  /**
+   * Tests passing a non-string value to Token::scan().
+   *
+   * @group legacy
+   */
+  public function testScanDeprecation2() {
+    $this->expectDeprecation('bingo-bongo');
+    $this->expectDeprecation('Calling Drupal\Core\Utility\Token::scan() with a $text parameter of type other than string is deprecated in drupal:10.1.0, a typehint will be added in drupal:11.0.0. See https://www.drupal.org/node/3334317');
+    $this->assertSame([], $this->token->scan(NULL));
+  }
+
   /**
    * Sets up the token library to return site tokens.
    */
