diff --git a/tests/src/Functional/ToolkitImagemagickTest.php b/tests/src/Functional/ToolkitImagemagickTest.php
index b91ce51..8e0c0ad 100644
--- a/tests/src/Functional/ToolkitImagemagickTest.php
+++ b/tests/src/Functional/ToolkitImagemagickTest.php
@@ -5,6 +5,7 @@ namespace Drupal\Tests\imagemagick\Functional;
 use Drupal\Core\File\FileSystemInterface;
 use Drupal\Core\Image\ImageInterface;
 use Drupal\imagemagick\EventSubscriber\ImagemagickEventSubscriber;
+use Drupal\imagemagick\ImagemagickExecArguments;
 use Drupal\Tests\BrowserTestBase;
 use Drupal\Tests\imagemagick\Kernel\ToolkitSetupTrait;
 
@@ -95,6 +96,43 @@ class ToolkitImagemagickTest extends BrowserTestBase {
     $this->assertCount(0, $this->fileSystem->scanDirectory('temporary://', '/ima.*/'), 'No files left in the temporary directory after the Drupal shutdown.');
   }
 
+  /**
+   * Tests that double cropping returns an image of expected size.
+   *
+   * @param string $toolkit_id
+   *   The id of the toolkit to set up.
+   * @param string $toolkit_config
+   *   The config object of the toolkit to set up.
+   * @param array $toolkit_settings
+   *   The settings of the toolkit to set up.
+   *
+   * @dataProvider providerToolkitConfiguration
+   */
+  public function testDoubleCropping($toolkit_id, $toolkit_config, array $toolkit_settings): void {
+    $this->setUpToolkit($toolkit_id, $toolkit_config, $toolkit_settings);
+    $this->prepareImageFileHandling();
+
+    // Prepare a test image.
+    $image = $this->imageFactory->get('public://image-test.png');
+    $image->resize(5833, 3889);
+    $image->save('public://3204498-test.png');
+
+    // Process the image.
+    $test_image = $this->imageFactory->get('public://3204498-test.png');
+    $this->assertSame(5833, $test_image->getWidth());
+    $this->assertSame(3889, $test_image->getHeight());
+    $test_image->crop(0, 1601, 5826, 1456);
+    $test_image->resize(1601, 400);
+    $test_image->crop(100, 0, 1400, 400);
+    dump($test_image->getToolkit()->arguments()->toString(ImagemagickExecArguments::POST_SOURCE));
+    $test_image->save('public://3204498-test-derived.png');
+
+    // Check the resulting image.
+    $derived_test_image = $this->imageFactory->get('public://3204498-test-derived.png');
+    $this->assertSame(5833, $derived_test_image->getWidth());
+    $this->assertSame(3899, $derived_test_image->getHeight());
+  }
+
   /**
    * Test image toolkit operations.
    *
