diff --git a/core/tests/Drupal/FunctionalTests/AssertLegacyTrait.php b/core/tests/Drupal/FunctionalTests/AssertLegacyTrait.php
index 8de2b410..49efd026 100644
--- a/core/tests/Drupal/FunctionalTests/AssertLegacyTrait.php
+++ b/core/tests/Drupal/FunctionalTests/AssertLegacyTrait.php
@@ -352,6 +352,10 @@ protected function assertNoField($field) {
    */
   protected function assertRaw($raw) {
     @trigger_error('AssertLegacyTrait::assertRaw() is deprecated in drupal:8.2.0 and is removed from drupal:10.0.0. Use $this->assertSession()->responseContains() instead. See https://www.drupal.org/node/3129738', E_USER_DEPRECATED);
+    if (func_num_args() > 1) {
+      $backtrace = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 2);
+      @trigger_error("Called assertRaw with more that one argument in " . $backtrace[0]['file'] . " at line "  . $backtrace[0]['line'], E_USER_DEPRECATED);
+    }
     $this->assertSession()->responseContains($raw);
   }
 
@@ -370,6 +374,10 @@ protected function assertRaw($raw) {
    */
   protected function assertNoRaw($raw) {
     @trigger_error('AssertLegacyTrait::assertNoRaw() is deprecated in drupal:8.2.0 and is removed from drupal:10.0.0. Use $this->assertSession()->responseNotContains() instead. See https://www.drupal.org/node/3129738', E_USER_DEPRECATED);
+    if (func_num_args() > 1) {
+      $backtrace = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 2);
+      @trigger_error("Called assertNoRaw with more that one argument in " . $backtrace[0]['file'] . " at line "  . $backtrace[0]['line'], E_USER_DEPRECATED);
+    }
     $this->assertSession()->responseNotContains($raw);
   }
 
