diff --git a/core/lib/Drupal/Core/ImageToolkit/ImageToolkitBase.php b/core/lib/Drupal/Core/ImageToolkit/ImageToolkitBase.php
index ce0e283c8e..33874d5c4a 100644
--- a/core/lib/Drupal/Core/ImageToolkit/ImageToolkitBase.php
+++ b/core/lib/Drupal/Core/ImageToolkit/ImageToolkitBase.php
@@ -127,8 +127,8 @@ public function apply($operation, array $arguments = []) {
       $this->logger->error("The selected image handling toolkit '@toolkit' can not process operation '@operation'.", ['@toolkit' => $this->getPluginId(), '@operation' => $operation]);
       return FALSE;
     }
-    catch (\InvalidArgumentException $e) {
-      $this->logger->warning($e->getMessage(), []);
+    catch (\Throwable $t) {
+      $this->logger->warning(get_class($t) . ': ' . $t->getMessage(), []);
       return FALSE;
     }
   }
diff --git a/core/tests/Drupal/KernelTests/Core/Image/ToolkitGdTest.php b/core/tests/Drupal/KernelTests/Core/Image/ToolkitGdTest.php
index b5dddf3959..89d5e3731b 100644
--- a/core/tests/Drupal/KernelTests/Core/Image/ToolkitGdTest.php
+++ b/core/tests/Drupal/KernelTests/Core/Image/ToolkitGdTest.php
@@ -459,6 +459,23 @@ public function testResourceDestruction() {
     $this->assertIsResource($new_res);
   }
 
+  /**
+   * Tests images that will exceed memory available.
+   */
+   public function testInsufficientMemory() {
+     $image = $this->imageFactory->get('core/tests/fixtures/files/image-test.png');
+
+     // Test creation of an image that will exceed the memory available.
+     $old_res = $image->getToolkit()->getResource();
+     $this->assertFalse($image->createNew(200000, 200000));
+     $new_res = $image->getToolkit()->getResource();
+
+     // Check that a new resource has not been created, and the old one is still
+     // valid.
+     $this->assertEquals($old_res, $new_res);
+     $this->assertTrue(is_resource($old_res));
+   }
+
   /**
    * Tests for GIF images with transparency.
    */
