diff --git a/core/.phpstan-baseline.php b/core/.phpstan-baseline.php
index a4b7ee050fb087535b9724dc0d5522ddeaa17135..3a747c497e330563f1654d37d698f8df03302faa 100644
--- a/core/.phpstan-baseline.php
+++ b/core/.phpstan-baseline.php
@@ -12193,12 +12193,6 @@
 	'count' => 1,
 	'path' => __DIR__ . '/modules/block/tests/src/Kernel/BlockConfigSchemaTest.php',
 ];
-$ignoreErrors[] = [
-	'message' => '#^@covers value \\:\\:block_rebuild references an invalid method\\.$#',
-	'identifier' => 'phpunit.coversMethod',
-	'count' => 3,
-	'path' => __DIR__ . '/modules/block/tests/src/Kernel/BlockRebuildTest.php',
-];
 $ignoreErrors[] = [
 	'message' => '#^Method Drupal\\\\Tests\\\\block\\\\Kernel\\\\Plugin\\\\migrate\\\\source\\\\BlockTest\\:\\:providerSource\\(\\) has no return type specified\\.$#',
 	'identifier' => 'missingType.return',
@@ -24541,12 +24535,6 @@
 	'count' => 1,
 	'path' => __DIR__ . '/modules/layout_builder/tests/src/Kernel/DefaultsSectionStorageTest.php',
 ];
-$ignoreErrors[] = [
-	'message' => '#^@covers value layout_builder_entity_view_alter references an invalid class or function\\.$#',
-	'identifier' => 'phpunit.covers',
-	'count' => 1,
-	'path' => __DIR__ . '/modules/layout_builder/tests/src/Kernel/EntityViewAlterTest.php',
-];
 $ignoreErrors[] = [
 	'message' => '#^Method Drupal\\\\Tests\\\\layout_builder\\\\Kernel\\\\FieldBlockTest\\:\\:providerTestBlockAccessEntityAllowedFieldHasValue\\(\\) has no return type specified\\.$#',
 	'identifier' => 'missingType.return',
@@ -28081,18 +28069,6 @@
 	'count' => 1,
 	'path' => __DIR__ . '/modules/migrate/tests/src/Kernel/MigrateEntityContentBaseTest.php',
 ];
-$ignoreErrors[] = [
-	'message' => '#^@dataProvider providerSource related method not found\\.$#',
-	'identifier' => 'phpunit.dataProviderMethod',
-	'count' => 1,
-	'path' => __DIR__ . '/modules/migrate/tests/src/Kernel/MigrateSourceTestBase.php',
-];
-$ignoreErrors[] = [
-	'message' => '#^@dataProvider providerSource related method not found\\.$#',
-	'identifier' => 'phpunit.dataProviderMethod',
-	'count' => 1,
-	'path' => __DIR__ . '/modules/migrate/tests/src/Kernel/MigrateSqlSourceTestBase.php',
-];
 $ignoreErrors[] = [
 	'message' => '#^Method Drupal\\\\Tests\\\\migrate\\\\Kernel\\\\MigrateSqlSourceTestBase\\:\\:register\\(\\) has no return type specified\\.$#',
 	'identifier' => 'missingType.return',
@@ -30517,12 +30493,6 @@
 	'count' => 1,
 	'path' => __DIR__ . '/modules/node/tests/src/Functional/NodeBlockFunctionalTest.php',
 ];
-$ignoreErrors[] = [
-	'message' => '#^@covers value node_local_tasks_alter references an invalid class or function\\.$#',
-	'identifier' => 'phpunit.covers',
-	'count' => 1,
-	'path' => __DIR__ . '/modules/node/tests/src/Functional/NodeRevisionsUiTest.php',
-];
 $ignoreErrors[] = [
 	'message' => '#^Method Drupal\\\\Tests\\\\node\\\\Functional\\\\NodeTestBase\\:\\:assertNodeAccess\\(\\) has no return type specified\\.$#',
 	'identifier' => 'missingType.return',
@@ -42513,12 +42483,6 @@
 	'count' => 1,
 	'path' => __DIR__ . '/modules/update/tests/src/Unit/ProjectCoreCompatibilityTest.php',
 ];
-$ignoreErrors[] = [
-	'message' => '#^@covers value \\\\update_mail references an invalid class or function\\.$#',
-	'identifier' => 'phpunit.covers',
-	'count' => 1,
-	'path' => __DIR__ . '/modules/update/tests/src/Unit/UpdateMailTest.php',
-];
 $ignoreErrors[] = [
 	'message' => '#^Function _update_authorize_clear_update_status\\(\\) has no return type specified\\.$#',
 	'identifier' => 'missingType.return',
@@ -54327,30 +54291,6 @@
 	'count' => 1,
 	'path' => __DIR__ . '/modules/workspaces/tests/src/Kernel/WorkspaceInformationTest.php',
 ];
-$ignoreErrors[] = [
-	'message' => '#^@covers value \\:\\:workspaces_entity_delete references an invalid method\\.$#',
-	'identifier' => 'phpunit.coversMethod',
-	'count' => 1,
-	'path' => __DIR__ . '/modules/workspaces/tests/src/Kernel/WorkspaceIntegrationTest.php',
-];
-$ignoreErrors[] = [
-	'message' => '#^@covers value \\:\\:workspaces_entity_insert references an invalid method\\.$#',
-	'identifier' => 'phpunit.coversMethod',
-	'count' => 1,
-	'path' => __DIR__ . '/modules/workspaces/tests/src/Kernel/WorkspaceIntegrationTest.php',
-];
-$ignoreErrors[] = [
-	'message' => '#^@covers value \\:\\:workspaces_entity_presave references an invalid method\\.$#',
-	'identifier' => 'phpunit.coversMethod',
-	'count' => 1,
-	'path' => __DIR__ . '/modules/workspaces/tests/src/Kernel/WorkspaceIntegrationTest.php',
-];
-$ignoreErrors[] = [
-	'message' => '#^@covers value \\:\\:workspaces_entity_revision_delete references an invalid method\\.$#',
-	'identifier' => 'phpunit.coversMethod',
-	'count' => 1,
-	'path' => __DIR__ . '/modules/workspaces/tests/src/Kernel/WorkspaceIntegrationTest.php',
-];
 $ignoreErrors[] = [
 	'message' => '#^Method Drupal\\\\Tests\\\\workspaces\\\\Kernel\\\\WorkspaceIntegrationTest\\:\\:assertWorkspaceAssociation\\(\\) has no return type specified\\.$#',
 	'identifier' => 'missingType.return',
diff --git a/core/modules/block/tests/src/Kernel/BlockRebuildTest.php b/core/modules/block/tests/src/Kernel/BlockRebuildTest.php
index 70daee8158acfafabf1beb2fc2e20c7e7ed34e29..9b203681bb8117da20c906062ccd701af11e820e 100644
--- a/core/modules/block/tests/src/Kernel/BlockRebuildTest.php
+++ b/core/modules/block/tests/src/Kernel/BlockRebuildTest.php
@@ -46,7 +46,7 @@ protected function setUp(): void {
   }
 
   /**
-   * @covers ::block_rebuild
+   * @covers \Drupal\block\Hook\BlockHooks::rebuild
    */
   public function testRebuildNoBlocks(): void {
     $blockRebuild = new BlockHooks();
@@ -57,7 +57,7 @@ public function testRebuildNoBlocks(): void {
   }
 
   /**
-   * @covers ::block_rebuild
+   * @covers \Drupal\block\Hook\BlockHooks::rebuild
    */
   public function testRebuildNoInvalidBlocks(): void {
     $this->placeBlock('system_powered_by_block', ['region' => 'content']);
@@ -70,7 +70,7 @@ public function testRebuildNoInvalidBlocks(): void {
   }
 
   /**
-   * @covers ::block_rebuild
+   * @covers \Drupal\block\Hook\BlockHooks::rebuild
    */
   public function testRebuildInvalidBlocks(): void {
     $this->placeBlock('system_powered_by_block', ['region' => 'content']);
diff --git a/core/modules/layout_builder/tests/src/Kernel/EntityViewAlterTest.php b/core/modules/layout_builder/tests/src/Kernel/EntityViewAlterTest.php
index ea166c6f1938243a36b5b39b3aec75f6bbd44cb0..25ff022576798f77a88dfedcf9df0f2b42f7239e 100644
--- a/core/modules/layout_builder/tests/src/Kernel/EntityViewAlterTest.php
+++ b/core/modules/layout_builder/tests/src/Kernel/EntityViewAlterTest.php
@@ -15,7 +15,7 @@
 use Drupal\layout_builder\Hook\LayoutBuilderHooks;
 
 /**
- * @covers layout_builder_entity_view_alter
+ * @covers \Drupal\layout_builder\Hook\LayoutBuilderHooks::entityViewAlter
  *
  * @group layout_builder
  */
diff --git a/core/modules/migrate/tests/src/Kernel/MigrateSourceTestBase.php b/core/modules/migrate/tests/src/Kernel/MigrateSourceTestBase.php
index 95d62ce50d499765f85bfb1b50572fd62a9cd16f..7d2a59e5297a9b042ced336f18566cccb4d9ec50 100644
--- a/core/modules/migrate/tests/src/Kernel/MigrateSourceTestBase.php
+++ b/core/modules/migrate/tests/src/Kernel/MigrateSourceTestBase.php
@@ -206,4 +206,12 @@ public function testSource(array $source_data, array $expected_data, $expected_c
     }
   }
 
+  /**
+   * Provides source data for ::testSource.
+   *
+   * @return iterable
+   *   The source data.
+   */
+  abstract public static function providerSource();
+
 }
diff --git a/core/modules/node/tests/src/Functional/NodeRevisionsUiTest.php b/core/modules/node/tests/src/Functional/NodeRevisionsUiTest.php
index 088af231afcde10b41c756581b3a1869d0879317..88fe3e34e3ed4723a2c864167b15a12d781f0f2b 100644
--- a/core/modules/node/tests/src/Functional/NodeRevisionsUiTest.php
+++ b/core/modules/node/tests/src/Functional/NodeRevisionsUiTest.php
@@ -201,7 +201,7 @@ public function testNodeRevisionsTabWithDefaultRevision(): void {
    * This can be removed after 'entity.node.version_history' local task is
    * removed by https://www.drupal.org/project/drupal/issues/3153559.
    *
-   * @covers node_local_tasks_alter
+   * @covers \Drupal\node\Hook\NodeHooks1::localTasksAlter
    */
   public function testNodeDuplicateRevisionsTab(): void {
     $this->drupalPlaceBlock('local_tasks_block');
diff --git a/core/modules/update/tests/src/Unit/UpdateMailTest.php b/core/modules/update/tests/src/Unit/UpdateMailTest.php
index b8caa9e5f675c3fc65e063518695a5d8f4a63150..4481c8badf4ea67cd1ac57887cc245e30f9584be 100644
--- a/core/modules/update/tests/src/Unit/UpdateMailTest.php
+++ b/core/modules/update/tests/src/Unit/UpdateMailTest.php
@@ -12,7 +12,7 @@
 /**
  * Tests text of update email.
  *
- * @covers \update_mail
+ * @covers \Drupal\update\Hook\UpdateHooks::mail
  *
  * @group update
  */
diff --git a/core/modules/workspaces/tests/src/Kernel/WorkspaceIntegrationTest.php b/core/modules/workspaces/tests/src/Kernel/WorkspaceIntegrationTest.php
index b7a117a4f3f9ab9807772f0628b65f1ab7331f98..83a70519eafc5231006aac6855d0a85dd31732ed 100644
--- a/core/modules/workspaces/tests/src/Kernel/WorkspaceIntegrationTest.php
+++ b/core/modules/workspaces/tests/src/Kernel/WorkspaceIntegrationTest.php
@@ -374,10 +374,10 @@ public function testWorkspaces(): void {
   /**
    * Tests the workspace association data integrity for entity CRUD operations.
    *
-   * @covers ::workspaces_entity_presave
-   * @covers ::workspaces_entity_insert
-   * @covers ::workspaces_entity_delete
-   * @covers ::workspaces_entity_revision_delete
+   * @covers \Drupal\workspaces\Hook\EntityOperations::entityPresave
+   * @covers \Drupal\workspaces\Hook\EntityOperations::entityInsert
+   * @covers \Drupal\workspaces\Hook\EntityOperations::entityDelete
+   * @covers \Drupal\workspaces\Hook\EntityOperations::entityRevisionDelete
    */
   public function testWorkspaceAssociationDataIntegrity(): void {
     $this->initializeWorkspacesModule();
