diff --git a/core/lib/Drupal/Core/Form/FormBuilderExtra.php b/core/lib/Drupal/Core/Form/FormBuilderExtra.php
new file mode 100644
index 00000000000..c2f41cb61fb
--- /dev/null
+++ b/core/lib/Drupal/Core/Form/FormBuilderExtra.php
@@ -0,0 +1,11 @@
+<?php
+
+namespace Drupal\Core\Form;
+
+class FormBuilderExtra extends FormBuilder {
+
+  public function getForm($form_arg) {
+    return parent::getForm($form_arg);
+  }
+
+}
diff --git a/core/tests/Drupal/Tests/Core/Form/FormTestBase.php b/core/tests/Drupal/Tests/Core/Form/FormTestBase.php
index f10f79be833..64c5622fabf 100644
--- a/core/tests/Drupal/Tests/Core/Form/FormTestBase.php
+++ b/core/tests/Drupal/Tests/Core/Form/FormTestBase.php
@@ -4,6 +4,7 @@
 
 use Drupal\Component\Utility\Html;
 use Drupal\Core\Form\FormBuilder;
+use Drupal\Core\Form\FormBuilderExtra;
 use Drupal\Core\Form\FormInterface;
 use Drupal\Core\Form\FormState;
 use Drupal\Core\Form\FormStateInterface;
@@ -188,6 +189,7 @@ protected function setUp(): void {
     $this->root = dirname(substr(__DIR__, 0, -strlen(__NAMESPACE__)), 2);
 
     $this->formBuilder = new FormBuilder($this->formValidator, $this->formSubmitter, $this->formCache, $this->moduleHandler, $this->eventDispatcher, $this->requestStack, $this->classResolver, $this->elementInfo, $this->themeManager, $this->csrfToken);
+    $formBuilderExtra = new FormBuilderExtra($this->formValidator, $this->formSubmitter, $this->formCache, $this->moduleHandler, $this->eventDispatcher, $this->requestStack, $this->classResolver, $this->elementInfo, $this->themeManager, $this->csrfToken);
   }
 
   /**
