diff --git a/core/modules/mysql/tests/src/Kernel/mysql/Cache/EndOfTransactionQueriesTest.php b/core/modules/mysql/tests/src/Kernel/mysql/Cache/EndOfTransactionQueriesTest.php
new file mode 100644
index 0000000000..51067a2922
--- /dev/null
+++ b/core/modules/mysql/tests/src/Kernel/mysql/Cache/EndOfTransactionQueriesTest.php
@@ -0,0 +1,28 @@
+<?php
+
+namespace Drupal\Tests\mysql\Kernel\mysql\Cache;
+
+use Drupal\KernelTests\Core\Cache\DriverSpecificEndOfTransactionQueriesTestBase;
+
+/**
+ * Tests that cache tag invalidation queries are delayed to the end of transactions.
+ *
+ * @group Cache
+ */
+class EndOfTransactionQueriesTest extends DriverSpecificEndOfTransactionQueriesTestBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function getDatabaseConnectionInfo() {
+    $info = parent::getDatabaseConnectionInfo();
+    // Override default database driver to one that does logging. Third-party
+    // (non-core) database drivers can achieve the same test coverage by
+    // subclassing this test class and overriding only this method.
+    // @see \Drupal\database_statement_monitoring_test\LoggedStatementsTrait
+    // @see \Drupal\database_statement_monitoring_test\mysql\Connection
+    $info['default']['namespace'] = '\Drupal\database_statement_monitoring_test\mysql';
+    return $info;
+  }
+
+}
diff --git a/core/modules/mysql/tests/src/Kernel/mysql/SchemaTest.php b/core/modules/mysql/tests/src/Kernel/mysql/SchemaTest.php
index f08a04df89..f219b37328 100644
--- a/core/modules/mysql/tests/src/Kernel/mysql/SchemaTest.php
+++ b/core/modules/mysql/tests/src/Kernel/mysql/SchemaTest.php
@@ -43,6 +43,24 @@ protected function assertCollation(): void {
     $this->assertSame('ascii_general_ci', $string_ascii_check, 'test_field_string_ascii should have a ascii_general_ci collation, but it has not.');
   }
 
+  /**
+   * {@inheritdoc}
+   */
+  public function testTableWithSpecificDataType(): void {
+    $table_specification = [
+      'description' => 'Schema table description.',
+      'fields' => [
+        'timestamp'  => [
+          'mysql_type' => 'timestamp',
+          'not null' => FALSE,
+          'default' => NULL,
+        ],
+      ],
+    ];
+    $this->schema->createTable('test_timestamp', $table_specification);
+    $this->assertTrue($this->schema->tableExists('test_timestamp'));
+  }
+
   /**
    * Tests that indexes on string fields are limited to 191 characters on MySQL.
    *
diff --git a/core/modules/mysql/tests/src/Kernel/mysql/SyntaxTest.php b/core/modules/mysql/tests/src/Kernel/mysql/SyntaxTest.php
new file mode 100644
index 0000000000..b5ad81267a
--- /dev/null
+++ b/core/modules/mysql/tests/src/Kernel/mysql/SyntaxTest.php
@@ -0,0 +1,13 @@
+<?php
+
+namespace Drupal\Tests\mysql\Kernel\mysql;
+
+use Drupal\KernelTests\Core\Database\DriverSpecificSyntaxTestBase;
+
+/**
+ * Tests MySql syntax interpretation.
+ *
+ * @group Database
+ */
+class SyntaxTest extends DriverSpecificSyntaxTestBase {
+}
diff --git a/core/modules/mysql/tests/src/Kernel/mysql/TransactionTest.php b/core/modules/mysql/tests/src/Kernel/mysql/TransactionTest.php
new file mode 100644
index 0000000000..d9d5dd513c
--- /dev/null
+++ b/core/modules/mysql/tests/src/Kernel/mysql/TransactionTest.php
@@ -0,0 +1,13 @@
+<?php
+
+namespace Drupal\Tests\mysql\Kernel\mysql;
+
+use Drupal\KernelTests\Core\Database\DriverSpecificTransactionTestBase;
+
+/**
+ * Tests transaction for the MySQL driver.
+ *
+ * @group Database
+ */
+class TransactionTest extends DriverSpecificTransactionTestBase {
+}
diff --git a/core/modules/pgsql/tests/src/Kernel/pgsql/Cache/EndOfTransactionQueriesTest.php b/core/modules/pgsql/tests/src/Kernel/pgsql/Cache/EndOfTransactionQueriesTest.php
new file mode 100644
index 0000000000..3c0882900f
--- /dev/null
+++ b/core/modules/pgsql/tests/src/Kernel/pgsql/Cache/EndOfTransactionQueriesTest.php
@@ -0,0 +1,28 @@
+<?php
+
+namespace Drupal\Tests\pgsql\Kernel\pgsql\Cache;
+
+use Drupal\KernelTests\Core\Cache\DriverSpecificEndOfTransactionQueriesTestBase;
+
+/**
+ * Tests that cache tag invalidation queries are delayed to the end of transactions.
+ *
+ * @group Cache
+ */
+class EndOfTransactionQueriesTest extends DriverSpecificEndOfTransactionQueriesTestBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function getDatabaseConnectionInfo() {
+    $info = parent::getDatabaseConnectionInfo();
+    // Override default database driver to one that does logging. Third-party
+    // (non-core) database drivers can achieve the same test coverage by
+    // subclassing this test class and overriding only this method.
+    // @see \Drupal\database_statement_monitoring_test\LoggedStatementsTrait
+    // @see \Drupal\database_statement_monitoring_test\pgsql\Connection
+    $info['default']['namespace'] = '\Drupal\database_statement_monitoring_test\pgsql';
+    return $info;
+  }
+
+}
diff --git a/core/modules/pgsql/tests/src/Kernel/pgsql/SchemaTest.php b/core/modules/pgsql/tests/src/Kernel/pgsql/SchemaTest.php
index 0c252ec529..9f39c75c07 100644
--- a/core/modules/pgsql/tests/src/Kernel/pgsql/SchemaTest.php
+++ b/core/modules/pgsql/tests/src/Kernel/pgsql/SchemaTest.php
@@ -39,6 +39,24 @@ protected function checkSequenceRenaming(string $tableName): void {
     $this->assertTrue($sequenceExists, 'Sequence was renamed.');
   }
 
+  /**
+   * {@inheritdoc}
+   */
+  public function testTableWithSpecificDataType(): void {
+    $table_specification = [
+      'description' => 'Schema table description.',
+      'fields' => [
+        'timestamp'  => [
+          'pgsql_type' => 'timestamp',
+          'not null' => FALSE,
+          'default' => NULL,
+        ],
+      ],
+    ];
+    $this->schema->createTable('test_timestamp', $table_specification);
+    $this->assertTrue($this->schema->tableExists('test_timestamp'));
+  }
+
   /**
    * @covers \Drupal\pgsql\Driver\Database\pgsql\Schema::introspectIndexSchema
    */
diff --git a/core/modules/pgsql/tests/src/Kernel/pgsql/SyntaxTest.php b/core/modules/pgsql/tests/src/Kernel/pgsql/SyntaxTest.php
new file mode 100644
index 0000000000..2da93a315c
--- /dev/null
+++ b/core/modules/pgsql/tests/src/Kernel/pgsql/SyntaxTest.php
@@ -0,0 +1,13 @@
+<?php
+
+namespace Drupal\Tests\pgsql\Kernel\pgsql;
+
+use Drupal\KernelTests\Core\Database\DriverSpecificSyntaxTestBase;
+
+/**
+ * Tests PostgreSQL syntax interpretation.
+ *
+ * @group Database
+ */
+class SyntaxTest extends DriverSpecificSyntaxTestBase {
+}
diff --git a/core/modules/pgsql/tests/src/Kernel/pgsql/TransactionTest.php b/core/modules/pgsql/tests/src/Kernel/pgsql/TransactionTest.php
new file mode 100644
index 0000000000..f9d7d0ed73
--- /dev/null
+++ b/core/modules/pgsql/tests/src/Kernel/pgsql/TransactionTest.php
@@ -0,0 +1,13 @@
+<?php
+
+namespace Drupal\Tests\pgsql\Kernel\pgsql;
+
+use Drupal\KernelTests\Core\Database\DriverSpecificTransactionTestBase;
+
+/**
+ * Tests transaction for the PostgreSQL driver.
+ *
+ * @group Database
+ */
+class TransactionTest extends DriverSpecificTransactionTestBase {
+}
diff --git a/core/modules/sqlite/tests/src/Kernel/sqlite/Cache/EndOfTransactionQueriesTest.php b/core/modules/sqlite/tests/src/Kernel/sqlite/Cache/EndOfTransactionQueriesTest.php
new file mode 100644
index 0000000000..08d02c32e4
--- /dev/null
+++ b/core/modules/sqlite/tests/src/Kernel/sqlite/Cache/EndOfTransactionQueriesTest.php
@@ -0,0 +1,28 @@
+<?php
+
+namespace Drupal\Tests\sqlite\Kernel\sqlite\Cache;
+
+use Drupal\KernelTests\Core\Cache\DriverSpecificEndOfTransactionQueriesTestBase;
+
+/**
+ * Tests that cache tag invalidation queries are delayed to the end of transactions.
+ *
+ * @group Cache
+ */
+class EndOfTransactionQueriesTest extends DriverSpecificEndOfTransactionQueriesTestBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function getDatabaseConnectionInfo() {
+    $info = parent::getDatabaseConnectionInfo();
+    // Override default database driver to one that does logging. Third-party
+    // (non-core) database drivers can achieve the same test coverage by
+    // subclassing this test class and overriding only this method.
+    // @see \Drupal\database_statement_monitoring_test\LoggedStatementsTrait
+    // @see \Drupal\database_statement_monitoring_test\sqlite\Connection
+    $info['default']['namespace'] = '\Drupal\database_statement_monitoring_test\sqlite';
+    return $info;
+  }
+
+}
diff --git a/core/modules/sqlite/tests/src/Kernel/sqlite/SchemaTest.php b/core/modules/sqlite/tests/src/Kernel/sqlite/SchemaTest.php
index 07b6624196..a3aec37553 100644
--- a/core/modules/sqlite/tests/src/Kernel/sqlite/SchemaTest.php
+++ b/core/modules/sqlite/tests/src/Kernel/sqlite/SchemaTest.php
@@ -5,7 +5,7 @@
 use Drupal\KernelTests\Core\Database\DriverSpecificSchemaTestBase;
 
 /**
- * Tests schema API for the PostgreSQL driver.
+ * Tests schema API for the SQLite driver.
  *
  * @group Database
  */
@@ -26,6 +26,24 @@ protected function tryInsertExpectsIntegrityConstraintViolationException(string
     // Sqlite does not throw an IntegrityConstraintViolationException here.
   }
 
+  /**
+   * {@inheritdoc}
+   */
+  public function testTableWithSpecificDataType(): void {
+    $table_specification = [
+      'description' => 'Schema table description.',
+      'fields' => [
+        'timestamp'  => [
+          'sqlite_type' => 'datetime',
+          'not null' => FALSE,
+          'default' => NULL,
+        ],
+      ],
+    ];
+    $this->schema->createTable('test_timestamp', $table_specification);
+    $this->assertTrue($this->schema->tableExists('test_timestamp'));
+  }
+
   /**
    * @covers \Drupal\sqlite\Driver\Database\sqlite\Schema::introspectIndexSchema
    */
diff --git a/core/modules/sqlite/tests/src/Kernel/sqlite/SyntaxTest.php b/core/modules/sqlite/tests/src/Kernel/sqlite/SyntaxTest.php
new file mode 100644
index 0000000000..fd495ffefe
--- /dev/null
+++ b/core/modules/sqlite/tests/src/Kernel/sqlite/SyntaxTest.php
@@ -0,0 +1,13 @@
+<?php
+
+namespace Drupal\Tests\sqlite\Kernel\sqlite;
+
+use Drupal\KernelTests\Core\Database\DriverSpecificSyntaxTestBase;
+
+/**
+ * Tests SQLite syntax interpretation.
+ *
+ * @group Database
+ */
+class SyntaxTest extends DriverSpecificSyntaxTestBase {
+}
diff --git a/core/modules/sqlite/tests/src/Kernel/sqlite/TransactionTest.php b/core/modules/sqlite/tests/src/Kernel/sqlite/TransactionTest.php
new file mode 100644
index 0000000000..1c69ab1592
--- /dev/null
+++ b/core/modules/sqlite/tests/src/Kernel/sqlite/TransactionTest.php
@@ -0,0 +1,13 @@
+<?php
+
+namespace Drupal\Tests\sqlite\Kernel\sqlite;
+
+use Drupal\KernelTests\Core\Database\DriverSpecificTransactionTestBase;
+
+/**
+ * Tests transaction for the SQLite driver.
+ *
+ * @group Database
+ */
+class TransactionTest extends DriverSpecificTransactionTestBase {
+}
diff --git a/core/tests/Drupal/KernelTests/Core/Cache/EndOfTransactionQueriesTest.php b/core/tests/Drupal/KernelTests/Core/Cache/DriverSpecificEndOfTransactionQueriesTestBase.php
similarity index 88%
rename from core/tests/Drupal/KernelTests/Core/Cache/EndOfTransactionQueriesTest.php
rename to core/tests/Drupal/KernelTests/Core/Cache/DriverSpecificEndOfTransactionQueriesTestBase.php
index 1d735d511d..356d1d7f3a 100644
--- a/core/tests/Drupal/KernelTests/Core/Cache/EndOfTransactionQueriesTest.php
+++ b/core/tests/Drupal/KernelTests/Core/Cache/DriverSpecificEndOfTransactionQueriesTestBase.php
@@ -6,16 +6,14 @@
 use Drupal\Core\Cache\DatabaseBackendFactory;
 use Drupal\Core\DependencyInjection\ContainerBuilder;
 use Drupal\entity_test\Entity\EntityTest;
-use Drupal\KernelTests\KernelTestBase;
+use Drupal\KernelTests\Core\Database\DriverSpecificKernelTestBase;
 use Drupal\user\Entity\User;
 use Symfony\Component\DependencyInjection\Reference;
 
 /**
  * Tests that cache tag invalidation queries are delayed to the end of transactions.
- *
- * @group Cache
  */
-class EndOfTransactionQueriesTest extends KernelTestBase {
+class DriverSpecificEndOfTransactionQueriesTestBase extends DriverSpecificKernelTestBase {
 
   /**
    * {@inheritdoc}
@@ -172,20 +170,4 @@ protected static function statementToTableName($statement) {
     }
   }
 
-  /**
-   * {@inheritdoc}
-   */
-  protected function getDatabaseConnectionInfo() {
-    $info = parent::getDatabaseConnectionInfo();
-    // Override default database driver to one that does logging. Third-party
-    // (non-core) database drivers can achieve the same test coverage by
-    // subclassing this test class and overriding only this method.
-    // @see \Drupal\database_statement_monitoring_test\LoggedStatementsTrait
-    // @see \Drupal\database_statement_monitoring_test\mysql\Connection
-    // @see \Drupal\database_statement_monitoring_test\pgsql\Connection
-    // @see \Drupal\database_statement_monitoring_test\sqlite\Connection
-    $info['default']['namespace'] = '\Drupal\database_statement_monitoring_test\\' . $info['default']['driver'];
-    return $info;
-  }
-
 }
diff --git a/core/tests/Drupal/KernelTests/Core/Database/BasicSyntaxTest.php b/core/tests/Drupal/KernelTests/Core/Database/BasicSyntaxTest.php
index 773ec6a99b..b1b3bb7a45 100644
--- a/core/tests/Drupal/KernelTests/Core/Database/BasicSyntaxTest.php
+++ b/core/tests/Drupal/KernelTests/Core/Database/BasicSyntaxTest.php
@@ -141,33 +141,4 @@ public function testGetFullQualifiedTableName() {
     $this->assertSame('4', $num_matches, 'Found 4 records.');
   }
 
-  /**
-   * Tests allowing square brackets in queries.
-   *
-   * @see \Drupal\Core\Database\Connection::prepareQuery()
-   */
-  public function testAllowSquareBrackets() {
-    $this->connection->insert('test')
-      ->fields(['name'])
-      ->values([
-        'name' => '[square]',
-      ])
-      ->execute();
-
-    // Note that this is a very bad example query because arguments should be
-    // passed in via the $args parameter.
-    $result = $this->connection->query("select name from {test} where name = '[square]'", [], ['allow_square_brackets' => TRUE]);
-    $this->assertSame('[square]', $result->fetchField());
-
-    // Test that allow_square_brackets has no effect on arguments.
-    $result = $this->connection->query("select [name] from {test} where [name] = :value", [':value' => '[square]']);
-    $this->assertSame('[square]', $result->fetchField());
-    $result = $this->connection->query("select name from {test} where name = :value", [':value' => '[square]'], ['allow_square_brackets' => TRUE]);
-    $this->assertSame('[square]', $result->fetchField());
-
-    // Test square brackets using the query builder.
-    $result = $this->connection->select('test')->fields('test', ['name'])->condition('name', '[square]')->execute();
-    $this->assertSame('[square]', $result->fetchField());
-  }
-
 }
diff --git a/core/tests/Drupal/KernelTests/Core/Database/DriverSpecificKernelTestBase.php b/core/tests/Drupal/KernelTests/Core/Database/DriverSpecificKernelTestBase.php
index a84046e7d9..332ebe4a18 100644
--- a/core/tests/Drupal/KernelTests/Core/Database/DriverSpecificKernelTestBase.php
+++ b/core/tests/Drupal/KernelTests/Core/Database/DriverSpecificKernelTestBase.php
@@ -27,11 +27,11 @@ abstract class DriverSpecificKernelTestBase extends KernelTestBase {
    * @inheritdoc
    */
   protected function setUp(): void {
-    parent::setUp();
-    $this->connection = Database::getConnection();
-
-    $running_provider = $this->connection->getProvider();
-    $running_driver = $this->connection->driver();
+    // Find the current SUT database driver from the connection info. If that is
+    // not the one the test requires, skip before test database initialization
+    // so to save cycles.
+    $this->root = static::getDrupalRoot();
+    $connectionInfo = $this->getDatabaseConnectionInfo();
     $test_class_parts = explode('\\', get_class($this));
     $expected_provider = $test_class_parts[2] ?? '';
     for ($i = 3; $i < count($test_class_parts); $i++) {
@@ -40,6 +40,17 @@ protected function setUp(): void {
         break;
       }
     }
+    if ($connectionInfo['default']['driver'] !== $expected_driver) {
+      $this->markTestSkipped("This test only runs for the database driver '$expected_driver'. Current database driver is '{$connectionInfo['default']['driver']}'.");
+    }
+
+    parent::setUp();
+    $this->connection = Database::getConnection();
+
+    // After database initialization, the database driver may be not provided by
+    // the expected module; skip test in that case.
+    $running_provider = $this->connection->getProvider();
+    $running_driver = $this->connection->driver();
     if ($running_provider !== $expected_provider || $running_driver !== $expected_driver) {
       $this->markTestSkipped("This test only runs for the database driver '$expected_driver' provided by the '$expected_provider' module. Connected database driver is '$running_driver' provided by '$running_provider'.");
     }
diff --git a/core/tests/Drupal/KernelTests/Core/Database/DriverSpecificSchemaTestBase.php b/core/tests/Drupal/KernelTests/Core/Database/DriverSpecificSchemaTestBase.php
index d950fc24c4..9098e244f9 100644
--- a/core/tests/Drupal/KernelTests/Core/Database/DriverSpecificSchemaTestBase.php
+++ b/core/tests/Drupal/KernelTests/Core/Database/DriverSpecificSchemaTestBase.php
@@ -308,28 +308,13 @@ public function testSchema(): void {
 
     // Check that the ID sequence gets renamed when the table is renamed.
     $this->checkSequenceRenaming($new_table_name);
-
-    // Use database specific data type and ensure that table is created.
-    $table_specification = [
-      'description' => 'Schema table description.',
-      'fields' => [
-        'timestamp'  => [
-          'mysql_type' => 'timestamp',
-          'pgsql_type' => 'timestamp',
-          'sqlite_type' => 'datetime',
-          'not null' => FALSE,
-          'default' => NULL,
-        ],
-      ],
-    ];
-    try {
-      $this->schema->createTable('test_timestamp', $table_specification);
-    }
-    catch (\Exception $e) {
-    }
-    $this->assertTrue($this->schema->tableExists('test_timestamp'), 'Table with database specific datatype was created.');
   }
 
+  /**
+   * Tests creating a table with database specific data type.
+   */
+  abstract public function testTableWithSpecificDataType(): void;
+
   /**
    * Tests creating unsigned columns and data integrity thereof.
    */
diff --git a/core/tests/Drupal/KernelTests/Core/Database/DriverSpecificSyntaxTestBase.php b/core/tests/Drupal/KernelTests/Core/Database/DriverSpecificSyntaxTestBase.php
new file mode 100644
index 0000000000..c028002734
--- /dev/null
+++ b/core/tests/Drupal/KernelTests/Core/Database/DriverSpecificSyntaxTestBase.php
@@ -0,0 +1,39 @@
+<?php
+
+namespace Drupal\KernelTests\Core\Database;
+
+/**
+ * Tests driver specific SQL syntax interpretation.
+ */
+abstract class DriverSpecificSyntaxTestBase extends DriverSpecificDatabaseTestBase {
+
+  /**
+   * Tests allowing square brackets in queries.
+   *
+   * @see \Drupal\Core\Database\Connection::prepareQuery()
+   */
+  public function testAllowSquareBrackets() {
+    $this->connection->insert('test')
+      ->fields(['name'])
+      ->values([
+        'name' => '[square]',
+      ])
+      ->execute();
+
+    // Note that this is a very bad example query because arguments should be
+    // passed in via the $args parameter.
+    $result = $this->connection->query("select name from {test} where name = '[square]'", [], ['allow_square_brackets' => TRUE]);
+    $this->assertSame('[square]', $result->fetchField());
+
+    // Test that allow_square_brackets has no effect on arguments.
+    $result = $this->connection->query("select [name] from {test} where [name] = :value", [':value' => '[square]']);
+    $this->assertSame('[square]', $result->fetchField());
+    $result = $this->connection->query("select name from {test} where name = :value", [':value' => '[square]'], ['allow_square_brackets' => TRUE]);
+    $this->assertSame('[square]', $result->fetchField());
+
+    // Test square brackets using the query builder.
+    $result = $this->connection->select('test')->fields('test', ['name'])->condition('name', '[square]')->execute();
+    $this->assertSame('[square]', $result->fetchField());
+  }
+
+}
diff --git a/core/tests/Drupal/KernelTests/Core/Database/TransactionTest.php b/core/tests/Drupal/KernelTests/Core/Database/DriverSpecificTransactionTestBase.php
similarity index 99%
rename from core/tests/Drupal/KernelTests/Core/Database/TransactionTest.php
rename to core/tests/Drupal/KernelTests/Core/Database/DriverSpecificTransactionTestBase.php
index 8adc584063..70856aba17 100644
--- a/core/tests/Drupal/KernelTests/Core/Database/TransactionTest.php
+++ b/core/tests/Drupal/KernelTests/Core/Database/DriverSpecificTransactionTestBase.php
@@ -25,10 +25,8 @@
  *       [Maybe decide to roll back]
  *     Do more stuff
  *     Should still be in transaction A
- *
- * @group Database
  */
-class TransactionTest extends DatabaseTestBase {
+class DriverSpecificTransactionTestBase extends DriverSpecificDatabaseTestBase {
 
   /**
    * Encapsulates a transaction's "inner layer" with an "outer layer".
