diff --git a/core/modules/comment/tests/src/Unit/CommentLinkBuilderTest.php b/core/modules/comment/tests/src/Unit/CommentLinkBuilderTest.php
index 1b00a4ff02..c83c1b26ea 100644
--- a/core/modules/comment/tests/src/Unit/CommentLinkBuilderTest.php
+++ b/core/modules/comment/tests/src/Unit/CommentLinkBuilderTest.php
@@ -5,10 +5,12 @@
 use Drupal\comment\CommentLinkBuilder;
 use Drupal\comment\Plugin\Field\FieldType\CommentItemInterface;
 use Drupal\Core\Entity\EntityTypeManagerInterface;
+use Drupal\Core\Field\FieldDefinitionInterface;
 use Drupal\Core\Url;
 use Drupal\node\NodeInterface;
 use Drupal\Tests\Traits\Core\GeneratePermutationsTrait;
 use Drupal\Tests\UnitTestCase;
+use Prophecy\Prophet;
 
 /**
  * @coversDefaultClass \Drupal\comment\CommentLinkBuilder
@@ -55,10 +57,8 @@ class CommentLinkBuilderTest extends UnitTestCase {
 
   /**
    * Timestamp used in test.
-   *
-   * @var int
    */
-  protected $timestamp;
+  protected static int $timestamp;
 
   /**
    * @var \Drupal\comment\CommentLinkBuilderInterface
@@ -131,6 +131,7 @@ public function testCommentLinkBuilder(NodeInterface $node, $context, $has_acces
     $this->currentUser->expects($this->any())
       ->method('isAnonymous')
       ->willReturn($is_anonymous);
+
     $links = $this->commentLinkBuilder->buildCommentedEntityLinks($node, $context);
     if (!empty($expected)) {
       if (!empty($links)) {
@@ -159,11 +160,11 @@ public function testCommentLinkBuilder(NodeInterface $node, $context, $has_acces
   /**
    * Data provider for ::testCommentLinkBuilder.
    */
-  public function getLinkCombinations() {
+  public static function getLinkCombinations() {
     $cases = [];
     // No links should be created if the entity doesn't have the field.
     $cases[] = [
-      $this->getMockNode(FALSE, CommentItemInterface::OPEN, CommentItemInterface::FORM_BELOW, 1),
+      static::getMockNode(FALSE, CommentItemInterface::OPEN, CommentItemInterface::FORM_BELOW, 1),
       ['view_mode' => 'teaser'],
       TRUE,
       TRUE,
@@ -174,7 +175,7 @@ public function getLinkCombinations() {
     foreach (['search_result', 'search_index', 'print'] as $view_mode) {
       // Nothing should be output in these view modes.
       $cases[] = [
-        $this->getMockNode(TRUE, CommentItemInterface::OPEN, CommentItemInterface::FORM_BELOW, 1),
+        static::getMockNode(TRUE, CommentItemInterface::OPEN, CommentItemInterface::FORM_BELOW, 1),
         ['view_mode' => $view_mode],
         TRUE,
         TRUE,
@@ -200,10 +201,10 @@ public function getLinkCombinations() {
         'teaser', 'rss', 'full',
       ],
     ];
-    $permutations = $this->generatePermutations($combinations);
+    $permutations = static::generatePermutations($combinations);
     foreach ($permutations as $combination) {
       $case = [
-        $this->getMockNode(TRUE, $combination['comments'], $combination['form_location'], $combination['comment_count']),
+        static::getMockNode(TRUE, $combination['comments'], $combination['form_location'], $combination['comment_count']),
         ['view_mode' => $combination['view_mode']],
         $combination['has_access_comments'],
         $combination['history_exists'],
@@ -267,11 +268,40 @@ public function getLinkCombinations() {
    * @param int $comment_count
    *   Number of comments against the field.
    *
-   * @return \Drupal\node\NodeInterface|\PHPUnit\Framework\MockObject\MockObject
+   * @return \Drupal\node\NodeInterface
    *   Mock node for testing.
    */
-  protected function getMockNode($has_field, $comment_status, $form_location, $comment_count) {
-    $node = $this->createMock('\Drupal\node\NodeInterface');
+  protected static function getMockNode($has_field, $comment_status, $form_location, $comment_count) {
+    $prophet = new Prophet();
+
+    $node = $prophet->prophesize(NodeInterface::class);
+    $node->hasField()->willReturn($has_field);
+
+    if (!isset(static::$timestamp)) {
+      static::$timestamp = time();
+    }
+    $field_item = (object) [
+      'status' => $comment_status,
+      'comment_count' => $comment_count,
+      'last_comment_timestamp' => static::$timestamp,
+    ];
+    $node->get('comment')->willReturn($field_item);
+
+    $field_definition = $prophet->prophesize(FieldDefinitionInterface::class);
+    $field_definition->getSetting('form_location')->willReturn($form_location);
+    $field_definition->reveal();
+
+    $node->getFieldDefinition('comment')->willReturn($field_definition);
+    $node->language()->willReturn('und');
+    $node->getEntityTypeId()->willReturn('node');
+    $node->id()->willReturn(1);
+
+    $url = Url::fromRoute('node.view');
+    $node->toUrl()->willReturn($url);
+
+    return $node->reveal();
+
+/*    $node = $this->createMock('\Drupal\node\NodeInterface');
     $node->expects($this->any())
       ->method('hasField')
       ->willReturn($has_field);
@@ -314,9 +344,8 @@ protected function getMockNode($has_field, $comment_status, $form_location, $com
     $url = Url::fromRoute('node.view');
     $node->expects($this->any())
       ->method('toUrl')
-      ->willReturn($url);
 
-    return $node;
+      ->willReturn($url);*/
   }
 
 }
diff --git a/core/modules/config_translation/tests/src/Unit/ConfigMapperManagerTest.php b/core/modules/config_translation/tests/src/Unit/ConfigMapperManagerTest.php
index 329647f2df..6aad4e192d 100644
--- a/core/modules/config_translation/tests/src/Unit/ConfigMapperManagerTest.php
+++ b/core/modules/config_translation/tests/src/Unit/ConfigMapperManagerTest.php
@@ -3,11 +3,13 @@
 namespace Drupal\Tests\config_translation\Unit;
 
 use Drupal\config_translation\ConfigMapperManager;
+use Drupal\Core\Config\Schema\Mapping;
 use Drupal\Core\Language\Language;
 use Drupal\Core\Language\LanguageInterface;
 use Drupal\Core\TypedData\TypedDataInterface;
 use Drupal\Tests\UnitTestCase;
 use Drupal\Core\TypedData\DataDefinition;
+use Prophecy\Prophet;
 
 /**
  * Tests the functionality provided by configuration translation mapper manager.
@@ -87,54 +89,54 @@ public function testHasTranslatable(TypedDataInterface $element, $expected) {
    *   to test as the first key and the expected result of
    *   ConfigMapperManager::hasTranslatable() as the second key.
    */
-  public function providerTestHasTranslatable() {
+  public static function providerTestHasTranslatable() {
     return [
-      [$this->getElement([]), FALSE],
-      [$this->getElement(['aaa' => 'bbb']), FALSE],
-      [$this->getElement(['translatable' => FALSE]), FALSE],
-      [$this->getElement(['translatable' => TRUE]), TRUE],
-      [$this->getNestedElement([$this->getElement([])]), FALSE],
-      [$this->getNestedElement([$this->getElement(['translatable' => TRUE])]), TRUE],
+      [static::getElement([]), FALSE],
+      [static::getElement(['aaa' => 'bbb']), FALSE],
+      [static::getElement(['translatable' => FALSE]), FALSE],
+      [static::getElement(['translatable' => TRUE]), TRUE],
+      [static::getNestedElement([static::getElement([])]), FALSE],
+      [static::getNestedElement([static::getElement(['translatable' => TRUE])]), TRUE],
       [
-        $this->getNestedElement([
-          $this->getElement(['aaa' => 'bbb']),
-          $this->getElement(['ccc' => 'ddd']),
-          $this->getElement(['eee' => 'fff']),
+        static::getNestedElement([
+          static::getElement(['aaa' => 'bbb']),
+          static::getElement(['ccc' => 'ddd']),
+          static::getElement(['eee' => 'fff']),
         ]),
         FALSE,
       ],
       [
-        $this->getNestedElement([
-          $this->getElement(['aaa' => 'bbb']),
-          $this->getElement(['ccc' => 'ddd']),
-          $this->getElement(['translatable' => TRUE]),
+        static::getNestedElement([
+          static::getElement(['aaa' => 'bbb']),
+          static::getElement(['ccc' => 'ddd']),
+          static::getElement(['translatable' => TRUE]),
         ]),
         TRUE,
       ],
       [
-        $this->getNestedElement([
-          $this->getElement(['aaa' => 'bbb']),
-          $this->getNestedElement([
-            $this->getElement(['ccc' => 'ddd']),
-            $this->getElement(['eee' => 'fff']),
+        static::getNestedElement([
+          static::getElement(['aaa' => 'bbb']),
+          static::getNestedElement([
+            static::getElement(['ccc' => 'ddd']),
+            static::getElement(['eee' => 'fff']),
           ]),
-          $this->getNestedElement([
-            $this->getElement(['ggg' => 'hhh']),
-            $this->getElement(['iii' => 'jjj']),
+          static::getNestedElement([
+            static::getElement(['ggg' => 'hhh']),
+            static::getElement(['iii' => 'jjj']),
           ]),
         ]),
         FALSE,
       ],
       [
-        $this->getNestedElement([
-          $this->getElement(['aaa' => 'bbb']),
-          $this->getNestedElement([
-            $this->getElement(['ccc' => 'ddd']),
-            $this->getElement(['eee' => 'fff']),
+        static::getNestedElement([
+          static::getElement(['aaa' => 'bbb']),
+          static::getNestedElement([
+            static::getElement(['ccc' => 'ddd']),
+            static::getElement(['eee' => 'fff']),
           ]),
-          $this->getNestedElement([
-            $this->getElement(['ggg' => 'hhh']),
-            $this->getElement(['translatable' => TRUE]),
+          static::getNestedElement([
+            static::getElement(['ggg' => 'hhh']),
+            static::getElement(['translatable' => TRUE]),
           ]),
         ]),
         TRUE,
@@ -151,13 +153,11 @@ public function providerTestHasTranslatable() {
    * @return \Drupal\Core\Config\Schema\Element
    *   The mocked schema element.
    */
-  protected function getElement(array $definition) {
+  protected static function getElement(array $definition) {
     $data_definition = new DataDefinition($definition);
-    $element = $this->createMock('Drupal\Core\TypedData\TypedDataInterface');
-    $element->expects($this->any())
-      ->method('getDataDefinition')
-      ->willReturn($data_definition);
-    return $element;
+    $element = (new Prophet())->prophesize(TypedDataInterface::class);
+    $element->getDataDefinition()->willReturn($data_definition);
+    return $element->reveal();
   }
 
   /**
@@ -169,20 +169,16 @@ protected function getElement(array $definition) {
    * @return \Drupal\Core\Config\Schema\Mapping
    *   A nested schema element, containing the passed-in elements.
    */
-  protected function getNestedElement(array $elements) {
+  protected static function getNestedElement(array $elements) {
     // ConfigMapperManager::findTranslatable() checks for
     // \Drupal\Core\TypedData\TraversableTypedDataInterface, but mocking that
     // directly does not work, because we need to implement \IteratorAggregate
     // in order for getIterator() to be called. Therefore we need to mock
     // \Drupal\Core\Config\Schema\ArrayElement, but that is abstract, so we
     // need to mock one of the subclasses of it.
-    $nested_element = $this->getMockBuilder('Drupal\Core\Config\Schema\Mapping')
-      ->disableOriginalConstructor()
-      ->getMock();
-    $nested_element->expects($this->once())
-      ->method('getIterator')
-      ->willReturn(new \ArrayIterator($elements));
-    return $nested_element;
+    $nested_element = (new Prophet())->prophesize(Mapping::class);
+    $nested_element->getIterator()->shouldBeCalledTimes(1)->willReturn(new \ArrayIterator($elements));
+    return $nested_element->reveal();
   }
 
 }
