diff --git a/core/.cspell.json b/core/.cspell.json
index 56df9c720214e7027c15b4b91d2c0a9d893b0250..22b94fc48b77dd788365dc8545736a4065310458 100644
--- a/core/.cspell.json
+++ b/core/.cspell.json
@@ -31,7 +31,8 @@
       "MAINTAINERS.txt",
       "package.json",
       "yarn.lock",
-      "misc/cspell/dictionary.txt"
+      "misc/cspell/dictionary.txt",
+      "phpstan*"
     ],
     "ignoreRegExpList": [
       "^msgstr .*",
diff --git a/core/phpstan.neon.dist b/core/phpstan.neon.dist
index d26cc73a128b37cbdec7074c76f191e1c23fe67e..f31d8d11555a9ea6bc893a6f011f2ce36e7e2f00 100644
--- a/core/phpstan.neon.dist
+++ b/core/phpstan.neon.dist
@@ -15,6 +15,7 @@ parameters:
     # Skip settings.
     - ../*/settings*.php
     # Skip test fixtures.
+    - ../*/node_modules/*
     - */tests/fixtures/*.php
     - */tests/fixtures/*.php.gz
     # Skip data files.
@@ -30,10 +31,6 @@ parameters:
     - "#^Unsafe usage of new static#"
 
     # Ignore common errors for now.
-    - "#^Access to an undefined property#"
-    - "#^Call to an undefined method#"
-    - "#^Cannot unset offset#"
-    - "#should return .* but return statement is missing#"
     - "#Drupal calls should be avoided in classes, use dependency injection instead#"
     - "#^Plugin definitions cannot be altered.#"
     - "#^Missing cache backend declaration for performance.#"
