diff --git a/core/modules/aggregator/tests/src/Functional/AggregatorRenderingTest.php b/core/modules/aggregator/tests/src/Functional/AggregatorRenderingTest.php
index faebea4a..de677975 100644
--- a/core/modules/aggregator/tests/src/Functional/AggregatorRenderingTest.php
+++ b/core/modules/aggregator/tests/src/Functional/AggregatorRenderingTest.php
@@ -68,6 +68,8 @@ public function testBlockLinks() {
     $href = $feed->toUrl()->toString();
     $links = $this->xpath('//a[@href = :href]', [':href' => $href]);
     $this->assert(isset($links[0]), new FormattableMarkup('Link to href %href found.', ['%href' => $href]));
+    $this->assertSession()->responseHeaderExists('X-Drupal-Cache-Tags');
+    $this->assertSession()->responseHeaderDoesNotExist('X-Drupal-Cache-Tags');
     $cache_tags_header = $this->drupalGetHeader('X-Drupal-Cache-Tags');
     $cache_tags = explode(' ', $cache_tags_header);
     $this->assertTrue(in_array('aggregator_feed:' . $feed->id(), $cache_tags));
diff --git a/core/modules/basic_auth/tests/src/Functional/BasicAuthTest.php b/core/modules/basic_auth/tests/src/Functional/BasicAuthTest.php
index d2ebf8d6..cc650945 100644
--- a/core/modules/basic_auth/tests/src/Functional/BasicAuthTest.php
+++ b/core/modules/basic_auth/tests/src/Functional/BasicAuthTest.php
@@ -51,7 +51,7 @@ public function testBasicAuth() {
     $this->assertText($account->getAccountName(), 'Account name is displayed.');
     $this->assertResponse('200', 'HTTP response is OK');
     $this->mink->resetSessions();
-    $this->assertNull($this->drupalGetHeader('X-Drupal-Cache'));
+    $this->assertSession()->responseHeaderDoesNotExist('X-Drupal-Cache');
     // Check that Cache-Control is not set to public.
     $this->assertSession()->responseHeaderNotContains('Cache-Control', 'public');
 
@@ -80,7 +80,7 @@ public function testBasicAuth() {
     $this->drupalGet($url);
     $this->assertEqual($this->drupalGetHeader('X-Drupal-Cache'), 'MISS');
     $this->basicAuthGet($url, $account->getAccountName(), $account->pass_raw);
-    $this->assertNull($this->drupalGetHeader('X-Drupal-Cache'));
+    $this->assertSession()->responseHeaderExists('X-Drupal-Cache');
     // Check that Cache-Control is not set to public.
     $this->assertSession()->responseHeaderNotContains('Cache-Control', 'public');
   }
diff --git a/core/tests/Drupal/FunctionalTests/BrowserTestBaseTest.php b/core/tests/Drupal/FunctionalTests/BrowserTestBaseTest.php
index 5bf2f49e..2862c678 100644
--- a/core/tests/Drupal/FunctionalTests/BrowserTestBaseTest.php
+++ b/core/tests/Drupal/FunctionalTests/BrowserTestBaseTest.php
@@ -56,6 +56,7 @@ public function testGoTo() {
     $this->assertStringNotContainsString('</html>', $text);
 
     // Response includes cache tags that we can assert.
+    $this->assertSession()->responseHeaderExists('X-Drupal-Cache-Tags');
     $this->assertSession()->responseHeaderEquals('X-Drupal-Cache-Tags', 'http_response rendered');
 
     // Test that we can read the JS settings.
@@ -74,6 +75,7 @@ public function testGoTo() {
     $this->drupalGet('system-test/header', [], [
       'Test-Header' => 'header value',
     ]);
+    $this->assertSession()->responseHeaderExists('Test-Header');
     $returned_header = $this->getSession()->getResponseHeader('Test-Header');
     $this->assertSame('header value', $returned_header);
   }
@@ -226,6 +228,16 @@ public function testLinkNotExistsExact() {
     $this->assertSession()->linkNotExistsExact('foo|bar');
   }
 
+  /**
+   * Tests responseHeaderDoesNotExist() functionality.
+   *
+   * @see \Drupal\Tests\WebAssert::responseHeaderDoesNotExist()
+   */
+  public function testResponseHeaderDoesNotExist() {
+    $this->drupalGet('test-pipe-char');
+    $this->assertSession()->responseHeaderDoesNotExist('Foo-Bar');
+  }
+
   /**
    * Tests linkNotExistsExact() functionality fail.
    *
diff --git a/core/tests/Drupal/Tests/WebAssert.php b/core/tests/Drupal/Tests/WebAssert.php
index 7b296263..a3bf6b07 100644
--- a/core/tests/Drupal/Tests/WebAssert.php
+++ b/core/tests/Drupal/Tests/WebAssert.php
@@ -10,6 +10,9 @@
 use Behat\Mink\Session;
 use Drupal\Component\Utility\Html;
 use Drupal\Core\Url;
+use PHPUnit\Framework\Assert;
+use PHPUnit\Framework\Constraint\ArrayHasKey;
+use PHPUnit\Framework\Constraint\LogicalNot;
 
 /**
  * Defines a class with methods for asserting presence of elements during tests.
@@ -55,6 +58,38 @@ protected function cleanUrl($url) {
     return parent::cleanUrl($url);
   }
 
+  /**
+   * Asserts that the current response header has a specific entry.
+   *
+   * @param string $name
+   *   The name of the header entry to check existence of.
+   */
+  public function responseHeaderExists(string $name, string $message = ''): void {
+    if ($message === '') {
+      $message = "Failed asserting that the response has a '$name' header.";
+    }
+    $headers = $this->session->getResponseHeaders();
+    $constraint = new ArrayHasKey($name);
+    Assert::assertThat($headers, $constraint, $message);
+  }
+
+  /**
+   * Asserts that the current response header does not have a specific entry.
+   *
+   * @param string $name
+   *   The name of the header entry to check existence of.
+   */
+  public function responseHeaderDoesNotExist(string $name, string $message = ''): void {
+    if ($message === '') {
+      $message = "Failed asserting that the response does not have a '$name' header.";
+    }
+    $headers = $this->session->getResponseHeaders();
+    $constraint = new LogicalNot(
+      new ArrayHasKey($name)
+    );
+    Assert::assertThat($headers, $constraint, $message);
+  }
+
   /**
    * Checks that specific button exists on the current page.
    *
