diff --git a/core/lib/Drupal/Core/Config/ConfigInstaller.php b/core/lib/Drupal/Core/Config/ConfigInstaller.php
index ec94fb331c..e6634aff8a 100644
--- a/core/lib/Drupal/Core/Config/ConfigInstaller.php
+++ b/core/lib/Drupal/Core/Config/ConfigInstaller.php
@@ -106,6 +106,7 @@ public function __construct(ConfigFactoryInterface $config_factory, StorageInter
    * {@inheritdoc}
    */
   public function installDefaultConfig($type, $name) {
+dump([__METHOD__, $type, $name]);
     $extension_path = $this->extensionPathResolver->getPath($type, $name);
     // Refresh the schema cache if the extension provides configuration schema
     // or is a theme.
@@ -326,6 +327,7 @@ protected function getConfigToCreate(StorageInterface $storage, $collection, $pr
    *   An array of configuration data to create, keyed by name.
    */
   protected function createConfiguration($collection, array $config_to_create) {
+//dump([__METHOD__, $collection, $config_to_create]);
     // Order the configuration to install in the order of dependencies.
     if ($collection == StorageInterface::DEFAULT_COLLECTION) {
       $dependency_manager = new ConfigDependencyManager();
