diff --git a/core/modules/views/src/Plugin/views/query/Sql.php b/core/modules/views/src/Plugin/views/query/Sql.php
index 14b335b63f..7f07d1a70e 100644
--- a/core/modules/views/src/Plugin/views/query/Sql.php
+++ b/core/modules/views/src/Plugin/views/query/Sql.php
@@ -1530,13 +1530,24 @@ public function execute(ViewExecutable $view) {
         }
 
         $result = $query->execute();
-        $result->setFetchMode(\PDO::FETCH_CLASS, 'Drupal\views\ResultRow');
+/*        $result->setFetchMode(\PDO::FETCH_CLASS, 'Drupal\views\ResultRow');
 
         // Setup the result row objects.
         $view->result = iterator_to_array($result);
         array_walk($view->result, function (ResultRow $row, $index) {
           $row->index = $index;
-        });
+        });*/
+        $result->setFetchMode(\PDO::FETCH_ASSOC);
+
+        // Setup the result row objects.
+        $i = 0;
+        $view->result = [];
+        foreach ($result as $row) {
+          $resultRow = new ResultRow($row + ['index' => $i]);
+          $i++;
+          $view->result[] = $resultRow;
+        }
+dump([__METHOD__, 'b', $view->result]);
 
         $view->pager->postExecute($view->result);
         $view->pager->updatePageInfo();
