diff --git a/core/modules/system/tests/modules/image_test/src/Plugin/ImageToolkit/Operation/gd/RunTimeException.php b/core/modules/system/tests/modules/image_test/src/Plugin/ImageToolkit/Operation/gd/RunTimeException.php
new file mode 100644
index 0000000000..fb5cd39bdb
--- /dev/null
+++ b/core/modules/system/tests/modules/image_test/src/Plugin/ImageToolkit/Operation/gd/RunTimeException.php
@@ -0,0 +1,35 @@
+<?php
+
+namespace Drupal\image_test\Plugin\ImageToolkit\Operation\gd;
+
+use Drupal\system\Plugin\ImageToolkit\Operation\gd\GDImageToolkitOperationBase;
+
+/**
+ * A GD2 operation that throws a RunTimeException.
+ *
+ * @ImageToolkitOperation(
+ *   id = "gd_run_time_exception",
+ *   toolkit = "gd",
+ *   operation = "run_time_exception",
+ *   label = @Translation("RunTimeException"),
+ *   description = @Translation("Throws a RunTimeException.")
+ * )
+ */
+class RunTimeException extends GDImageToolkitOperationBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function arguments() {
+    // This operation does not use any parameters.
+    return [];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function execute(array $arguments) {
+    throw new \RunTimeException('This operation throws a RunTimeException on purpose for testing');
+  }
+
+}
diff --git a/core/tests/Drupal/KernelTests/Core/Image/ToolkitGdTest.php b/core/tests/Drupal/KernelTests/Core/Image/ToolkitGdTest.php
index 0496973a7d..c7f3f8d19f 100644
--- a/core/tests/Drupal/KernelTests/Core/Image/ToolkitGdTest.php
+++ b/core/tests/Drupal/KernelTests/Core/Image/ToolkitGdTest.php
@@ -43,7 +43,7 @@ class ToolkitGdTest extends KernelTestBase {
    *
    * @var array
    */
-  public static $modules = ['system', 'simpletest'];
+  public static $modules = ['system', 'simpletest', 'image_test'];
 
   /**
    * {@inheritdoc}
@@ -531,4 +531,21 @@ public function testMissingOperation() {
     $this->assertFalse($image->apply('missing_op', []), 'Calling a missing image toolkit operation plugin fails.');
   }
 
+  /**
+   * Tests creating an image that exceeds memory available.
+   */
+  public function testImageTooBig() {
+    $image = $this->imageFactory->get();
+    $this->assertFalse($image->createNew(50000, 50000));
+  }
+
+  /**
+   * Tests an operation that throws a RunTimeException.
+   */
+  public function testRunTimeExceptionThrowingOperation() {
+    $image = $this->imageFactory->get();
+    $image->createNew(50, 50);
+    $this->assertFalse($image->apply('run_time_exception'));
+  }
+
 }
