diff --git a/core/modules/aggregator/tests/src/Functional/Rest/FeedXmlAnonTest.php b/core/modules/aggregator/tests/src/Functional/Rest/FeedXmlAnonTest.php
index 41b03a6a8a..2e341fa120 100644
--- a/core/modules/aggregator/tests/src/Functional/Rest/FeedXmlAnonTest.php
+++ b/core/modules/aggregator/tests/src/Functional/Rest/FeedXmlAnonTest.php
@@ -21,6 +21,6 @@ class FeedXmlAnonTest extends FeedResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/modules/aggregator/tests/src/Functional/Rest/FeedXmlBasicAuthTest.php b/core/modules/aggregator/tests/src/Functional/Rest/FeedXmlBasicAuthTest.php
index 450cc49eff..a8a6a37fe7 100644
--- a/core/modules/aggregator/tests/src/Functional/Rest/FeedXmlBasicAuthTest.php
+++ b/core/modules/aggregator/tests/src/Functional/Rest/FeedXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class FeedXmlBasicAuthTest extends FeedResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/aggregator/tests/src/Functional/Rest/FeedXmlCookieTest.php b/core/modules/aggregator/tests/src/Functional/Rest/FeedXmlCookieTest.php
index 3fee585d07..ef491abff8 100644
--- a/core/modules/aggregator/tests/src/Functional/Rest/FeedXmlCookieTest.php
+++ b/core/modules/aggregator/tests/src/Functional/Rest/FeedXmlCookieTest.php
@@ -21,7 +21,7 @@ class FeedXmlCookieTest extends FeedResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/aggregator/tests/src/Functional/Rest/ItemXmlAnonTest.php b/core/modules/aggregator/tests/src/Functional/Rest/ItemXmlAnonTest.php
index f61a878413..94d86fd7e0 100644
--- a/core/modules/aggregator/tests/src/Functional/Rest/ItemXmlAnonTest.php
+++ b/core/modules/aggregator/tests/src/Functional/Rest/ItemXmlAnonTest.php
@@ -21,6 +21,6 @@ class ItemXmlAnonTest extends ItemResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/modules/aggregator/tests/src/Functional/Rest/ItemXmlBasicAuthTest.php b/core/modules/aggregator/tests/src/Functional/Rest/ItemXmlBasicAuthTest.php
index 438a4233ca..8bcd2435e2 100644
--- a/core/modules/aggregator/tests/src/Functional/Rest/ItemXmlBasicAuthTest.php
+++ b/core/modules/aggregator/tests/src/Functional/Rest/ItemXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class ItemXmlBasicAuthTest extends ItemResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/aggregator/tests/src/Functional/Rest/ItemXmlCookieTest.php b/core/modules/aggregator/tests/src/Functional/Rest/ItemXmlCookieTest.php
index 69144066ad..8c623a8ce9 100644
--- a/core/modules/aggregator/tests/src/Functional/Rest/ItemXmlCookieTest.php
+++ b/core/modules/aggregator/tests/src/Functional/Rest/ItemXmlCookieTest.php
@@ -21,7 +21,7 @@ class ItemXmlCookieTest extends ItemResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/block/tests/src/Functional/Rest/BlockXmlAnonTest.php b/core/modules/block/tests/src/Functional/Rest/BlockXmlAnonTest.php
index 923f02d83f..42ded67024 100644
--- a/core/modules/block/tests/src/Functional/Rest/BlockXmlAnonTest.php
+++ b/core/modules/block/tests/src/Functional/Rest/BlockXmlAnonTest.php
@@ -21,6 +21,6 @@ class BlockXmlAnonTest extends BlockResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/modules/block/tests/src/Functional/Rest/BlockXmlBasicAuthTest.php b/core/modules/block/tests/src/Functional/Rest/BlockXmlBasicAuthTest.php
index 2b720d1d9f..249b62ebb0 100644
--- a/core/modules/block/tests/src/Functional/Rest/BlockXmlBasicAuthTest.php
+++ b/core/modules/block/tests/src/Functional/Rest/BlockXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class BlockXmlBasicAuthTest extends BlockResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/block/tests/src/Functional/Rest/BlockXmlCookieTest.php b/core/modules/block/tests/src/Functional/Rest/BlockXmlCookieTest.php
index 9474dbdf0d..22a1a5e7fa 100644
--- a/core/modules/block/tests/src/Functional/Rest/BlockXmlCookieTest.php
+++ b/core/modules/block/tests/src/Functional/Rest/BlockXmlCookieTest.php
@@ -21,7 +21,7 @@ class BlockXmlCookieTest extends BlockResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/block_content/tests/src/Functional/Rest/BlockContentTypeXmlAnonTest.php b/core/modules/block_content/tests/src/Functional/Rest/BlockContentTypeXmlAnonTest.php
index 2277d6b978..7571455aa8 100644
--- a/core/modules/block_content/tests/src/Functional/Rest/BlockContentTypeXmlAnonTest.php
+++ b/core/modules/block_content/tests/src/Functional/Rest/BlockContentTypeXmlAnonTest.php
@@ -21,6 +21,6 @@ class BlockContentTypeXmlAnonTest extends BlockContentTypeResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/modules/block_content/tests/src/Functional/Rest/BlockContentTypeXmlBasicAuthTest.php b/core/modules/block_content/tests/src/Functional/Rest/BlockContentTypeXmlBasicAuthTest.php
index a878c7d5b3..7617706028 100644
--- a/core/modules/block_content/tests/src/Functional/Rest/BlockContentTypeXmlBasicAuthTest.php
+++ b/core/modules/block_content/tests/src/Functional/Rest/BlockContentTypeXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class BlockContentTypeXmlBasicAuthTest extends BlockContentTypeResourceTestBase
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/block_content/tests/src/Functional/Rest/BlockContentTypeXmlCookieTest.php b/core/modules/block_content/tests/src/Functional/Rest/BlockContentTypeXmlCookieTest.php
index 836c330dce..c1b72905ec 100644
--- a/core/modules/block_content/tests/src/Functional/Rest/BlockContentTypeXmlCookieTest.php
+++ b/core/modules/block_content/tests/src/Functional/Rest/BlockContentTypeXmlCookieTest.php
@@ -21,7 +21,7 @@ class BlockContentTypeXmlCookieTest extends BlockContentTypeResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/block_content/tests/src/Functional/Rest/BlockContentXmlAnonTest.php b/core/modules/block_content/tests/src/Functional/Rest/BlockContentXmlAnonTest.php
index fba08ed473..21e2a9d707 100644
--- a/core/modules/block_content/tests/src/Functional/Rest/BlockContentXmlAnonTest.php
+++ b/core/modules/block_content/tests/src/Functional/Rest/BlockContentXmlAnonTest.php
@@ -21,6 +21,6 @@ class BlockContentXmlAnonTest extends BlockContentResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/modules/block_content/tests/src/Functional/Rest/BlockContentXmlBasicAuthTest.php b/core/modules/block_content/tests/src/Functional/Rest/BlockContentXmlBasicAuthTest.php
index 199abea08d..422f04ece6 100644
--- a/core/modules/block_content/tests/src/Functional/Rest/BlockContentXmlBasicAuthTest.php
+++ b/core/modules/block_content/tests/src/Functional/Rest/BlockContentXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class BlockContentXmlBasicAuthTest extends BlockContentResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/block_content/tests/src/Functional/Rest/BlockContentXmlCookieTest.php b/core/modules/block_content/tests/src/Functional/Rest/BlockContentXmlCookieTest.php
index 6251455952..c92a144744 100644
--- a/core/modules/block_content/tests/src/Functional/Rest/BlockContentXmlCookieTest.php
+++ b/core/modules/block_content/tests/src/Functional/Rest/BlockContentXmlCookieTest.php
@@ -21,7 +21,7 @@ class BlockContentXmlCookieTest extends BlockContentResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/comment/tests/src/Functional/Rest/CommentTypeXmlAnonTest.php b/core/modules/comment/tests/src/Functional/Rest/CommentTypeXmlAnonTest.php
index 96d6c41a0e..10b3f5e1e0 100644
--- a/core/modules/comment/tests/src/Functional/Rest/CommentTypeXmlAnonTest.php
+++ b/core/modules/comment/tests/src/Functional/Rest/CommentTypeXmlAnonTest.php
@@ -21,6 +21,6 @@ class CommentTypeXmlAnonTest extends CommentTypeResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/modules/comment/tests/src/Functional/Rest/CommentTypeXmlBasicAuthTest.php b/core/modules/comment/tests/src/Functional/Rest/CommentTypeXmlBasicAuthTest.php
index e92f2bcc65..b122c44e0a 100644
--- a/core/modules/comment/tests/src/Functional/Rest/CommentTypeXmlBasicAuthTest.php
+++ b/core/modules/comment/tests/src/Functional/Rest/CommentTypeXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class CommentTypeXmlBasicAuthTest extends CommentTypeResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/comment/tests/src/Functional/Rest/CommentTypeXmlCookieTest.php b/core/modules/comment/tests/src/Functional/Rest/CommentTypeXmlCookieTest.php
index 01f42db448..6d913eb87f 100644
--- a/core/modules/comment/tests/src/Functional/Rest/CommentTypeXmlCookieTest.php
+++ b/core/modules/comment/tests/src/Functional/Rest/CommentTypeXmlCookieTest.php
@@ -21,7 +21,7 @@ class CommentTypeXmlCookieTest extends CommentTypeResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/comment/tests/src/Functional/Rest/CommentXmlAnonTest.php b/core/modules/comment/tests/src/Functional/Rest/CommentXmlAnonTest.php
index 22c1b858f5..2d539fe96f 100644
--- a/core/modules/comment/tests/src/Functional/Rest/CommentXmlAnonTest.php
+++ b/core/modules/comment/tests/src/Functional/Rest/CommentXmlAnonTest.php
@@ -21,7 +21,7 @@ class CommentXmlAnonTest extends CommentResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/comment/tests/src/Functional/Rest/CommentXmlBasicAuthTest.php b/core/modules/comment/tests/src/Functional/Rest/CommentXmlBasicAuthTest.php
index d1f38e44a4..926d9c91f1 100644
--- a/core/modules/comment/tests/src/Functional/Rest/CommentXmlBasicAuthTest.php
+++ b/core/modules/comment/tests/src/Functional/Rest/CommentXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class CommentXmlBasicAuthTest extends CommentResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/comment/tests/src/Functional/Rest/CommentXmlCookieTest.php b/core/modules/comment/tests/src/Functional/Rest/CommentXmlCookieTest.php
index e06cf26521..b2a5232763 100644
--- a/core/modules/comment/tests/src/Functional/Rest/CommentXmlCookieTest.php
+++ b/core/modules/comment/tests/src/Functional/Rest/CommentXmlCookieTest.php
@@ -21,7 +21,7 @@ class CommentXmlCookieTest extends CommentResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/config/tests/config_test/tests/src/Functional/Rest/ConfigTestXmlAnonTest.php b/core/modules/config/tests/config_test/tests/src/Functional/Rest/ConfigTestXmlAnonTest.php
index 451e86df22..2567f72b60 100644
--- a/core/modules/config/tests/config_test/tests/src/Functional/Rest/ConfigTestXmlAnonTest.php
+++ b/core/modules/config/tests/config_test/tests/src/Functional/Rest/ConfigTestXmlAnonTest.php
@@ -21,6 +21,6 @@ class ConfigTestXmlAnonTest extends ConfigTestResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/modules/config/tests/config_test/tests/src/Functional/Rest/ConfigTestXmlBasicAuthTest.php b/core/modules/config/tests/config_test/tests/src/Functional/Rest/ConfigTestXmlBasicAuthTest.php
index c65028f212..ca853e2ca9 100644
--- a/core/modules/config/tests/config_test/tests/src/Functional/Rest/ConfigTestXmlBasicAuthTest.php
+++ b/core/modules/config/tests/config_test/tests/src/Functional/Rest/ConfigTestXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class ConfigTestXmlBasicAuthTest extends ConfigTestResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/config/tests/config_test/tests/src/Functional/Rest/ConfigTestXmlCookieTest.php b/core/modules/config/tests/config_test/tests/src/Functional/Rest/ConfigTestXmlCookieTest.php
index 540b14e8d3..a53e19cf1c 100644
--- a/core/modules/config/tests/config_test/tests/src/Functional/Rest/ConfigTestXmlCookieTest.php
+++ b/core/modules/config/tests/config_test/tests/src/Functional/Rest/ConfigTestXmlCookieTest.php
@@ -21,7 +21,7 @@ class ConfigTestXmlCookieTest extends ConfigTestResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/contact/tests/src/Functional/Rest/ContactFormXmlAnonTest.php b/core/modules/contact/tests/src/Functional/Rest/ContactFormXmlAnonTest.php
index 2931edda8d..bb47644d83 100644
--- a/core/modules/contact/tests/src/Functional/Rest/ContactFormXmlAnonTest.php
+++ b/core/modules/contact/tests/src/Functional/Rest/ContactFormXmlAnonTest.php
@@ -21,6 +21,6 @@ class ContactFormXmlAnonTest extends ContactFormResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/modules/contact/tests/src/Functional/Rest/ContactFormXmlBasicAuthTest.php b/core/modules/contact/tests/src/Functional/Rest/ContactFormXmlBasicAuthTest.php
index dfc4b6498a..c3125f04e0 100644
--- a/core/modules/contact/tests/src/Functional/Rest/ContactFormXmlBasicAuthTest.php
+++ b/core/modules/contact/tests/src/Functional/Rest/ContactFormXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class ContactFormXmlBasicAuthTest extends ContactFormResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/contact/tests/src/Functional/Rest/ContactFormXmlCookieTest.php b/core/modules/contact/tests/src/Functional/Rest/ContactFormXmlCookieTest.php
index cdcd2daa3e..75528c82b7 100644
--- a/core/modules/contact/tests/src/Functional/Rest/ContactFormXmlCookieTest.php
+++ b/core/modules/contact/tests/src/Functional/Rest/ContactFormXmlCookieTest.php
@@ -21,7 +21,7 @@ class ContactFormXmlCookieTest extends ContactFormResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/contact/tests/src/Functional/Rest/MessageXmlAnonTest.php b/core/modules/contact/tests/src/Functional/Rest/MessageXmlAnonTest.php
index 6b9b10de8a..3e734475e7 100644
--- a/core/modules/contact/tests/src/Functional/Rest/MessageXmlAnonTest.php
+++ b/core/modules/contact/tests/src/Functional/Rest/MessageXmlAnonTest.php
@@ -21,6 +21,6 @@ class MessageXmlAnonTest extends MessageResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/modules/contact/tests/src/Functional/Rest/MessageXmlBasicAuthTest.php b/core/modules/contact/tests/src/Functional/Rest/MessageXmlBasicAuthTest.php
index 152a94a86f..c76dddefcf 100644
--- a/core/modules/contact/tests/src/Functional/Rest/MessageXmlBasicAuthTest.php
+++ b/core/modules/contact/tests/src/Functional/Rest/MessageXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class MessageXmlBasicAuthTest extends MessageResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/contact/tests/src/Functional/Rest/MessageXmlCookieTest.php b/core/modules/contact/tests/src/Functional/Rest/MessageXmlCookieTest.php
index ecdf0dba55..b94c2ae78c 100644
--- a/core/modules/contact/tests/src/Functional/Rest/MessageXmlCookieTest.php
+++ b/core/modules/contact/tests/src/Functional/Rest/MessageXmlCookieTest.php
@@ -21,7 +21,7 @@ class MessageXmlCookieTest extends MessageResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/editor/tests/src/Functional/Rest/EditorXmlAnonTest.php b/core/modules/editor/tests/src/Functional/Rest/EditorXmlAnonTest.php
index d5be1f6deb..6b0b033184 100644
--- a/core/modules/editor/tests/src/Functional/Rest/EditorXmlAnonTest.php
+++ b/core/modules/editor/tests/src/Functional/Rest/EditorXmlAnonTest.php
@@ -21,6 +21,6 @@ class EditorXmlAnonTest extends EditorResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/modules/editor/tests/src/Functional/Rest/EditorXmlBasicAuthTest.php b/core/modules/editor/tests/src/Functional/Rest/EditorXmlBasicAuthTest.php
index acab40b9a4..9728218980 100644
--- a/core/modules/editor/tests/src/Functional/Rest/EditorXmlBasicAuthTest.php
+++ b/core/modules/editor/tests/src/Functional/Rest/EditorXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class EditorXmlBasicAuthTest extends EditorResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/editor/tests/src/Functional/Rest/EditorXmlCookieTest.php b/core/modules/editor/tests/src/Functional/Rest/EditorXmlCookieTest.php
index 8940095162..33ffca7ebe 100644
--- a/core/modules/editor/tests/src/Functional/Rest/EditorXmlCookieTest.php
+++ b/core/modules/editor/tests/src/Functional/Rest/EditorXmlCookieTest.php
@@ -21,7 +21,7 @@ class EditorXmlCookieTest extends EditorResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/field/tests/src/Functional/Rest/FieldConfigXmlAnonTest.php b/core/modules/field/tests/src/Functional/Rest/FieldConfigXmlAnonTest.php
index 8809e60e93..d280b1e322 100644
--- a/core/modules/field/tests/src/Functional/Rest/FieldConfigXmlAnonTest.php
+++ b/core/modules/field/tests/src/Functional/Rest/FieldConfigXmlAnonTest.php
@@ -21,6 +21,6 @@ class FieldConfigXmlAnonTest extends FieldConfigResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/modules/field/tests/src/Functional/Rest/FieldConfigXmlBasicAuthTest.php b/core/modules/field/tests/src/Functional/Rest/FieldConfigXmlBasicAuthTest.php
index 109f2dc94a..b50930342a 100644
--- a/core/modules/field/tests/src/Functional/Rest/FieldConfigXmlBasicAuthTest.php
+++ b/core/modules/field/tests/src/Functional/Rest/FieldConfigXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class FieldConfigXmlBasicAuthTest extends FieldConfigResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/field/tests/src/Functional/Rest/FieldConfigXmlCookieTest.php b/core/modules/field/tests/src/Functional/Rest/FieldConfigXmlCookieTest.php
index d8aebdd1c8..7504f8f2e7 100644
--- a/core/modules/field/tests/src/Functional/Rest/FieldConfigXmlCookieTest.php
+++ b/core/modules/field/tests/src/Functional/Rest/FieldConfigXmlCookieTest.php
@@ -21,7 +21,7 @@ class FieldConfigXmlCookieTest extends FieldConfigResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/field/tests/src/Functional/Rest/FieldStorageConfigXmlAnonTest.php b/core/modules/field/tests/src/Functional/Rest/FieldStorageConfigXmlAnonTest.php
index 4af918b03f..e6982a4ced 100644
--- a/core/modules/field/tests/src/Functional/Rest/FieldStorageConfigXmlAnonTest.php
+++ b/core/modules/field/tests/src/Functional/Rest/FieldStorageConfigXmlAnonTest.php
@@ -21,6 +21,6 @@ class FieldStorageConfigXmlAnonTest extends FieldStorageConfigResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/modules/field/tests/src/Functional/Rest/FieldStorageConfigXmlBasicAuthTest.php b/core/modules/field/tests/src/Functional/Rest/FieldStorageConfigXmlBasicAuthTest.php
index d6fe68f8e7..ced7fd1f6c 100644
--- a/core/modules/field/tests/src/Functional/Rest/FieldStorageConfigXmlBasicAuthTest.php
+++ b/core/modules/field/tests/src/Functional/Rest/FieldStorageConfigXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class FieldStorageConfigXmlBasicAuthTest extends FieldStorageConfigResourceTestB
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/field/tests/src/Functional/Rest/FieldStorageConfigXmlCookieTest.php b/core/modules/field/tests/src/Functional/Rest/FieldStorageConfigXmlCookieTest.php
index 6672a0df02..79c7918b66 100644
--- a/core/modules/field/tests/src/Functional/Rest/FieldStorageConfigXmlCookieTest.php
+++ b/core/modules/field/tests/src/Functional/Rest/FieldStorageConfigXmlCookieTest.php
@@ -21,7 +21,7 @@ class FieldStorageConfigXmlCookieTest extends FieldStorageConfigResourceTestBase
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/file/tests/src/Functional/Rest/FileXmlAnonTest.php b/core/modules/file/tests/src/Functional/Rest/FileXmlAnonTest.php
index 8360096fb9..35960a13df 100644
--- a/core/modules/file/tests/src/Functional/Rest/FileXmlAnonTest.php
+++ b/core/modules/file/tests/src/Functional/Rest/FileXmlAnonTest.php
@@ -21,6 +21,6 @@ class FileXmlAnonTest extends FileResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/modules/file/tests/src/Functional/Rest/FileXmlBasicAuthTest.php b/core/modules/file/tests/src/Functional/Rest/FileXmlBasicAuthTest.php
index 66e9986dbc..e4750fb301 100644
--- a/core/modules/file/tests/src/Functional/Rest/FileXmlBasicAuthTest.php
+++ b/core/modules/file/tests/src/Functional/Rest/FileXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class FileXmlBasicAuthTest extends FileResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/file/tests/src/Functional/Rest/FileXmlCookieTest.php b/core/modules/file/tests/src/Functional/Rest/FileXmlCookieTest.php
index 4b7db031e5..2cefb5126f 100644
--- a/core/modules/file/tests/src/Functional/Rest/FileXmlCookieTest.php
+++ b/core/modules/file/tests/src/Functional/Rest/FileXmlCookieTest.php
@@ -21,7 +21,7 @@ class FileXmlCookieTest extends FileResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/filter/tests/src/Functional/Rest/FilterFormatXmlAnonTest.php b/core/modules/filter/tests/src/Functional/Rest/FilterFormatXmlAnonTest.php
index 25b84cf621..a58e7d003f 100644
--- a/core/modules/filter/tests/src/Functional/Rest/FilterFormatXmlAnonTest.php
+++ b/core/modules/filter/tests/src/Functional/Rest/FilterFormatXmlAnonTest.php
@@ -21,6 +21,6 @@ class FilterFormatXmlAnonTest extends FilterFormatResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/modules/filter/tests/src/Functional/Rest/FilterFormatXmlBasicAuthTest.php b/core/modules/filter/tests/src/Functional/Rest/FilterFormatXmlBasicAuthTest.php
index 6c0c62d52b..fd61cdc32f 100644
--- a/core/modules/filter/tests/src/Functional/Rest/FilterFormatXmlBasicAuthTest.php
+++ b/core/modules/filter/tests/src/Functional/Rest/FilterFormatXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class FilterFormatXmlBasicAuthTest extends FilterFormatResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/filter/tests/src/Functional/Rest/FilterFormatXmlCookieTest.php b/core/modules/filter/tests/src/Functional/Rest/FilterFormatXmlCookieTest.php
index 1c3be314dd..7849d3caab 100644
--- a/core/modules/filter/tests/src/Functional/Rest/FilterFormatXmlCookieTest.php
+++ b/core/modules/filter/tests/src/Functional/Rest/FilterFormatXmlCookieTest.php
@@ -21,7 +21,7 @@ class FilterFormatXmlCookieTest extends FilterFormatResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/image/tests/src/Functional/Rest/ImageStyleXmlAnonTest.php b/core/modules/image/tests/src/Functional/Rest/ImageStyleXmlAnonTest.php
index 427641aa4b..3af5594d87 100644
--- a/core/modules/image/tests/src/Functional/Rest/ImageStyleXmlAnonTest.php
+++ b/core/modules/image/tests/src/Functional/Rest/ImageStyleXmlAnonTest.php
@@ -21,7 +21,7 @@ class ImageStyleXmlAnonTest extends ImageStyleResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/image/tests/src/Functional/Rest/ImageStyleXmlBasicAuthTest.php b/core/modules/image/tests/src/Functional/Rest/ImageStyleXmlBasicAuthTest.php
index cd376a3140..111d324420 100644
--- a/core/modules/image/tests/src/Functional/Rest/ImageStyleXmlBasicAuthTest.php
+++ b/core/modules/image/tests/src/Functional/Rest/ImageStyleXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class ImageStyleXmlBasicAuthTest extends ImageStyleResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/image/tests/src/Functional/Rest/ImageStyleXmlCookieTest.php b/core/modules/image/tests/src/Functional/Rest/ImageStyleXmlCookieTest.php
index 7ca459a0ae..497f72b741 100644
--- a/core/modules/image/tests/src/Functional/Rest/ImageStyleXmlCookieTest.php
+++ b/core/modules/image/tests/src/Functional/Rest/ImageStyleXmlCookieTest.php
@@ -21,7 +21,7 @@ class ImageStyleXmlCookieTest extends ImageStyleResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/language/tests/src/Functional/Rest/ConfigurableLanguageXmlAnonTest.php b/core/modules/language/tests/src/Functional/Rest/ConfigurableLanguageXmlAnonTest.php
index 882b7bf458..09530d7122 100644
--- a/core/modules/language/tests/src/Functional/Rest/ConfigurableLanguageXmlAnonTest.php
+++ b/core/modules/language/tests/src/Functional/Rest/ConfigurableLanguageXmlAnonTest.php
@@ -21,6 +21,6 @@ class ConfigurableLanguageXmlAnonTest extends ConfigurableLanguageResourceTestBa
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/modules/language/tests/src/Functional/Rest/ConfigurableLanguageXmlBasicAuthTest.php b/core/modules/language/tests/src/Functional/Rest/ConfigurableLanguageXmlBasicAuthTest.php
index afc017f7a0..a221e24534 100644
--- a/core/modules/language/tests/src/Functional/Rest/ConfigurableLanguageXmlBasicAuthTest.php
+++ b/core/modules/language/tests/src/Functional/Rest/ConfigurableLanguageXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class ConfigurableLanguageXmlBasicAuthTest extends ConfigurableLanguageResourceT
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/language/tests/src/Functional/Rest/ConfigurableLanguageXmlCookieTest.php b/core/modules/language/tests/src/Functional/Rest/ConfigurableLanguageXmlCookieTest.php
index 7755d76e91..775dd90a0b 100644
--- a/core/modules/language/tests/src/Functional/Rest/ConfigurableLanguageXmlCookieTest.php
+++ b/core/modules/language/tests/src/Functional/Rest/ConfigurableLanguageXmlCookieTest.php
@@ -21,7 +21,7 @@ class ConfigurableLanguageXmlCookieTest extends ConfigurableLanguageResourceTest
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/language/tests/src/Functional/Rest/ContentLanguageSettingsXmlAnonTest.php b/core/modules/language/tests/src/Functional/Rest/ContentLanguageSettingsXmlAnonTest.php
index fd6519387e..fb4791e348 100644
--- a/core/modules/language/tests/src/Functional/Rest/ContentLanguageSettingsXmlAnonTest.php
+++ b/core/modules/language/tests/src/Functional/Rest/ContentLanguageSettingsXmlAnonTest.php
@@ -21,6 +21,6 @@ class ContentLanguageSettingsXmlAnonTest extends ContentLanguageSettingsResource
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/modules/language/tests/src/Functional/Rest/ContentLanguageSettingsXmlBasicAuthTest.php b/core/modules/language/tests/src/Functional/Rest/ContentLanguageSettingsXmlBasicAuthTest.php
index dee7ad27d4..69a8ecc550 100644
--- a/core/modules/language/tests/src/Functional/Rest/ContentLanguageSettingsXmlBasicAuthTest.php
+++ b/core/modules/language/tests/src/Functional/Rest/ContentLanguageSettingsXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class ContentLanguageSettingsXmlBasicAuthTest extends ContentLanguageSettingsRes
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/language/tests/src/Functional/Rest/ContentLanguageSettingsXmlCookieTest.php b/core/modules/language/tests/src/Functional/Rest/ContentLanguageSettingsXmlCookieTest.php
index a5b49684f7..943881ab53 100644
--- a/core/modules/language/tests/src/Functional/Rest/ContentLanguageSettingsXmlCookieTest.php
+++ b/core/modules/language/tests/src/Functional/Rest/ContentLanguageSettingsXmlCookieTest.php
@@ -21,7 +21,7 @@ class ContentLanguageSettingsXmlCookieTest extends ContentLanguageSettingsResour
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/layout_builder/tests/src/Functional/Rest/LayoutBuilderEntityViewDisplayXmlAnonTest.php b/core/modules/layout_builder/tests/src/Functional/Rest/LayoutBuilderEntityViewDisplayXmlAnonTest.php
index 8c337c6552..d9fff74fb8 100644
--- a/core/modules/layout_builder/tests/src/Functional/Rest/LayoutBuilderEntityViewDisplayXmlAnonTest.php
+++ b/core/modules/layout_builder/tests/src/Functional/Rest/LayoutBuilderEntityViewDisplayXmlAnonTest.php
@@ -22,6 +22,6 @@ class LayoutBuilderEntityViewDisplayXmlAnonTest extends LayoutBuilderEntityViewD
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/modules/layout_builder/tests/src/Functional/Rest/LayoutBuilderEntityViewDisplayXmlBasicAuthTest.php b/core/modules/layout_builder/tests/src/Functional/Rest/LayoutBuilderEntityViewDisplayXmlBasicAuthTest.php
index 6ec5880699..a8d9d0de29 100644
--- a/core/modules/layout_builder/tests/src/Functional/Rest/LayoutBuilderEntityViewDisplayXmlBasicAuthTest.php
+++ b/core/modules/layout_builder/tests/src/Functional/Rest/LayoutBuilderEntityViewDisplayXmlBasicAuthTest.php
@@ -27,7 +27,7 @@ class LayoutBuilderEntityViewDisplayXmlBasicAuthTest extends LayoutBuilderEntity
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/layout_builder/tests/src/Functional/Rest/LayoutBuilderEntityViewDisplayXmlCookieTest.php b/core/modules/layout_builder/tests/src/Functional/Rest/LayoutBuilderEntityViewDisplayXmlCookieTest.php
index 92925eca90..cc9c5aa972 100644
--- a/core/modules/layout_builder/tests/src/Functional/Rest/LayoutBuilderEntityViewDisplayXmlCookieTest.php
+++ b/core/modules/layout_builder/tests/src/Functional/Rest/LayoutBuilderEntityViewDisplayXmlCookieTest.php
@@ -22,7 +22,7 @@ class LayoutBuilderEntityViewDisplayXmlCookieTest extends LayoutBuilderEntityVie
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/media/tests/src/Functional/Rest/MediaTypeXmlAnonTest.php b/core/modules/media/tests/src/Functional/Rest/MediaTypeXmlAnonTest.php
index 083d80a1c1..ebd28d6b3b 100644
--- a/core/modules/media/tests/src/Functional/Rest/MediaTypeXmlAnonTest.php
+++ b/core/modules/media/tests/src/Functional/Rest/MediaTypeXmlAnonTest.php
@@ -21,6 +21,6 @@ class MediaTypeXmlAnonTest extends MediaTypeResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/modules/media/tests/src/Functional/Rest/MediaTypeXmlBasicAuthTest.php b/core/modules/media/tests/src/Functional/Rest/MediaTypeXmlBasicAuthTest.php
index 95de21197d..023a5a0b45 100644
--- a/core/modules/media/tests/src/Functional/Rest/MediaTypeXmlBasicAuthTest.php
+++ b/core/modules/media/tests/src/Functional/Rest/MediaTypeXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class MediaTypeXmlBasicAuthTest extends MediaTypeResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/media/tests/src/Functional/Rest/MediaTypeXmlCookieTest.php b/core/modules/media/tests/src/Functional/Rest/MediaTypeXmlCookieTest.php
index 3b6a53d96f..984019a6fb 100644
--- a/core/modules/media/tests/src/Functional/Rest/MediaTypeXmlCookieTest.php
+++ b/core/modules/media/tests/src/Functional/Rest/MediaTypeXmlCookieTest.php
@@ -21,7 +21,7 @@ class MediaTypeXmlCookieTest extends MediaTypeResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/media/tests/src/Functional/Rest/MediaXmlAnonTest.php b/core/modules/media/tests/src/Functional/Rest/MediaXmlAnonTest.php
index 787794996a..6989f84568 100644
--- a/core/modules/media/tests/src/Functional/Rest/MediaXmlAnonTest.php
+++ b/core/modules/media/tests/src/Functional/Rest/MediaXmlAnonTest.php
@@ -21,6 +21,6 @@ class MediaXmlAnonTest extends MediaResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/modules/media/tests/src/Functional/Rest/MediaXmlBasicAuthTest.php b/core/modules/media/tests/src/Functional/Rest/MediaXmlBasicAuthTest.php
index 971e75f42c..dd6de0a9fc 100644
--- a/core/modules/media/tests/src/Functional/Rest/MediaXmlBasicAuthTest.php
+++ b/core/modules/media/tests/src/Functional/Rest/MediaXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class MediaXmlBasicAuthTest extends MediaResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/media/tests/src/Functional/Rest/MediaXmlCookieTest.php b/core/modules/media/tests/src/Functional/Rest/MediaXmlCookieTest.php
index f67f7b4f8c..18c5be6269 100644
--- a/core/modules/media/tests/src/Functional/Rest/MediaXmlCookieTest.php
+++ b/core/modules/media/tests/src/Functional/Rest/MediaXmlCookieTest.php
@@ -21,7 +21,7 @@ class MediaXmlCookieTest extends MediaResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/menu_link_content/tests/src/Functional/Rest/MenuLinkContentXmlAnonTest.php b/core/modules/menu_link_content/tests/src/Functional/Rest/MenuLinkContentXmlAnonTest.php
index 7c39dae81d..d575a59640 100644
--- a/core/modules/menu_link_content/tests/src/Functional/Rest/MenuLinkContentXmlAnonTest.php
+++ b/core/modules/menu_link_content/tests/src/Functional/Rest/MenuLinkContentXmlAnonTest.php
@@ -21,6 +21,6 @@ class MenuLinkContentXmlAnonTest extends MenuLinkContentResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/modules/menu_link_content/tests/src/Functional/Rest/MenuLinkContentXmlBasicAuthTest.php b/core/modules/menu_link_content/tests/src/Functional/Rest/MenuLinkContentXmlBasicAuthTest.php
index 7bbf966170..74e9702534 100644
--- a/core/modules/menu_link_content/tests/src/Functional/Rest/MenuLinkContentXmlBasicAuthTest.php
+++ b/core/modules/menu_link_content/tests/src/Functional/Rest/MenuLinkContentXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class MenuLinkContentXmlBasicAuthTest extends MenuLinkContentResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/menu_link_content/tests/src/Functional/Rest/MenuLinkContentXmlCookieTest.php b/core/modules/menu_link_content/tests/src/Functional/Rest/MenuLinkContentXmlCookieTest.php
index 8b56dde815..2aa1ba595a 100644
--- a/core/modules/menu_link_content/tests/src/Functional/Rest/MenuLinkContentXmlCookieTest.php
+++ b/core/modules/menu_link_content/tests/src/Functional/Rest/MenuLinkContentXmlCookieTest.php
@@ -21,7 +21,7 @@ class MenuLinkContentXmlCookieTest extends MenuLinkContentResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/node/tests/src/Functional/Rest/NodeTypeXmlAnonTest.php b/core/modules/node/tests/src/Functional/Rest/NodeTypeXmlAnonTest.php
index c21a9d458b..d919a89b73 100644
--- a/core/modules/node/tests/src/Functional/Rest/NodeTypeXmlAnonTest.php
+++ b/core/modules/node/tests/src/Functional/Rest/NodeTypeXmlAnonTest.php
@@ -21,6 +21,6 @@ class NodeTypeXmlAnonTest extends NodeTypeResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/modules/node/tests/src/Functional/Rest/NodeTypeXmlBasicAuthTest.php b/core/modules/node/tests/src/Functional/Rest/NodeTypeXmlBasicAuthTest.php
index 29413dc9c2..735cd429d0 100644
--- a/core/modules/node/tests/src/Functional/Rest/NodeTypeXmlBasicAuthTest.php
+++ b/core/modules/node/tests/src/Functional/Rest/NodeTypeXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class NodeTypeXmlBasicAuthTest extends NodeTypeResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/node/tests/src/Functional/Rest/NodeTypeXmlCookieTest.php b/core/modules/node/tests/src/Functional/Rest/NodeTypeXmlCookieTest.php
index 1326f291ff..62ce3b1285 100644
--- a/core/modules/node/tests/src/Functional/Rest/NodeTypeXmlCookieTest.php
+++ b/core/modules/node/tests/src/Functional/Rest/NodeTypeXmlCookieTest.php
@@ -21,7 +21,7 @@ class NodeTypeXmlCookieTest extends NodeTypeResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/node/tests/src/Functional/Rest/NodeXmlAnonTest.php b/core/modules/node/tests/src/Functional/Rest/NodeXmlAnonTest.php
index 83e98a1712..7f9096acf5 100644
--- a/core/modules/node/tests/src/Functional/Rest/NodeXmlAnonTest.php
+++ b/core/modules/node/tests/src/Functional/Rest/NodeXmlAnonTest.php
@@ -21,7 +21,7 @@ class NodeXmlAnonTest extends NodeResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/node/tests/src/Functional/Rest/NodeXmlBasicAuthTest.php b/core/modules/node/tests/src/Functional/Rest/NodeXmlBasicAuthTest.php
index ce809747aa..d961d42f22 100644
--- a/core/modules/node/tests/src/Functional/Rest/NodeXmlBasicAuthTest.php
+++ b/core/modules/node/tests/src/Functional/Rest/NodeXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class NodeXmlBasicAuthTest extends NodeResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/node/tests/src/Functional/Rest/NodeXmlCookieTest.php b/core/modules/node/tests/src/Functional/Rest/NodeXmlCookieTest.php
index 288d5a143c..e4fdf321b6 100644
--- a/core/modules/node/tests/src/Functional/Rest/NodeXmlCookieTest.php
+++ b/core/modules/node/tests/src/Functional/Rest/NodeXmlCookieTest.php
@@ -21,7 +21,7 @@ class NodeXmlCookieTest extends NodeResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/rdf/tests/src/Functional/Rest/RdfMappingXmlAnonTest.php b/core/modules/rdf/tests/src/Functional/Rest/RdfMappingXmlAnonTest.php
index ab6922b8e7..42f32e5663 100644
--- a/core/modules/rdf/tests/src/Functional/Rest/RdfMappingXmlAnonTest.php
+++ b/core/modules/rdf/tests/src/Functional/Rest/RdfMappingXmlAnonTest.php
@@ -21,6 +21,6 @@ class RdfMappingXmlAnonTest extends RdfMappingResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/modules/rdf/tests/src/Functional/Rest/RdfMappingXmlBasicAuthTest.php b/core/modules/rdf/tests/src/Functional/Rest/RdfMappingXmlBasicAuthTest.php
index 9cf6b97d51..b5eab3f28f 100644
--- a/core/modules/rdf/tests/src/Functional/Rest/RdfMappingXmlBasicAuthTest.php
+++ b/core/modules/rdf/tests/src/Functional/Rest/RdfMappingXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class RdfMappingXmlBasicAuthTest extends RdfMappingResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/rdf/tests/src/Functional/Rest/RdfMappingXmlCookieTest.php b/core/modules/rdf/tests/src/Functional/Rest/RdfMappingXmlCookieTest.php
index 1519e5ac72..c67d239902 100644
--- a/core/modules/rdf/tests/src/Functional/Rest/RdfMappingXmlCookieTest.php
+++ b/core/modules/rdf/tests/src/Functional/Rest/RdfMappingXmlCookieTest.php
@@ -21,7 +21,7 @@ class RdfMappingXmlCookieTest extends RdfMappingResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/responsive_image/tests/src/Functional/Rest/ResponsiveImageStyleXmlAnonTest.php b/core/modules/responsive_image/tests/src/Functional/Rest/ResponsiveImageStyleXmlAnonTest.php
index 8493eddabe..567cf9dc54 100644
--- a/core/modules/responsive_image/tests/src/Functional/Rest/ResponsiveImageStyleXmlAnonTest.php
+++ b/core/modules/responsive_image/tests/src/Functional/Rest/ResponsiveImageStyleXmlAnonTest.php
@@ -21,6 +21,6 @@ class ResponsiveImageStyleXmlAnonTest extends ResponsiveImageStyleResourceTestBa
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/modules/responsive_image/tests/src/Functional/Rest/ResponsiveImageStyleXmlBasicAuthTest.php b/core/modules/responsive_image/tests/src/Functional/Rest/ResponsiveImageStyleXmlBasicAuthTest.php
index e850e99d57..0f1fe73937 100644
--- a/core/modules/responsive_image/tests/src/Functional/Rest/ResponsiveImageStyleXmlBasicAuthTest.php
+++ b/core/modules/responsive_image/tests/src/Functional/Rest/ResponsiveImageStyleXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class ResponsiveImageStyleXmlBasicAuthTest extends ResponsiveImageStyleResourceT
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/responsive_image/tests/src/Functional/Rest/ResponsiveImageStyleXmlCookieTest.php b/core/modules/responsive_image/tests/src/Functional/Rest/ResponsiveImageStyleXmlCookieTest.php
index f63f91cd54..66c2d4997b 100644
--- a/core/modules/responsive_image/tests/src/Functional/Rest/ResponsiveImageStyleXmlCookieTest.php
+++ b/core/modules/responsive_image/tests/src/Functional/Rest/ResponsiveImageStyleXmlCookieTest.php
@@ -21,7 +21,7 @@ class ResponsiveImageStyleXmlCookieTest extends ResponsiveImageStyleResourceTest
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/rest/tests/src/Functional/EntityResource/EntityResourceTestBase.php b/core/modules/rest/tests/src/Functional/EntityResource/EntityResourceTestBase.php
index e090e2c0bd..9bd03f0b89 100644
--- a/core/modules/rest/tests/src/Functional/EntityResource/EntityResourceTestBase.php
+++ b/core/modules/rest/tests/src/Functional/EntityResource/EntityResourceTestBase.php
@@ -1538,8 +1538,12 @@ protected function assert406Response(ResponseInterface $response) {
     else {
       // This is the desired response.
       $this->assertSame(406, $response->getStatusCode());
-      $this->stringContains('?_format=' . static::$format . '>; rel="alternate"; type="' . static::$mimeType . '"', $response->getHeader('Link'));
-      $this->stringContains('?_format=foobar>; rel="alternate"', $response->getHeader('Link'));
+      $actual_link_header = $response->getHeader('Link');
+      if ($actual_link_header) {
+        $actual_link_header_string = is_array($actual_link_header) ? reset($actual_link_header) : $actual_link_header;
+        $this->assertContains('?_format=' . static::$format . '>; rel="alternate"; type="' . static::$mimeType . '"', $actual_link_header_string);
+        $this->assertContains('?_format=foobar>; rel="alternate"', $actual_link_header_string);
+      }
     }
   }
 
diff --git a/core/modules/rest/tests/src/Functional/EntityResource/ModeratedNode/ModeratedNodeXmlAnonTest.php b/core/modules/rest/tests/src/Functional/EntityResource/ModeratedNode/ModeratedNodeXmlAnonTest.php
index 4b91d766c6..eda136e07b 100644
--- a/core/modules/rest/tests/src/Functional/EntityResource/ModeratedNode/ModeratedNodeXmlAnonTest.php
+++ b/core/modules/rest/tests/src/Functional/EntityResource/ModeratedNode/ModeratedNodeXmlAnonTest.php
@@ -21,7 +21,7 @@ class ModeratedNodeXmlAnonTest extends ModeratedNodeResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/rest/tests/src/Functional/EntityResource/ModeratedNode/ModeratedNodeXmlBasicAuthTest.php b/core/modules/rest/tests/src/Functional/EntityResource/ModeratedNode/ModeratedNodeXmlBasicAuthTest.php
index a321e03cb2..b77e431f00 100644
--- a/core/modules/rest/tests/src/Functional/EntityResource/ModeratedNode/ModeratedNodeXmlBasicAuthTest.php
+++ b/core/modules/rest/tests/src/Functional/EntityResource/ModeratedNode/ModeratedNodeXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class ModeratedNodeXmlBasicAuthTest extends ModeratedNodeResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/rest/tests/src/Functional/EntityResource/ModeratedNode/ModeratedNodeXmlCookieTest.php b/core/modules/rest/tests/src/Functional/EntityResource/ModeratedNode/ModeratedNodeXmlCookieTest.php
index 2014a56ff4..09c803f67a 100644
--- a/core/modules/rest/tests/src/Functional/EntityResource/ModeratedNode/ModeratedNodeXmlCookieTest.php
+++ b/core/modules/rest/tests/src/Functional/EntityResource/ModeratedNode/ModeratedNodeXmlCookieTest.php
@@ -21,7 +21,7 @@ class ModeratedNodeXmlCookieTest extends ModeratedNodeResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/rest/tests/src/Functional/ResourceTestBase.php b/core/modules/rest/tests/src/Functional/ResourceTestBase.php
index d7ca1e34f3..1d1cedb62b 100644
--- a/core/modules/rest/tests/src/Functional/ResourceTestBase.php
+++ b/core/modules/rest/tests/src/Functional/ResourceTestBase.php
@@ -376,8 +376,11 @@ protected function request($method, Url $url, array $request_options) {
    *   (optional) The expected X-Drupal-Dynamic-Cache response header value, or
    *   FALSE if that header should be absent. Possible strings: 'MISS', 'HIT'.
    *   Defaults to FALSE.
+   * @param string|false $expected_content_type_charset
+   *   The expected charset of the content. FALSE in case this should not be
+   *   asserted.
    */
-  protected function assertResourceResponse($expected_status_code, $expected_body, ResponseInterface $response, $expected_cache_tags = FALSE, $expected_cache_contexts = FALSE, $expected_page_cache_header_value = FALSE, $expected_dynamic_page_cache_header_value = FALSE) {
+  protected function assertResourceResponse($expected_status_code, $expected_body, ResponseInterface $response, $expected_cache_tags = FALSE, $expected_cache_contexts = FALSE, $expected_page_cache_header_value = FALSE, $expected_dynamic_page_cache_header_value = FALSE, $expected_content_type_charset = FALSE) {
     $this->assertSame($expected_status_code, $response->getStatusCode());
     if ($expected_status_code === 204) {
       // DELETE responses should not include a Content-Type header. But Apache
@@ -388,7 +391,11 @@ protected function assertResourceResponse($expected_status_code, $expected_body,
       $this->assertSame('', (string) $response->getBody());
     }
     else {
-      $this->assertSame([static::$mimeType], $response->getHeader('Content-Type'));
+      $expected_content_type = static::$mimeType;
+      if ($expected_content_type_charset !== FALSE) {
+        $expected_content_type .=  '; charset=' . $expected_content_type_charset;
+      }
+      $this->assertSame([$expected_content_type], $response->getHeader('Content-Type'));
       if ($expected_body !== FALSE) {
         $this->assertSame($expected_body, (string) $response->getBody());
       }
diff --git a/core/modules/rest/tests/src/Functional/Rest/RestResourceConfigXmlAnonTest.php b/core/modules/rest/tests/src/Functional/Rest/RestResourceConfigXmlAnonTest.php
index a508b55409..6816edfdff 100644
--- a/core/modules/rest/tests/src/Functional/Rest/RestResourceConfigXmlAnonTest.php
+++ b/core/modules/rest/tests/src/Functional/Rest/RestResourceConfigXmlAnonTest.php
@@ -21,6 +21,6 @@ class RestResourceConfigXmlAnonTest extends RestResourceConfigResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/modules/rest/tests/src/Functional/Rest/RestResourceConfigXmlBasicAuthTest.php b/core/modules/rest/tests/src/Functional/Rest/RestResourceConfigXmlBasicAuthTest.php
index c5034b4ee5..ba19782c53 100644
--- a/core/modules/rest/tests/src/Functional/Rest/RestResourceConfigXmlBasicAuthTest.php
+++ b/core/modules/rest/tests/src/Functional/Rest/RestResourceConfigXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class RestResourceConfigXmlBasicAuthTest extends RestResourceConfigResourceTestB
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/rest/tests/src/Functional/Rest/RestResourceConfigXmlCookieTest.php b/core/modules/rest/tests/src/Functional/Rest/RestResourceConfigXmlCookieTest.php
index 377a4497e9..d36dce12d7 100644
--- a/core/modules/rest/tests/src/Functional/Rest/RestResourceConfigXmlCookieTest.php
+++ b/core/modules/rest/tests/src/Functional/Rest/RestResourceConfigXmlCookieTest.php
@@ -21,7 +21,7 @@ class RestResourceConfigXmlCookieTest extends RestResourceConfigResourceTestBase
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/search/tests/src/Functional/Rest/SearchPageXmlAnonTest.php b/core/modules/search/tests/src/Functional/Rest/SearchPageXmlAnonTest.php
index 5f4698ec1b..9346da77b4 100644
--- a/core/modules/search/tests/src/Functional/Rest/SearchPageXmlAnonTest.php
+++ b/core/modules/search/tests/src/Functional/Rest/SearchPageXmlAnonTest.php
@@ -21,6 +21,6 @@ class SearchPageXmlAnonTest extends SearchPageResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/modules/search/tests/src/Functional/Rest/SearchPageXmlBasicAuthTest.php b/core/modules/search/tests/src/Functional/Rest/SearchPageXmlBasicAuthTest.php
index 7a2c3c8804..f87e14013f 100644
--- a/core/modules/search/tests/src/Functional/Rest/SearchPageXmlBasicAuthTest.php
+++ b/core/modules/search/tests/src/Functional/Rest/SearchPageXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class SearchPageXmlBasicAuthTest extends SearchPageResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/search/tests/src/Functional/Rest/SearchPageXmlCookieTest.php b/core/modules/search/tests/src/Functional/Rest/SearchPageXmlCookieTest.php
index 3ad901d6ef..9f4f51badf 100644
--- a/core/modules/search/tests/src/Functional/Rest/SearchPageXmlCookieTest.php
+++ b/core/modules/search/tests/src/Functional/Rest/SearchPageXmlCookieTest.php
@@ -21,7 +21,7 @@ class SearchPageXmlCookieTest extends SearchPageResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/shortcut/tests/src/Functional/Rest/ShortcutSetXmlAnonTest.php b/core/modules/shortcut/tests/src/Functional/Rest/ShortcutSetXmlAnonTest.php
index 75ad14efd4..d6566761b8 100644
--- a/core/modules/shortcut/tests/src/Functional/Rest/ShortcutSetXmlAnonTest.php
+++ b/core/modules/shortcut/tests/src/Functional/Rest/ShortcutSetXmlAnonTest.php
@@ -21,6 +21,6 @@ class ShortcutSetXmlAnonTest extends ShortcutSetResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/modules/shortcut/tests/src/Functional/Rest/ShortcutSetXmlBasicAuthTest.php b/core/modules/shortcut/tests/src/Functional/Rest/ShortcutSetXmlBasicAuthTest.php
index 50c6e71c23..467fdd4e2d 100644
--- a/core/modules/shortcut/tests/src/Functional/Rest/ShortcutSetXmlBasicAuthTest.php
+++ b/core/modules/shortcut/tests/src/Functional/Rest/ShortcutSetXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class ShortcutSetXmlBasicAuthTest extends ShortcutSetResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/shortcut/tests/src/Functional/Rest/ShortcutSetXmlCookieTest.php b/core/modules/shortcut/tests/src/Functional/Rest/ShortcutSetXmlCookieTest.php
index 71dccd16c3..3d4939cead 100644
--- a/core/modules/shortcut/tests/src/Functional/Rest/ShortcutSetXmlCookieTest.php
+++ b/core/modules/shortcut/tests/src/Functional/Rest/ShortcutSetXmlCookieTest.php
@@ -21,7 +21,7 @@ class ShortcutSetXmlCookieTest extends ShortcutSetResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/shortcut/tests/src/Functional/Rest/ShortcutXmlAnonTest.php b/core/modules/shortcut/tests/src/Functional/Rest/ShortcutXmlAnonTest.php
index 56c5c681a3..7bf7655965 100644
--- a/core/modules/shortcut/tests/src/Functional/Rest/ShortcutXmlAnonTest.php
+++ b/core/modules/shortcut/tests/src/Functional/Rest/ShortcutXmlAnonTest.php
@@ -21,6 +21,6 @@ class ShortcutXmlAnonTest extends ShortcutResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/modules/shortcut/tests/src/Functional/Rest/ShortcutXmlBasicAuthTest.php b/core/modules/shortcut/tests/src/Functional/Rest/ShortcutXmlBasicAuthTest.php
index 9645aedaea..d03a8933ee 100644
--- a/core/modules/shortcut/tests/src/Functional/Rest/ShortcutXmlBasicAuthTest.php
+++ b/core/modules/shortcut/tests/src/Functional/Rest/ShortcutXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class ShortcutXmlBasicAuthTest extends ShortcutResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/shortcut/tests/src/Functional/Rest/ShortcutXmlCookieTest.php b/core/modules/shortcut/tests/src/Functional/Rest/ShortcutXmlCookieTest.php
index 899b51d88a..1c29ff1eb6 100644
--- a/core/modules/shortcut/tests/src/Functional/Rest/ShortcutXmlCookieTest.php
+++ b/core/modules/shortcut/tests/src/Functional/Rest/ShortcutXmlCookieTest.php
@@ -21,7 +21,7 @@ class ShortcutXmlCookieTest extends ShortcutResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/system/tests/modules/entity_test/tests/src/Functional/Rest/EntityTestBundleXmlAnonTest.php b/core/modules/system/tests/modules/entity_test/tests/src/Functional/Rest/EntityTestBundleXmlAnonTest.php
index e88314b425..b01ee60761 100644
--- a/core/modules/system/tests/modules/entity_test/tests/src/Functional/Rest/EntityTestBundleXmlAnonTest.php
+++ b/core/modules/system/tests/modules/entity_test/tests/src/Functional/Rest/EntityTestBundleXmlAnonTest.php
@@ -21,6 +21,6 @@ class EntityTestBundleXmlAnonTest extends EntityTestBundleResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/modules/system/tests/modules/entity_test/tests/src/Functional/Rest/EntityTestBundleXmlBasicAuthTest.php b/core/modules/system/tests/modules/entity_test/tests/src/Functional/Rest/EntityTestBundleXmlBasicAuthTest.php
index 701c15ba9f..b77f828cb3 100644
--- a/core/modules/system/tests/modules/entity_test/tests/src/Functional/Rest/EntityTestBundleXmlBasicAuthTest.php
+++ b/core/modules/system/tests/modules/entity_test/tests/src/Functional/Rest/EntityTestBundleXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class EntityTestBundleXmlBasicAuthTest extends EntityTestBundleResourceTestBase
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/system/tests/modules/entity_test/tests/src/Functional/Rest/EntityTestBundleXmlCookieTest.php b/core/modules/system/tests/modules/entity_test/tests/src/Functional/Rest/EntityTestBundleXmlCookieTest.php
index 2f3ee3310c..595ccd18cc 100644
--- a/core/modules/system/tests/modules/entity_test/tests/src/Functional/Rest/EntityTestBundleXmlCookieTest.php
+++ b/core/modules/system/tests/modules/entity_test/tests/src/Functional/Rest/EntityTestBundleXmlCookieTest.php
@@ -21,7 +21,7 @@ class EntityTestBundleXmlCookieTest extends EntityTestBundleResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/system/tests/modules/entity_test/tests/src/Functional/Rest/EntityTestLabelXmlAnonTest.php b/core/modules/system/tests/modules/entity_test/tests/src/Functional/Rest/EntityTestLabelXmlAnonTest.php
index ac1bd8685c..445e366ff1 100644
--- a/core/modules/system/tests/modules/entity_test/tests/src/Functional/Rest/EntityTestLabelXmlAnonTest.php
+++ b/core/modules/system/tests/modules/entity_test/tests/src/Functional/Rest/EntityTestLabelXmlAnonTest.php
@@ -21,6 +21,6 @@ class EntityTestLabelXmlAnonTest extends EntityTestLabelResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/modules/system/tests/modules/entity_test/tests/src/Functional/Rest/EntityTestLabelXmlBasicAuthTest.php b/core/modules/system/tests/modules/entity_test/tests/src/Functional/Rest/EntityTestLabelXmlBasicAuthTest.php
index e671098f2f..1ed8e608e9 100644
--- a/core/modules/system/tests/modules/entity_test/tests/src/Functional/Rest/EntityTestLabelXmlBasicAuthTest.php
+++ b/core/modules/system/tests/modules/entity_test/tests/src/Functional/Rest/EntityTestLabelXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class EntityTestLabelXmlBasicAuthTest extends EntityTestLabelResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/system/tests/modules/entity_test/tests/src/Functional/Rest/EntityTestLabelXmlCookieTest.php b/core/modules/system/tests/modules/entity_test/tests/src/Functional/Rest/EntityTestLabelXmlCookieTest.php
index f859c7ff8b..691d704186 100644
--- a/core/modules/system/tests/modules/entity_test/tests/src/Functional/Rest/EntityTestLabelXmlCookieTest.php
+++ b/core/modules/system/tests/modules/entity_test/tests/src/Functional/Rest/EntityTestLabelXmlCookieTest.php
@@ -21,7 +21,7 @@ class EntityTestLabelXmlCookieTest extends EntityTestLabelResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/system/tests/modules/entity_test/tests/src/Functional/Rest/EntityTestXmlAnonTest.php b/core/modules/system/tests/modules/entity_test/tests/src/Functional/Rest/EntityTestXmlAnonTest.php
index a5548f9a78..033b1d4834 100644
--- a/core/modules/system/tests/modules/entity_test/tests/src/Functional/Rest/EntityTestXmlAnonTest.php
+++ b/core/modules/system/tests/modules/entity_test/tests/src/Functional/Rest/EntityTestXmlAnonTest.php
@@ -23,6 +23,6 @@ class EntityTestXmlAnonTest extends EntityTestResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/modules/system/tests/modules/entity_test/tests/src/Functional/Rest/EntityTestXmlBasicAuthTest.php b/core/modules/system/tests/modules/entity_test/tests/src/Functional/Rest/EntityTestXmlBasicAuthTest.php
index d0501afb97..6a54646147 100644
--- a/core/modules/system/tests/modules/entity_test/tests/src/Functional/Rest/EntityTestXmlBasicAuthTest.php
+++ b/core/modules/system/tests/modules/entity_test/tests/src/Functional/Rest/EntityTestXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class EntityTestXmlBasicAuthTest extends EntityTestResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/system/tests/modules/entity_test/tests/src/Functional/Rest/EntityTestXmlCookieTest.php b/core/modules/system/tests/modules/entity_test/tests/src/Functional/Rest/EntityTestXmlCookieTest.php
index 7fe4930fcb..d38b3bfd14 100644
--- a/core/modules/system/tests/modules/entity_test/tests/src/Functional/Rest/EntityTestXmlCookieTest.php
+++ b/core/modules/system/tests/modules/entity_test/tests/src/Functional/Rest/EntityTestXmlCookieTest.php
@@ -21,7 +21,7 @@ class EntityTestXmlCookieTest extends EntityTestResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/system/tests/src/Functional/Rest/ActionXmlAnonTest.php b/core/modules/system/tests/src/Functional/Rest/ActionXmlAnonTest.php
index e45af980ea..f673502aac 100644
--- a/core/modules/system/tests/src/Functional/Rest/ActionXmlAnonTest.php
+++ b/core/modules/system/tests/src/Functional/Rest/ActionXmlAnonTest.php
@@ -21,6 +21,6 @@ class ActionXmlAnonTest extends ActionResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/modules/system/tests/src/Functional/Rest/ActionXmlBasicAuthTest.php b/core/modules/system/tests/src/Functional/Rest/ActionXmlBasicAuthTest.php
index f11459bc6d..f15fe4797c 100644
--- a/core/modules/system/tests/src/Functional/Rest/ActionXmlBasicAuthTest.php
+++ b/core/modules/system/tests/src/Functional/Rest/ActionXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class ActionXmlBasicAuthTest extends ActionResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/system/tests/src/Functional/Rest/ActionXmlCookieTest.php b/core/modules/system/tests/src/Functional/Rest/ActionXmlCookieTest.php
index dbee6fa1af..10bf1a4394 100644
--- a/core/modules/system/tests/src/Functional/Rest/ActionXmlCookieTest.php
+++ b/core/modules/system/tests/src/Functional/Rest/ActionXmlCookieTest.php
@@ -21,7 +21,7 @@ class ActionXmlCookieTest extends ActionResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/system/tests/src/Functional/Rest/MenuXmlAnonTest.php b/core/modules/system/tests/src/Functional/Rest/MenuXmlAnonTest.php
index e1c23b5f70..e2dd3012c0 100644
--- a/core/modules/system/tests/src/Functional/Rest/MenuXmlAnonTest.php
+++ b/core/modules/system/tests/src/Functional/Rest/MenuXmlAnonTest.php
@@ -21,6 +21,6 @@ class MenuXmlAnonTest extends MenuResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/modules/system/tests/src/Functional/Rest/MenuXmlBasicAuthTest.php b/core/modules/system/tests/src/Functional/Rest/MenuXmlBasicAuthTest.php
index 0910430042..9bd03f2427 100644
--- a/core/modules/system/tests/src/Functional/Rest/MenuXmlBasicAuthTest.php
+++ b/core/modules/system/tests/src/Functional/Rest/MenuXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class MenuXmlBasicAuthTest extends MenuResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/system/tests/src/Functional/Rest/MenuXmlCookieTest.php b/core/modules/system/tests/src/Functional/Rest/MenuXmlCookieTest.php
index a229b6a3b2..49c53a6fc8 100644
--- a/core/modules/system/tests/src/Functional/Rest/MenuXmlCookieTest.php
+++ b/core/modules/system/tests/src/Functional/Rest/MenuXmlCookieTest.php
@@ -21,7 +21,7 @@ class MenuXmlCookieTest extends MenuResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/taxonomy/tests/src/Functional/Rest/TermXmlAnonTest.php b/core/modules/taxonomy/tests/src/Functional/Rest/TermXmlAnonTest.php
index 69abeaf210..0197e95b06 100644
--- a/core/modules/taxonomy/tests/src/Functional/Rest/TermXmlAnonTest.php
+++ b/core/modules/taxonomy/tests/src/Functional/Rest/TermXmlAnonTest.php
@@ -21,7 +21,7 @@ class TermXmlAnonTest extends TermResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/taxonomy/tests/src/Functional/Rest/TermXmlBasicAuthTest.php b/core/modules/taxonomy/tests/src/Functional/Rest/TermXmlBasicAuthTest.php
index 8eacb85c1c..d1629ac319 100644
--- a/core/modules/taxonomy/tests/src/Functional/Rest/TermXmlBasicAuthTest.php
+++ b/core/modules/taxonomy/tests/src/Functional/Rest/TermXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class TermXmlBasicAuthTest extends TermResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/taxonomy/tests/src/Functional/Rest/TermXmlCookieTest.php b/core/modules/taxonomy/tests/src/Functional/Rest/TermXmlCookieTest.php
index 3de95fdaba..290ed00ec4 100644
--- a/core/modules/taxonomy/tests/src/Functional/Rest/TermXmlCookieTest.php
+++ b/core/modules/taxonomy/tests/src/Functional/Rest/TermXmlCookieTest.php
@@ -21,7 +21,7 @@ class TermXmlCookieTest extends TermResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/taxonomy/tests/src/Functional/Rest/VocabularyXmlAnonTest.php b/core/modules/taxonomy/tests/src/Functional/Rest/VocabularyXmlAnonTest.php
index ac8da3d741..e47d0536e5 100644
--- a/core/modules/taxonomy/tests/src/Functional/Rest/VocabularyXmlAnonTest.php
+++ b/core/modules/taxonomy/tests/src/Functional/Rest/VocabularyXmlAnonTest.php
@@ -21,6 +21,6 @@ class VocabularyXmlAnonTest extends VocabularyResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/modules/taxonomy/tests/src/Functional/Rest/VocabularyXmlBasicAuthTest.php b/core/modules/taxonomy/tests/src/Functional/Rest/VocabularyXmlBasicAuthTest.php
index 52b3741a9a..45d402bada 100644
--- a/core/modules/taxonomy/tests/src/Functional/Rest/VocabularyXmlBasicAuthTest.php
+++ b/core/modules/taxonomy/tests/src/Functional/Rest/VocabularyXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class VocabularyXmlBasicAuthTest extends VocabularyResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/taxonomy/tests/src/Functional/Rest/VocabularyXmlCookieTest.php b/core/modules/taxonomy/tests/src/Functional/Rest/VocabularyXmlCookieTest.php
index 51e7e20e49..e27c417b12 100644
--- a/core/modules/taxonomy/tests/src/Functional/Rest/VocabularyXmlCookieTest.php
+++ b/core/modules/taxonomy/tests/src/Functional/Rest/VocabularyXmlCookieTest.php
@@ -21,7 +21,7 @@ class VocabularyXmlCookieTest extends VocabularyResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/tour/tests/src/Functional/Rest/TourXmlAnonTest.php b/core/modules/tour/tests/src/Functional/Rest/TourXmlAnonTest.php
index 8f44c87bbd..8c30a1982d 100644
--- a/core/modules/tour/tests/src/Functional/Rest/TourXmlAnonTest.php
+++ b/core/modules/tour/tests/src/Functional/Rest/TourXmlAnonTest.php
@@ -21,6 +21,6 @@ class TourXmlAnonTest extends TourResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/modules/tour/tests/src/Functional/Rest/TourXmlBasicAuthTest.php b/core/modules/tour/tests/src/Functional/Rest/TourXmlBasicAuthTest.php
index 618b41defd..846b79260b 100644
--- a/core/modules/tour/tests/src/Functional/Rest/TourXmlBasicAuthTest.php
+++ b/core/modules/tour/tests/src/Functional/Rest/TourXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class TourXmlBasicAuthTest extends TourResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/tour/tests/src/Functional/Rest/TourXmlCookieTest.php b/core/modules/tour/tests/src/Functional/Rest/TourXmlCookieTest.php
index 4d65bfff3d..1290a84740 100644
--- a/core/modules/tour/tests/src/Functional/Rest/TourXmlCookieTest.php
+++ b/core/modules/tour/tests/src/Functional/Rest/TourXmlCookieTest.php
@@ -21,7 +21,7 @@ class TourXmlCookieTest extends TourResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/user/tests/src/Functional/Rest/RoleXmlAnonTest.php b/core/modules/user/tests/src/Functional/Rest/RoleXmlAnonTest.php
index 14910a6202..a78addf390 100644
--- a/core/modules/user/tests/src/Functional/Rest/RoleXmlAnonTest.php
+++ b/core/modules/user/tests/src/Functional/Rest/RoleXmlAnonTest.php
@@ -21,6 +21,6 @@ class RoleXmlAnonTest extends RoleResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/modules/user/tests/src/Functional/Rest/RoleXmlBasicAuthTest.php b/core/modules/user/tests/src/Functional/Rest/RoleXmlBasicAuthTest.php
index fa98edd47c..2d7fe2202e 100644
--- a/core/modules/user/tests/src/Functional/Rest/RoleXmlBasicAuthTest.php
+++ b/core/modules/user/tests/src/Functional/Rest/RoleXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class RoleXmlBasicAuthTest extends RoleResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/user/tests/src/Functional/Rest/RoleXmlCookieTest.php b/core/modules/user/tests/src/Functional/Rest/RoleXmlCookieTest.php
index 5e04d7c350..3df7c8def7 100644
--- a/core/modules/user/tests/src/Functional/Rest/RoleXmlCookieTest.php
+++ b/core/modules/user/tests/src/Functional/Rest/RoleXmlCookieTest.php
@@ -21,7 +21,7 @@ class RoleXmlCookieTest extends RoleResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/user/tests/src/Functional/Rest/UserXmlAnonTest.php b/core/modules/user/tests/src/Functional/Rest/UserXmlAnonTest.php
index 4629192ce0..dcdcc1bc5d 100644
--- a/core/modules/user/tests/src/Functional/Rest/UserXmlAnonTest.php
+++ b/core/modules/user/tests/src/Functional/Rest/UserXmlAnonTest.php
@@ -21,7 +21,7 @@ class UserXmlAnonTest extends UserResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/user/tests/src/Functional/Rest/UserXmlBasicAuthTest.php b/core/modules/user/tests/src/Functional/Rest/UserXmlBasicAuthTest.php
index fc985980c3..54f762b490 100644
--- a/core/modules/user/tests/src/Functional/Rest/UserXmlBasicAuthTest.php
+++ b/core/modules/user/tests/src/Functional/Rest/UserXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class UserXmlBasicAuthTest extends UserResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/user/tests/src/Functional/Rest/UserXmlCookieTest.php b/core/modules/user/tests/src/Functional/Rest/UserXmlCookieTest.php
index 9a0c724f4d..3c2a9ccfed 100644
--- a/core/modules/user/tests/src/Functional/Rest/UserXmlCookieTest.php
+++ b/core/modules/user/tests/src/Functional/Rest/UserXmlCookieTest.php
@@ -21,7 +21,7 @@ class UserXmlCookieTest extends UserResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/views/tests/src/Functional/Rest/ViewXmlAnonTest.php b/core/modules/views/tests/src/Functional/Rest/ViewXmlAnonTest.php
index 7582a54f05..a4e7946486 100644
--- a/core/modules/views/tests/src/Functional/Rest/ViewXmlAnonTest.php
+++ b/core/modules/views/tests/src/Functional/Rest/ViewXmlAnonTest.php
@@ -21,6 +21,6 @@ class ViewXmlAnonTest extends ViewResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/modules/views/tests/src/Functional/Rest/ViewXmlBasicAuthTest.php b/core/modules/views/tests/src/Functional/Rest/ViewXmlBasicAuthTest.php
index ba9e58b3e6..aaa4d9c799 100644
--- a/core/modules/views/tests/src/Functional/Rest/ViewXmlBasicAuthTest.php
+++ b/core/modules/views/tests/src/Functional/Rest/ViewXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class ViewXmlBasicAuthTest extends ViewResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/views/tests/src/Functional/Rest/ViewXmlCookieTest.php b/core/modules/views/tests/src/Functional/Rest/ViewXmlCookieTest.php
index 8f362a06c1..dacdb4516b 100644
--- a/core/modules/views/tests/src/Functional/Rest/ViewXmlCookieTest.php
+++ b/core/modules/views/tests/src/Functional/Rest/ViewXmlCookieTest.php
@@ -21,7 +21,7 @@ class ViewXmlCookieTest extends ViewResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/workflows/tests/src/Functional/Rest/WorkflowXmlAnonTest.php b/core/modules/workflows/tests/src/Functional/Rest/WorkflowXmlAnonTest.php
index a66cc7763b..a3061e367c 100644
--- a/core/modules/workflows/tests/src/Functional/Rest/WorkflowXmlAnonTest.php
+++ b/core/modules/workflows/tests/src/Functional/Rest/WorkflowXmlAnonTest.php
@@ -21,6 +21,6 @@ class WorkflowXmlAnonTest extends WorkflowResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/modules/workflows/tests/src/Functional/Rest/WorkflowXmlBasicAuthTest.php b/core/modules/workflows/tests/src/Functional/Rest/WorkflowXmlBasicAuthTest.php
index db443e4763..5433a2c0d3 100644
--- a/core/modules/workflows/tests/src/Functional/Rest/WorkflowXmlBasicAuthTest.php
+++ b/core/modules/workflows/tests/src/Functional/Rest/WorkflowXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class WorkflowXmlBasicAuthTest extends WorkflowResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/workflows/tests/src/Functional/Rest/WorkflowXmlCookieTest.php b/core/modules/workflows/tests/src/Functional/Rest/WorkflowXmlCookieTest.php
index 50ea707ba8..35c11efb2d 100644
--- a/core/modules/workflows/tests/src/Functional/Rest/WorkflowXmlCookieTest.php
+++ b/core/modules/workflows/tests/src/Functional/Rest/WorkflowXmlCookieTest.php
@@ -21,7 +21,7 @@ class WorkflowXmlCookieTest extends WorkflowResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/workspaces/tests/src/Functional/EntityResource/WorkspaceXmlAnonTest.php b/core/modules/workspaces/tests/src/Functional/EntityResource/WorkspaceXmlAnonTest.php
index c87b438a23..d978cc459e 100644
--- a/core/modules/workspaces/tests/src/Functional/EntityResource/WorkspaceXmlAnonTest.php
+++ b/core/modules/workspaces/tests/src/Functional/EntityResource/WorkspaceXmlAnonTest.php
@@ -23,7 +23,7 @@ class WorkspaceXmlAnonTest extends WorkspaceResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/workspaces/tests/src/Functional/EntityResource/WorkspaceXmlBasicAuthTest.php b/core/modules/workspaces/tests/src/Functional/EntityResource/WorkspaceXmlBasicAuthTest.php
index 8769d9e8ba..a929633216 100644
--- a/core/modules/workspaces/tests/src/Functional/EntityResource/WorkspaceXmlBasicAuthTest.php
+++ b/core/modules/workspaces/tests/src/Functional/EntityResource/WorkspaceXmlBasicAuthTest.php
@@ -28,7 +28,7 @@ class WorkspaceXmlBasicAuthTest extends WorkspaceResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/modules/workspaces/tests/src/Functional/EntityResource/WorkspaceXmlCookieTest.php b/core/modules/workspaces/tests/src/Functional/EntityResource/WorkspaceXmlCookieTest.php
index c37a28376d..22d8b33ba2 100644
--- a/core/modules/workspaces/tests/src/Functional/EntityResource/WorkspaceXmlCookieTest.php
+++ b/core/modules/workspaces/tests/src/Functional/EntityResource/WorkspaceXmlCookieTest.php
@@ -23,7 +23,7 @@ class WorkspaceXmlCookieTest extends WorkspaceResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/tests/Drupal/FunctionalTests/Rest/BaseFieldOverrideXmlAnonTest.php b/core/tests/Drupal/FunctionalTests/Rest/BaseFieldOverrideXmlAnonTest.php
index 7b995b4250..1e100aa4ca 100644
--- a/core/tests/Drupal/FunctionalTests/Rest/BaseFieldOverrideXmlAnonTest.php
+++ b/core/tests/Drupal/FunctionalTests/Rest/BaseFieldOverrideXmlAnonTest.php
@@ -21,6 +21,6 @@ class BaseFieldOverrideXmlAnonTest extends BaseFieldOverrideResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/tests/Drupal/FunctionalTests/Rest/BaseFieldOverrideXmlBasicAuthTest.php b/core/tests/Drupal/FunctionalTests/Rest/BaseFieldOverrideXmlBasicAuthTest.php
index 1e3f833567..1adf2a9ba8 100644
--- a/core/tests/Drupal/FunctionalTests/Rest/BaseFieldOverrideXmlBasicAuthTest.php
+++ b/core/tests/Drupal/FunctionalTests/Rest/BaseFieldOverrideXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class BaseFieldOverrideXmlBasicAuthTest extends BaseFieldOverrideResourceTestBas
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/tests/Drupal/FunctionalTests/Rest/BaseFieldOverrideXmlCookieTest.php b/core/tests/Drupal/FunctionalTests/Rest/BaseFieldOverrideXmlCookieTest.php
index ba1d166c82..8f0db8b289 100644
--- a/core/tests/Drupal/FunctionalTests/Rest/BaseFieldOverrideXmlCookieTest.php
+++ b/core/tests/Drupal/FunctionalTests/Rest/BaseFieldOverrideXmlCookieTest.php
@@ -21,7 +21,7 @@ class BaseFieldOverrideXmlCookieTest extends BaseFieldOverrideResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/tests/Drupal/FunctionalTests/Rest/DateFormatXmlAnonTest.php b/core/tests/Drupal/FunctionalTests/Rest/DateFormatXmlAnonTest.php
index 8eec7d4e8c..72a17a973b 100644
--- a/core/tests/Drupal/FunctionalTests/Rest/DateFormatXmlAnonTest.php
+++ b/core/tests/Drupal/FunctionalTests/Rest/DateFormatXmlAnonTest.php
@@ -21,6 +21,6 @@ class DateFormatXmlAnonTest extends DateFormatResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/tests/Drupal/FunctionalTests/Rest/DateFormatXmlBasicAuthTest.php b/core/tests/Drupal/FunctionalTests/Rest/DateFormatXmlBasicAuthTest.php
index eb787a99b8..d520d62b89 100644
--- a/core/tests/Drupal/FunctionalTests/Rest/DateFormatXmlBasicAuthTest.php
+++ b/core/tests/Drupal/FunctionalTests/Rest/DateFormatXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class DateFormatXmlBasicAuthTest extends DateFormatResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/tests/Drupal/FunctionalTests/Rest/DateFormatXmlCookieTest.php b/core/tests/Drupal/FunctionalTests/Rest/DateFormatXmlCookieTest.php
index 124614d27d..dcfdad3c24 100644
--- a/core/tests/Drupal/FunctionalTests/Rest/DateFormatXmlCookieTest.php
+++ b/core/tests/Drupal/FunctionalTests/Rest/DateFormatXmlCookieTest.php
@@ -21,7 +21,7 @@ class DateFormatXmlCookieTest extends DateFormatResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/tests/Drupal/FunctionalTests/Rest/EntityFormDisplayXmlAnonTest.php b/core/tests/Drupal/FunctionalTests/Rest/EntityFormDisplayXmlAnonTest.php
index da1437ed12..9c37730019 100644
--- a/core/tests/Drupal/FunctionalTests/Rest/EntityFormDisplayXmlAnonTest.php
+++ b/core/tests/Drupal/FunctionalTests/Rest/EntityFormDisplayXmlAnonTest.php
@@ -21,6 +21,6 @@ class EntityFormDisplayXmlAnonTest extends EntityFormDisplayResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/tests/Drupal/FunctionalTests/Rest/EntityFormDisplayXmlBasicAuthTest.php b/core/tests/Drupal/FunctionalTests/Rest/EntityFormDisplayXmlBasicAuthTest.php
index 2edf2460a9..f81291d533 100644
--- a/core/tests/Drupal/FunctionalTests/Rest/EntityFormDisplayXmlBasicAuthTest.php
+++ b/core/tests/Drupal/FunctionalTests/Rest/EntityFormDisplayXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class EntityFormDisplayXmlBasicAuthTest extends EntityFormDisplayResourceTestBas
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/tests/Drupal/FunctionalTests/Rest/EntityFormDisplayXmlCookieTest.php b/core/tests/Drupal/FunctionalTests/Rest/EntityFormDisplayXmlCookieTest.php
index e9e93b0d47..509047aad4 100644
--- a/core/tests/Drupal/FunctionalTests/Rest/EntityFormDisplayXmlCookieTest.php
+++ b/core/tests/Drupal/FunctionalTests/Rest/EntityFormDisplayXmlCookieTest.php
@@ -21,7 +21,7 @@ class EntityFormDisplayXmlCookieTest extends EntityFormDisplayResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/tests/Drupal/FunctionalTests/Rest/EntityFormModeXmlAnonTest.php b/core/tests/Drupal/FunctionalTests/Rest/EntityFormModeXmlAnonTest.php
index 2da724f804..207dd211bb 100644
--- a/core/tests/Drupal/FunctionalTests/Rest/EntityFormModeXmlAnonTest.php
+++ b/core/tests/Drupal/FunctionalTests/Rest/EntityFormModeXmlAnonTest.php
@@ -21,6 +21,6 @@ class EntityFormModeXmlAnonTest extends EntityFormModeResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/tests/Drupal/FunctionalTests/Rest/EntityFormModeXmlBasicAuthTest.php b/core/tests/Drupal/FunctionalTests/Rest/EntityFormModeXmlBasicAuthTest.php
index 721ded8fd9..e82127b47f 100644
--- a/core/tests/Drupal/FunctionalTests/Rest/EntityFormModeXmlBasicAuthTest.php
+++ b/core/tests/Drupal/FunctionalTests/Rest/EntityFormModeXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class EntityFormModeXmlBasicAuthTest extends EntityFormModeResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/tests/Drupal/FunctionalTests/Rest/EntityFormModeXmlCookieTest.php b/core/tests/Drupal/FunctionalTests/Rest/EntityFormModeXmlCookieTest.php
index 6f426570a5..b8076f70c1 100644
--- a/core/tests/Drupal/FunctionalTests/Rest/EntityFormModeXmlCookieTest.php
+++ b/core/tests/Drupal/FunctionalTests/Rest/EntityFormModeXmlCookieTest.php
@@ -21,7 +21,7 @@ class EntityFormModeXmlCookieTest extends EntityFormModeResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/tests/Drupal/FunctionalTests/Rest/EntityViewDisplayXmlAnonTest.php b/core/tests/Drupal/FunctionalTests/Rest/EntityViewDisplayXmlAnonTest.php
index 561f2374fd..69b67a349c 100644
--- a/core/tests/Drupal/FunctionalTests/Rest/EntityViewDisplayXmlAnonTest.php
+++ b/core/tests/Drupal/FunctionalTests/Rest/EntityViewDisplayXmlAnonTest.php
@@ -21,6 +21,6 @@ class EntityViewDisplayXmlAnonTest extends EntityViewDisplayResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/tests/Drupal/FunctionalTests/Rest/EntityViewDisplayXmlBasicAuthTest.php b/core/tests/Drupal/FunctionalTests/Rest/EntityViewDisplayXmlBasicAuthTest.php
index 7c7ce5bb03..1d1db975e8 100644
--- a/core/tests/Drupal/FunctionalTests/Rest/EntityViewDisplayXmlBasicAuthTest.php
+++ b/core/tests/Drupal/FunctionalTests/Rest/EntityViewDisplayXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class EntityViewDisplayXmlBasicAuthTest extends EntityViewDisplayResourceTestBas
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/tests/Drupal/FunctionalTests/Rest/EntityViewDisplayXmlCookieTest.php b/core/tests/Drupal/FunctionalTests/Rest/EntityViewDisplayXmlCookieTest.php
index 9024ff71ae..5e1c8cba94 100644
--- a/core/tests/Drupal/FunctionalTests/Rest/EntityViewDisplayXmlCookieTest.php
+++ b/core/tests/Drupal/FunctionalTests/Rest/EntityViewDisplayXmlCookieTest.php
@@ -21,7 +21,7 @@ class EntityViewDisplayXmlCookieTest extends EntityViewDisplayResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/tests/Drupal/FunctionalTests/Rest/EntityViewModeXmlAnonTest.php b/core/tests/Drupal/FunctionalTests/Rest/EntityViewModeXmlAnonTest.php
index 6e35948417..3db9247986 100644
--- a/core/tests/Drupal/FunctionalTests/Rest/EntityViewModeXmlAnonTest.php
+++ b/core/tests/Drupal/FunctionalTests/Rest/EntityViewModeXmlAnonTest.php
@@ -21,6 +21,6 @@ class EntityViewModeXmlAnonTest extends EntityViewModeResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
 }
diff --git a/core/tests/Drupal/FunctionalTests/Rest/EntityViewModeXmlBasicAuthTest.php b/core/tests/Drupal/FunctionalTests/Rest/EntityViewModeXmlBasicAuthTest.php
index 1b1e7181d0..14ed1f42ca 100644
--- a/core/tests/Drupal/FunctionalTests/Rest/EntityViewModeXmlBasicAuthTest.php
+++ b/core/tests/Drupal/FunctionalTests/Rest/EntityViewModeXmlBasicAuthTest.php
@@ -26,7 +26,7 @@ class EntityViewModeXmlBasicAuthTest extends EntityViewModeResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
diff --git a/core/tests/Drupal/FunctionalTests/Rest/EntityViewModeXmlCookieTest.php b/core/tests/Drupal/FunctionalTests/Rest/EntityViewModeXmlCookieTest.php
index bde3e24664..a7901b64a6 100644
--- a/core/tests/Drupal/FunctionalTests/Rest/EntityViewModeXmlCookieTest.php
+++ b/core/tests/Drupal/FunctionalTests/Rest/EntityViewModeXmlCookieTest.php
@@ -21,7 +21,7 @@ class EntityViewModeXmlCookieTest extends EntityViewModeResourceTestBase {
   /**
    * {@inheritdoc}
    */
-  protected static $mimeType = 'text/xml; charset=UTF-8';
+  protected static $mimeType = 'text/xml';
 
   /**
    * {@inheritdoc}
