diff --git a/core/phpunit.xml.dist b/core/phpunit.xml.dist
index 4f81a2d533..71e66110e0 100644
--- a/core/phpunit.xml.dist
+++ b/core/phpunit.xml.dist
@@ -48,6 +48,13 @@
     <testsuite name="functional-javascript">
       <file>./tests/TestSuites/FunctionalJavascriptTestSuite.php</file>
     </testsuite>
+    <testsuite name="drupal-directories">
+      <directory>./tests/Drupal</directory>
+      <directory>./modules/*/tests/src</directory>
+      <directory>../modules/*/tests/src</directory>
+      <directory>../profiles/*/tests/src</directory>
+      <directory>../sites/*/modules/*/tests/src</directory>
+    </testsuite>
   </testsuites>
   <listeners>
     <listener class="\Drupal\Tests\Listeners\DrupalListener">
