diff --git a/core/tests/Drupal/Tests/Component/Diff/DiffFormatterTest.php b/core/tests/Drupal/Tests/Component/Diff/DiffFormatterTest.php
index 9016dd483f..1709da336d 100644
--- a/core/tests/Drupal/Tests/Component/Diff/DiffFormatterTest.php
+++ b/core/tests/Drupal/Tests/Component/Diff/DiffFormatterTest.php
@@ -52,6 +52,7 @@ public static function provideTestDiff() {
    */
   public function testDiff($expected, $from, $to) {
     $diff = new Diff($from, $to);
+    dump($diff);
     $formatter = new DiffFormatter();
     $output = $formatter->format($diff);
     $this->assertEquals($expected, $output);
