diff --git a/core/lib/Drupal/Core/Config/ExportStorageManager.php b/core/lib/Drupal/Core/Config/ExportStorageManager.php
index fdd34cad54c..21c386be581 100644
--- a/core/lib/Drupal/Core/Config/ExportStorageManager.php
+++ b/core/lib/Drupal/Core/Config/ExportStorageManager.php
@@ -68,7 +68,7 @@ public function getStorage() {
     // Wrapping the queries in a transaction for performance gain.
     $transaction = $this->connection->startTransaction();
     self::replaceStorageContents($this->active, $this->storage);
-    unset($transaction);
+    $transaction->commitOrRelease();
 
     $this->eventDispatcher->dispatch(new StorageTransformEvent($this->storage), ConfigEvents::STORAGE_TRANSFORM_EXPORT);
 
diff --git a/core/lib/Drupal/Core/Config/ImportStorageTransformer.php b/core/lib/Drupal/Core/Config/ImportStorageTransformer.php
index df02d748794..b9765674b3b 100644
--- a/core/lib/Drupal/Core/Config/ImportStorageTransformer.php
+++ b/core/lib/Drupal/Core/Config/ImportStorageTransformer.php
@@ -117,7 +117,7 @@ public function transform(StorageInterface $storage) {
     // Wrapping the queries in a transaction for performance gain.
     $transaction = $this->connection->startTransaction();
     self::replaceStorageContents($storage, $mutable);
-    unset($transaction);
+    $transaction->commitOrRelease();
 
     // Dispatch the event so that event listeners can alter the configuration.
     $this->eventDispatcher->dispatch(new StorageTransformEvent($mutable), ConfigEvents::STORAGE_TRANSFORM_IMPORT);
diff --git a/core/lib/Drupal/Core/Database/Connection.php b/core/lib/Drupal/Core/Database/Connection.php
index 489b2f5f94d..8f55a9e0469 100644
--- a/core/lib/Drupal/Core/Database/Connection.php
+++ b/core/lib/Drupal/Core/Database/Connection.php
@@ -223,8 +223,14 @@ public function __destruct() {
    *   This method exists only to work around a bug caused by Drupal incorrectly
    *   relying on object destruction order to commit transactions. Xdebug 3.3.0
    *   changes the order of object destruction when the develop mode is enabled.
+   *
+   * @deprecated in drupal:11.3.0 and is removed from drupal:13.0.0. There is no
+   *   replacement.
+   *
+   * @see https://www.drupal.org/node/3524461
    */
   public function commitAll() {
+    @trigger_error(__METHOD__ . '() is deprecated in drupal:11.3.0 and is removed from drupal:13.0.0. There is no replacement. See https://www.drupal.org/node/3524461', E_USER_DEPRECATED);
     $manager = $this->transactionManager();
     if ($manager->inTransaction() && method_exists($manager, 'commitAll')) {
       $this->transactionManager()->commitAll();
diff --git a/core/lib/Drupal/Core/Database/Database.php b/core/lib/Drupal/Core/Database/Database.php
index e76bc2d6991..7f09ea4e4ad 100644
--- a/core/lib/Drupal/Core/Database/Database.php
+++ b/core/lib/Drupal/Core/Database/Database.php
@@ -448,17 +448,9 @@ public static function closeConnection($target = NULL, $key = NULL) {
       $key = self::$activeKey;
     }
     if (isset($target) && isset(self::$connections[$key][$target])) {
-      if (self::$connections[$key][$target] instanceof Connection) {
-        self::$connections[$key][$target]->commitAll();
-      }
       unset(self::$connections[$key][$target]);
     }
     elseif (isset(self::$connections[$key])) {
-      foreach (self::$connections[$key] as $connection) {
-        if ($connection instanceof Connection) {
-          $connection->commitAll();
-        }
-      }
       unset(self::$connections[$key]);
     }
 
@@ -626,8 +618,14 @@ public static function getConnectionInfoAsUrl($key = 'default') {
    *   This method exists only to work around a bug caused by Drupal incorrectly
    *   relying on object destruction order to commit transactions. Xdebug 3.3.0
    *   changes the order of object destruction when the develop mode is enabled.
+   *
+   * @deprecated in drupal:11.3.0 and is removed from drupal:13.0.0. There is no
+   *   replacement.
+   *
+   * @see https://www.drupal.org/node/3524461
    */
   public static function commitAllOnShutdown(bool $shutdown = FALSE): void {
+    @trigger_error(__METHOD__ . '() is deprecated in drupal:11.3.0 and is removed from drupal:13.0.0. There is no replacement. See https://www.drupal.org/node/3524461', E_USER_DEPRECATED);
     static $registered = FALSE;
 
     if ($shutdown) {
diff --git a/core/lib/Drupal/Core/Database/Query/Insert.php b/core/lib/Drupal/Core/Database/Query/Insert.php
index 4657e5c3e5e..08829deca37 100644
--- a/core/lib/Drupal/Core/Database/Query/Insert.php
+++ b/core/lib/Drupal/Core/Database/Query/Insert.php
@@ -84,6 +84,7 @@ public function execute() {
         $stmt->execute($insert_values, $this->queryOptions);
         $last_insert_id = $this->connection->lastInsertId();
       }
+      $transaction->commitOrRelease();
     }
     catch (\Exception $e) {
       if (isset($transaction)) {
diff --git a/core/lib/Drupal/Core/Database/Transaction.php b/core/lib/Drupal/Core/Database/Transaction.php
index b8693e4bb76..f0c222c4032 100644
--- a/core/lib/Drupal/Core/Database/Transaction.php
+++ b/core/lib/Drupal/Core/Database/Transaction.php
@@ -5,9 +5,6 @@
 /**
  * A wrapper class for creating and managing database transactions.
  *
- * To begin a transaction, simply start it. When the object goes out of scope
- * and is destroyed it will automatically commit.
- *
  * In the vast majority of cases, you should not instantiate this class
  * directly. Instead, call ->startTransaction(), from the appropriate connection
  * object.
@@ -21,10 +18,6 @@ public function __construct(
     protected readonly string $name,
     protected readonly string $id,
   ) {
-    // Transactions rely on objects being destroyed in order to be committed.
-    // PHP makes no guarantee about the order in which objects are destroyed so
-    // ensure all transactions are committed on shutdown.
-    Database::commitAllOnShutdown();
   }
 
   /**
diff --git a/core/lib/Drupal/Core/Database/Transaction/TransactionManagerBase.php b/core/lib/Drupal/Core/Database/Transaction/TransactionManagerBase.php
index fa1a309a767..fb1a4fb5829 100644
--- a/core/lib/Drupal/Core/Database/Transaction/TransactionManagerBase.php
+++ b/core/lib/Drupal/Core/Database/Transaction/TransactionManagerBase.php
@@ -101,6 +101,14 @@ abstract class TransactionManagerBase implements TransactionManagerInterface {
    */
   private ClientConnectionTransactionState $connectionTransactionState;
 
+  /**
+   * Indicates if a exception was thrown.
+   *
+   * This is needed to manage destruction of objects after a transaction was
+   * thrown on an active stack.
+   */
+  private bool $exceptionThrown = FALSE;
+
   /**
    * Whether to trigger warnings when unpiling a void transaction.
    *
@@ -165,8 +173,14 @@ protected function stack(): array {
    *   This method exists only to work around a bug caused by Drupal incorrectly
    *   relying on object destruction order to commit transactions. Xdebug 3.3.0
    *   changes the order of object destruction when the develop mode is enabled.
+   *
+   * @deprecated in drupal:11.3.0 and is removed from drupal:13.0.0. There is no
+   *   replacement.
+   *
+   * @see https://www.drupal.org/node/3524461
    */
   public function commitAll(): void {
+    @trigger_error(__METHOD__ . '() is deprecated in drupal:11.3.0 and is removed from drupal:13.0.0. There is no replacement. See https://www.drupal.org/node/3524461', E_USER_DEPRECATED);
     foreach (array_reverse($this->stack()) as $id => $item) {
       $this->unpile($item->name, $id);
     }
@@ -264,17 +278,20 @@ public function push(string $name = ''): Transaction {
     }
 
     if ($this->has($name)) {
+      $this->exceptionThrown = TRUE;
       throw new TransactionNameNonUniqueException("A transaction named {$name} is already in use. Active stack: " . $this->dumpStackItemsAsString());
     }
 
     // Define a unique ID for the transaction.
     $id = uniqid('', TRUE);
+dump([__METHOD__, $this->connectionTransactionState ?? 'nah', $this->dumpStackItemsAsString(), $name, $id]);
 
     // Do the client-level processing.
     if ($this->stackDepth() === 0) {
       $this->beginClientTransaction();
       $type = StackItemType::Root;
       $this->setConnectionTransactionState(ClientConnectionTransactionState::Active);
+      $this->exceptionThrown = FALSE;
       // Only set ::rootId if there's not one set already, which may happen in
       // case of broken transactions.
       if ($this->rootId === NULL) {
@@ -315,6 +332,7 @@ public function push(string $name = ''): Transaction {
    * @internal
    */
   public function purge(string $name, string $id): void {
+dump([__METHOD__, $this->getConnectionTransactionState(), $this->dumpStackItemsAsString(), $name, $id]);
     // If this is a 'root' transaction, and it is voided (that is, no longer in
     // the stack), then the transaction on the database is no longer active. An
     // action such as a commit, a release savepoint, a rollback, or a DDL
@@ -336,6 +354,15 @@ public function purge(string $name, string $id): void {
       return;
     }
 
+    // We deprecate auto-commit on destruction, but in case an exception is
+    // thrown, we still need to destruct and cleanup objects in relation to the
+    // underlying database transaction, so we do not trigger the error in such
+    // circumstance.
+    if (!$this->exceptionThrown) {
+      @trigger_error('Database commit by letting a Transaction object go out of scope is deprecated in drupal:11.3.0 and is removed from drupal:13.0.0. Commit explicitly via Transaction::commitOrRelease() instead. See https://www.drupal.org/node/3524461', E_USER_DEPRECATED);
+    }
+
+    // Commit the transaction.
     // When we get here, the transaction (or savepoint) is still active on the
     // database. We can unpile it, and if we are left with no more items in the
     // stack, we can also process the post transaction callbacks.
@@ -351,6 +378,7 @@ public function purge(string $name, string $id): void {
    * {@inheritdoc}
    */
   public function unpile(string $name, string $id): void {
+dump([__METHOD__, $this->getConnectionTransactionState(), $this->dumpStackItemsAsString(), $name, $id]);
     // If the transaction was voided, we cannot unpile. Skip but trigger a user
     // warning if requested.
     if ($this->getConnectionTransactionState() === ClientConnectionTransactionState::Voided) {
@@ -360,9 +388,16 @@ public function unpile(string $name, string $id): void {
       return;
     }
 
+    // If the client transaction was already committed, there's no longer
+    // anything to do on an explicit commit/savepoint release.
+    if ($this->getConnectionTransactionState() === ClientConnectionTransactionState::Committed) {
+      return;
+    }
+
     // If there is no $id to commit, or if $id does not correspond to the one
     // in the stack for that $name, the commit is out of order.
     if (!isset($this->stack()[$id]) || $this->stack()[$id]->name !== $name) {
+      $this->exceptionThrown = TRUE;
       throw new TransactionOutOfOrderException("Error attempting commit of {$id}\\{$name}. Active stack: " . $this->dumpStackItemsAsString());
     }
 
@@ -387,8 +422,10 @@ public function unpile(string $name, string $id): void {
    *   If the commit of the root transaction failed.
    */
   protected function commit(string $name, string $id): void {
+#dump([__METHOD__, $this->getConnectionTransactionState(), $this->dumpStackItemsAsString(), $name, $id]);
     if ($this->getConnectionTransactionState() !== ClientConnectionTransactionState::Active) {
       // The stack got corrupted.
+      $this->exceptionThrown = TRUE;
       throw new TransactionOutOfOrderException("Transaction {$id}\\{$name} is out of order. Active stack: " . $this->dumpStackItemsAsString());
     }
 
@@ -412,6 +449,7 @@ protected function commit(string $name, string $id): void {
     }
     else {
       // The stack got corrupted.
+      $this->exceptionThrown = TRUE;
       throw new TransactionOutOfOrderException("Transaction {$id}/{$name} is out of order. Active stack: " . $this->dumpStackItemsAsString());
     }
   }
@@ -420,16 +458,18 @@ protected function commit(string $name, string $id): void {
    * {@inheritdoc}
    */
   public function rollback(string $name, string $id): void {
+dump([__METHOD__, $this->getConnectionTransactionState(), $this->dumpStackItemsAsString(), $name, $id]);
     // If the transaction was voided, we cannot rollback. Fail silently but
     // trigger a user warning.
     if ($this->getConnectionTransactionState() === ClientConnectionTransactionState::Voided) {
-      $this->connectionTransactionState = ClientConnectionTransactionState::RollbackFailed;
+      $this->setConnectionTransactionState(ClientConnectionTransactionState::RollbackFailed);
       trigger_error('Transaction::rollBack() failed because of a prior execution of a DDL statement.', E_USER_WARNING);
       return;
     }
 
     // Rolled back item should match the last one in stack.
     if ($id != array_key_last($this->stack()) || $name !== $this->stack()[$id]->name) {
+      $this->exceptionThrown = TRUE;
       throw new TransactionOutOfOrderException("Error attempting rollback of {$id}\\{$name}. Active stack: " . $this->dumpStackItemsAsString());
     }
 
@@ -456,12 +496,14 @@ public function rollback(string $name, string $id): void {
       }
       else {
         // The stack got corrupted.
+        $this->exceptionThrown = TRUE;
         throw new TransactionOutOfOrderException("Error attempting rollback of {$id}\\{$name}. Active stack: " . $this->dumpStackItemsAsString());
       }
       return;
     }
 
     // The stack got corrupted.
+    $this->exceptionThrown = TRUE;
     throw new TransactionOutOfOrderException("Error attempting rollback of {$id}\\{$name}. Active stack: " . $this->dumpStackItemsAsString());
   }
 
@@ -470,6 +512,7 @@ public function rollback(string $name, string $id): void {
    */
   public function addPostTransactionCallback(callable $callback): void {
     if (!$this->inTransaction()) {
+      $this->exceptionThrown = TRUE;
       throw new \LogicException('Root transaction end callbacks can only be added when there is an active transaction.');
     }
     $this->postTransactionCallbacks[] = $callback;
@@ -537,6 +580,7 @@ protected function processRootRollback(): void {
   protected function processRootCommit(): void {
     $clientCommit = $this->commitClientTransaction();
     if (!$clientCommit) {
+      $this->exceptionThrown = TRUE;
       throw new TransactionCommitFailedException();
     }
   }
diff --git a/core/lib/Drupal/Core/Database/database.api.php b/core/lib/Drupal/Core/Database/database.api.php
index 62d63d36716..a0edf0114ac 100644
--- a/core/lib/Drupal/Core/Database/database.api.php
+++ b/core/lib/Drupal/Core/Database/database.api.php
@@ -159,18 +159,16 @@
  * @endcode
  *
  * @section sec_transaction Transactions
- * Drupal supports transactions, including a transparent fallback for
- * databases that do not support transactions. To start a new transaction,
- * call startTransaction(), like this:
+ * Drupal supports transactions. To start a new transaction, call
+ * startTransaction(), like this:
  * @code
  * $transaction = \Drupal::database()->startTransaction();
  * @endcode
- * The transaction will remain open for as long as the variable $transaction
- * remains in scope; when $transaction is destroyed, the transaction will be
- * committed. If your transaction is nested inside of another then Drupal will
- * track each transaction and only commit the outer-most transaction when the
- * last transaction object goes out of scope (when all relevant queries have
- * completed successfully).
+ * The transaction needs to be explicitly committed or rolled back once you
+ * have completed the database operations requested. If your transaction is
+ * nested inside of another then Drupal will track each transaction and only
+ * commit the outer-most transaction when the last transaction object is
+ * committed (when all relevant queries have completed successfully).
  *
  * Example:
  * @code
@@ -190,6 +188,9 @@
  *
  *     my_other_function($id);
  *
+ *     // The transaction has completed, commit it here.
+ *     $transaction->commitOrRelease();
+ *
  *     return $id;
  *   }
  *   catch (Exception $e) {
diff --git a/core/lib/Drupal/Core/Entity/Sql/SqlContentEntityStorage.php b/core/lib/Drupal/Core/Entity/Sql/SqlContentEntityStorage.php
index 459743cfc46..b001167d386 100644
--- a/core/lib/Drupal/Core/Entity/Sql/SqlContentEntityStorage.php
+++ b/core/lib/Drupal/Core/Entity/Sql/SqlContentEntityStorage.php
@@ -8,6 +8,7 @@
 use Drupal\Core\Database\DatabaseExceptionWrapper;
 use Drupal\Core\Database\SchemaException;
 use Drupal\Core\Database\Statement\FetchAs;
+use Drupal\Core\Database\TransactionOutOfOrderException;
 use Drupal\Core\Entity\ContentEntityInterface;
 use Drupal\Core\Entity\ContentEntityStorageBase;
 use Drupal\Core\Entity\ContentEntityTypeInterface;
@@ -751,13 +752,19 @@ public function delete(array $entities) {
     try {
       $transaction = $this->database->startTransaction();
       parent::delete($entities);
+      $transaction->commitOrRelease();
 
       // Ignore replica server temporarily.
       \Drupal::service('database.replica_kill_switch')->trigger();
     }
     catch (\Exception $e) {
       if (isset($transaction)) {
-        $transaction->rollBack();
+        try {
+          $transaction->rollBack();
+        }
+        catch (TransactionOutOfOrderException $rollbackException) {
+          Error::logException(\Drupal::logger($this->entityTypeId), $rollbackException);
+        }
       }
       Error::logException(\Drupal::logger($this->entityTypeId), $e);
       throw new EntityStorageException($e->getMessage(), $e->getCode(), $e);
@@ -802,6 +809,7 @@ public function save(EntityInterface $entity) {
     try {
       $transaction = $this->database->startTransaction();
       $return = parent::save($entity);
+      $transaction->commitOrRelease();
 
       // Ignore replica server temporarily.
       \Drupal::service('database.replica_kill_switch')->trigger();
@@ -809,7 +817,12 @@ public function save(EntityInterface $entity) {
     }
     catch (\Exception $e) {
       if (isset($transaction)) {
-        $transaction->rollBack();
+        try {
+          $transaction->rollBack();
+        }
+        catch (TransactionOutOfOrderException $rollbackException) {
+          Error::logException(\Drupal::logger($this->entityTypeId), $rollbackException);
+        }
       }
       Error::logException(\Drupal::logger($this->entityTypeId), $e);
       throw new EntityStorageException($e->getMessage(), $e->getCode(), $e);
@@ -853,12 +866,19 @@ public function restore(EntityInterface $entity) {
       // Insert the entity data in the dedicated tables.
       $this->saveToDedicatedTables($entity, FALSE, []);
 
+      $transaction->commitOrRelease();
+
       // Ignore replica server temporarily.
       \Drupal::service('database.replica_kill_switch')->trigger();
     }
     catch (\Exception $e) {
       if (isset($transaction)) {
-        $transaction->rollBack();
+        try {
+          $transaction->rollBack();
+        }
+        catch (TransactionOutOfOrderException $rollbackException) {
+          Error::logException(\Drupal::logger($this->entityTypeId), $rollbackException);
+        }
       }
       Error::logException(\Drupal::logger($this->entityTypeId), $e);
       throw new EntityStorageException($e->getMessage(), $e->getCode(), $e);
diff --git a/core/lib/Drupal/Core/Entity/Sql/SqlContentEntityStorageSchema.php b/core/lib/Drupal/Core/Entity/Sql/SqlContentEntityStorageSchema.php
index 6473a846133..4d230f23969 100644
--- a/core/lib/Drupal/Core/Entity/Sql/SqlContentEntityStorageSchema.php
+++ b/core/lib/Drupal/Core/Entity/Sql/SqlContentEntityStorageSchema.php
@@ -789,6 +789,10 @@ public function onFieldStorageDefinitionDelete(FieldStorageDefinitionInterface $
             ->from($this->getSelectQueryForFieldStorageDeletion($revision_table, $shared_table_field_columns, $dedicated_table_field_columns, $field_table_name))
             ->execute();
         }
+
+        if (isset($transaction)) {
+          $transaction->commitOrRelease();
+        }
       }
       catch (\Exception $e) {
         if ($this->database->supportsTransactionalDDL()) {
@@ -1742,6 +1746,10 @@ protected function updateDedicatedTableSchema(FieldStorageDefinitionInterface $s
         // to a dedicated table schema, hence we should use the proper API.
         $this->performFieldSchemaOperation('delete', $original);
         $this->performFieldSchemaOperation('create', $storage_definition);
+
+        if (isset($transaction)) {
+          $transaction->commitOrRelease();
+        }
       }
       catch (\Exception $e) {
         if ($this->database->supportsTransactionalDDL()) {
@@ -1836,6 +1844,10 @@ protected function updateSharedTableSchema(FieldStorageDefinitionInterface $stor
         // to a schema table schema, hence we should use the proper API.
         $this->performFieldSchemaOperation('delete', $original);
         $this->performFieldSchemaOperation('create', $storage_definition);
+
+        if (isset($transaction)) {
+          $transaction->commitOrRelease();
+        }
       }
       catch (\Exception $e) {
         if ($this->database->supportsTransactionalDDL()) {
diff --git a/core/lib/Drupal/Core/EventSubscriber/MenuRouterRebuildSubscriber.php b/core/lib/Drupal/Core/EventSubscriber/MenuRouterRebuildSubscriber.php
index 72e49111bac..0c0b31080f6 100644
--- a/core/lib/Drupal/Core/EventSubscriber/MenuRouterRebuildSubscriber.php
+++ b/core/lib/Drupal/Core/EventSubscriber/MenuRouterRebuildSubscriber.php
@@ -57,14 +57,15 @@ public function onRouterRebuild($event) {
   protected function menuLinksRebuild() {
     if ($this->lock->acquire(__FUNCTION__)) {
       try {
-        $transaction = $this->connection->startTransaction();
         // Ensure the menu links are up to date.
+        $transaction = $this->connection->startTransaction();
         $this->menuLinkManager->rebuild();
+        $transaction->commitOrRelease();
         // Ignore any database replicas temporarily.
         $this->replicaKillSwitch->trigger();
       }
       catch (\Exception $e) {
-        if (isset($transaction)) {
+        if (isset($transaction) && $this->connection->inTransaction()) {
           $transaction->rollBack();
         }
         Error::logException($this->logger, $e);
diff --git a/core/lib/Drupal/Core/Menu/MenuTreeStorage.php b/core/lib/Drupal/Core/Menu/MenuTreeStorage.php
index fb5d75a9c2d..2c1cef40e69 100644
--- a/core/lib/Drupal/Core/Menu/MenuTreeStorage.php
+++ b/core/lib/Drupal/Core/Menu/MenuTreeStorage.php
@@ -304,6 +304,7 @@ protected function doSave(array $link) {
         $this->updateParentalStatus($original);
       }
       $this->updateParentalStatus($link);
+      $transaction->commitOrRelease();
     }
     catch (\Exception $e) {
       if (isset($transaction)) {
diff --git a/core/lib/Drupal/Core/Routing/MatcherDumper.php b/core/lib/Drupal/Core/Routing/MatcherDumper.php
index 4907172f7df..b22464d9a5b 100644
--- a/core/lib/Drupal/Core/Routing/MatcherDumper.php
+++ b/core/lib/Drupal/Core/Routing/MatcherDumper.php
@@ -159,6 +159,8 @@ public function dump(array $options = []): string {
         }
         $insert->execute();
       }
+
+      $transaction->commitOrRelease();
     }
     catch (\Exception $e) {
       if (isset($transaction)) {
diff --git a/core/modules/pgsql/src/Driver/Database/pgsql/Connection.php b/core/modules/pgsql/src/Driver/Database/pgsql/Connection.php
index 40b2de75cf0..1de3caa18ea 100644
--- a/core/modules/pgsql/src/Driver/Database/pgsql/Connection.php
+++ b/core/modules/pgsql/src/Driver/Database/pgsql/Connection.php
@@ -85,6 +85,12 @@ class Connection extends DatabaseConnection implements SupportsTemporaryTablesIn
    * @see ::addSavepoint()
    * @see ::releaseSavepoint()
    * @see ::rollbackSavepoint()
+   *
+   * @deprecated in drupal:11.3.0 and is removed from drupal:13.0.0. Use
+   *   TransactionManager to start a transaction then call ::commitOrRelease()
+   *   or ::rollback() on it.
+   *
+   * @see https://www.drupal.org/node/3524461
    */
   protected array $savepoints = [];
 
@@ -227,13 +233,19 @@ public function query($query, array $args = [], $options = []) {
       // mimic MySQL and SQLite transactions which don't fail if a single query
       // fails. This is important for tables that are created on demand. For
       // example, \Drupal\Core\Cache\DatabaseBackend.
-      $this->addSavepoint();
+      if ($this->inTransaction()) {
+        $savepoint = $this->startTransaction('mimic_implicit_commit');
+      }
       try {
         $return = parent::query($query, $args, $options);
-        $this->releaseSavepoint();
+        if (isset($savepoint)) {
+          $savepoint->commitOrRelease();
+        }
       }
       catch (\Exception $e) {
-        $this->rollbackSavepoint();
+        if (isset($savepoint)) {
+          $savepoint->rollback();
+        }
         throw $e;
       }
     }
@@ -378,8 +390,15 @@ public function getFullQualifiedTableName($table) {
    * @param string $savepoint_name
    *   A string representing the savepoint name. By default,
    *   "mimic_implicit_commit" is used.
+   *
+   * @deprecated in drupal:11.3.0 and is removed from drupal:13.0.0. Use
+   *   TransactionManager to start a transaction then call ::commitOrRelease()
+   *   or ::rollback() on it.
+   *
+   * @see https://www.drupal.org/node/3524461
    */
   public function addSavepoint($savepoint_name = 'mimic_implicit_commit') {
+    @trigger_error(__METHOD__ . '() is deprecated in drupal:11.3.0 and is removed from drupal:13.0.0. Use TransactionManager to start a transaction then call ::commitOrRelease() or ::rollback() on it. See https://www.drupal.org/node/3524461', E_USER_DEPRECATED);
     if ($this->inTransaction()) {
       $this->savepoints[$savepoint_name] = $this->startTransaction($savepoint_name);
     }
@@ -391,8 +410,15 @@ public function addSavepoint($savepoint_name = 'mimic_implicit_commit') {
    * @param string $savepoint_name
    *   A string representing the savepoint name. By default,
    *   "mimic_implicit_commit" is used.
+   *
+   * @deprecated in drupal:11.3.0 and is removed from drupal:13.0.0. Use
+   *   TransactionManager to start a transaction then call ::commitOrRelease()
+   *   or ::rollback() on it.
+   *
+   * @see https://www.drupal.org/node/3524461
    */
   public function releaseSavepoint($savepoint_name = 'mimic_implicit_commit') {
+    @trigger_error(__METHOD__ . '() is deprecated in drupal:11.3.0 and is removed from drupal:13.0.0. Use TransactionManager to start a transaction then call ::commitOrRelease() or ::rollback() on it. See https://www.drupal.org/node/3524461', E_USER_DEPRECATED);
     if ($this->inTransaction() && $this->transactionManager()->has($savepoint_name)) {
       unset($this->savepoints[$savepoint_name]);
     }
@@ -404,8 +430,15 @@ public function releaseSavepoint($savepoint_name = 'mimic_implicit_commit') {
    * @param string $savepoint_name
    *   A string representing the savepoint name. By default,
    *   "mimic_implicit_commit" is used.
+   *
+   * @deprecated in drupal:11.3.0 and is removed from drupal:13.0.0. Use
+   *   TransactionManager to start a transaction then call ::commitOrRelease()
+   *   or ::rollback() on it.
+   *
+   * @see https://www.drupal.org/node/3524461
    */
   public function rollbackSavepoint($savepoint_name = 'mimic_implicit_commit') {
+    @trigger_error(__METHOD__ . '() is deprecated in drupal:11.3.0 and is removed from drupal:13.0.0. Use TransactionManager to start a transaction then call ::commitOrRelease() or ::rollback() on it. See https://www.drupal.org/node/3524461', E_USER_DEPRECATED);
     if ($this->inTransaction() && $this->transactionManager()->has($savepoint_name)) {
       $this->savepoints[$savepoint_name]->rollBack();
       unset($this->savepoints[$savepoint_name]);
diff --git a/core/modules/pgsql/src/Driver/Database/pgsql/Delete.php b/core/modules/pgsql/src/Driver/Database/pgsql/Delete.php
index 9585a9c4bc5..1502d229096 100644
--- a/core/modules/pgsql/src/Driver/Database/pgsql/Delete.php
+++ b/core/modules/pgsql/src/Driver/Database/pgsql/Delete.php
@@ -13,17 +13,22 @@ class Delete extends QueryDelete {
    * {@inheritdoc}
    */
   public function execute() {
-    $this->connection->addSavepoint();
+    if ($this->connection->inTransaction()) {
+      $savepoint = $this->connection->startTransaction('mimic_implicit_commit');
+    }
     try {
       $result = parent::execute();
+      if (isset($savepoint)) {
+        $savepoint->commitOrRelease();
+      }
+      return $result;
     }
     catch (\Exception $e) {
-      $this->connection->rollbackSavepoint();
+      if (isset($savepoint)) {
+        $savepoint->rollback();
+      }
       throw $e;
     }
-    $this->connection->releaseSavepoint();
-
-    return $result;
   }
 
 }
diff --git a/core/modules/pgsql/src/Driver/Database/pgsql/Insert.php b/core/modules/pgsql/src/Driver/Database/pgsql/Insert.php
index 223daeb8d8a..e51b8c8a3cc 100644
--- a/core/modules/pgsql/src/Driver/Database/pgsql/Insert.php
+++ b/core/modules/pgsql/src/Driver/Database/pgsql/Insert.php
@@ -86,16 +86,22 @@ public function execute() {
     // mimic MySQL and SQLite transactions which don't fail if a single query
     // fails. This is important for tables that are created on demand. For
     // example, \Drupal\Core\Cache\DatabaseBackend.
-    $this->connection->addSavepoint();
+    if ($this->connection->inTransaction()) {
+      $savepoint = $this->connection->startTransaction('mimic_implicit_commit');
+    }
     try {
       $stmt->execute(NULL, $this->queryOptions);
       if (isset($table_information->serial_fields[0])) {
         $last_insert_id = $stmt->fetchField();
       }
-      $this->connection->releaseSavepoint();
+      if (isset($savepoint)) {
+        $savepoint->commitOrRelease();
+      }
     }
     catch (\Exception $e) {
-      $this->connection->rollbackSavepoint();
+      if (isset($savepoint)) {
+        $savepoint->rollback();
+      }
       $this->connection->exceptionHandler()->handleExecutionException($e, $stmt, [], $this->queryOptions);
     }
 
diff --git a/core/modules/pgsql/src/Driver/Database/pgsql/Schema.php b/core/modules/pgsql/src/Driver/Database/pgsql/Schema.php
index 440dc7179db..a7eade89dd9 100644
--- a/core/modules/pgsql/src/Driver/Database/pgsql/Schema.php
+++ b/core/modules/pgsql/src/Driver/Database/pgsql/Schema.php
@@ -147,7 +147,10 @@ public function queryTableInformation($table) {
         'blob_fields' => [],
         'sequences' => [],
       ];
-      $this->connection->addSavepoint();
+
+      if ($this->connection->inTransaction()) {
+        $savepoint = $this->connection->startTransaction('mimic_implicit_commit');
+      }
 
       try {
         // The bytea columns and sequences for a table can be found in
@@ -171,10 +174,14 @@ public function queryTableInformation($table) {
         ]);
       }
       catch (\Exception $e) {
-        $this->connection->rollbackSavepoint();
+        if (isset($savepoint)) {
+          $savepoint->rollback();
+        }
         throw $e;
       }
-      $this->connection->releaseSavepoint();
+      if (isset($savepoint)) {
+        $savepoint->commitOrRelease();
+      }
 
       // If the table information does not yet exist in the PostgreSQL
       // metadata, then return the default table information here, so that it
@@ -260,7 +267,9 @@ public function queryFieldInformation($table, $field, $constraint_type = 'c') {
     $schema = $prefixInfo['schema'];
     $table_name = $prefixInfo['table'];
 
-    $this->connection->addSavepoint();
+    if ($this->connection->inTransaction()) {
+      $savepoint = $this->connection->startTransaction('mimic_implicit_commit');
+    }
 
     try {
       $checks = $this->connection->query("SELECT conname FROM pg_class cl INNER JOIN pg_constraint co ON co.conrelid = cl.oid INNER JOIN pg_attribute attr ON attr.attrelid = cl.oid AND attr.attnum = ANY (co.conkey) INNER JOIN pg_namespace ns ON cl.relnamespace = ns.oid WHERE co.contype = :constraint_type AND ns.nspname = :schema AND cl.relname = :table AND attr.attname = :column", [
@@ -271,11 +280,15 @@ public function queryFieldInformation($table, $field, $constraint_type = 'c') {
       ]);
     }
     catch (\Exception $e) {
-      $this->connection->rollbackSavepoint();
+      if (isset($savepoint)) {
+        $savepoint->rollback();
+      }
       throw $e;
     }
 
-    $this->connection->releaseSavepoint();
+    if (isset($savepoint)) {
+      $savepoint->commitOrRelease();
+    }
 
     $field_information = $checks->fetchCol();
 
diff --git a/core/modules/pgsql/src/Driver/Database/pgsql/Select.php b/core/modules/pgsql/src/Driver/Database/pgsql/Select.php
index ac3adaceaa5..a0f15be8083 100644
--- a/core/modules/pgsql/src/Driver/Database/pgsql/Select.php
+++ b/core/modules/pgsql/src/Driver/Database/pgsql/Select.php
@@ -143,15 +143,21 @@ public function addExpression($expression, $alias = NULL, $arguments = []) {
    * {@inheritdoc}
    */
   public function execute() {
-    $this->connection->addSavepoint();
+    if ($this->connection->inTransaction()) {
+      $savepoint = $this->connection->startTransaction('mimic_implicit_commit');
+    }
     try {
       $result = parent::execute();
     }
     catch (\Exception $e) {
-      $this->connection->rollbackSavepoint();
+      if (isset($savepoint)) {
+        $savepoint->rollback();
+      }
       throw $e;
     }
-    $this->connection->releaseSavepoint();
+    if (isset($savepoint)) {
+      $savepoint->commitOrRelease();
+    }
 
     return $result;
   }
diff --git a/core/modules/pgsql/src/Driver/Database/pgsql/Truncate.php b/core/modules/pgsql/src/Driver/Database/pgsql/Truncate.php
index 18115e0a7d0..d35336c56c9 100644
--- a/core/modules/pgsql/src/Driver/Database/pgsql/Truncate.php
+++ b/core/modules/pgsql/src/Driver/Database/pgsql/Truncate.php
@@ -13,15 +13,21 @@ class Truncate extends QueryTruncate {
    * {@inheritdoc}
    */
   public function execute() {
-    $this->connection->addSavepoint();
+    if ($this->connection->inTransaction()) {
+      $savepoint = $this->connection->startTransaction('mimic_implicit_commit');
+    }
     try {
       $result = parent::execute();
     }
     catch (\Exception $e) {
-      $this->connection->rollbackSavepoint();
+      if (isset($savepoint)) {
+        $savepoint->rollback();
+      }
       throw $e;
     }
-    $this->connection->releaseSavepoint();
+    if (isset($savepoint)) {
+      $savepoint->commitOrRelease();
+    }
 
     return $result;
   }
diff --git a/core/modules/pgsql/src/Driver/Database/pgsql/Update.php b/core/modules/pgsql/src/Driver/Database/pgsql/Update.php
index 0b4f9a0f4f6..19206f85e6a 100644
--- a/core/modules/pgsql/src/Driver/Database/pgsql/Update.php
+++ b/core/modules/pgsql/src/Driver/Database/pgsql/Update.php
@@ -71,14 +71,20 @@ public function execute() {
       }
     }
 
-    $this->connection->addSavepoint();
+    if ($this->connection->inTransaction()) {
+      $savepoint = $this->connection->startTransaction('mimic_implicit_commit');
+    }
     try {
       $stmt->execute(NULL, $this->queryOptions);
-      $this->connection->releaseSavepoint();
+      if (isset($savepoint)) {
+        $savepoint->commitOrRelease();
+      }
       return $stmt->rowCount();
     }
     catch (\Exception $e) {
-      $this->connection->rollbackSavepoint();
+      if (isset($savepoint)) {
+        $savepoint->rollback();
+      }
       $this->connection->exceptionHandler()->handleExecutionException($e, $stmt, [], $this->queryOptions);
     }
   }
diff --git a/core/modules/pgsql/src/Driver/Database/pgsql/Upsert.php b/core/modules/pgsql/src/Driver/Database/pgsql/Upsert.php
index 35823a270b1..a26555b2a00 100644
--- a/core/modules/pgsql/src/Driver/Database/pgsql/Upsert.php
+++ b/core/modules/pgsql/src/Driver/Database/pgsql/Upsert.php
@@ -78,14 +78,20 @@ public function execute() {
     // mimic MySQL and SQLite transactions which don't fail if a single query
     // fails. This is important for tables that are created on demand. For
     // example, \Drupal\Core\Cache\DatabaseBackend.
-    $this->connection->addSavepoint();
+    if ($this->connection->inTransaction()) {
+      $savepoint = $this->connection->startTransaction('mimic_implicit_commit');
+    }
     try {
       $stmt->execute(NULL, $options);
-      $this->connection->releaseSavepoint();
+      if (isset($savepoint)) {
+        $savepoint->commitOrRelease();
+      }
       return $stmt->rowCount();
     }
     catch (\Exception $e) {
-      $this->connection->rollbackSavepoint();
+      if (isset($savepoint)) {
+        $savepoint->rollback();
+      }
       $this->connection->exceptionHandler()->handleExecutionException($e, $stmt, [], $options);
     }
   }
diff --git a/core/modules/sqlite/src/Driver/Database/sqlite/Insert.php b/core/modules/sqlite/src/Driver/Database/sqlite/Insert.php
index 301b6e5bb4f..6bc3e7362f2 100644
--- a/core/modules/sqlite/src/Driver/Database/sqlite/Insert.php
+++ b/core/modules/sqlite/src/Driver/Database/sqlite/Insert.php
@@ -65,6 +65,7 @@ public function execute() {
             $this->connection->exceptionHandler()->handleExecutionException($e, $stmt, $insert_values, $this->queryOptions);
           }
         }
+        $transaction->commitOrRelease();
       }
       // Re-initialize the values array so that we can re-use this query.
       $this->insertValues = [];
diff --git a/core/modules/workspaces/src/WorkspaceAssociation.php b/core/modules/workspaces/src/WorkspaceAssociation.php
index 8ed8fbb2ad2..d19ebeebe28 100644
--- a/core/modules/workspaces/src/WorkspaceAssociation.php
+++ b/core/modules/workspaces/src/WorkspaceAssociation.php
@@ -102,6 +102,8 @@ public function trackEntity(RevisionableInterface $entity, WorkspaceInterface $w
         }
         $insert_query->execute();
       }
+
+      $transaction->commitOrRelease();
     }
     catch (\Exception $e) {
       if (isset($transaction)) {
diff --git a/core/modules/workspaces/src/WorkspaceMerger.php b/core/modules/workspaces/src/WorkspaceMerger.php
index 30792ed0c6c..2ac5db3c4a3 100644
--- a/core/modules/workspaces/src/WorkspaceMerger.php
+++ b/core/modules/workspaces/src/WorkspaceMerger.php
@@ -65,6 +65,7 @@ public function merge() {
           }
         }
       }
+      $transaction->commitOrRelease();
     }
     catch (\Exception $e) {
       if (isset($transaction)) {
diff --git a/core/modules/workspaces/src/WorkspacePublisher.php b/core/modules/workspaces/src/WorkspacePublisher.php
index a964234daf1..78e6c1d9873 100644
--- a/core/modules/workspaces/src/WorkspacePublisher.php
+++ b/core/modules/workspaces/src/WorkspacePublisher.php
@@ -84,6 +84,7 @@ public function publish() {
           }
         }
       });
+      $transaction->commitOrRelease();
     }
     catch (\Exception $e) {
       if (isset($transaction)) {
diff --git a/core/tests/Drupal/KernelTests/Core/Database/DeleteTruncateTest.php b/core/tests/Drupal/KernelTests/Core/Database/DeleteTruncateTest.php
index 0c5da802929..d8fffb5c1a1 100644
--- a/core/tests/Drupal/KernelTests/Core/Database/DeleteTruncateTest.php
+++ b/core/tests/Drupal/KernelTests/Core/Database/DeleteTruncateTest.php
@@ -97,10 +97,9 @@ public function testTruncateInTransaction(): void {
     $num_records_after = $this->connection->select('test')->countQuery()->execute()->fetchField();
     $this->assertEquals(0, $num_records_after);
 
-    // Close the transaction, and check that there are still no records in the
+    // Commit the transaction, and check that there are still no records in the
     // table.
-    // phpcs:ignore DrupalPractice.CodeAnalysis.VariableAnalysis.UnusedVariable
-    $transaction = NULL;
+    $transaction->commitOrRelease();
     $this->assertFalse($this->connection->inTransaction());
     $num_records_after = $this->connection->select('test')->countQuery()->execute()->fetchField();
     $this->assertEquals(0, $num_records_after);
diff --git a/core/tests/Drupal/KernelTests/Core/Database/DriverSpecificTransactionTestBase.php b/core/tests/Drupal/KernelTests/Core/Database/DriverSpecificTransactionTestBase.php
index cbda6e3d7f7..660b67e39c1 100644
--- a/core/tests/Drupal/KernelTests/Core/Database/DriverSpecificTransactionTestBase.php
+++ b/core/tests/Drupal/KernelTests/Core/Database/DriverSpecificTransactionTestBase.php
@@ -12,6 +12,7 @@
 use Drupal\Core\Database\Transaction\TransactionManagerBase;
 use Drupal\Core\Database\TransactionNameNonUniqueException;
 use Drupal\Core\Database\TransactionOutOfOrderException;
+use PHPUnit\Framework\Attributes\IgnoreDeprecations;
 
 /**
  * Tests the transaction abstraction system.
@@ -251,6 +252,7 @@ public function testRollbackRootWithActiveSavepoint(): void {
   /**
    * Tests savepoint transaction rollback.
    */
+  #[IgnoreDeprecations]
   public function testRollbackSavepoint(): void {
     $transaction = $this->createRootTransaction();
     $savepoint = $this->createFirstSavepointTransaction();
@@ -343,6 +345,7 @@ public function testRollbackSavepointWithLaterSavepoint(): void {
    * The behavior of this test should be identical for connections that support
    * transactions and those that do not.
    */
+  #[IgnoreDeprecations]
   public function testCommittedTransaction(): void {
     try {
       // Create two nested transactions. The changes should be committed.
@@ -362,6 +365,7 @@ public function testCommittedTransaction(): void {
   /**
    * Tests the compatibility of transactions with DDL statements.
    */
+  #[IgnoreDeprecations]
   public function testTransactionWithDdlStatement(): void {
     // First, test that a commit works normally, even with DDL statements.
     $transaction = $this->createRootTransaction('', FALSE);
@@ -542,6 +546,7 @@ public function assertRowAbsent(string $name, ?string $message = NULL): void {
   /**
    * Tests transaction stacking, commit, and rollback.
    */
+  #[IgnoreDeprecations]
   public function testTransactionStacking(): void {
     // Standard case: pop the inner transaction before the outer transaction.
     $transaction = $this->createRootTransaction('', FALSE);
@@ -579,6 +584,7 @@ public function testTransactionStacking(): void {
   /**
    * Tests that transactions can continue to be used if a query fails.
    */
+  #[IgnoreDeprecations]
   public function testQueryFailureInTransaction(): void {
     $transaction = $this->createRootTransaction('test_transaction', FALSE);
     $this->connection->schema()->dropTable('test');
@@ -697,6 +703,7 @@ public function testQueryFailureInTransaction(): void {
   /**
    * Tests releasing a savepoint before last is safe.
    */
+  #[IgnoreDeprecations]
   public function testReleaseIntermediateSavepoint(): void {
     $transaction = $this->createRootTransaction();
     $savepoint1 = $this->createFirstSavepointTransaction('', FALSE);
@@ -737,6 +744,7 @@ public function testReleaseIntermediateSavepoint(): void {
   /**
    * Tests committing a transaction while savepoints are active.
    */
+  #[IgnoreDeprecations]
   public function testCommitWithActiveSavepoint(): void {
     $transaction = $this->createRootTransaction();
     // phpcs:ignore DrupalPractice.CodeAnalysis.VariableAnalysis
@@ -792,6 +800,7 @@ public function testRootTransactionEndCallbackAddedWithoutTransaction(): void {
   /**
    * Tests post-transaction callback executes after transaction commit.
    */
+  #[IgnoreDeprecations]
   public function testRootTransactionEndCallbackCalledOnCommit(): void {
     $transaction = $this->createRootTransaction('', FALSE);
     $this->connection->transactionManager()->addPostTransactionCallback([$this, 'rootTransactionCallback']);
@@ -835,6 +844,7 @@ public function testRootTransactionEndCallbackCalledAfterRollbackAndDestruction(
   /**
    * Tests post-transaction callback executes after a DDL statement.
    */
+  #[IgnoreDeprecations]
   public function testRootTransactionEndCallbackCalledAfterDdlAndDestruction(): void {
     $transaction = $this->createRootTransaction('', FALSE);
     $this->connection->transactionManager()->addPostTransactionCallback([$this, 'rootTransactionCallback']);
diff --git a/core/tests/Drupal/KernelTests/Core/Database/TransactionTest.php b/core/tests/Drupal/KernelTests/Core/Database/TransactionTest.php
index f40a977f6f4..769b93be823 100644
--- a/core/tests/Drupal/KernelTests/Core/Database/TransactionTest.php
+++ b/core/tests/Drupal/KernelTests/Core/Database/TransactionTest.php
@@ -509,9 +509,10 @@ public function testTransactionWithDdlStatement(): void {
       }
     }
     $this->assertRowPresent('row');
+    unset($transaction);
+    $this->cleanUp();
 
     // Even in different order.
-    $this->cleanUp();
     $transaction = $this->createRootTransaction('', FALSE);
     $this->executeDDLStatement();
     $this->insertRow('row');
@@ -533,9 +534,10 @@ public function testTransactionWithDdlStatement(): void {
       }
     }
     $this->assertRowPresent('row');
+    unset($transaction);
+    $this->cleanUp();
 
     // Even with stacking.
-    $this->cleanUp();
     $transaction = $this->createRootTransaction('', FALSE);
     $transaction2 = $this->createFirstSavepointTransaction('', FALSE);
     $this->executeDDLStatement();
@@ -560,23 +562,18 @@ public function testTransactionWithDdlStatement(): void {
     $this->insertRow('row');
     $transaction3->commitOrRelease();
 
-    if ($this->connection->supportsTransactionalDDL()) {
-      $transaction->commitOrRelease();
-    }
-    else {
-      try {
-        $transaction->commitOrRelease();
-        $this->fail('TransactionOutOfOrderException was expected, but did not throw.');
-      }
-      catch (TransactionOutOfOrderException) {
-        // Just continue, this is out or order since $transaction3 started a
-        // new root.
-      }
-    }
+    // If this is running on a database supporting transactional DDL,
+    // $transaction is still active. Otherwise, the initial root
+    // transaction's commit is a no-op now since the DDL statement forced a
+    // commit and then $transaction3 was fully committed too: we do not raise
+    // any error in this case.
+    $transaction->commitOrRelease();
+
     $this->assertRowPresent('row');
+    unset($transaction, $transaction2, $transaction3);
+    $this->cleanUp();
 
     // A transaction after a DDL statement should still work the same.
-    $this->cleanUp();
     $transaction = $this->createRootTransaction('', FALSE);
     $transaction2 = $this->createFirstSavepointTransaction('', FALSE);
     $this->executeDDLStatement();
@@ -614,12 +611,13 @@ public function testTransactionWithDdlStatement(): void {
       }
     }
     $this->assertRowAbsent('row');
+    unset($transaction, $transaction2, $transaction3);
+    $this->cleanUp();
 
     // The behavior of a rollback depends on the type of database server.
     if ($this->connection->supportsTransactionalDDL()) {
       // For database servers that support transactional DDL, a rollback
       // of a transaction including DDL statements should be possible.
-      $this->cleanUp();
       $transaction = $this->createRootTransaction('', FALSE);
       $this->insertRow('row');
       $this->executeDDLStatement();
@@ -994,11 +992,9 @@ public function testCommitWithActiveSavepoint(): void {
     $this->assertSame(0, $this->connection->transactionManager()->stackDepth());
     $this->assertFalse($this->connection->inTransaction());
     $this->assertRowPresent('row');
-    // Trying to release the inner (savepoint) Transaction object, throws an
-    // exception since it was dropped by the database already, and removed from
-    // our transaction stack.
-    $this->expectException(TransactionOutOfOrderException::class);
-    $this->expectExceptionMessageMatches("/^Error attempting commit of .*\\\\savepoint_2\\. Active stack: .* empty/");
+    // Trying to release the inner (savepoint) Transaction object is a no-op
+    // at this point since the root transaction was fully committed already: we
+    // do not raise any error in this case.
     $savepoint2->commitOrRelease();
   }
 
@@ -1117,7 +1113,26 @@ public function testRootTransactionEndCallbackCalledAfterDdlAndDestruction(): vo
     $this->assertRowAbsent('rtcRollback');
     $this->assertRowPresent('row');
 
-    // Destruct the transaction.
+    // Commit and destruct the transaction. Non-transactional DDL database
+    // will have committed already, so the explicit commit would be a no-op in
+    // that case.
+    if ($this->connection->supportsTransactionalDDL()) {
+      $transaction->commitOrRelease();
+    }
+    else {
+      set_error_handler(static function (int $errno, string $errstr): bool {
+        throw new \ErrorException($errstr);
+      });
+      try {
+        $transaction->commitOrRelease();
+      }
+      catch (\ErrorException $e) {
+        $this->assertSame('Transaction::commitOrRelease() was not processed because a prior execution of a DDL statement already committed the transaction.', $e->getMessage());
+      }
+      finally {
+        restore_error_handler();
+      }
+    }
     unset($transaction);
 
     // The post-transaction callback should now have inserted a 'rtcCommit'
