diff --git a/core/tests/Drupal/KernelTests/Core/Database/InsertTest.php b/core/tests/Drupal/KernelTests/Core/Database/InsertTest.php
index 1ba5d1a74b..b205285bd9 100644
--- a/core/tests/Drupal/KernelTests/Core/Database/InsertTest.php
+++ b/core/tests/Drupal/KernelTests/Core/Database/InsertTest.php
@@ -232,4 +232,54 @@ public function testInsertIntegrityViolation() {
       ->execute();
   }
 
+  /**
+   * Tests insertion integrity violation with no default value for a column.
+   */
+  public function testInsertIntegrityViolationMk2() {
+    $this->connection->schema()->createTable('testmk2', [
+      'fields' => [
+        'bid' => [
+          'description' => 'Primary Key: Unique batch ID.',
+          // This is not a serial column, to allow both progressive and
+          // non-progressive batches. See batch_process().
+          'type' => 'int',
+          'unsigned' => TRUE,
+          'not null' => TRUE,
+        ],
+        'token' => [
+          'description' => "A string token generated against the current user's session id and the batch id, used to ensure that only the user who submitted the batch can effectively access it.",
+          'type' => 'varchar_ascii',
+          'length' => 64,
+          'not null' => TRUE,
+        ],
+        'timestamp' => [
+          'description' => 'A Unix timestamp indicating when this batch was submitted for processing. Stale batches are purged at cron time.',
+          'type' => 'int',
+          'not null' => TRUE,
+        ],
+        'batch' => [
+          'description' => 'A serialized array containing the processing data for the batch.',
+          'type' => 'blob',
+          'not null' => FALSE,
+          'size' => 'big',
+        ],
+      ],
+      'primary key' => ['bid'],
+      'indexes' => [
+        'token' => ['token'],
+      ],
+    ]);
+
+    // Try inserting a record that misses the value for the 'age' column,
+    // should raise an IntegrityConstraintViolationException.
+    // $this->expectException(IntegrityConstraintViolationException::class);
+    $this->connection->insert('testmk2')
+      ->fields([
+        'timestamp' => REQUEST_TIME,
+        'token' => '',
+        'batch' => NULL,
+      ])
+      ->execute();
+  }
+
 }
