diff --git a/core/modules/menu_link_content/src/Form/MenuLinkContentForm.php b/core/modules/menu_link_content/src/Form/MenuLinkContentForm.php
index 6bffb46a37..9c70c6d085 100644
--- a/core/modules/menu_link_content/src/Form/MenuLinkContentForm.php
+++ b/core/modules/menu_link_content/src/Form/MenuLinkContentForm.php
@@ -35,10 +35,8 @@ class MenuLinkContentForm extends ContentEntityForm {
 
   /**
    * The path validator.
-   *
-   * @var \Drupal\Core\Path\PathValidatorInterface
    */
-  protected $pathValidator;
+  protected PathValidatorInterface $pathValidator;
 
   /**
    * Constructs a MenuLinkContentForm object.
@@ -47,16 +45,20 @@ class MenuLinkContentForm extends ContentEntityForm {
    *   The entity repository.
    * @param \Drupal\Core\Menu\MenuParentFormSelectorInterface $menu_parent_selector
    *   The menu parent form selector service.
-   * @param \Drupal\Core\Language\LanguageManagerInterface $language_manager
-   *   The language manager.
-   * @param \Drupal\Core\Path\PathValidatorInterface $path_validator
+   * @param \Drupal\Core\Path\PathValidatorInterface|\Drupal\Core\Language\LanguageManagerInterface $path_validator
    *   The path validator.
-   * @param \Drupal\Core\Entity\EntityTypeBundleInfoInterface $entity_type_bundle_info
+   * @param \Drupal\Core\Entity\EntityTypeBundleInfoInterface|\Drupal\Core\Path\PathValidatorInterface $entity_type_bundle_info
    *   The entity type bundle service.
-   * @param \Drupal\Component\Datetime\TimeInterface $time
+   * @param \Drupal\Component\Datetime\TimeInterface|\Drupal\Core\Entity\EntityTypeBundleInfoInterface $time
    *   The time service.
    */
-  public function __construct(EntityRepositoryInterface $entity_repository, MenuParentFormSelectorInterface $menu_parent_selector, LanguageManagerInterface $language_manager, PathValidatorInterface $path_validator, EntityTypeBundleInfoInterface $entity_type_bundle_info = NULL, TimeInterface $time = NULL) {
+  public function __construct(EntityRepositoryInterface $entity_repository, MenuParentFormSelectorInterface $menu_parent_selector, PathValidatorInterface|LanguageManagerInterface $path_validator, EntityTypeBundleInfoInterface|PathValidatorInterface $entity_type_bundle_info, TimeInterface|EntityTypeBundleInfoInterface $time) {
+    if ($path_validator instanceof LanguageManagerInterface) {
+      $path_validator = func_get_arg(3);
+      $entity_type_bundle_info = func_get_arg(4);
+      $time = func_num_args() >= 6 ? func_get_arg(5) : NULL;
+      @trigger_error('Calling ' . __CLASS__ . '::__construct() with the $language_manager argument is deprecated in drupal:10.1.0 and is removed in drupal:11.0.0. See https://www.drupal.org/node/3325178', E_USER_DEPRECATED);
+    }
     parent::__construct($entity_repository, $entity_type_bundle_info, $time);
     $this->menuParentSelector = $menu_parent_selector;
     $this->pathValidator = $path_validator;
@@ -69,7 +71,6 @@ public static function create(ContainerInterface $container) {
     return new static(
       $container->get('entity.repository'),
       $container->get('menu.parent_form_selector'),
-      $container->get('language_manager'),
       $container->get('path.validator'),
       $container->get('entity_type.bundle.info'),
       $container->get('datetime.time')
diff --git a/core/modules/menu_link_content/tests/src/Kernel/MenuLinkContentDeprecationsTest.php b/core/modules/menu_link_content/tests/src/Kernel/MenuLinkContentDeprecationsTest.php
new file mode 100644
index 0000000000..0d8b45bc39
--- /dev/null
+++ b/core/modules/menu_link_content/tests/src/Kernel/MenuLinkContentDeprecationsTest.php
@@ -0,0 +1,39 @@
+<?php
+
+namespace Drupal\Tests\menu_link_content\Kernel;
+
+use Drupal\Component\Datetime\TimeInterface;
+use Drupal\Core\Entity\EntityRepositoryInterface;
+use Drupal\Core\Entity\EntityTypeBundleInfoInterface;
+use Drupal\Core\Language\LanguageManagerInterface;
+use Drupal\Core\Menu\MenuParentFormSelectorInterface;
+use Drupal\Core\Path\PathValidatorInterface;
+use Drupal\KernelTests\KernelTestBase;
+use Drupal\menu_link_content\Form\MenuLinkContentForm;
+
+/**
+ * Tests the deprecation notices of the menu_link_content module.
+ *
+ * @coversDefaultClass \Drupal\menu_link_content\Form\MenuLinkContentForm
+ *
+ * @group block
+ * @group legacy
+ */
+class MenuLinkContentDeprecationsTest extends KernelTestBase {
+
+  /**
+   * Tests the deprecation of MenuLinkContentForm constructor.
+   */
+  public function testMenuLinkContentFormConstructorDeprecation(): void {
+    $this->expectDeprecation('Calling Drupal\menu_link_content\Form\MenuLinkContentForm::__construct() with the $language_manager argument is deprecated in drupal:10.1.0 and is removed in drupal:11.0.0. See https://www.drupal.org/node/3325178');
+    new MenuLinkContentForm(
+      $this->createMock(EntityRepositoryInterface::class),
+      $this->createMock(MenuParentFormSelectorInterface::class),
+      $this->createMock(LanguageManagerInterface::class),
+      $this->createMock(PathValidatorInterface::class),
+      $this->createMock(EntityTypeBundleInfoInterface::class),
+      $this->createMock(TimeInterface::class)
+    );
+  }
+
+}
diff --git a/core/phpstan-baseline.neon b/core/phpstan-baseline.neon
index cf84abce27..1ee464e277 100644
--- a/core/phpstan-baseline.neon
+++ b/core/phpstan-baseline.neon
@@ -1590,11 +1590,6 @@ parameters:
 			count: 1
 			path: modules/media_library/tests/src/FunctionalJavascript/EmbeddedFormWidgetTest.php
 
-		-
-			message: "#^Constructor of class Drupal\\\\menu_link_content\\\\Form\\\\MenuLinkContentForm has an unused parameter \\$language_manager\\.$#"
-			count: 1
-			path: modules/menu_link_content/src/Form/MenuLinkContentForm.php
-
 		-
 			message: "#^Method Drupal\\\\menu_link_content\\\\Form\\\\MenuLinkContentForm\\:\\:save\\(\\) should return int but return statement is missing\\.$#"
 			count: 1
